From c9c79c331341d7e9f10aad5d73c8946beb118428 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 31 Dec 2015 15:35:47 -0500
Subject: [PATCH] Implement UUID based Skull Owner API

---
 .../bukkit/craftbukkit/inventory/CraftMetaSkull.java   | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
index ce5425f..5afde67 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
@@ -1,6 +1,7 @@
 package org.bukkit.craftbukkit.inventory;
 
 import java.util.Map;
+import java.util.UUID;
 
 import net.minecraft.server.GameProfileSerializer;
 import net.minecraft.server.NBTBase;
@@ -124,15 +125,26 @@ class CraftMetaSkull extends CraftMetaItem implements SkullMeta {
         return hasOwner() ? profile.getName() : null;
     }
 
-    public boolean setOwner(String name) {
+    // EMC start
+    public UUID getOwnerUUID() {
+        return hasOwner() ? profile.getId() : null;
+    }
+    public boolean setOwnerUUID(UUID uuid) {
+        return setOwner(null, uuid);
+    }
+    public boolean setOwner(String owner) {
+        return setOwner(owner, null);
+    }
+    public boolean setOwner(String name, UUID uuid) {
+        // EMC end
         if (name != null && name.length() > MAX_OWNER_LENGTH) {
             return false;
         }
 
-        if (name == null) {
+        if (name == null && uuid == null) { // EMC
             profile = null;
         } else {
-            profile = new GameProfile(null, name);
+            profile = new GameProfile(uuid, name); // EMC
         }
 
         return true;
-- 
2.6.4

