--- a/src/main/java/com/destroystokyo/paper/event/player/PlayerLaunchProjectileEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerLaunchProjectileEvent.java
@@ -25,6 +_,7 @@
     private final Projectile projectile;
     private final ItemStack itemStack;
     private boolean consumeItem = true;
+    private boolean playLaunchSound = true; // EMC - Add playLaunchSound API
 
     private boolean cancelled;
 
@@ -71,6 +_,15 @@
         this.consumeItem = consumeItem;
     }
 
+    // EMC start - Add playLaunchSound API
+    public boolean shouldPlayLaunchSound() {
+        return this.playLaunchSound;
+    }
+
+    public void setShouldPlayLaunchSound(boolean playLaunchSound) {
+        this.playLaunchSound = playLaunchSound;
+    }
+    // EMC end - Add playLaunchSound API
     @Override
     public boolean isCancelled() {
         return this.cancelled;
