--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -55,6 +_,7 @@
 import net.minecraft.world.level.block.state.properties.RotationSegment;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.slf4j.Logger;
 
 public interface DispenseItemBehavior {
@@ -569,6 +_,11 @@
                             blockStateBase -> blockStateBase.hasProperty(BeehiveBlock.HONEY_LEVEL) && blockStateBase.getBlock() instanceof BeehiveBlock
                         )
                         && blockState.getValue(BeehiveBlock.HONEY_LEVEL) >= 5) {
+                        // EMC start - BlockHarvestBeehiveEvent
+                        if (!(new com.empireminecraft.empirecraft.event.block.BlockHarvestBeehiveEvent(org.bukkit.craftbukkit.block.CraftBlock.at(serverLevel, blockSource.pos()), org.bukkit.craftbukkit.block.CraftBlock.at(serverLevel, blockPos), CraftItemStack.asCraftMirror(item.copyWithCount(1))).callEvent())) {
+                            return item;
+                        }
+                        // EMC end - BlockHarvestBeehiveEvent
                         ((BeehiveBlock)blockState.getBlock())
                             .releaseBeesAndResetHoneyLevel(serverLevel, blockState, blockPos, null, BeehiveBlockEntity.BeeReleaseStatus.BEE_RELEASED);
                         this.setSuccess(true);
