--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -584,6 +_,11 @@
                             blockStateBase -> blockStateBase.hasProperty(BeehiveBlock.HONEY_LEVEL) && blockStateBase.getBlock() instanceof BeehiveBlock
                         )
                         && blockState.getValue(BeehiveBlock.HONEY_LEVEL) >= 5) {
+                        // EMC start - BlockHarvestBeehiveEvent
+                        if (!(new com.empireminecraft.empirecraft.event.block.BlockHarvestBeehiveEvent(org.bukkit.craftbukkit.block.CraftBlock.at(serverLevel, blockSource.pos()), org.bukkit.craftbukkit.block.CraftBlock.at(serverLevel, blockPos), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(item.copyWithCount(1))).callEvent())) {
+                            return item;
+                        }
+                        // EMC end - BlockHarvestBeehiveEvent
                         ((BeehiveBlock)blockState.getBlock())
                             .releaseBeesAndResetHoneyLevel(serverLevel, blockState, blockPos, null, BeehiveBlockEntity.BeeReleaseStatus.BEE_RELEASED);
                         this.setSuccess(true);
