--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -523,6 +_,11 @@
                     // Paper end - Call BlockDispenseEvent
                     if (state.is(BlockTags.BEEHIVES, s -> s.hasProperty(BeehiveBlock.HONEY_LEVEL) && s.getBlock() instanceof BeehiveBlock)
                         && state.getValue(BeehiveBlock.HONEY_LEVEL) >= 5) {
+                        // EMC start - BlockHarvestBeehiveEvent
+                        if (!(new com.empireminecraft.empirecraft.event.block.BlockHarvestBeehiveEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, source.pos()), org.bukkit.craftbukkit.block.CraftBlock.at(level, target), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(dispensed.copyWithCount(1))).callEvent())) {
+                            return dispensed;
+                        }
+                        // EMC end - BlockHarvestBeehiveEvent
                         ((BeehiveBlock)state.getBlock())
                             .releaseBeesAndResetHoneyLevel(level, state, target, null, BeehiveBlockEntity.BeeReleaseStatus.BEE_RELEASED);
                         this.setSuccess(true);
