--- a/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java
+++ b/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java
@@ -55,6 +_,11 @@
             } catch (Exception e) {
                 LOGGER.error("Error trying to place shulker box at {}", relativePos, e);
             }
+            // EMC start - ShulkerBoxDispensedEvent
+            if (this.isSuccess()) {
+                (new com.empireminecraft.empirecraft.event.block.ShulkerBoxDispensedEvent(bukkitBlock, craftItem.clone())).callEvent();
+            }
+            // EMC end - ShulkerBoxDispensedEvent
         }
 
         return dispensed;
