--- a/net/minecraft/core/dispenser/SulfurCubeBlockDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/SulfurCubeBlockDispenseItemBehavior.java
@@ -26,7 +_,7 @@
         // Paper end - Call BlockDispenseEvent
 
         for (SulfurCube entity : level.getEntitiesOfClass(SulfurCube.class, new AABB(pos))) {
-            if (entity.equipItem(dispensed)) {
+            if (SulfurCube.isSwallowableItem(dispensed) && entity.equipItem(dispensed)) { // EMC - add isSwallableItem
                 dispensed.shrink(1);
                 return true;
             }
