--- a/src/main/java/org/bukkit/inventory/Inventory.java
+++ b/src/main/java/org/bukkit/inventory/Inventory.java
@@ -25,7 +_,7 @@
  * @see #getContents()
  * @see #getStorageContents()
  */
-public interface Inventory extends Iterable<ItemStack> {
+public interface Inventory extends Iterable<ItemStack>, com.empireminecraft.empirecraft.meta.TemporaryDataHolder { // EMC - Add TemporaryDataHolder
 
     /**
      * Returns the size of the inventory
