--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -583,7 +_,20 @@
             }
         }
 
-        return result.ensureServerConversions(); // Paper
+        // EMC start - Re-set name and lore to make ItemStacks more consistent internally
+        result = result.ensureServerConversions(); // Paper
+        if (result.hasItemMeta()) {
+            ItemMeta itemMeta = result.getItemMeta();
+            if (itemMeta.hasDisplayName()) {
+                itemMeta.displayName(itemMeta.displayName());
+            }
+            if (itemMeta.hasLore()) {
+                itemMeta.lore(itemMeta.lore());
+            }
+            result.setItemMeta(itemMeta);
+        }
+        return result;
+        // EMC end - Re-set name and lore to make ItemStacks more consistent internally
     }
 
     // Paper start
