--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/meta/ItemStackPersistentDataType.java
@@ -1,0 +_,29 @@
+package com.empireminecraft.empirecraft.meta;
+
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.persistence.PersistentDataAdapterContext;
+import org.bukkit.persistence.PersistentDataType;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class ItemStackPersistentDataType implements PersistentDataType<byte[], ItemStack> {
+    @Override
+    public Class<byte[]> getPrimitiveType() {
+        return byte[].class;
+    }
+
+    @Override
+    public Class<ItemStack> getComplexType() {
+        return ItemStack.class;
+    }
+
+    @Override
+    public byte [] toPrimitive(ItemStack complex, PersistentDataAdapterContext context) {
+        return complex.serializeAsBytes();
+    }
+
+    @Override
+    public ItemStack fromPrimitive(byte [] primitive, PersistentDataAdapterContext context) {
+        return ItemStack.deserializeBytes(primitive);
+    }
+}
