--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/meta/MetaKey.java
@@ -1,0 +_,27 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.empirecraft.meta;
+
+import org.bukkit.NamespacedKey;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public interface MetaKey<H, K, C> {
+
+    @NotNull K key();
+
+    @Nullable C get(@NotNull H holder);
+
+    @NotNull C get(@NotNull H holder, @NotNull C def);
+
+    boolean has(@NotNull H holder);
+
+    @Nullable C set(@NotNull H holder, @NotNull C value);
+
+    @Nullable C remove(@NotNull H holder);
+
+}
