--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/meta/Meta.java
@@ -1,0 +_,31 @@
+package com.empireminecraft.empirecraft.meta;
+
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public final class Meta {
+
+    public static <H, K, P, C> C getMeta(H holder, MetaKey<H, K, C> key, C def) {
+        return key.get(holder, def);
+    }
+
+    public static @Nullable <H, K, P, C> C getMeta(H holder, MetaKey<H, K, C> key) {
+        return key.get(holder);
+    }
+
+    public static <H, K, P, C> boolean hasMeta(H holder, MetaKey<H, K, C> key) {
+        return key.has(holder);
+    }
+
+    public static <H, K, P, C> @Nullable C setMeta(H holder, MetaKey<H, K, C> key, C value) {
+        return key.set(holder, value);
+    }
+
+    public static <H, K, P, C> @Nullable C removeMeta(H holder, MetaKey<H, K, C> key) {
+        return key.remove(holder);
+    }
+
+    private Meta() {
+    }
+}
