--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/meta/MetaKey.java
@@ -1,0 +_,28 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.empirecraft.meta;
+
+import org.bukkit.NamespacedKey;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public interface MetaKey<H, K, C> {
+
+    K key();
+
+    @Nullable C get(H holder);
+
+    C get(H holder, C def);
+
+    boolean has(H holder);
+
+    @Nullable C set(H holder, C value);
+
+    @Nullable C remove(H holder);
+
+}
