--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/meta/PersistentMetaKey.java
@@ -1,0 +_,12 @@
+package com.empireminecraft.empirecraft.meta;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.persistence.PersistentDataType;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public interface PersistentMetaKey<H, P, C> extends MetaKey<H, NamespacedKey, C> {
+
+    PersistentDataType<P, C> type();
+
+}
