--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/meta/PersistentListMetaKey.java
@@ -1,0 +_,15 @@
+package com.empireminecraft.empirecraft.meta;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.persistence.ListPersistentDataType;
+import org.bukkit.persistence.PersistentDataHolder;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.List;
+
+@NullMarked
+public interface PersistentListMetaKey<H, P, C> extends MetaKey<H, NamespacedKey, List<C>> {
+
+    ListPersistentDataType<P, C> type();
+
+}
