--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/meta/TemporaryDataHolder.java
@@ -1,0 +_,18 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.empirecraft.meta;
+
+import org.bukkit.NamespacedKey;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Map;
+
+public interface TemporaryDataHolder {
+
+    @NotNull Map<NamespacedKey, Object> getTemporaryMeta();
+
+}
