--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/meta/TemporaryMetaKey.java
@@ -1,0 +_,12 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.empirecraft.meta;
+
+import org.bukkit.NamespacedKey;
+
+public interface TemporaryMetaKey<H, C> extends MetaKey<H, NamespacedKey, C> {
+}
