--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1478,7 +_,7 @@
                 BookMeta meta = editBookEvent.getNewBookMeta();
                 CraftItemStack.setItemMeta(itemInHand, meta);
             } else {
-                player.containerMenu.forceSlot(player.getInventory(), itemInHandIndex); // SPIGOT-7484
+                player.containerMenu.sendAllDataToRemote(); // SPIGOT-7484
             }
         }
 
