--- a/net/minecraft/server/players/OldUsersConverter.java
+++ b/net/minecraft/server/players/OldUsersConverter.java
@@ -390,6 +_,7 @@
     }
 
     public static boolean serverReadyAfterUserconversion(MinecraftServer server) {
+        if (true) return true; // EMC - Disable UUID conversion
         boolean flag = areOldUserlistsRemoved();
         return flag && areOldPlayersConverted(server);
     }
