--- a/net/minecraft/world/item/ProjectileWeaponItem.java
+++ b/net/minecraft/world/item/ProjectileWeaponItem.java
@@ -140,6 +_,11 @@
         int i = !intangable && !shooter.hasInfiniteMaterials() && shooter.level() instanceof ServerLevel serverLevel
             ? EnchantmentHelper.processAmmoUse(serverLevel, weapon, ammo, 1)
             : 0;
+        // EMC start - ProjectileUseAmmoEvent
+        com.empireminecraft.empirecraft.event.entity.ProjectileUseAmmoEvent event = new com.empireminecraft.empirecraft.event.entity.ProjectileUseAmmoEvent(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(weapon), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(ammo), shooter.getBukkitEntity(), i);
+        event.callEvent();
+        i = event.getAmmoCount();
+        // EMC end - ProjectileUseAmmoEvent
         if (i > ammo.getCount()) {
             return ItemStack.EMPTY;
         } else if (i == 0) {
