--- a/net/minecraft/world/item/SnowballItem.java
+++ b/net/minecraft/world/item/SnowballItem.java
@@ -37,6 +_,7 @@
                 }
                 // Paper end - PlayerLaunchProjectileEvent
 
+                if (event.shouldPlayLaunchSound()) { // EMC - Incomplete play launch sound control for projectiles
                 level.playSound(
                     null,
                     player.getX(),
@@ -47,6 +_,7 @@
                     0.5F,
                     0.4F / (level.getRandom().nextFloat() * 0.4F + 0.8F)
                 );
+                } // EMC - Incomplete play launch sound control for projectiles
                 // Paper start - PlayerLaunchProjectileEvent - return fail
             } else {
                 player.containerMenu.sendAllDataToRemote();
