--- a/net/minecraft/world/item/TridentItem.java
+++ b/net/minecraft/world/item/TridentItem.java
@@ -94,6 +_,7 @@
                             return false;
                         }
                         ThrownTrident trident = tridentDelayed.projectile(); // Paper - PlayerLaunchProjectileEvent
+                        (new com.empireminecraft.empirecraft.event.player.PlayerThrowTridentEvent((org.bukkit.entity.Player) player.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemStack), trident.getBukkitEntity())).callEvent(); // EMC - PlayerThrowTridentEven
                         if (event.shouldConsume()) {
                             thrownItemStack.hurtWithoutBreaking(1, player); // Paper - PlayerLaunchProjectileEvent - use thrownItemStack; pickup item damage
                         }
