--- a/net/minecraft/world/item/TridentItem.java
+++ b/net/minecraft/world/item/TridentItem.java
@@ -93,6 +_,7 @@
                                 return false;
                             }
                             ThrownTrident thrownTrident = tridentDelayed.projectile(); // Paper - PlayerLaunchProjectileEvent
+                            (new com.empireminecraft.empirecraft.event.player.PlayerThrowTridentEvent((org.bukkit.entity.Player) player.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack), thrownTrident.getBukkitEntity())).callEvent(); // EMC - PlayerThrowTridentEvent
                             if (event.shouldConsume()) {
                                 itemStack.hurtWithoutBreaking(1, player); // Paper - PlayerLaunchProjectileEvent - use itemStack; pickup item damage
                             }
