--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -373,11 +_,11 @@
 
             );
 
-            if (this.cserver.getServer().isRunning()) {
+            if (true) { // EMC - Always fire PlayerKickEvent even in shutdown
                 this.cserver.getPluginManager().callEvent(event);
             }
 
-            if (event.isCancelled()) {
+            if (!cserver.getServer().hasStopped() && event.isCancelled()) { // EMC - Always fire PlayerKickEvent even in shutdown
                 // Do not kick the player
                 return;
             }
