--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -556,9 +_,9 @@
                 // Paper end - Prevent moving into unloaded chunks
                 if (d7 - d6 > Math.max(100.0, Math.pow((double) (org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed), 2)) && !this.isSingleplayerOwner()) {
                     // CraftBukkit end
-                    LOGGER.warn(
-                        "{} (vehicle of {}) moved too quickly! {},{},{}", rootVehicle.getName().getString(), this.player.getName().getString(), d3, d4, d5
-                    );
+                    //LOGGER.warn( // EMC - Comment out movement logs
+                    //    "{} (vehicle of {}) moved too quickly! {},{},{}", rootVehicle.getName().getString(), this.player.getName().getString(), d3, d4, d5
+                    //);
                     this.send(ClientboundMoveVehiclePacket.fromEntity(rootVehicle));
                     return;
                 }
@@ -586,7 +_,7 @@
                 boolean flag2 = false;
                 if (d7 > org.spigotmc.SpigotConfig.movedWronglyThreshold) { // Spigot
                     flag2 = true; // Paper - diff on change, this should be moved wrongly
-                    LOGGER.warn("{} (vehicle of {}) moved wrongly! {}", rootVehicle.getName().getString(), this.player.getName().getString(), Math.sqrt(d7));
+                    //LOGGER.warn("{} (vehicle of {}) moved wrongly! {}", rootVehicle.getName().getString(), this.player.getName().getString(), Math.sqrt(d7)); // EMC - Comment out movement logs
                 }
 
                 rootVehicle.absMoveTo(d, d1, d2, f, f1);
