--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -361,6 +_,11 @@
         uniqueId = gameprofile.getId();
         // Paper end - Add more fields to AsyncPlayerPreLoginEvent
 
+        // EMC start - Allow changing of player name in Prelogin
+        GameProfile old = gameprofile;
+        gameprofile = new GameProfile(uniqueId, asyncEvent.getName());
+        gameprofile.getProperties().putAll(old.getProperties());
+        // EMC end - Allow changing of player name in Prelogin
         if (PlayerPreLoginEvent.getHandlerList().getRegisteredListeners().length != 0) {
             final PlayerPreLoginEvent event = new PlayerPreLoginEvent(playerName, address, uniqueId);
             if (asyncEvent.getResult() != PlayerPreLoginEvent.Result.ALLOWED) {
