--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -509,6 +_,11 @@
         double d = this.firedFromWeapon != null && this.level() instanceof ServerLevel serverLevel
             ? EnchantmentHelper.modifyKnockback(serverLevel, this.firedFromWeapon, entity, damageSource, 0.0F)
             : 0.0F;
+        // EMC start - ArrowKnockbackEvent
+        com.empireminecraft.empirecraft.event.entity.ArrowKnockbackEvent event = new com.empireminecraft.empirecraft.event.entity.ArrowKnockbackEvent(this.getBukkitEntity(), entity.getBukkitEntity(), d);
+        event.callEvent();
+        d = event.getKnockback();
+        // EMC end - ArrowKnockbackEvent
         if (d > 0.0) {
             double max = Math.max(0.0, 1.0 - entity.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE));
             Vec3 vec3 = this.getDeltaMovement().multiply(1.0, 0.0, 1.0).normalize().scale(d * 0.6 * max);
