--- a/net/minecraft/world/entity/projectile/EyeOfEnder.java
+++ b/net/minecraft/world/entity/projectile/EyeOfEnder.java
@@ -28,6 +_,11 @@
     public Vec3 target;
     public int life;
     public boolean surviveAfterDeath;
+    // EMC start - EyeOfEnder API
+    public int lifeTime = 80;
+    public double speed = 0.0025D;
+    public boolean exactTarget = false;
+    // EMC end - EyeOfEnder API
 
     public EyeOfEnder(EntityType<? extends EyeOfEnder> entityType, Level level) {
         super(entityType, level);
@@ -80,6 +_,7 @@
         // Paper end - Change EnderEye target without changing other things
         Vec3 vec3 = pos.subtract(this.position());
         double d = vec3.horizontalDistance();
+        if (exactTarget) d = 0; // EMC
         if (d > 12.0) {
             this.target = this.position().add(vec3.x / d * 12.0, 8.0, vec3.z / d * 12.0);
         } else {
@@ -108,7 +_,8 @@
         this.setPos(vec3);
         if (!this.level().isClientSide()) {
             this.life++;
-            if (this.life > 80 && !this.level().isClientSide) {
+            if (this.life > this.lifeTime && !this.level().isClientSide) { // EMC - add lifeTime
+                if (!(new com.empireminecraft.empirecraft.event.entity.EnderSignalArriveEvent((org.bukkit.entity.EnderSignal) this.getBukkitEntity()).callEvent())) {return;} // EMC - EnderSignalArriveEvent
                 this.playSound(SoundEvents.ENDER_EYE_DEATH, 1.0F, 1.0F);
                 this.discard(this.surviveAfterDeath ? org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP : org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                 if (this.surviveAfterDeath) {
@@ -139,10 +_,10 @@
         }
     }
 
-    private static Vec3 updateDeltaMovement(Vec3 deltaMovement, Vec3 pos, Vec3 target) {
-        Vec3 vec3 = new Vec3(target.x - pos.x, 0.0, target.z - pos.z);
+    private Vec3 updateDeltaMovement(Vec3 deltaMovement, Vec3 pos, Vec3 target) { // EMC - de-static
+        Vec3 vec3 = new Vec3(target.x - pos.x, exactTarget ? (target.y - pos.y) : 0.0, target.z - pos.z); // EMC - Set Exact Target
         double len = vec3.length();
-        double d = Mth.lerp(0.0025, deltaMovement.horizontalDistance(), len);
+        double d = Mth.lerp(speed, deltaMovement.horizontalDistance(), len); // EMC - Set Speed
         double d1 = deltaMovement.y;
         if (len < 1.0) {
             d *= 0.8;
