--- a/net/minecraft/world/entity/projectile/EyeOfEnder.java
+++ b/net/minecraft/world/entity/projectile/EyeOfEnder.java
@@ -28,6 +_,11 @@
     public @Nullable Vec3 target;
     public int life;
     public boolean surviveAfterDeath;
+    // EMC start - EyeOfEnder API
+    public int lifeTime = 80;
+    public double speed = 0.0025D;
+    public boolean exactTarget = false;
+    // EMC end - EyeOfEnder API
 
     public EyeOfEnder(final EntityType<? extends EyeOfEnder> type, final Level level) {
         super(type, level);
@@ -80,6 +_,7 @@
         // Paper end - Change EyeOfEnder target without changing other things
         Vec3 delta = target.subtract(this.position());
         double horizontalDistance = delta.horizontalDistance();
+        if (exactTarget) horizontalDistance = 0; // EMC
         if (horizontalDistance > 12.0) {
             this.target = this.position().add(delta.x / horizontalDistance * 12.0, 8.0, delta.z / horizontalDistance * 12.0);
         } else {
@@ -108,7 +_,8 @@
         this.setPos(newPosition);
         if (!this.level().isClientSide()) {
             this.life++;
-            if (this.life > 80 && !this.level().isClientSide()) {
+            if (this.life > this.lifeTime && !this.level().isClientSide()) { // EMC - add lifeTime
+                if (!(new com.empireminecraft.empirecraft.event.entity.EnderSignalArriveEvent((org.bukkit.entity.EnderSignal) this.getBukkitEntity()).callEvent())) { return; } // EMC - EnderSignalArriveEvent
                 this.playSound(SoundEvents.ENDER_EYE_DEATH, 1.0F, 1.0F);
                 this.discard(this.surviveAfterDeath ? org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP : org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                 if (this.surviveAfterDeath) {
@@ -139,10 +_,10 @@
         }
     }
 
-    private static Vec3 updateDeltaMovement(final Vec3 oldMovement, final Vec3 position, final Vec3 target) {
-        Vec3 horizontalDelta = new Vec3(target.x - position.x, 0.0, target.z - position.z);
+    private Vec3 updateDeltaMovement(final Vec3 oldMovement, final Vec3 position, final Vec3 target) { // EMC - de-static
+        Vec3 horizontalDelta = new Vec3(target.x - position.x, exactTarget ? (target.y - position.y) : 0.0, target.z - position.z); // EMC - Set Exact Target
         double horizontalLength = horizontalDelta.length();
-        double wantedSpeed = Mth.lerp(0.0025, oldMovement.horizontalDistance(), horizontalLength);
+        double wantedSpeed = Mth.lerp(speed, oldMovement.horizontalDistance(), horizontalLength); // EMC - Set Speed
         double movementY = oldMovement.y;
         if (horizontalLength < 1.0) {
             wantedSpeed *= 0.8;
