--- a/net/minecraft/world/entity/animal/equine/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/equine/AbstractHorse.java
@@ -968,7 +_,7 @@
     }
 
     protected static double generateSpeed(final DoubleSupplier probabilityProvider) {
-        return (0.45F + probabilityProvider.getAsDouble() * 0.3 + probabilityProvider.getAsDouble() * 0.3 + probabilityProvider.getAsDouble() * 0.3) * 0.25;
+        return (0.45F + probabilityProvider.getAsDouble() * 0.315 + probabilityProvider.getAsDouble() * 0.315 + probabilityProvider.getAsDouble() * 0.315) * 0.25; // EMC - increase factor multiplier by 0.015
     }
 
     @Override
