--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -20,7 +_,7 @@
  * then the chunk will only be generated as far as it needs to provide the
  * requested data.
  */
-public interface Chunk extends PersistentDataHolder {
+public interface Chunk extends PersistentDataHolder, com.empireminecraft.empirecraft.meta.TemporaryDataHolder { // EMC - Add TemporaryDataHolder
 
     /**
      * Gets the X-coordinate of this chunk
@@ -165,6 +_,7 @@
     @NotNull
     Collection<BlockState> getTileEntities(java.util.function.@NotNull Predicate<? super Block> blockPredicate, boolean useSnapshot);
     // Paper end
+    @NotNull com.empireminecraft.empirecraft.Identifier getIdentifier(); // EMC - Add Identifier API
 
     /**
      * Checks if the chunk is fully generated.
