--- a/net/minecraft/world/entity/ambient/Bat.java
+++ b/net/minecraft/world/entity/ambient/Bat.java
@@ -157,10 +_,21 @@
 
             if (this.targetPosition == null || this.random.nextInt(30) == 0 || this.targetPosition.closerToCenterThan(this.position(), 2.0)) {
                 this.targetPosition = BlockPos.containing(
-                    this.getX() + this.random.nextInt(7) - this.random.nextInt(7),
-                    this.getY() + this.random.nextInt(6) - 2.0,
-                    this.getZ() + this.random.nextInt(7) - this.random.nextInt(7)
+                        this.getX() + this.random.nextInt(7) - this.random.nextInt(7),
+                        this.getY() + this.random.nextInt(6) - 2.0,
+                        this.getZ() + this.random.nextInt(7) - this.random.nextInt(7)
                 );
+                // EMC start - BatMoveEvent
+                org.bukkit.entity.Bat bukkitBat = (org.bukkit.entity.Bat) this.getBukkitEntity();
+                org.bukkit.Location loc = new org.bukkit.Location(bukkitBat.getWorld(), this.targetPosition.getX(), this.targetPosition.getY(), this.targetPosition.getZ());
+                com.empireminecraft.empirecraft.event.entity.BatMoveEvent event = new com.empireminecraft.empirecraft.event.entity.BatMoveEvent(bukkitBat, loc);
+                if (!event.callEvent()) {
+                    this.targetPosition = new BlockPos(this.getBlockX(), this.getBlockY(), this.getBlockZ());
+                } else {
+                    org.bukkit.Location target = event.getMoveTarget();
+                    this.targetPosition = new BlockPos(target.getBlockX(), target.getBlockY(), target.getBlockZ());
+                }
+                // EMC end - BatMoveEvent
             }
 
             double d = this.targetPosition.getX() + 0.5 - this.getX();
