--- a/net/minecraft/world/entity/animal/wolf/Wolf.java
+++ b/net/minecraft/world/entity/animal/wolf/Wolf.java
@@ -582,6 +_,13 @@
         this.entityData.set(DATA_REMAINING_ANGER_TIME, remainingPersistentAngerTime);
     }
 
+    // EMC start - Always Angry API
+    public boolean alwaysAngry = false;
+    @Override
+    public boolean isAngry() {
+        return alwaysAngry || NeutralMob.super.isAngry();
+    }
+    // EMC end - Always Angry API
     @Override
     public void startPersistentAngerTimer() {
         this.setRemainingPersistentAngerTime(PERSISTENT_ANGER_TIME.sample(this.random));
