--- a/src/main/java/org/bukkit/entity/EnderSignal.java
+++ b/src/main/java/org/bukkit/entity/EnderSignal.java
@@ -91,4 +_,11 @@
      * @param timer how long (in ticks) this EnderSignal has been alive.
      */
     public void setDespawnTimer(int timer);
+    // EMC start - add EnderSignal API
+    void setExactTarget(boolean exact);
+    void setLife(int lifeTime);
+    int getLife();
+    void setSpeed(double speed);
+    double getSpeed();
+    // EMC end - add EnderSignal API
 }
