--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2074,6 +_,38 @@
      */
     public boolean canSee(Player player);
 
+    // EMC start - Fake Player API
+    /**
+     * Checks to see if a fake player has been hidden from this player
+     *
+     * @param fakePlayer FakePlayer to check
+     * @return True if the provided fakePlayer is not being hidden from this
+     *     player
+     */
+    boolean canSee(com.empireminecraft.empirecraft.api.FakePlayer fakePlayer);
+    /**
+     * Hides a fake player from this player
+     *
+     * @param plugin Plugin that wants to hide the fake player
+     * @param fakePlayer FakePlayer to hide
+     */
+    void hideFakePlayer(Plugin plugin, com.empireminecraft.empirecraft.api.FakePlayer fakePlayer);
+    /**
+     * Allows this player to see a fake player that was previously hidden. If
+     * another plugin had hidden the fake player too, then the player will
+     * remain hidden until the other plugin calls this method too.
+     *
+     * @param plugin Plugin that wants to show the fake player
+     * @param fakePlayer FakePlayer to show
+     */
+    void showFakePlayer(Plugin plugin, com.empireminecraft.empirecraft.api.FakePlayer fakePlayer);
+    /**
+     * Gets all fake players hidden with {@link #hideFakePlayer(Plugin, com.empireminecraft.empirecraft.api.FakePlayer)}.
+     *
+     * @return a Set with all hidden fake players
+     */
+    java.util.Set<com.empireminecraft.empirecraft.api.FakePlayer> getHiddenFakePlayers();
+    // EMC end - Fake Player API
     /**
      * Visually hides an entity from this player.
      *
