From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 3 Jun 2015 21:39:34 -0400
Subject: [PATCH] Ignore completion starting with _


diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index 5df19bd701c67506689fc7f49d91f99ebfbc83f0..ba43107777c9558e083d88f7da0a1e0b90d6cc07 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -224,7 +224,7 @@ public class SimpleCommandMap implements CommandMap {
 
                 String name = commandEntry.getKey(); // Use the alias, not command name
 
-                if (StringUtil.startsWithIgnoreCase(name, cmdLine)) {
+                if (name.charAt(0) != '_' && StringUtil.startsWithIgnoreCase(name, cmdLine)) { // EMC
                     completions.add(prefix + name);
                 }
             }
