--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -108,7 +_,7 @@
         try {
             JsonElement jsonElement = StrictJsonParser.parse(json);
             if (jsonElement.isJsonNull()) {
-                LOGGER.error("Unable to parse Stat data from {}", this.file);
+                //LOGGER.error("Unable to parse Stat data from {}", this.file); // EMC - Ignore statistics errors
                 return;
             }
 
@@ -117,11 +_,13 @@
             this.stats
                 .putAll(
                     STATS_CODEC.parse(dynamic.get("stats").orElseEmptyMap())
-                        .resultOrPartial(string -> LOGGER.error("Failed to parse statistics for {}: {}", this.file, string))
+                        .resultOrPartial(string -> {
+                            //LOGGER.error("Failed to parse statistics for {}: {}", this.file, string) // EMC - Ignore statistics errors
+                        })
                         .orElse(Map.of())
                 );
         } catch (JsonParseException var5) {
-            LOGGER.error("Unable to parse Stat data from {}", this.file, var5);
+            //LOGGER.error("Unable to parse Stat data from {}", this.file, var5); // EMC - Ignore statistics errors
         }
     }
 
