--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -113,16 +_,18 @@
                                             Util.ifElse(
                                                 this.getStat(type, string1),
                                                 stat -> this.stats.put(stat, compound1.getInt(string1)),
-                                                () -> LOGGER.warn("Invalid statistic in {}: Don't know what {} is", this.file, string1)
+                                                () -> {} /*LOGGER.warn("Invalid statistic in {}: Don't know what {} is", this.file, string1)*/ // EMC - Ignore statistics warnings
                                             );
                                         } else {
-                                            LOGGER.warn(
-                                                "Invalid statistic value in {}: Don't know what {} is for key {}", this.file, compound1.get(string1), string1
-                                            );
+                                            // EMC start - Ignore statistics warnings
+                                            //LOGGER.warn(
+                                            //    "Invalid statistic value in {}: Don't know what {} is for key {}", this.file, compound1.get(string1), string1
+                                            //);
+                                            // EMC end - Ignore statistics warnings
                                         }
                                     }
                                 },
-                                () -> LOGGER.warn("Invalid statistic type in {}: Don't know what {} is", this.file, string)
+                                () -> {} /*LOGGER.warn("Invalid statistic type in {}: Don't know what {} is", this.file, string)*/ // EMC - Ignore statistics warnings
                             );
                         }
                     }
