--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -138,7 +_,7 @@
             try {
                 ResourceKey<Recipe<?>> resourceKey = ResourceKey.create(Registries.RECIPE, ResourceLocation.parse(string));
                 if (!isRecognized.test(resourceKey)) {
-                    LOGGER.error("Tried to load unrecognized recipe: {} removed now.", resourceKey);
+                    //LOGGER.error("Tried to load unrecognized recipe: {} removed now.", resourceKey); // EMC - Undesired log
                 } else {
                     output.accept(resourceKey);
                 }
