--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -121,7 +_,9 @@
         this.stats
             .putAll(
                 STATS_CODEC.parse(data.get("stats").orElseEmptyMap())
-                    .resultOrPartial(error -> LOGGER.error("Failed to parse statistics for {}: {}", this.file, error))
+                    .resultOrPartial(error -> {
+                        //LOGGER.error("Failed to parse statistics for {}: {}", this.file, error) // EMC - Ignore statistics errors
+                    })
                     .orElse(Map.of())
             );
     }
