--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -121,7 +_,9 @@
         this.stats
             .putAll(
                 STATS_CODEC.parse(dynamic.get("stats").orElseEmptyMap())
-                    .resultOrPartial(string -> LOGGER.error("Failed to parse statistics for {}: {}", this.file, string))
+                    .resultOrPartial(string -> {
+                        //LOGGER.error("Failed to parse statistics for {}: {}", this.file, string) // EMC - Ignore statistics errors
+                    })
                     .orElse(Map.of())
             );
     }
