--- a/net/minecraft/world/entity/monster/Blaze.java
+++ b/net/minecraft/world/entity/monster/Blaze.java
@@ -31,6 +_,10 @@
     private float allowedHeightOffset = 0.5F;
     private int nextHeightOffsetChangeTick;
     private static final EntityDataAccessor<Byte> DATA_FLAGS_ID = SynchedEntityData.defineId(Blaze.class, EntityDataSerializers.BYTE);
+    // EMC start
+    public int numFireballs = 4;
+    public int timeBetweenFireballs = 6;
+    // EMC end
 
     public Blaze(EntityType<? extends Blaze> type, Level level) {
         super(type, level);
@@ -154,7 +_,7 @@
         this.entityData.set(DATA_FLAGS_ID, b);
     }
 
-    static class BlazeAttackGoal extends Goal {
+    public static class BlazeAttackGoal extends Goal { // EMC - make public
         private final Blaze blaze;
         private int attackStep;
         private int attackTime;
@@ -220,8 +_,8 @@
                         if (this.attackStep == 1) {
                             this.attackTime = 60;
                             this.blaze.setCharged(true);
-                        } else if (this.attackStep <= 4) {
-                            this.attackTime = 6;
+                        } else if (this.attackStep <= this.blaze.numFireballs) { // EMC - setFireballs
+                            this.attackTime = this.blaze.timeBetweenFireballs; // EMC - setTimeBetweenFireballs
                         } else {
                             this.attackTime = 100;
                             this.attackStep = 0;
@@ -238,7 +_,9 @@
                                 Vec3 vec3 = new Vec3(this.blaze.getRandom().triangle(d1, 2.297 * d4), d2, this.blaze.getRandom().triangle(d3, 2.297 * d4));
                                 SmallFireball smallFireball = new SmallFireball(this.blaze.level(), this.blaze, vec3.normalize());
                                 smallFireball.setPos(smallFireball.getX(), this.blaze.getY(0.5) + 0.5, smallFireball.getZ());
+                                if (new com.empireminecraft.empirecraft.event.entity.BlazeLaunchFireballEvent((org.bukkit.entity.Blaze) this.blaze.getBukkitEntity(), (org.bukkit.entity.SmallFireball) smallFireball.getBukkitEntity()).callEvent()) { // EMC - BlazeLaunchFireballEvent
                                 this.blaze.level().addFreshEntity(smallFireball);
+                                } else { smallFireball.kill(this.blaze.level().getMinecraftWorld()); } // EMC - BlazeLaunchFireballEvent
                             }
                         }
                     }
