--- a/net/minecraft/world/entity/monster/Ghast.java
+++ b/net/minecraft/world/entity/monster/Ghast.java
@@ -35,6 +_,7 @@
 public class Ghast extends FlyingMob implements Enemy {
     private static final EntityDataAccessor<Boolean> DATA_IS_CHARGING = SynchedEntityData.defineId(Ghast.class, EntityDataSerializers.BOOLEAN);
     private int explosionPower = 1;
+    public int fireballCooldown = 40; // EMC - Ghast API
 
     public Ghast(EntityType<? extends Ghast> entityType, Level level) {
         super(entityType, level);
@@ -286,7 +_,7 @@
                         largeFireball.bukkitYield = largeFireball.explosionPower = this.ghast.getExplosionPower(); // CraftBukkit - set bukkitYield when setting explosionPower
                         largeFireball.setPos(this.ghast.getX() + viewVector.x * 4.0, this.ghast.getY(0.5) + 0.5, largeFireball.getZ() + viewVector.z * 4.0);
                         level.addFreshEntity(largeFireball);
-                        this.chargeTime = -40;
+                        this.chargeTime = -ghast.fireballCooldown; // EMC - Fireball Cooldown
                     }
                 } else if (this.chargeTime > 0) {
                     this.chargeTime--;
