--- a/net/minecraft/world/entity/monster/Spider.java
+++ b/net/minecraft/world/entity/monster/Spider.java
@@ -177,6 +_,7 @@
         return entity.getBbWidth() <= this.getBbWidth() ? new Vec3(0.0, 0.3125 * this.getScale(), 0.0) : super.getVehicleAttachmentPoint(entity);
     }
 
+    public boolean attackDuringDay = false; // EMC - Attack During Day API
     static class SpiderAttackGoal extends MeleeAttackGoal {
         public SpiderAttackGoal(Spider spider) {
             super(spider, 1.0, true);
@@ -190,6 +_,7 @@
         @Override
         public boolean canContinueToUse() {
             float lightLevelDependentMagicValue = this.mob.getLightLevelDependentMagicValue();
+            if (((Spider) this.mob).attackDuringDay) { lightLevelDependentMagicValue = 0.0F; } // EMC - Attack During Day API
             if (lightLevelDependentMagicValue >= 0.5F && this.mob.getRandom().nextInt(100) == 0) {
                 this.mob.setTarget(null);
                 return false;
@@ -225,6 +_,7 @@
         @Override
         public boolean canUse() {
             float lightLevelDependentMagicValue = this.mob.getLightLevelDependentMagicValue();
+            if (((Spider) this.mob).attackDuringDay) { lightLevelDependentMagicValue = 0.0F; } // EMC - Attack During Day API
             return !(lightLevelDependentMagicValue >= 0.5F) && super.canUse();
         }
     }
