--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
@@ -13,4 +_,11 @@
     public net.minecraft.world.entity.monster.spider.Spider getHandle() {
         return (net.minecraft.world.entity.monster.spider.Spider) this.entity;
     }
+
+    // EMC start - Spider API
+    @Override
+    public void makeAggressiveDuringDay(boolean attack) {
+        getHandle().attackDuringDay = attack;
+    }
+    // EMC end - Spider API
 }
