--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftGhast.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftGhast.java
@@ -41,4 +_,15 @@
         this.getHandle().setExplosionPower(explosionPower);
     }
     // Paper end
+    // EMC start - Ghast API
+    @Override
+    public int getFireballCooldown() {
+        return getHandle().fireballCooldown;
+    }
+
+    @Override
+    public void setFireballCooldown(int cooldown) {
+        getHandle().fireballCooldown = cooldown;
+    }
+    // EMC end - Ghast API
 }
