--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftParrot.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftParrot.java
@@ -37,4 +_,10 @@
     public boolean isDancing() {
         return this.getHandle().isPartyParrot();
     }
+    // EMC start - Parrot API
+    @Override
+    public void setCanSitOnShoulders(boolean canSitOnShoulders) {
+        this.getHandle().setCanSitOnShoulders(canSitOnShoulders);
+    }
+    // EMC end - Parrot API
 }
