--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftWolf.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftWolf.java
@@ -66,6 +_,16 @@
     public void setInterested(boolean interested) {
         this.getHandle().setIsInterested(interested);
     }
+        // EMC start - Wolf API
+        @Override
+        public void setAlwaysAngry(boolean alwaysAngry) {
+            net.minecraft.world.entity.animal.wolf.Wolf entity = getHandle();
+            entity.alwaysAngry = alwaysAngry;
+            if (alwaysAngry) {
+                entity.startPersistentAngerTimer();
+            }
+        }
+    // EMC end - Wolf API
 
     @Override
     public Variant getVariant() {
