--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -103,6 +_,7 @@
         return this.getClass().getSimpleName() + "{name=" + this.getName() + ", uuid=" + this.getUniqueId() + '}';
     }
 
+    @Override public float getAttackPct() { return getHandle().lastAttackPct; } // EMC - Human#getAttackPct API
     @Override
     public PlayerInventory getInventory() {
         return this.inventory;
