--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -92,6 +_,7 @@
         this.enderChest = new CraftInventory(entity.getEnderChestInventory());
     }
 
+    @Override public float getAttackPct() { return getHandle().lastAttackPct; } // EMC - Human#getAttackPct API
     @Override
     public PlayerInventory getInventory() {
         return this.inventory;
