--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
@@ -18,4 +_,10 @@
     public String toString() {
         return "CraftSpider";
     }
+    // EMC start - Spider API
+    @Override
+    public void makeAggressiveDuringDay(boolean attack) {
+        getHandle().attackDuringDay = attack;
+    }
+    // EMC end - Spider API
 }
