--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftBlaze.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftBlaze.java
@@ -12,4 +_,25 @@
     public net.minecraft.world.entity.monster.Blaze getHandle() {
         return (net.minecraft.world.entity.monster.Blaze) this.entity;
     }
+    // EMC start - Blaze API
+    @Override
+    public void setFireballs(int num) {
+        getHandle().numFireballs = num;
+    }
+
+    @Override
+    public int getNumFireballs() {
+        return getHandle().numFireballs;
+    }
+
+    @Override
+    public int getTimeBetweenFireballs() {
+        return getHandle().timeBetweenFireballs;
+    }
+
+    @Override
+    public void setTimeBetweenFireballs(int timeBetweenFireballs) {
+        getHandle().timeBetweenFireballs = timeBetweenFireballs;
+    }
+    // EMC end - Blaze API
 }
