--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderSignal.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderSignal.java
@@ -72,4 +_,30 @@
     public void setDespawnTimer(int time) {
         this.getHandle().life = time;
     }
+    // EMC start - EnderSignal API
+    @Override
+    public void setExactTarget(boolean exact) {
+        getHandle().exactTarget = exact;
+    }
+
+    @Override
+    public void setSpeed(double speed) {
+        getHandle().speed = speed / 10000D;
+    }
+
+    @Override
+    public double getSpeed() {
+        return getHandle().speed * 10000D;
+    }
+
+    @Override
+    public void setLife(int lifeTime) {
+        getHandle().lifeTime = lifeTime;
+    }
+
+    @Override
+    public int getLife() {
+        return getHandle().lifeTime;
+    }
+    // EMC end - EnderSignal API
 }
