From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 29 Dec 2024 02:30:48 -0600
Subject: [PATCH] Fix spawn egg deserialization


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
index fe3f45fcdd8981102b6e85eb1895139289520bd5..fe350b86c39e94f8491598d39163f7945948e73b 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
@@ -2,6 +2,8 @@ package org.bukkit.craftbukkit.inventory;
 
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableMap.Builder;
+
+import java.util.HashMap;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
@@ -24,6 +26,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
     static final ItemMetaKeyType<CustomData> ENTITY_TAG = new ItemMetaKeyType<>(DataComponents.ENTITY_DATA, "entity-tag");
     @ItemMetaKey.Specific(ItemMetaKey.Specific.To.NBT)
     static final ItemMetaKey ENTITY_ID = new ItemMetaKey("id");
+    private EntityType legacyType;
 
     private CompoundTag entityTag;
 
@@ -47,6 +50,11 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
 
     CraftMetaSpawnEgg(Map<String, Object> map) {
         super(map);
+
+        EntityType entityType = ENTITY_ID_TO_ENTITY_TYPE.get((String) map.get("id"));
+        if (entityType != null) {
+            legacyType = entityType;
+        }
     }
 
     @Override
@@ -83,7 +91,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
 
     @Override
     public EntityType getSpawnedType() {
-        throw new UnsupportedOperationException("Must check item type to get spawned type");
+        return legacyType; // EMC
     }
 
     @Override
@@ -162,7 +170,58 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
 
         return builder;
     }
-
+    private static final Map<String, EntityType> ENTITY_ID_TO_ENTITY_TYPE = new HashMap<>();
+    static {
+        ENTITY_ID_TO_ENTITY_TYPE.put("bat", EntityType.BAT);
+        ENTITY_ID_TO_ENTITY_TYPE.put("blaze", EntityType.BLAZE);
+        ENTITY_ID_TO_ENTITY_TYPE.put("cave_spider", EntityType.CAVE_SPIDER);
+        ENTITY_ID_TO_ENTITY_TYPE.put("chicken", EntityType.CHICKEN);
+        ENTITY_ID_TO_ENTITY_TYPE.put("cow", EntityType.COW);
+        ENTITY_ID_TO_ENTITY_TYPE.put("creeper", EntityType.CREEPER);
+        ENTITY_ID_TO_ENTITY_TYPE.put("donkey", EntityType.DONKEY);
+        ENTITY_ID_TO_ENTITY_TYPE.put("elder_guardian", EntityType.ELDER_GUARDIAN);
+        ENTITY_ID_TO_ENTITY_TYPE.put("enderman", EntityType.ENDERMAN);
+        ENTITY_ID_TO_ENTITY_TYPE.put("endermite", EntityType.ENDERMITE);
+        ENTITY_ID_TO_ENTITY_TYPE.put("evocation_illager", EntityType.EVOKER);
+        ENTITY_ID_TO_ENTITY_TYPE.put("ghast", EntityType.GHAST);
+        ENTITY_ID_TO_ENTITY_TYPE.put("guardian", EntityType.GUARDIAN);
+        ENTITY_ID_TO_ENTITY_TYPE.put("ender_dragon", EntityType.ENDER_DRAGON);
+        ENTITY_ID_TO_ENTITY_TYPE.put("horse", EntityType.HORSE);
+        ENTITY_ID_TO_ENTITY_TYPE.put("husk", EntityType.HUSK);
+        ENTITY_ID_TO_ENTITY_TYPE.put("iron_golem", EntityType.IRON_GOLEM);
+        ENTITY_ID_TO_ENTITY_TYPE.put("llama", EntityType.LLAMA);
+        ENTITY_ID_TO_ENTITY_TYPE.put("magma_cube", EntityType.MAGMA_CUBE);
+        ENTITY_ID_TO_ENTITY_TYPE.put("mooshroom", EntityType.MOOSHROOM);
+        ENTITY_ID_TO_ENTITY_TYPE.put("mule", EntityType.MULE);
+        ENTITY_ID_TO_ENTITY_TYPE.put("ocelot", EntityType.OCELOT);
+        ENTITY_ID_TO_ENTITY_TYPE.put("pufferfish", EntityType.PUFFERFISH);
+        ENTITY_ID_TO_ENTITY_TYPE.put("parrot", EntityType.PARROT);
+        ENTITY_ID_TO_ENTITY_TYPE.put("pig", EntityType.PIG);
+        ENTITY_ID_TO_ENTITY_TYPE.put("polar_bear", EntityType.POLAR_BEAR);
+        ENTITY_ID_TO_ENTITY_TYPE.put("rabbit", EntityType.RABBIT);
+        ENTITY_ID_TO_ENTITY_TYPE.put("sheep", EntityType.SHEEP);
+        ENTITY_ID_TO_ENTITY_TYPE.put("shulker", EntityType.SHULKER);
+        ENTITY_ID_TO_ENTITY_TYPE.put("silverfish", EntityType.SILVERFISH);
+        ENTITY_ID_TO_ENTITY_TYPE.put("skeleton", EntityType.SKELETON);
+        ENTITY_ID_TO_ENTITY_TYPE.put("skeleton_horse", EntityType.SKELETON_HORSE);
+        ENTITY_ID_TO_ENTITY_TYPE.put("slime", EntityType.SLIME);
+        ENTITY_ID_TO_ENTITY_TYPE.put("snow_golem", EntityType.SNOW_GOLEM);
+        ENTITY_ID_TO_ENTITY_TYPE.put("spider", EntityType.SPIDER);
+        ENTITY_ID_TO_ENTITY_TYPE.put("squid", EntityType.SQUID);
+        ENTITY_ID_TO_ENTITY_TYPE.put("stray", EntityType.STRAY);
+        ENTITY_ID_TO_ENTITY_TYPE.put("turtle", EntityType.TURTLE);
+        ENTITY_ID_TO_ENTITY_TYPE.put("vex", EntityType.VEX);
+        ENTITY_ID_TO_ENTITY_TYPE.put("villager", EntityType.VILLAGER);
+        ENTITY_ID_TO_ENTITY_TYPE.put("vindication_illager", EntityType.VINDICATOR);
+        ENTITY_ID_TO_ENTITY_TYPE.put("witch", EntityType.WITCH);
+        ENTITY_ID_TO_ENTITY_TYPE.put("wither", EntityType.WITHER);
+        ENTITY_ID_TO_ENTITY_TYPE.put("wither_skeleton", EntityType.WITHER_SKELETON);
+        ENTITY_ID_TO_ENTITY_TYPE.put("wolf", EntityType.WOLF);
+        ENTITY_ID_TO_ENTITY_TYPE.put("zombie", EntityType.ZOMBIE);
+        ENTITY_ID_TO_ENTITY_TYPE.put("zombie_horse", EntityType.ZOMBIE_HORSE);
+        ENTITY_ID_TO_ENTITY_TYPE.put("zombie_pigman", EntityType.ZOMBIFIED_PIGLIN);
+        ENTITY_ID_TO_ENTITY_TYPE.put("zombie_villager", EntityType.ZOMBIE_VILLAGER);
+    }
     @Override
     public CraftMetaSpawnEgg clone() {
         CraftMetaSpawnEgg clone = (CraftMetaSpawnEgg) super.clone();
