--- a/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/net/minecraft/world/entity/animal/goat/Goat.java
@@ -347,6 +_,11 @@
             } else {
                 entityDataAccessor = this.random.nextBoolean() ? DATA_HAS_LEFT_HORN : DATA_HAS_RIGHT_HORN;
             }
+            // EMC start - GoatBreakHornEvent - could make this more configurable: which horn, etc
+            if (!(new com.empireminecraft.empirecraft.event.entity.GoatBreakHornEvent((org.bukkit.entity.Goat) this.getBukkitEntity()).callEvent())) {
+                return false;
+            }
+            // EMC end
 
             this.entityData.set(entityDataAccessor, false);
             Vec3 vec3 = this.position();
