--- a/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -965,7 +_,7 @@
     }
 
     protected static double generateSpeed(DoubleSupplier supplier) {
-        return (0.45F + supplier.getAsDouble() * 0.3 + supplier.getAsDouble() * 0.3 + supplier.getAsDouble() * 0.3) * 0.25;
+        return (0.45F + supplier.getAsDouble() * 0.315 + supplier.getAsDouble() * 0.315 + supplier.getAsDouble() * 0.315) * 0.25; // EMC - increase factor multiplier by 0.015
     }
 
     @Override
