From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 11 Jun 2022 20:08:46 -0500
Subject: [PATCH] Add a temporary api hack to obtain the instrument type of a
 goat horn


diff --git a/src/main/java/com/empireminecraft/api/EAPI_Misc.java b/src/main/java/com/empireminecraft/api/EAPI_Misc.java
index 663e376d633a18acf6782d9c9df955df495b8bd6..b6c39208ba0ed8d1c7679e386a1511d20656d0b8 100644
--- a/src/main/java/com/empireminecraft/api/EAPI_Misc.java
+++ b/src/main/java/com/empireminecraft/api/EAPI_Misc.java
@@ -56,4 +56,7 @@ public interface EAPI_Misc {
      */
     @Nullable
     List<ItemStack> getShulkerInventory(@NotNull ItemStack shulker);
+
+    @Deprecated
+    @Nullable String getInstrumentId(@NotNull ItemStack itemStack);
 }
