--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -374,7 +_,7 @@
             // Paper start - Add DragonEggFormEvent
             BlockPos eggPosition = this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.getLocation(this.origin));
             org.bukkit.craftbukkit.block.CraftBlockState eggState = org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(this.level, eggPosition);
-            eggState.setData(Blocks.DRAGON_EGG.defaultBlockState());
+            eggState.setData(Blocks.DRAGON_HEAD.defaultBlockState()); // EMC - Give a Dragon Head instead of egg Every Dragon Kill
             io.papermc.paper.event.block.DragonEggFormEvent eggEvent = new io.papermc.paper.event.block.DragonEggFormEvent(org.bukkit.craftbukkit.block.CraftBlock.at(this.level, eggPosition), eggState,
                 new org.bukkit.craftbukkit.boss.CraftDragonBattle(this));
             // Paper end - Add DragonEggFormEvent
