--- a/src/main/java/org/bukkit/craftbukkit/legacy/MaterialRerouting.java
+++ b/src/main/java/org/bukkit/craftbukkit/legacy/MaterialRerouting.java
@@ -276,6 +_,10 @@
         return MaterialRerouting.transformFromItemType(furnaceExtractEvent.getItemType(), version);
     }
 
+    public static Material getResultType(FurnaceExtractEvent furnaceExtractEvent, @InjectPluginVersion ApiVersion version) {
+        return MaterialRerouting.transformFromItemType(furnaceExtractEvent.getResultType(), version);
+    }
+
     public static Material getBucket(PlayerBucketEvent playerBucketEvent, @InjectPluginVersion ApiVersion version) {
         return MaterialRerouting.transformFromItemType(playerBucketEvent.getBucket(), version);
     }
