--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/CheckSpawnPatternBlockEvent.java
@@ -1,0 +_,50 @@
+package com.empireminecraft.empirecraft.event;
+
+import org.bukkit.block.Block;
+import org.bukkit.entity.EntityType;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.List;
+
+@NullMarked
+public class CheckSpawnPatternBlockEvent extends Event implements Cancellable {
+    private final List<Block> blocks;
+    private final EntityType entityType;
+    private boolean cancelled;
+
+    public CheckSpawnPatternBlockEvent(List<Block> blocks, EntityType entityType) {
+        this.blocks = blocks;
+        this.entityType = entityType;
+    }
+
+    public List<Block> getBlocks() {
+        return blocks;
+    }
+
+    public EntityType getEntityType() {
+        return entityType;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    @Override
+    public void setCancelled(boolean cancelled) {
+        this.cancelled = cancelled;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
