--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/ServerShutdownEvent.java
@@ -1,0 +_,35 @@
+package com.empireminecraft.empirecraft.event;
+
+import org.bukkit.Bukkit;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class ServerShutdownEvent extends Event {
+    private String reason;
+    public ServerShutdownEvent(@NotNull String reason) {
+        super(!Bukkit.isPrimaryThread());
+        this.reason = reason;
+    }
+
+    @NotNull
+    public String getReason() {
+        return reason;
+    }
+
+    public void setReason(@NotNull String reason) {
+        this.reason = reason;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
