--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/ServerShutdownEvent.java
@@ -1,0 +_,33 @@
+package com.empireminecraft.empirecraft.event;
+
+import org.bukkit.Bukkit;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class ServerShutdownEvent extends Event {
+    private String reason;
+    public ServerShutdownEvent(String reason) {
+        super(!Bukkit.isPrimaryThread());
+        this.reason = reason;
+    }
+
+    public String getReason() {
+        return reason;
+    }
+
+    public void setReason(String reason) {
+        this.reason = reason;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
