--- a/net/minecraft/server/commands/GiveCommand.java
+++ b/net/minecraft/server/commands/GiveCommand.java
@@ -62,6 +_,7 @@
                 remaining -= size;
                 ItemStack copyToDrop = prototypeItemStack.copyWithCount(size);
                 boolean added = player.getInventory().add(copyToDrop);
+                if (true) { continue; } // EMC - never drop items
                 if (added && copyToDrop.isEmpty()) {
                     ItemEntity drop = player.drop(prototypeItemStack.copy(), false, false, false, null); // Paper - do not fire PlayerDropItemEvent for /give command
                     if (drop != null) {
