--- a/net/minecraft/world/level/storage/loot/entries/LootPoolSingletonContainer.java
+++ b/net/minecraft/world/level/storage/loot/entries/LootPoolSingletonContainer.java
@@ -18,6 +_,7 @@
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
 
 public abstract class LootPoolSingletonContainer extends LootPoolEntryContainer {
+    public static final boolean USE_NEW_LUCK = Boolean.getBoolean("use-new-luck"); // EMC - Use System Property for new luck
     public static final int DEFAULT_WEIGHT = 1;
     public static final int DEFAULT_QUALITY = 0;
     protected final int weight;
@@ -138,7 +_,7 @@
             // This is vanilla
             float qualityModifer = (float) LootPoolSingletonContainer.this.quality * luck;
             double baseWeight = (LootPoolSingletonContainer.this.weight + qualityModifer);
-            if (io.papermc.paper.configuration.GlobalConfiguration.get().misc.useAlternativeLuckFormula) {
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().misc.useAlternativeLuckFormula || USE_NEW_LUCK) { // EMC - Use System Property for new luck
                 // Random boost to avoid losing precision in the final int cast on return
                 final int weightBoost = 100;
                 baseWeight *= weightBoost;
