--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1076,7 +_,7 @@
         if (!keepInventory) {
             for (ItemStack item : this.getInventory().getContents()) {
                 if (!item.isEmpty() && !EnchantmentHelper.has(item, net.minecraft.world.item.enchantment.EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP)) {
-                    loot.add(new DefaultDrop(item, stack -> this.drop(stack, true, false, false))); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event)
+                    loot.add(new DefaultDrop(item, stack -> this.drop(stack, true, false, true))); // Paper - Restore vanilla drops behavior; drop function taken from Inventory#dropAll (don't fire drop event) // EMC - Do fire drop event for player deaths
                 }
             }
         }
@@ -2612,6 +_,7 @@
                 this.level().getCraftServer().getPluginManager().callEvent(event);
 
                 if (event.isCancelled()) {
+                    if (true) { return null; } // EMC - Do not return item to player on cancel drop item event
                     org.bukkit.inventory.ItemStack cur = player.getInventory().getItemInHand();
                     if (traceItem && (cur == null || cur.getAmount() == 0)) {
                         // The complete stack was dropped
