--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -2605,6 +_,7 @@
                 this.level().getCraftServer().getPluginManager().callEvent(event);
 
                 if (event.isCancelled()) {
+                    if (true) { return null; } // EMC - Do not return item to player on cancel drop item event
                     org.bukkit.inventory.ItemStack cur = player.getInventory().getItemInHand();
                     if (traceItem && (cur == null || cur.getAmount() == 0)) {
                         // The complete stack was dropped
