From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 4 Mar 2025 21:03:39 -0600
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 652ff54e7c50412503725d628bfe72ed03059790..37df401caee7cf666ec4f3ecba78bbaccaa7c5b1 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -19,6 +19,12 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
 
+    // EMC start
+    /**
+     * The brand id for EmpireCraft.
+     */
+    Key BRAND_EMPIRECRAFT_ID = Key.key("empireminecraft", "empirecraft");
+    // EMC end
     /**
      * Gets the {@code ServerBuildInfo}.
      *
