From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 4 Mar 2025 21:03:39 -0600
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 4e099b584c3435ba1ebce492011d35ae1ac0cdf8..7967a207c04f40f5cb6b231a851522953028f228 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -17,6 +17,12 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
 
+    // EMC start
+    /**
+     * The brand id for EmpireCraft.
+     */
+    Key BRAND_EMPIRECRAFT_ID = Key.key("empireminecraft", "empirecraft");
+    // EMC end
     /**
      * Gets the {@code ServerBuildInfo}.
      *
