From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 29 Dec 2020 01:31:22 -0600
Subject: [PATCH] Add FakePlayer api for faking tablist users


diff --git a/src/main/java/com/empireminecraft/empirecraft/api/EAPI_Server.java b/src/main/java/com/empireminecraft/empirecraft/api/EAPI_Server.java
index 6ee46a5abf4b76725ee4d2bc651b5796d9277caa..681ec68e050f4fa279061550949a2b1a37261c95 100644
--- a/src/main/java/com/empireminecraft/empirecraft/api/EAPI_Server.java
+++ b/src/main/java/com/empireminecraft/empirecraft/api/EAPI_Server.java
@@ -23,6 +23,21 @@
 
 package com.empireminecraft.empirecraft.api;
 
+import com.destroystokyo.paper.profile.PlayerProfile;
+import net.kyori.adventure.text.Component;
+import org.bukkit.GameMode;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+import java.util.Set;
+import java.util.UUID;
+
+@NullMarked
 public interface EAPI_Server {
 
+    @Nullable FakePlayer getFakePlayer(UUID uuid);
+
+    FakePlayer createFakePlayer(PlayerProfile profile, int ping, GameMode gameMode, Component playerListName) throws IllegalArgumentException;
+
+    Set<FakePlayer> getJoinedFakePlayers();
 }
\ No newline at end of file
diff --git a/src/main/java/com/empireminecraft/empirecraft/api/FakePlayer.java b/src/main/java/com/empireminecraft/empirecraft/api/FakePlayer.java
new file mode 100644
index 0000000000000000000000000000000000000000..3e9ba7879e7fe835690352a30f8fef6e6c3a3210
--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/api/FakePlayer.java
@@ -0,0 +1,44 @@
+package com.empireminecraft.empirecraft.api;
+
+import com.destroystokyo.paper.profile.PlayerProfile;
+import net.kyori.adventure.text.Component;
+import org.bukkit.GameMode;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.UUID;
+
+@NullMarked
+public interface FakePlayer {
+
+    PlayerProfile getProfile();
+
+    Component getPlayerListName();
+
+    int getPing();
+
+    GameMode getGameMode();
+
+    UUID getUniqueId();
+
+    void join();
+    void quit();
+
+    default void setPing(int ping) {
+        setPing(ping, true);
+    }
+    void setPing(int ping, boolean update);
+
+    default void setGameMode(GameMode gameMode) {
+        setGameMode(gameMode, true);
+    }
+    void setGameMode(GameMode gameMode, boolean update);
+
+    default void setPlayerListName(Component playerListName) {
+        setPlayerListName(playerListName, true);
+    }
+    void setPlayerListName(Component playerListName, boolean update);
+
+    void update();
+
+    boolean isJoined();
+}
diff --git a/src/main/java/com/empireminecraft/empirecraft/event/FakePlayerEvent.java b/src/main/java/com/empireminecraft/empirecraft/event/FakePlayerEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..281b08b1d6cbd3c5a9537ef33f7ce8c3ce33e2d5
--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/FakePlayerEvent.java
@@ -0,0 +1,26 @@
+package com.empireminecraft.empirecraft.event;
+
+import com.empireminecraft.empirecraft.api.FakePlayer;
+import org.bukkit.event.Event;
+import org.jspecify.annotations.NullMarked;
+
+/**
+ * Represents a fakeplayer related event
+ */
+@NullMarked
+public abstract class FakePlayerEvent extends Event {
+    protected FakePlayer fakePlayer;
+
+    public FakePlayerEvent(FakePlayer who) {
+        fakePlayer = who;
+    }
+
+    /**
+     * Returns the fakeplayer involved in this event
+     *
+     * @return Player who is involved in this event
+     */
+    public final FakePlayer getFakePlayer() {
+        return fakePlayer;
+    }
+}
diff --git a/src/main/java/com/empireminecraft/empirecraft/event/FakePlayerJoinEvent.java b/src/main/java/com/empireminecraft/empirecraft/event/FakePlayerJoinEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..39b95f5e21bcb30025e41c7e0852c25100300140
--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/FakePlayerJoinEvent.java
@@ -0,0 +1,22 @@
+package com.empireminecraft.empirecraft.event;
+
+import com.empireminecraft.empirecraft.api.FakePlayer;
+import org.bukkit.event.HandlerList;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class FakePlayerJoinEvent extends FakePlayerEvent {
+    public FakePlayerJoinEvent( FakePlayer who) {
+        super(who);
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/com/empireminecraft/empirecraft/event/FakePlayerQuitEvent.java b/src/main/java/com/empireminecraft/empirecraft/event/FakePlayerQuitEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..0ec4291e1ccaed66857f65aaf20ef8a547a40d45
--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/FakePlayerQuitEvent.java
@@ -0,0 +1,22 @@
+package com.empireminecraft.empirecraft.event;
+
+import com.empireminecraft.empirecraft.api.FakePlayer;
+import org.bukkit.event.HandlerList;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class FakePlayerQuitEvent extends FakePlayerEvent {
+    public FakePlayerQuitEvent(FakePlayer who) {
+        super(who);
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
