From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 30 Dec 2025 19:11:49 -0600
Subject: [PATCH] Nautilus should extend vehicle


diff --git a/src/main/java/org/bukkit/entity/AbstractNautilus.java b/src/main/java/org/bukkit/entity/AbstractNautilus.java
index 266e91b755213da39bb78322dfdd4570007c7dde..89587163c308f12bc399fc62caf067de40f4c5f1 100644
--- a/src/main/java/org/bukkit/entity/AbstractNautilus.java
+++ b/src/main/java/org/bukkit/entity/AbstractNautilus.java
@@ -5,7 +5,7 @@ import org.bukkit.inventory.InventoryHolder;
 import org.jspecify.annotations.NullMarked;
 
 @NullMarked
-public interface AbstractNautilus extends Animals, InventoryHolder, Tameable {
+public interface AbstractNautilus extends Animals, InventoryHolder, Tameable, Vehicle {
 
     @Override
     ArmoredSaddledMountInventory getInventory();
