From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 29 Nov 2025 20:09:48 -0600
Subject: [PATCH] Add SpawnPatternBlocks events

Naming convention isn't great - but accomplishes what I want

diff --git a/src/main/java/com/empireminecraft/empirecraft/event/CheckSpawnPatternBlockEvent.java b/src/main/java/com/empireminecraft/empirecraft/event/CheckSpawnPatternBlockEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..3b47b7fedc03d5e445ead38e8ae3ec230575d110
--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/CheckSpawnPatternBlockEvent.java
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2025. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.empirecraft.event;
+
+import org.bukkit.block.Block;
+import org.bukkit.entity.EntityType;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.List;
+
+@NullMarked
+public class CheckSpawnPatternBlockEvent extends Event implements Cancellable {
+    private final List<Block> blocks;
+    private final EntityType entityType;
+    private boolean cancelled;
+
+    public CheckSpawnPatternBlockEvent(List<Block> blocks, EntityType entityType) {
+        this.blocks = blocks;
+        this.entityType = entityType;
+    }
+
+    public List<Block> getBlocks() {
+        return blocks;
+    }
+
+    public EntityType getEntityType() {
+        return entityType;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    @Override
+    public void setCancelled(boolean cancelled) {
+        this.cancelled = cancelled;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/com/empireminecraft/empirecraft/event/ClearSpawnPatternBlocksEvent.java b/src/main/java/com/empireminecraft/empirecraft/event/ClearSpawnPatternBlocksEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..c12b7e0b6bc8ff646d97dcfef1c897c89e9b20f6
--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/ClearSpawnPatternBlocksEvent.java
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2025. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.empirecraft.event;
+
+import org.bukkit.block.Block;
+import org.bukkit.entity.EntityType;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+import java.util.List;
+
+public class ClearSpawnPatternBlocksEvent extends Event {
+    private final List<Block> blocks;
+    private final EntityType entityType;
+    private boolean cancelled;
+
+    public ClearSpawnPatternBlocksEvent(List<Block> blocks, EntityType entityType) {
+        this.blocks = blocks;
+        this.entityType = entityType;
+    }
+
+    public List<Block> getBlocks() {
+        return blocks;
+    }
+
+    public EntityType getEntityType() {
+        return entityType;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
