--- a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
@@ -24,6 +_,12 @@
     @Override
     public boolean execute(@NotNull CommandSender sender, @NotNull String currentAlias, @NotNull String[] args) { // Paper
         if (!testPermission(sender)) return true;
+        // EMC start - Disable the reload command
+        if (true) {
+            Command.broadcastCommandMessage(sender, net.kyori.adventure.text.Component.text("Reloading is disabled.", net.kyori.adventure.text.format.NamedTextColor.RED));
+            return true;
+        }
+        // EMC end - Disable the reload command
 
         // Paper start - Reload permissions.yml & require confirm
         boolean confirmed = System.getProperty("LetMeReload") != null;
