--- a/net/minecraft/world/entity/animal/ShoulderRidingEntity.java
+++ b/net/minecraft/world/entity/animal/ShoulderRidingEntity.java
@@ -39,6 +_,12 @@
     }
 
     public boolean canSitOnShoulder() {
-        return this.rideCooldownCounter > 100;
-    }
+        // EMC start - Parrot API
+        return this.rideCooldownCounter > 100 && canSitOnShoulders; // EMC - add && canSitOnShoulders
+    }
+    private boolean canSitOnShoulders = true;
+    public void setCanSitOnShoulders(boolean canSitOnShoulders) {
+        this.canSitOnShoulders = canSitOnShoulders;
+    }
+    // EMC end - Parrot API
 }
