--- a/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/SnowGolem.java
@@ -122,8 +_,25 @@
         double d3 = Math.sqrt(d * d + d2 * d2) * 0.2F;
         if (this.level() instanceof ServerLevel serverLevel) {
             ItemStack itemStack = new ItemStack(Items.SNOWBALL);
+            // EMC start - SnowmanThrowSnowballEvent
+            Snowball entitysnowball = new Snowball(serverLevel, this, itemStack);
+            org.bukkit.entity.Snowman bukkitSnowman = (org.bukkit.entity.Snowman) this.getBukkitEntity();
+            org.bukkit.entity.Snowball bukkitSnowball = (org.bukkit.entity.Snowball) entitysnowball.getBukkitEntity();
+            entitysnowball.projectileSource = bukkitSnowman;
+            if (!(new org.bukkit.event.entity.ProjectileLaunchEvent(bukkitSnowball).callEvent())) {
+                return;
+            }
+            com.empireminecraft.empirecraft.event.entity.SnowmanThrowSnowballEvent event = new com.empireminecraft.empirecraft.event.entity.SnowmanThrowSnowballEvent(
+                    bukkitSnowman,
+                    bukkitSnowball,
+                    target.getBukkitLivingEntity()
+            );
+            if (!event.callEvent()) {
+                return;
+            }
             Projectile.spawnProjectile(
-                new Snowball(serverLevel, this, itemStack), serverLevel, itemStack, snowball -> snowball.shoot(d, d1 + d3 - snowball.getY(), d2, 1.6F, 12.0F)
+                /*new Snowball(serverLevel, this, itemStack)*/ entitysnowball, serverLevel, itemStack, snowball -> snowball.shoot(d, d1 + d3 - snowball.getY(), d2, 1.6F, 12.0F)
+            // EMC end - SnowmanThrowSnowballEvent
             );
         }
 
