--- a/net/minecraft/world/level/block/CakeBlock.java
+++ b/net/minecraft/world/level/block/CakeBlock.java
@@ -63,6 +_,7 @@
     ) {
         Item item = itemStack.getItem();
         if (itemStack.is(ItemTags.CANDLES) && state.getValue(BITES) == 0 && Block.byItem(item) instanceof CandleBlock candleBlock) {
+            new com.empireminecraft.empirecraft.event.player.PlayerCandleCakeEvent((org.bukkit.entity.Player) player.getBukkitEntity(), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level), itemStack.getBukkitStack()).callEvent(); // EMC - PlayerCandleCakeEvent
             // Paper start - call change block event
             if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(player, pos, CandleCakeBlock.byCandle(candleBlock))) {
                 player.containerMenu.forceHeldSlot(hand); // update inv because candle could decrease
