--- a/net/minecraft/world/level/block/CakeBlock.java
+++ b/net/minecraft/world/level/block/CakeBlock.java
@@ -67,6 +_,7 @@
     ) {
         Item item = stack.getItem();
         if (stack.is(ItemTags.CANDLES) && state.getValue(BITES) == 0 && Block.byItem(item) instanceof CandleBlock candleBlock) {
+            new com.empireminecraft.empirecraft.event.player.PlayerCandleCakeEvent((org.bukkit.entity.Player) player.getBukkitEntity(), io.papermc.paper.util.MCUtil.toLocation(level, pos), stack.getBukkitStack()).callEvent(); // EMC - PlayerCandleCakeEvent
             // Paper start - call change block event
             if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(player, pos, CandleCakeBlock.byCandle(candleBlock))) {
                 player.containerMenu.sendAllDataToRemote(); // update inv because candle could decrease
