--- a/net/minecraft/world/level/block/DriedGhastBlock.java
+++ b/net/minecraft/world/level/block/DriedGhastBlock.java
@@ -108,6 +_,11 @@
             level.setBlock(pos, state.setValue(HYDRATION_LEVEL, this.getHydrationLevel(state) + 1), 2);
             level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(state));
         } else {
+            // EMC start - Call BlockFadeEvent
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, Blocks.AIR.defaultBlockState()).isCancelled()) {
+                return;
+            }
+            // EMC end
             this.spawnGhastling(level, pos, state);
         }
     }
