--- a/net/minecraft/world/level/block/FireBlock.java
+++ b/net/minecraft/world/level/block/FireBlock.java
@@ -208,7 +_,7 @@
                         return;
                     }
 
-                    if (ageValue == 15 && random.nextInt(4) == 0 && !this.canBurn(level.getBlockState(pos.below()))) {
+                    if (ageValue == 15 && random.nextInt(4) == 0 /*&& !this.canBurn(level.getBlockState(pos.below()))*/) { // EMC - Make all blocks burn at the same rate
                         this.fireExtinguished(level, pos); // CraftBukkit
                         return;
                     }
