--- a/net/minecraft/world/level/block/FlowerPotBlock.java
+++ b/net/minecraft/world/level/block/FlowerPotBlock.java
@@ -104,6 +_,7 @@
                 return InteractionResult.PASS;
             }
             // Paper end - Add PlayerFlowerPotManipulateEvent
+            itemStack = ItemStack.fromBukkitCopy(event.getItem()); // EMC - Use FlowerPotManipulateEvent item for flowerpot removal
             if (!player.addItem(itemStack)) {
                 player.drop(itemStack, false);
             }
