--- a/net/minecraft/world/level/block/FlowerPotBlock.java
+++ b/net/minecraft/world/level/block/FlowerPotBlock.java
@@ -115,6 +_,7 @@
             return InteractionResult.PASS;
         }
         // Paper end - Add PlayerFlowerPotManipulateEvent
+        plant = ItemStack.fromBukkitCopy(event.getItem()); // EMC - Use FlowerPotManipulateEvent item for flowerpot removal
         if (!player.addItem(plant)) {
             player.drop(plant, false);
         }
