--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/api/CraftEAPI_Server.java
@@ -1,0 +_,33 @@
+package com.empireminecraft.empirecraft.api;
+
+import com.destroystokyo.paper.profile.PlayerProfile;
+import com.google.common.collect.Maps;
+import net.kyori.adventure.text.Component;
+import org.bukkit.GameMode;
+import org.jspecify.annotations.NullMarked;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+
+@NullMarked
+public class CraftEAPI_Server implements EAPI_Server {
+    public final Map<UUID, FakePlayer> fakePlayerMap = Maps.newHashMap();
+
+    @Override
+    public @Nullable FakePlayer getFakePlayer(UUID uuid) {
+        return fakePlayerMap.get(uuid);
+    }
+
+    @Override
+    public FakePlayer createFakePlayer(PlayerProfile profile, int ping, GameMode gameMode, Component playerListName) throws IllegalArgumentException {
+        return new CraftFakePlayer(profile, ping, gameMode, playerListName);
+    }
+
+    @Override
+    public Set<FakePlayer> getJoinedFakePlayers() {
+        return Set.copyOf(fakePlayerMap.values());
+    }
+
+}
