--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/api/EntityTasksHandler.java
@@ -1,0 +_,38 @@
+package com.empireminecraft.empirecraft.api;
+
+import net.minecraft.world.entity.Entity;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public final class EntityTasksHandler {
+
+    private EntityTasksHandler() {
+    }
+
+    public static void tickHandler(Entity entity) {
+        if (entity.entityTasks.isEmpty()) {
+            return;
+        }
+        List < EntityTask < ? >> tasksToRun = new ArrayList<>();
+        final Iterator<EntityTask<?>> it = entity.entityTasks.iterator();
+        while (it.hasNext()) {
+            EntityTask < ? > task = it.next();
+            if (!task.isValid()) {
+                it.remove();
+                continue;
+            }
+            if (task.isReady()) {
+                tasksToRun.add(task);
+
+            }
+        }
+
+        tasksToRun.forEach(EntityTask::tick);
+
+    }
+
+    public static class TaskList extends ArrayList<EntityTask<?>> {
+    }
+}
