--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/player/PlayerThrowTridentEvent.java
@@ -1,0 +_,46 @@
+package com.empireminecraft.empirecraft.event.player;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.bukkit.inventory.ItemStack;
+import org.jspecify.annotations.NullMarked;
+
+/**
+ * Called when a Player throws a trident
+ */
+@NullMarked
+public class PlayerThrowTridentEvent extends PlayerEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private final Entity projectile;
+    private final ItemStack tridentItem;
+
+    public ItemStack getTridentItem() {
+        return tridentItem;
+    }
+
+    public PlayerThrowTridentEvent(Player shooter, ItemStack tridentItem, Entity projectile) {
+        super(shooter);
+        this.tridentItem = tridentItem;
+        this.projectile = projectile;
+    }
+
+    /**
+     * Gets the projectile which will be launched by this event
+     *
+     * @return the launched projectile
+     */
+    public Entity getProjectile() {
+        return projectile;
+    }
+
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
