--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/player/PlayerCandleCakeEvent.java
@@ -1,0 +_,39 @@
+package com.empireminecraft.empirecraft.event.player;
+
+import org.bukkit.Location;
+import org.bukkit.block.Block;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.bukkit.inventory.ItemStack;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class PlayerCandleCakeEvent extends PlayerEvent {
+    private final Block cake;
+    private final ItemStack candle;
+
+    public PlayerCandleCakeEvent(Player player, Location cakeLoc, ItemStack candle) {
+        super(player);
+        this.cake = cakeLoc.getBlock();
+        this.candle = candle;
+    }
+
+    public Block getCakeBlock() {
+        return cake;
+    }
+
+    public ItemStack getItem() {
+        return candle;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
