--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -121,8 +_,19 @@
         final boolean tryMoveDown,
         final boolean movedUp
     ) {
+        // EMC start - MonsterEggSpawnEvent
+        PostSpawnProcessor<T> postSpawn = null;
+        final com.empireminecraft.empirecraft.event.MonsterEggSpawnEvent event = new com.empireminecraft.empirecraft.event.MonsterEggSpawnEvent(user != null ? user.getBukkitLivingEntity() : null, org.bukkit.craftbukkit.util.CraftLocation.toBukkit(spawnPos, level), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemStack), spawnReason == EntitySpawnReason.DISPENSER ? org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DISPENSE_EGG : org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG);
+        if (!event.callEvent()) {
+            return null;
+        }
+        var afterSpawn = event.getAfterSpawn();
+        if (afterSpawn != null) {
+            postSpawn = entity -> afterSpawn.accept(entity.getBukkitEntity());
+        }
+        // EMC end - MonsterEggSpawnEvent
         // CraftBukkit start
-        return this.spawn(level, itemStack, user, spawnPos, spawnReason, tryMoveDown, movedUp, spawnReason == EntitySpawnReason.DISPENSER ? org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DISPENSE_EGG : org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG); // Paper - use correct spawn reason for dispenser spawn eggs
+        return this.spawn(level, itemStack, user, spawnPos, spawnReason, tryMoveDown, movedUp, spawnReason == EntitySpawnReason.DISPENSER ? org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DISPENSE_EGG : org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG, postSpawn); // Paper - use correct spawn reason for dispenser spawn eggs // EMC - Add Consumer to spawn method
     }
 
     @Nullable
@@ -136,6 +_,21 @@
         final boolean movedUp,
         final org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason createSpawnReason
     ) {
+        // EMC start - Add PostSpawnProcessor to spawn method
+        return this.spawn(level, itemStack, user, spawnPos, spawnReason, tryMoveDown, movedUp, createSpawnReason, null);
+    }
+    @Nullable
+    public T spawn(
+        final ServerLevel level,
+        final @Nullable ItemStack itemStack,
+        final @Nullable LivingEntity user,
+        final BlockPos spawnPos,
+        final EntitySpawnReason spawnReason,
+        final boolean tryMoveDown,
+        final boolean movedUp,
+        final org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason createSpawnReason,
+        @Nullable PostSpawnProcessor<T> postSpawn
+    ) {
         // CraftBukkit end
         PostSpawnProcessor<T> postSpawnConfig;
         if (itemStack != null) {
@@ -143,6 +_,11 @@
         } else {
             postSpawnConfig = PostSpawnProcessor.nop();
         }
+        // EMC start - Add PostSpawnProcessor to spawn method
+        if (postSpawn != null) {
+            postSpawnConfig = postSpawnConfig.andThen(postSpawn);
+        }
+        // EMC end - Add PostSpawnProcessor to spawn method
 
         return this.spawn(level, postSpawnConfig, spawnPos, spawnReason, tryMoveDown, movedUp, createSpawnReason); // CraftBukkit
     }
