--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1291,8 +_,18 @@
         boolean shouldOffsetY,
         boolean shouldOffsetYMore
     ) {
+        // EMC start - MonsterEggSpawnEvent
+        Consumer<T> consumer = null;
+        final com.empireminecraft.empirecraft.event.MonsterEggSpawnEvent event = new com.empireminecraft.empirecraft.event.MonsterEggSpawnEvent(owner != null ? owner.getBukkitLivingEntity() : null, org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(spawnedFrom), spawnReason == EntitySpawnReason.DISPENSER ? org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DISPENSE_EGG : org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG);
+        if (!event.callEvent()) {
+            return null;
+        }
+        if (event.getAfterSpawn() != null) {
+            consumer = entity -> event.getAfterSpawn().accept(entity.getBukkitEntity());
+        }
+        // EMC end - MonsterEggSpawnEvent
         // CraftBukkit start
-        return this.spawn(level, spawnedFrom, owner, pos, spawnReason, shouldOffsetY, shouldOffsetYMore, spawnReason == EntitySpawnReason.DISPENSER ? org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DISPENSE_EGG : org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG); // Paper - use correct spawn reason for dispenser spawn eggs
+        return this.spawn(level, spawnedFrom, owner, pos, spawnReason, shouldOffsetY, shouldOffsetYMore, spawnReason == EntitySpawnReason.DISPENSER ? org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DISPENSE_EGG : org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG, consumer); // Paper - use correct spawn reason for dispenser spawn eggs // EMC - Add Consumer to spawn method
     }
 
     @Nullable
@@ -1306,6 +_,22 @@
         boolean shouldOffsetYMore,
         org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason createSpawnReason
     ) {
+        // EMC start - Add After Consumer to spawn method
+        return this.spawn(level, spawnedFrom, owner, pos, spawnReason, shouldOffsetY, shouldOffsetYMore, createSpawnReason, null);
+    }
+    @Nullable
+    public T spawn(
+        ServerLevel level,
+        @Nullable ItemStack spawnedFrom,
+        @Nullable LivingEntity owner,
+        BlockPos pos,
+        EntitySpawnReason spawnReason,
+        boolean shouldOffsetY,
+        boolean shouldOffsetYMore,
+        org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason createSpawnReason,
+        @Nullable Consumer<T> afterConsumer
+    ) {
+        // EMC end - Add After Consumer to spawn method
         // CraftBukkit end
         Consumer<T> consumer;
         if (spawnedFrom != null) {
@@ -1313,6 +_,11 @@
         } else {
             consumer = entity -> {};
         }
+        // EMC start - Add After Consumer to spawn method
+        if (afterConsumer != null) {
+            consumer = consumer.andThen(afterConsumer);
+        }
+        // EMC end - Add After Consumer to spawn method
 
         return this.spawn(level, consumer, pos, spawnReason, shouldOffsetY, shouldOffsetYMore, createSpawnReason); // CraftBukkit
     }
