--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -367,6 +_,7 @@
             .add(Attributes.EXPLOSION_KNOCKBACK_RESISTANCE)
             .add(Attributes.WATER_MOVEMENT_EFFICIENCY)
             .add(Attributes.MOVEMENT_EFFICIENCY)
+            .add(Attributes.ATTACK_DAMAGE, 2.0D) // EMC - make all mobs able to attack
             .add(Attributes.ATTACK_KNOCKBACK)
             .add(Attributes.CAMERA_DISTANCE)
             .add(Attributes.WAYPOINT_TRANSMIT_RANGE);
@@ -825,6 +_,7 @@
                     this.level().getCraftServer().getPluginManager().callEvent(event);
 
                     if (event.isCancelled()) {
+                        if (true) { return null; } // EMC - Do not return item to player on cancel drop item event
                         org.bukkit.inventory.ItemStack inHandItem = player.getInventory().getItemInMainHand();
                         if (includeThrower && inHandItem.getAmount() == 0) {
                             // The complete stack was dropped
@@ -1165,6 +_,13 @@
             return true;
         }
         // CraftBukkit end
+        // EMC start - EntityEffectAddedEvent
+        com.empireminecraft.empirecraft.event.entity.EntityEffectAddedEvent addedEvent = new com.empireminecraft.empirecraft.event.entity.EntityEffectAddedEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), org.bukkit.craftbukkit.potion.CraftPotionUtil.toBukkit(effectInstance));
+        if (!addedEvent.callEvent()) {
+            return false;
+        }
+        effectInstance = org.bukkit.craftbukkit.potion.CraftPotionUtil.fromBukkit(addedEvent.getEffect());
+        // EMC end - EntityEffectAddedEvent
         if (!this.canBeAffected(effectInstance)) {
             return false;
         } else {
@@ -1978,9 +_,14 @@
 
     protected float getKnockback(Entity attacker, DamageSource damageSource) {
         float f = (float)this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
-        return this.level() instanceof ServerLevel serverLevel
+        // EMC start - EntityKnockbackEvent
+        f = this.level() instanceof ServerLevel serverLevel
             ? EnchantmentHelper.modifyKnockback(serverLevel, this.getWeaponItem(), attacker, damageSource, f) / 2.0F
             : f / 2.0F;
+        var event = new com.empireminecraft.empirecraft.event.entity.EntityKnockbackEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), attacker.getBukkitEntity(), f);
+        event.callEvent();
+        return event.getLevel();
+        // EMC end - EntityKnockbackEvent
     }
 
     protected void dropFromLootTable(ServerLevel level, DamageSource damageSource, boolean playerKill) {
@@ -2273,7 +_,12 @@
     }
 
     public int getArmorValue() {
-        return Mth.floor(this.getAttributeValue(Attributes.ARMOR));
+        // EMC start
+        final double value = this.getAttributeValue(Attributes.ARMOR);
+        com.empireminecraft.empirecraft.event.entity.LivingEntityArmorProtectEvent event = new com.empireminecraft.empirecraft.event.entity.LivingEntityArmorProtectEvent(this.getBukkitLivingEntity(), value);
+        event.callEvent();
+        return Mth.floor(event.getArmorValue());
+        // EMC end
     }
 
     protected void hurtArmor(DamageSource damageSource, float damageAmount) {
@@ -3304,8 +_,15 @@
     }
 
     public boolean doHurtTarget(ServerLevel level, Entity target) {
-        this.setLastHurtMob(target);
-        return false;
+        // EMC start - make all mobs able to attack
+        boolean flag = target.hurtServer(level, this.damageSources().mobAttack(this), (float) ((int) this.getAttribute(Attributes.ATTACK_DAMAGE).getValue()));
+        if (flag) {
+            DamageSource damagesource = this.damageSources().mobAttack(this);
+            EnchantmentHelper.doPostAttackEffects(level, target, damagesource);
+            this.setLastHurtMob(target);
+        }
+        return flag;
+        // EMC end - make all mobs able to attack
     }
 
     public void causeExtraKnockback(Entity target, float strength, Vec3 currentMovement) {
