--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -361,6 +_,7 @@
             .add(Attributes.EXPLOSION_KNOCKBACK_RESISTANCE)
             .add(Attributes.WATER_MOVEMENT_EFFICIENCY)
             .add(Attributes.MOVEMENT_EFFICIENCY)
+            .add(Attributes.ATTACK_DAMAGE, 2.0D) // EMC - make all mobs able to attack
             .add(Attributes.ATTACK_KNOCKBACK)
             .add(Attributes.CAMERA_DISTANCE)
             .add(Attributes.WAYPOINT_TRANSMIT_RANGE);
@@ -820,6 +_,7 @@
                     this.level().getCraftServer().getPluginManager().callEvent(event);
 
                     if (event.isCancelled()) {
+                        if (true) { return null; } // EMC - Do not return item to player on cancel drop item event
                         org.bukkit.inventory.ItemStack inHandItem = player.getInventory().getItemInMainHand();
                         if (thrownFromHand && inHandItem.getAmount() == 0) {
                             // The complete stack was dropped
@@ -1154,7 +_,7 @@
         // Paper start - Don't fire sync event during generation
         return this.addEffect(newEffect, source, cause, true);
     }
-    public boolean addEffect(final MobEffectInstance newEffect, final @Nullable Entity source, final EntityPotionEffectEvent.Cause cause, final boolean fireEvent) {
+    public boolean addEffect(MobEffectInstance newEffect, final @Nullable Entity source, final EntityPotionEffectEvent.Cause cause, final boolean fireEvent) { // EMC - Remove final
         // Paper end - Don't fire sync event during generation
         // org.spigotmc.AsyncCatcher.catchOp("effect add"); // Spigot // Paper - move to API
         if (this.isTickingEffects) {
@@ -1162,6 +_,13 @@
             return true;
         }
         // CraftBukkit end
+        // EMC start - EntityEffectAddedEvent
+        com.empireminecraft.empirecraft.event.entity.EntityEffectAddedEvent addedEvent = new com.empireminecraft.empirecraft.event.entity.EntityEffectAddedEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), org.bukkit.craftbukkit.potion.CraftPotionUtil.toBukkit(newEffect));
+        if (!addedEvent.callEvent()) {
+            return false;
+        }
+        newEffect = org.bukkit.craftbukkit.potion.CraftPotionUtil.fromBukkit(addedEvent.getEffect());
+        // EMC end - EntityEffectAddedEvent
         if (!this.canBeAffected(newEffect)) {
             return false;
         } else {
@@ -1979,9 +_,14 @@
 
     protected float getKnockback(final Entity target, final DamageSource damageSource) {
         float knockback = (float)this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
-        return this.level() instanceof ServerLevel level
+        // EMC start - EntityKnockbackEvent
+        knockback = this.level() instanceof ServerLevel level
             ? EnchantmentHelper.modifyKnockback(level, this.getWeaponItem(), target, damageSource, knockback) / 2.0F
             : knockback / 2.0F;
+        var event = new com.empireminecraft.empirecraft.event.entity.EntityKnockbackEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), target.getBukkitEntity(), knockback);
+        event.callEvent();
+        return event.getLevel();
+        // EMC end - EntityKnockbackEvent
     }
 
     protected void dropFromLootTable(final ServerLevel level, final DamageSource source, final boolean playerKilled) {
@@ -2337,7 +_,12 @@
     }
 
     public int getArmorValue() {
-        return Mth.floor(this.getAttributeValue(Attributes.ARMOR));
+        // EMC start
+        final double value = this.getAttributeValue(Attributes.ARMOR);
+        com.empireminecraft.empirecraft.event.entity.LivingEntityArmorProtectEvent event = new com.empireminecraft.empirecraft.event.entity.LivingEntityArmorProtectEvent(this.getBukkitLivingEntity(), value);
+        event.callEvent();
+        return Mth.floor(event.getArmorValue());
+        // EMC end
     }
 
     protected void hurtArmor(final DamageSource damageSource, final float damage) {
@@ -3358,8 +_,15 @@
     }
 
     public boolean doHurtTarget(final ServerLevel level, final Entity target) {
-        this.setLastHurtMob(target);
-        return false;
+        // EMC start - make all mobs able to attack
+        boolean flag = target.hurtServer(level, this.damageSources().mobAttack(this), (float) ((int) this.getAttribute(Attributes.ATTACK_DAMAGE).getValue()));
+        if (flag) {
+            DamageSource damagesource = this.damageSources().mobAttack(this);
+            EnchantmentHelper.doPostAttackEffects(level, target, damagesource);
+            this.setLastHurtMob(target);
+        }
+        return flag;
+        // EMC end - make all mobs able to attack
     }
 
     public void causeExtraKnockback(final Entity target, final float knockback, final Vec3 oldMovement) {
