--- a/net/minecraft/world/entity/NeutralMob.java
+++ b/net/minecraft/world/entity/NeutralMob.java
@@ -51,8 +_,15 @@
             this.stopBeingAngry();
         } else {
             if (target != null && !Objects.equals(persistentAngerTarget, target.getUUID())) {
-                this.setPersistentAngerTarget(target.getUUID());
-                this.startPersistentAngerTimer();
+                // EMC start - EntityAngerEvent
+                if (this instanceof Entity) {
+                    com.empireminecraft.empirecraft.event.entity.EntityAngerEvent event = new com.empireminecraft.empirecraft.event.entity.EntityAngerEvent(((Entity)this).getBukkitEntity(), target.getBukkitLivingEntity());
+                    if (event.callEvent()) {
+                        this.setPersistentAngerTarget(target.getUUID());
+                        this.startPersistentAngerTimer();
+                    }
+                }
+                // EMC end - EntityAngerEvent
             }
 
             if (this.getRemainingPersistentAngerTime() > 0 && (target == null || target.getType() != EntityType.PLAYER || !updateAnger)) {
