--- a/net/minecraft/world/entity/NeutralMob.java
+++ b/net/minecraft/world/entity/NeutralMob.java
@@ -65,11 +_,21 @@
             this.stopBeingAngry();
         } else {
             if (target != null) {
+                // EMC start - EntityAngerEvent
+                boolean cancelStartAnger = false;
                 if (persistentAngerTarget == null || !persistentAngerTarget.matches(target)) {
-                    this.setPersistentAngerTarget(EntityReference.of(target));
+                    if (this instanceof Entity) {
+                        cancelStartAnger = !(new com.empireminecraft.empirecraft.event.entity.EntityAngerEvent(((Entity) this).getBukkitEntity(), target.getBukkitLivingEntity()).callEvent());
+                    }
+                    if (!cancelStartAnger) {
+                        this.setPersistentAngerTarget(EntityReference.of(target));
+                    }
                 }
 
-                this.startPersistentAngerTimer();
+                if (!cancelStartAnger) {
+                    this.startPersistentAngerTimer();
+                }
+                // EMC end - EntityAngerEvent
             }
 
             if (persistentAngerTarget != null && !this.isAngry() && (target == null || !isValidPlayerTarget(target) || !updateAnger)) {
