--- a/net/minecraft/world/entity/NeutralMob.java
+++ b/net/minecraft/world/entity/NeutralMob.java
@@ -71,12 +_,20 @@
         } else {
             LivingEntity target = this.getTarget();
             if (target != null) {
+                boolean cancelStartAnger = false; // EMC - EntityAngerEvent
                 boolean newTarget = persistentAngerTarget == null || !persistentAngerTarget.matches(target);
                 if (newTarget) {
+                    // EMC start - EntityAngerEvent
+                    if (this instanceof Entity) {
+                        cancelStartAnger = !(new com.empireminecraft.empirecraft.event.entity.EntityAngerEvent(((Entity) this).getBukkitEntity(), target.getBukkitLivingEntity()).callEvent());
+                    }
+                    if (!cancelStartAnger) {
+                        // EMC end - EntityAngerEvent
                     this.setPersistentAngerTarget(EntityReference.of(target));
+                    } // EMC - EntityAngerEvent
                 }
 
-                if (newTarget || stayAngryIfTargetPresent) {
+                if ((newTarget && !cancelStartAnger) || stayAngryIfTargetPresent) { // EMC - EntityAngerEvent
                     this.startPersistentAngerTimer();
                 }
             }
