--- a/net/minecraft/world/entity/PathfinderMob.java
+++ b/net/minecraft/world/entity/PathfinderMob.java
@@ -18,6 +_,7 @@
     }
 
     public BlockPos movingTarget; public BlockPos getMovingTarget() { return movingTarget; } // Paper
+    public boolean markedAggressive = false; // EMC - Make Aggressive API
 
     public float getWalkTargetValue(BlockPos pos) {
         return this.getWalkTargetValue(pos, this.level());
