--- a/net/minecraft/world/entity/TamableAnimal.java
+++ b/net/minecraft/world/entity/TamableAnimal.java
@@ -188,12 +_,14 @@
         if (playerTeam != null) {
             return playerTeam;
         } else {
-            if (this.isTame()) {
-                LivingEntity rootOwner = this.getRootOwner();
-                if (rootOwner != null) {
-                    return rootOwner.getTeam();
-                }
-            }
+            // EMC start - Don't use Players scoreboard team for Tamed Animals
+            //if (this.isTame()) {
+            //    LivingEntity rootOwner = this.getRootOwner();
+            //    if (rootOwner != null) {
+            //        return rootOwner.getTeam();
+            //    }
+            //}
+            // EMC end - Don't use Players scoreboard team for Tamed Animals
 
             return null;
         }
