--- a/net/minecraft/world/entity/TamableAnimal.java
+++ b/net/minecraft/world/entity/TamableAnimal.java
@@ -200,7 +_,7 @@
             return ownTeam;
         }
 
-        if (this.isTame()) {
+        if (false && this.isTame()) { // EMC - Don't use Players scoreboard team for Tamed Animals
             LivingEntity owner = this.getRootOwner();
             if (owner != null) {
                 return owner.getTeam();
