--- a/net/minecraft/world/entity/TamableAnimal.java
+++ b/net/minecraft/world/entity/TamableAnimal.java
@@ -208,17 +_,19 @@
         return true;
     }
 
-    @Override
-    public PlayerTeam getTeam() {
-        if (this.isTame()) {
-            LivingEntity owner = this.getOwner();
-            if (owner != null) {
-                return owner.getTeam();
-            }
-        }
+    // EMC start - Don't use Players scoreboard team for Tamed Animals
+    //@Override
+    //public PlayerTeam getTeam() {
+    //    if (this.isTame()) {
+    //        LivingEntity owner = this.getOwner();
+    //        if (owner != null) {
+    //            return owner.getTeam();
+    //        }
+    //    }
 
-        return super.getTeam();
-    }
+    //    return super.getTeam();
+    //}
+    // EMC end - Don't use Players scoreboard team for Tamed Animals
 
     @Override
     protected boolean considersEntityAsAlly(Entity entity) {
