--- a/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -156,7 +_,7 @@
                     if (instance != null) {
                         instance.load(compound);
                     }
-                }, () -> LOGGER.warn("Ignoring unknown attribute '{}'", resourceLocation));
+                }, () -> LOGGER.debug("Ignoring unknown attribute '{}'", resourceLocation)); // EMC - Suppress log warning
             } else {
                 LOGGER.warn("Ignoring malformed attribute '{}'", string);
             }
