--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/entity/ArrowKnockbackEvent.java
@@ -1,0 +_,47 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.empirecraft.event.entity;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class ArrowKnockbackEvent extends EntityEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private final Entity target;
+    private double knockback;
+
+    public ArrowKnockbackEvent(@NotNull Entity arrow, @NotNull Entity target, double knockback) {
+        super(arrow);
+        this.target = target;
+        this.knockback = knockback;
+    }
+
+    @NotNull
+    public Entity getTarget() {
+        return this.target;
+    }
+
+    public void setKnockback(double knockback) {
+        this.knockback = knockback;
+    }
+
+    public double getKnockback() {
+        return knockback;
+    }
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
