--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/entity/EntityEffectAddedEvent.java
@@ -1,0 +_,54 @@
+package com.empireminecraft.empirecraft.event.entity;
+
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.bukkit.potion.PotionEffect;
+import org.jetbrains.annotations.NotNull;
+
+public class EntityEffectAddedEvent extends Event implements Cancellable {
+    private static final HandlerList handlers = new HandlerList();
+    private boolean canceled;
+
+    private final LivingEntity entity;
+    private PotionEffect effect;
+
+    public EntityEffectAddedEvent(@NotNull LivingEntity entity, @NotNull PotionEffect effect) {
+        this.entity = entity;
+        this.effect = effect;
+    }
+
+    @NotNull
+    public PotionEffect getEffect() {
+        return effect;
+    }
+
+    public void setEffect(@NotNull PotionEffect effect) {
+        this.effect = effect;
+    }
+
+    @NotNull
+    public LivingEntity getEntity() {
+        return entity;
+    }
+
+    public boolean isCancelled() {
+        return canceled;
+    }
+
+    public void setCancelled(boolean cancel) {
+        canceled = cancel;
+    }
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+}
