--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/entity/EntityTriggerSculkCatalystEvent.java
@@ -1,0 +_,49 @@
+package com.empireminecraft.empirecraft.event.entity;
+
+import org.bukkit.block.Block;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class EntityTriggerSculkCatalystEvent extends EntityEvent implements Cancellable {
+    private final Block sculkCatalyst;
+    private boolean cancelled = false;
+
+    public EntityTriggerSculkCatalystEvent(LivingEntity entity, Block sculkCatalyst) {
+        super(entity);
+        this.sculkCatalyst = sculkCatalyst;
+    }
+
+    @Override
+    public LivingEntity getEntity() {
+        return (LivingEntity) entity;
+    }
+
+    public Block getSculkCatalyst() {
+        return sculkCatalyst;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+        cancelled = cancel;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+}
