--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/entity/EntityKnockbackEvent.java
@@ -1,0 +_,48 @@
+package com.empireminecraft.empirecraft.event.entity;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public class EntityKnockbackEvent extends Event {
+    private final LivingEntity attackingEntity;
+    private final @Nullable Entity targetEntity;
+    private float level;
+
+    public EntityKnockbackEvent(LivingEntity attackingEntity, @Nullable Entity targetEntity, float level) {
+        this.attackingEntity = attackingEntity;
+        this.targetEntity = targetEntity;
+        this.level = level;
+    }
+
+    public LivingEntity getAttackingEntity() {
+        return attackingEntity;
+    }
+
+
+    public @Nullable Entity getTargetEntity() {
+        return targetEntity;
+    }
+
+    public float getLevel() {
+        return level;
+    }
+
+    public void setLevel(float level) {
+        this.level = level;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
