--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/entity/GhastMoveTargetEvent.java
@@ -1,0 +_,60 @@
+/*
+ * Copyright (c) 2025. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.empirecraft.event.entity;
+
+import org.bukkit.Location;
+import org.bukkit.entity.Bat;
+import org.bukkit.entity.Ghast;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.Random;
+
+@NullMarked
+public class GhastMoveTargetEvent extends EntityEvent implements Cancellable {
+    private Location moveTarget;
+    private static final Random random = new Random();
+
+    public GhastMoveTargetEvent(LivingEntity entity, Location moveTarget) {
+        super(entity);
+        this.moveTarget = moveTarget;
+    }
+
+    public Location getMoveTarget() {
+        return moveTarget;
+    }
+
+    public void setMoveTarget(Location moveTarget) {
+        this.moveTarget = moveTarget;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+    private boolean cancelled = false;
+
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+        cancelled = cancel;
+    }
+
+}
