--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/entity/LivingEntityArmorProtectEvent.java
@@ -1,0 +_,39 @@
+package com.empireminecraft.empirecraft.event.entity;
+
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class LivingEntityArmorProtectEvent extends EntityEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private final LivingEntity entity;
+
+    private double armorValue;
+
+    public LivingEntityArmorProtectEvent(LivingEntity entity, double armorValue) {
+        super(entity);
+        this.entity = entity;
+        this.armorValue = armorValue;
+    }
+    public double getArmorValue() {
+        return armorValue;
+    }
+
+    public void setArmorValue(double armorValue) {
+        this.armorValue = armorValue;
+    }
+
+    public LivingEntity getEntity() {
+        return entity;
+    }
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
