--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/entity/ProjectileUseAmmoEvent.java
@@ -1,0 +_,55 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.empirecraft.event.entity;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+public class ProjectileUseAmmoEvent extends EntityEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private final ItemStack weapon;
+    private final ItemStack projectile;
+    private int ammoCount;
+
+    public ProjectileUseAmmoEvent(@NotNull ItemStack weapon, @NotNull ItemStack projectile, @NotNull Entity entity, int ammoCount) {
+        super(entity);
+        this.weapon = weapon;
+        this.projectile = projectile;
+        this.ammoCount = ammoCount;
+    }
+
+    @NotNull
+    public ItemStack getWeapon() {
+        return this.weapon;
+    }
+
+    @NotNull
+    public ItemStack getProjectile() {
+        return this.projectile;
+    }
+
+    public void setAmmoCount(int count) {
+        this.ammoCount = count;
+    }
+
+    public int getAmmoCount() {
+        return ammoCount;
+    }
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
