--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/entity/ZombieReinforcementEvent.java
@@ -1,0 +_,55 @@
+package com.empireminecraft.empirecraft.event.entity;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class ZombieReinforcementEvent extends EntityEvent implements Cancellable {
+    private static final HandlerList handlers = new HandlerList();
+    private final LivingEntity targetEntity;
+    private double chance;
+
+    public ZombieReinforcementEvent(@NotNull Entity sourceZombie, @NotNull LivingEntity targetEntity, double chance) {
+        super(sourceZombie);
+        this.targetEntity = targetEntity;
+        this.chance = chance;
+    }
+
+    @NotNull
+    public LivingEntity getTargetEntity() {
+        return targetEntity;
+    }
+
+    public double getChance() {
+        return chance;
+    }
+
+    public void setChance(double chance) {
+        this.chance = chance;
+    }
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+    private boolean cancelled = false;
+
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+        cancelled = cancel;
+    }
+}
