--- a/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
@@ -67,6 +_,15 @@
                         int i3 = config.depth.sample(random)
                             + (config.extraBottomBlockChance > 0.0F && random.nextFloat() < config.extraBottomBlockChance ? 1 : 0);
                         BlockPos blockPos = mutableBlockPos1.immutable();
+                        // EMC start - VegetationPatchPlaceEvent
+                        if (org.bukkit.Bukkit.isPrimaryThread()) {
+                        org.bukkit.block.Block origin = org.bukkit.craftbukkit.block.CraftBlock.at(level.getLevel(), pos);
+                        org.bukkit.block.Block place = org.bukkit.craftbukkit.block.CraftBlock.at(level.getLevel(), blockPos);
+                        if (!(new com.empireminecraft.empirecraft.event.block.VegetationPatchPlaceEvent(place, origin).callEvent())) {
+                            continue;
+                        }
+                        }
+                        // EMC end - VegetationPatchPlaceEvent
                         boolean flag5 = this.placeGround(level, config, state, random, mutableBlockPos1, i3);
                         if (flag5) {
                             set.add(blockPos);
@@ -90,6 +_,15 @@
     ) {
         for (BlockPos blockPos : possiblePositions) {
             if (config.vegetationChance > 0.0F && random.nextFloat() < config.vegetationChance) {
+                // EMC start - VegetationPatchPlaceEvent
+                if (org.bukkit.Bukkit.isPrimaryThread()) {
+                    org.bukkit.block.Block origin = org.bukkit.craftbukkit.block.CraftBlock.at(context.level().getLevel(), context.origin());
+                    org.bukkit.block.Block place = org.bukkit.craftbukkit.block.CraftBlock.at(context.level().getLevel(), blockPos);
+                    if (!(new com.empireminecraft.empirecraft.event.block.VegetationPatchPlaceEvent(place, origin).callEvent())) {
+                        continue;
+                    }
+                }
+                // EMC end - VegetationPatchPlaceEvent
                 this.placeVegetation(level, config, context.chunkGenerator(), random, blockPos);
             }
         }
