--- a/net/minecraft/world/level/portal/PortalForcer.java
+++ b/net/minecraft/world/level/portal/PortalForcer.java
@@ -246,6 +_,11 @@
                 if (height >= 0 && !this.canPortalReplaceBlock(mutable)) {
                     return false;
                 }
+                // EMC start - PortalFrameCreateBlockCheckEvent
+                if (!(new com.empireminecraft.empirecraft.event.block.PortalFrameCreateBlockCheckEvent(org.bukkit.craftbukkit.block.CraftBlock.at(this.level, mutable))).callEvent()) {
+                    return false;
+                }
+                // EMC end - PortalFrameCreateBlockCheckEvent
             }
         }
 
