--- a/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
+++ b/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
@@ -33,4 +_,8 @@
     default boolean isChargingAttack() {
         return this.isHandRaised();
     }
+    // EMC start - Ranged Attack API
+    void setRangedAttackDistance(float range);
+    void setRangedAttackSpeed(@org.jetbrains.annotations.Nullable Integer min, @org.jetbrains.annotations.Nullable Integer max);
+    // EMC end - Ranged Attack API
 }
