--- a/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -211,7 +_,7 @@
         if (compound.contains("Offers")) {
             MerchantOffers.CODEC
                 .parse(this.registryAccess().createSerializationContext(NbtOps.INSTANCE), compound.get("Offers"))
-                .resultOrPartial(Util.prefix("Failed to load offers: ", LOGGER::warn))
+                .resultOrPartial(Util.prefix("Failed to load offers: ", LOGGER::debug)) // EMC - Suppress log warning
                 .ifPresent(list -> this.offers = list);
         }
 
