--- a/src/main/java/io/papermc/paper/event/player/PlayerFlowerPotManipulateEvent.java
+++ b/src/main/java/io/papermc/paper/event/player/PlayerFlowerPotManipulateEvent.java
@@ -18,7 +_,7 @@
     private static final HandlerList HANDLER_LIST = new HandlerList();
 
     private final Block flowerpot;
-    private final ItemStack item;
+    private ItemStack item; // EMC - Add setRemovedItem; remove final
     private final boolean placing;
 
     private boolean cancelled;
@@ -49,6 +_,16 @@
     public ItemStack getItem() {
         return this.item;
     }
+    // EMC start - Add setRemovedItem
+    /**
+     * Sets a different item to be taken from the flower pot
+     *
+     * @param item the ItemStack to be given when removed from the flowerpot
+     */
+    public void setRemovedItem(ItemStack item) {
+        this.item = item;
+    }
+    // EMC end - Add setRemovedItem
 
     /**
      * Gets if the item is being placed into the flowerpot.
