From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 4 Mar 2013 23:35:02 -0500
Subject: [PATCH] Temporary Meta API


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index c8289d78d1c0e9b876462bc2dc2556b986255c52..241560ede3b03a5b8609308834b0540979008d6e 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -194,6 +194,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         return this.bukkitEntity;
     }
     // Paper end
+    public final java.util.Map<org.bukkit.NamespacedKey, Object> tempMeta = com.google.common.collect.Maps.newHashMap(); // EMC - Temporary Meta API
 
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final String ID_TAG = "id";
diff --git a/net/minecraft/world/level/chunk/ChunkAccess.java b/net/minecraft/world/level/chunk/ChunkAccess.java
index 6d565b52552534ce9cacfc35ad1bf4adcb69eac3..659e79f925a0dc0a65442df9eaff28e994397dcf 100644
--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -87,6 +87,7 @@ public abstract class ChunkAccess implements BiomeManager.NoiseBiomeSource, Ligh
     public org.bukkit.craftbukkit.persistence.DirtyCraftPersistentDataContainer persistentDataContainer = new org.bukkit.craftbukkit.persistence.DirtyCraftPersistentDataContainer(ChunkAccess.DATA_TYPE_REGISTRY);
     // CraftBukkit end
     public final Registry<Biome> biomeRegistry; // CraftBukkit
+    public final java.util.Map<org.bukkit.NamespacedKey, Object> tempMeta = com.google.common.collect.Maps.newHashMap(); // EMC - Temporary Meta API
 
     // Paper start - rewrite chunk system
     private volatile ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray[] blockNibbles;
diff --git a/net/minecraft/world/level/storage/PrimaryLevelData.java b/net/minecraft/world/level/storage/PrimaryLevelData.java
index 2f4922a2035cc7d628dd2c979137163feb90e19f..b01fae9d25fcbd4063d6b9211f199ab754d59a57 100644
--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -88,6 +88,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
         this.pdc = null;
     }
     // CraftBukkit end
+    public final java.util.Map<org.bukkit.NamespacedKey, Object> tempMeta = com.google.common.collect.Maps.newHashMap(); // EMC - Temporary Meta API
 
     private PrimaryLevelData(
         @Nullable CompoundTag loadedPlayerTag,
