From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 4 Mar 2013 23:35:02 -0500
Subject: [PATCH] Temporary Meta API


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index dbfa528e1cd42dac7e99dc0c61ffd13593fd27ac..35e95f34e8ad49eb64b764944d4a2f97f8fdfce4 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -173,6 +173,7 @@ public abstract class Entity
     static boolean isLevelAtLeast(ValueInput input, int level) {
         return input.getIntOr("Bukkit.updateLevel", CURRENT_LEVEL) >= level;
     }
+    public final java.util.Map<org.bukkit.NamespacedKey, Object> tempMeta = com.google.common.collect.Maps.newHashMap(); // EMC - Temporary Meta API
 
     // Paper start - Share random for entities to make them more random
     public static RandomSource SHARED_RANDOM = new RandomRandomSource();
diff --git a/net/minecraft/world/level/chunk/ChunkAccess.java b/net/minecraft/world/level/chunk/ChunkAccess.java
index c73b9871faa87493e081cb54e1bf1dce487359c6..420f4a391d26d92ba30294dd3221f1730dba5307 100644
--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -83,6 +83,7 @@ public abstract class ChunkAccess implements LightChunk, StructureAccess, BiomeM
     private static final org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry();
     public org.bukkit.craftbukkit.persistence.DirtyCraftPersistentDataContainer persistentDataContainer = new org.bukkit.craftbukkit.persistence.DirtyCraftPersistentDataContainer(ChunkAccess.DATA_TYPE_REGISTRY);
     // CraftBukkit end
+    public final java.util.Map<org.bukkit.NamespacedKey, Object> tempMeta = com.google.common.collect.Maps.newHashMap(); // EMC - Temporary Meta API
 
     // Paper start - rewrite chunk system
     private volatile ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray[] blockNibbles;
