From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 4 Mar 2013 23:35:02 -0500
Subject: [PATCH] Temporary Meta API


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 41ceb258bc5992e9252b48dd13a1d6b386460204..be7510fd6feea7d50c7430f24733b32bf43cec29 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -154,6 +154,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     static boolean isLevelAtLeast(ValueInput input, int level) {
         return input.getIntOr("Bukkit.updateLevel", CURRENT_LEVEL) >= level;
     }
+    public final java.util.Map<org.bukkit.NamespacedKey, Object> tempMeta = com.google.common.collect.Maps.newHashMap(); // EMC - Temporary Meta API
 
     // Paper start - Share random for entities to make them more random
     public static RandomSource SHARED_RANDOM = new RandomRandomSource();
diff --git a/net/minecraft/world/level/chunk/ChunkAccess.java b/net/minecraft/world/level/chunk/ChunkAccess.java
index 182c14b660f8860bed627eed4e01fd4002153e9a..9f0e4f14279d3992cd99892836f850ca222478b3 100644
--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -88,6 +88,7 @@ public abstract class ChunkAccess implements BiomeManager.NoiseBiomeSource, Ligh
     public org.bukkit.craftbukkit.persistence.DirtyCraftPersistentDataContainer persistentDataContainer = new org.bukkit.craftbukkit.persistence.DirtyCraftPersistentDataContainer(ChunkAccess.DATA_TYPE_REGISTRY);
     // CraftBukkit end
     public final Registry<Biome> biomeRegistry; // CraftBukkit
+    public final java.util.Map<org.bukkit.NamespacedKey, Object> tempMeta = com.google.common.collect.Maps.newHashMap(); // EMC - Temporary Meta API
 
     // Paper start - rewrite chunk system
     private volatile ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray[] blockNibbles;
diff --git a/net/minecraft/world/level/storage/PrimaryLevelData.java b/net/minecraft/world/level/storage/PrimaryLevelData.java
index d83ef97f6dc71e85c853699e65ad2e76c9c69661..d835a6314bc63c87ee9c8500651c0fbf3614bcb5 100644
--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -88,6 +88,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
         this.pdc = null;
     }
     // CraftBukkit end
+    public final java.util.Map<org.bukkit.NamespacedKey, Object> tempMeta = com.google.common.collect.Maps.newHashMap(); // EMC - Temporary Meta API
 
     private PrimaryLevelData(
         @Nullable CompoundTag loadedPlayerTag,
