From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 8 Mar 2025 17:56:53 -0600
Subject: [PATCH] Target Range API


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 1ed07fd23985a6bf8cf8300f74c92b7531a79fc6..fc49aab3a88a16dd456478d5949839464f8a4404 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -144,6 +144,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     private Leashable.LeashData leashData;
     private BlockPos restrictCenter = BlockPos.ZERO;
     private float restrictRadius = -1.0F;
+    public @Nullable Double targetRange; // EMC - Target Range API
     public boolean aware = true; // CraftBukkit
 
     protected Mob(EntityType<? extends Mob> entityType, Level level) {
diff --git a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
index ae585f10827f52db80f048b84d6096737d12fee2..beb075a877debeef2e2a834830506b7a57c53453 100644
--- a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -71,6 +71,7 @@ public abstract class TargetGoal extends Goal {
     }
 
     protected double getFollowDistance() {
+        if (this.mob.targetRange != null) { return this.mob.targetRange; } // EMC - Target Range API
         return this.mob.getAttributeValue(Attributes.FOLLOW_RANGE);
     }
 
