From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 8 Mar 2025 17:56:53 -0600
Subject: [PATCH] Target Range API


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index a6ed274cb0cf86a1ca93588f946ed52a58354546..6acd5927788ba675e0028f13fc6c9c2f2b795d34 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -148,6 +148,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     private Leashable.@Nullable LeashData leashData;
     private BlockPos homePosition = BlockPos.ZERO;
     private int homeRadius = -1;
+    public @Nullable Double targetRange; // EMC - Target Range API
     public boolean aware = true; // CraftBukkit
     public net.kyori.adventure.util.TriState despawnInPeacefulOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - allow changing despawnInPeaceful
 
diff --git a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
index 404c5304c0a5ccdcfddde981d909d508f9164ce3..9cb12f3eadb7133bbe11acec1822aa175fd00d1b 100644
--- a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -70,6 +70,7 @@ public abstract class TargetGoal extends Goal {
     }
 
     protected double getFollowDistance() {
+        if (this.mob.targetRange != null) { return this.mob.targetRange; } // EMC - Target Range API
         return this.mob.getAttributeValue(Attributes.FOLLOW_RANGE);
     }
 
