From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 8 Mar 2025 17:56:53 -0600
Subject: [PATCH] Target Range API


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 72aee46db9686d5e417be7c930ec78beeeda5f9d..717a391bb0465a7aa87c807ffb2a7c3409c4916f 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -153,6 +153,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     private Leashable.@Nullable LeashData leashData;
     private BlockPos homePosition = BlockPos.ZERO;
     private int homeRadius = -1;
+    public @Nullable Double targetRange; // EMC - Target Range API
     public boolean aware = true; // CraftBukkit
 
     protected Mob(final EntityType<? extends Mob> type, final Level level) {
diff --git a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
index 89765d720fdd810545e8c2010e05a002f87b0c7f..ed0db555a3bbeafff06b71269fba17bac650620f 100644
--- a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -72,6 +72,7 @@ public abstract class TargetGoal extends Goal {
     }
 
     protected double getFollowDistance() {
+        if (this.mob.targetRange != null) { return this.mob.targetRange; } // EMC - Target Range API
         return this.mob.getAttributeValue(Attributes.FOLLOW_RANGE);
     }
 
