From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 8 Mar 2025 18:03:03 -0600
Subject: [PATCH] Panic Proof API


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index d48ecb8a7743a5ed031d891ec7508966a664378c..cbcf7ecd5eb7af7d85ef0684841c4ccd3d756644 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -139,6 +139,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     private BlockPos homePosition = BlockPos.ZERO;
     private int homeRadius = -1;
     public @Nullable Double targetRange; // EMC - Target Range API
+    public boolean panicProof = false; // EMC - Panic Proof API
     public boolean aware = true; // CraftBukkit
     public net.kyori.adventure.util.TriState despawnInPeacefulOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - allow changing despawnInPeaceful
 
diff --git a/net/minecraft/world/entity/ai/goal/FleeSunGoal.java b/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
index b384b1981e0307f6683f88692af1532ec5968406..007cff7c8f56f9e2750dfa9095ebf4c5c9077f36 100644
--- a/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
@@ -29,6 +29,7 @@ public class FleeSunGoal extends Goal {
         return this.mob.getTarget() == null
             && this.level.isBrightOutside()
             && this.mob.isOnFire()
+            && !this.mob.panicProof // EMC - Panic Proof API
             && this.level.canSeeSky(this.mob.blockPosition())
             && this.mob.getItemBySlot(EquipmentSlot.HEAD).isEmpty()
             && this.setWantedPos();
diff --git a/net/minecraft/world/entity/ai/goal/PanicGoal.java b/net/minecraft/world/entity/ai/goal/PanicGoal.java
index 190052ffdd8078ae8eb77f721f9f17191e3a5a61..8146e649cf0735f57ca032eb1d0a1379397cdced 100644
--- a/net/minecraft/world/entity/ai/goal/PanicGoal.java
+++ b/net/minecraft/world/entity/ai/goal/PanicGoal.java
@@ -41,6 +41,7 @@ public class PanicGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (this.mob.panicProof) { return false; } // EMC - Panic Proof API
         if (!this.shouldPanic()) {
             return false;
         } else {
