From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 8 Mar 2025 18:03:03 -0600
Subject: [PATCH] Panic Proof API


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 890a6c9d3975a6e923d625639df607c22bf44722..edc3c8a6f429fd07e626444fe452a5d5ebbf8b47 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -149,6 +149,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     private BlockPos homePosition = BlockPos.ZERO;
     private int homeRadius = -1;
     public @Nullable Double targetRange; // EMC - Target Range API
+    public boolean panicProof = false; // EMC - Panic Proof API
     public boolean aware = true; // CraftBukkit
     public net.kyori.adventure.util.TriState despawnInPeacefulOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - allow changing despawnInPeaceful
 
diff --git a/net/minecraft/world/entity/ai/goal/FleeSunGoal.java b/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
index 78fda22eb673c83f88382269754baed0ad2b21cb..5798b3b49043e68c5d7416e9039642d25f0d75ed 100644
--- a/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
@@ -29,6 +29,7 @@ public class FleeSunGoal extends Goal {
         return this.mob.getTarget() == null
             && this.level.isBrightOutside()
             && this.mob.isOnFire()
+            && !this.mob.panicProof // EMC - Panic Proof API
             && this.level.canSeeSky(this.mob.blockPosition())
             && this.mob.getItemBySlot(EquipmentSlot.HEAD).isEmpty()
             && this.setWantedPos();
diff --git a/net/minecraft/world/entity/ai/goal/PanicGoal.java b/net/minecraft/world/entity/ai/goal/PanicGoal.java
index 491aca1f5ab0c01035ffce98140388f566e61135..12f98543d626c1812237a957867a42794013b202 100644
--- a/net/minecraft/world/entity/ai/goal/PanicGoal.java
+++ b/net/minecraft/world/entity/ai/goal/PanicGoal.java
@@ -41,6 +41,7 @@ public class PanicGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (this.mob.panicProof) { return false; } // EMC - Panic Proof API
         if (!this.shouldPanic()) {
             return false;
         }
