From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 8 Mar 2025 18:03:03 -0600
Subject: [PATCH] Panic Proof API


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index d1fe802a95c73c2f843e7cd29af18e8d552c7954..147a2d8b2fe9671a33b5e251316bb9ede6ad031c 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -149,6 +149,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     private BlockPos homePosition = BlockPos.ZERO;
     private int homeRadius = -1;
     public @Nullable Double targetRange; // EMC - Target Range API
+    public boolean panicProof = false; // EMC - Panic Proof API
     public boolean aware = true; // CraftBukkit
     public net.kyori.adventure.util.TriState despawnInPeacefulOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - allow changing despawnInPeaceful
 
diff --git a/net/minecraft/world/entity/ai/goal/FleeSunGoal.java b/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
index 29b7bab1391dae95e56819264d0b58429e6b642d..ca40fbfe59d6eec4fd8068d6f615c26ea4405e6c 100644
--- a/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
@@ -29,6 +29,7 @@ public class FleeSunGoal extends Goal {
         return this.mob.getTarget() == null
             && this.level.isBrightOutside()
             && this.mob.isOnFire()
+            && !this.mob.panicProof // EMC - Panic Proof API
             && this.level.canSeeSky(this.mob.blockPosition())
             && this.mob.getItemBySlot(EquipmentSlot.HEAD).isEmpty()
             && this.setWantedPos();
diff --git a/net/minecraft/world/entity/ai/goal/PanicGoal.java b/net/minecraft/world/entity/ai/goal/PanicGoal.java
index 07365e490831351ab6e6ce88abe13ed5de974dbb..0217e6a898b94185ec85d68582f8f42e03dd8f80 100644
--- a/net/minecraft/world/entity/ai/goal/PanicGoal.java
+++ b/net/minecraft/world/entity/ai/goal/PanicGoal.java
@@ -41,6 +41,7 @@ public class PanicGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (this.mob.panicProof) { return false; } // EMC - Panic Proof API
         if (!this.shouldPanic()) {
             return false;
         } else {
