From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 12:14:27 -0500
Subject: [PATCH] RangedEntity-API


diff --git a/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
index aecc93f5a989a5ce97f22bed8b894c75076778f8..d300ba3c540d9006ddfcc5eee99af28d2070ccca 100644
--- a/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
@@ -14,10 +14,12 @@ public class RangedAttackGoal extends Goal {
     private int attackTime = -1;
     private final double speedModifier;
     private int seeTime;
-    private final int attackIntervalMin;
-    private final int attackIntervalMax;
-    private final float attackRadius;
-    private final float attackRadiusSqr;
+    // EMC start - RangedEntity API - make non-final
+    private int attackIntervalMin; public void setMinInterval(int time) { this.attackIntervalMin = time; }
+    private int attackIntervalMax; public void setMaxInterval(int time) { this.attackIntervalMax = time; }
+    private float attackRadius;  public void setAttackRadius(float radius) { this.attackRadius = radius; this.attackRadiusSqr = radius * radius; }
+    private float attackRadiusSqr;
+    // EMC end - RangedEntity API
 
     public RangedAttackGoal(RangedAttackMob rangedAttackMob, double speedModifier, int attackInterval, float attackRadius) {
         this(rangedAttackMob, speedModifier, attackInterval, attackInterval, attackRadius);
diff --git a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
index 722f378bb3726200b6fe88948bf1e1d3514a7b5c..b20fef130dfbfe92aaf64ad467223dce3e9962b5 100644
--- a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -13,7 +13,7 @@ public class RangedBowAttackGoal<T extends Monster & RangedAttackMob> extends Go
     private final T mob;
     private final double speedModifier;
     private int attackIntervalMin;
-    private final float attackRadiusSqr;
+    private float attackRadiusSqr; public void setAttackRadius(float radius) { this.attackRadiusSqr = radius * radius; } // EMC - RangedEntity API - make non-final
     private int attackTime = -1;
     private int seeTime;
     private boolean strafingClockwise;
