From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 12:14:27 -0500
Subject: [PATCH] RangedEntity-API


diff --git a/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
index 0c4378a3f6e4aef8579410ba2b2b19edaddb56d5..c59c05c666ec97e8be12c26c53a33cb53d9c34f5 100644
--- a/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
@@ -14,10 +14,12 @@ public class RangedAttackGoal extends Goal {
     private int attackTime = -1;
     private final double speedModifier;
     private int seeTime;
-    private final int attackIntervalMin;
-    private final int attackIntervalMax;
-    private final float attackRadius;
-    private final float attackRadiusSqr;
+    // EMC start - RangedEntity API - make non-final
+    private int attackIntervalMin; public void setMinInterval(int time) { this.attackIntervalMin = time; }
+    private int attackIntervalMax; public void setMaxInterval(int time) { this.attackIntervalMax = time; }
+    private float attackRadius;  public void setAttackRadius(float radius) { this.attackRadius = radius; this.attackRadiusSqr = radius * radius; }
+    private float attackRadiusSqr;
+    // EMC end - RangedEntity API
 
     public RangedAttackGoal(final RangedAttackMob mob, final double speedModifier, final int attackInterval, final float attackRadius) {
         this(mob, speedModifier, attackInterval, attackInterval, attackRadius);
diff --git a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
index 50f12f2581267827c394134cd51a09016a5fa2b0..13a815c753f01088af53ba63c07ee9293ca0b440 100644
--- a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -13,7 +13,7 @@ public class RangedBowAttackGoal<T extends Monster & RangedAttackMob> extends Go
     private final T mob;
     private final double speedModifier;
     private int attackIntervalMin;
-    private final float attackRadiusSqr;
+    private float attackRadiusSqr; public void setAttackRadius(float radius) { this.attackRadiusSqr = radius * radius; } // EMC - RangedEntity API - make non-final
     private int attackTime = -1;
     private int seeTime;
     private boolean strafingClockwise;
