From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 7 Aug 2013 19:52:14 -0400
Subject: [PATCH] EntityDismountEvent Dismount Reason API


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index acd97f0fcf00a3b8712cfa464e256cd6f357bbf0..5a53dd0201d6836644f05b0f8b9dcb181da4b1cf 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -846,6 +846,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
                                         if (!vehicle.isRemoved() && vehicle.hasPassenger(entity)) {
                                             return;
                                         }
+                                        if (vehicle.isRemoved()) { entity.dismountReason =  org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; } // EMC - DismountReason API
 
                                         entity.stopRiding();
                                     }
@@ -1371,6 +1372,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
 
     private void tickPassenger(Entity ridingEntity, Entity passengerEntity, final boolean isActive) { // Paper - EAR 2
         if (passengerEntity.isRemoved() || passengerEntity.getVehicle() != ridingEntity) {
+            if (passengerEntity.isRemoved()) { passengerEntity.dismountReason =  org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; } // EMC - DismountReason API
             passengerEntity.stopRiding();
         } else if (passengerEntity instanceof Player || this.entityTickList.contains(passengerEntity)) {
             passengerEntity.setOldPosAndRot();
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 19037dcb668ecce8a80f19c43b38cdc5fb015dd7..c6bf4e70ac74880f92116cc9e04888be422c38a9 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1516,6 +1516,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             }
             // CraftBukkit end
             if (!teleportTransition.asPassenger()) {
+                this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TELEPORT; // EMC - DismountReason API
                 this.removeVehicle();
             }
 
@@ -2099,10 +2100,11 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
     public void disconnect() {
         this.disconnected = true;
-        this.ejectPassengers();
+        this.ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason.DISCONNECT); // EMC - DismountReason API
 
         // Paper start - Workaround vehicle not tracking the passenger disconnection dismount
         if (this.isPassenger() && this.getVehicle() instanceof ServerPlayer) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DISCONNECT; // EMC - DismountReason API
             this.stopRiding();
         }
         // Paper end - Workaround vehicle not tracking the passenger disconnection dismount
@@ -2316,6 +2318,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             this.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.CHANGE_GAME_MODE, gameMode.getId()));
             if (gameMode == GameType.SPECTATOR) {
                 this.removeEntitiesOnShoulder();
+                this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.SPECTATE; // EMC - DismountReason API
                 this.stopRiding();
                 this.stopUsingItem();
                 EnchantmentHelper.stopLocationBasedEffects(this);
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 989ac565c47a70c7947cb7315d0f5c2cfecd0363..bd4368b0a805ded8ee58c762e224785e8622d345 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -468,6 +468,7 @@ public abstract class PlayerList {
             Entity rootVehicle = player.getRootVehicle();
             if (rootVehicle.hasExactlyOnePlayerPassenger()) {
                 LOGGER.debug("Removing player mount");
+                player.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DISCONNECT; // EMC - DismountReason API
                 player.stopRiding();
                 rootVehicle.getPassengersAndSelf().forEach(entity -> {
                     // Paper start - Fix villager boat exploit
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 1f2cec3936ca3cc82e227fd747f8e09a093ed2ed..b82f9e23c8bf2073c0251df54ad40c4739b0d9cb 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -860,6 +860,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         this.computeSpeed();
         this.inBlockState = null;
         if (this.isPassenger() && this.getVehicle().isRemoved()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; // EMC - DismountReason API
             this.stopRiding();
         }
 
@@ -3283,6 +3284,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
                 }
                 // CraftBukkit end
                 if (this.isPassenger()) {
+                    this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TRANSFER; // EMC - DismountReason API
                     this.stopRiding();
                 }
 
@@ -3308,7 +3310,13 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public void ejectPassengers() {
+        // EMC start - DismountReason API
+        ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN);
+    }
+    public void ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason reason) {
         for (int i = this.passengers.size() - 1; i >= 0; i--) {
+            this.passengers.get(i).dismountReason = reason;
+            // EMC end - DismountReason API
             this.passengers.get(i).stopRiding();
         }
     }
@@ -3366,6 +3374,10 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
     protected boolean removePassenger(Entity passenger, boolean suppressCancellation) { // CraftBukkit
         // Paper end - Force entity dismount during teleportation
+        // EMC start - DismountReason API
+        org.bukkit.event.entity.EntityDismountEvent.DismountReason dismountReason = passenger.dismountReason;
+        passenger.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN;
+        // EMC end - DismountReason API
         if (passenger.getVehicle() == this) {
             throw new IllegalStateException("Use x.stopRiding(y), not y.removePassenger(x)");
         } else {
@@ -3388,7 +3400,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
                 }
             }
 
-            org.bukkit.event.entity.EntityDismountEvent event = new org.bukkit.event.entity.EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity(), !suppressCancellation); // Paper - Force entity dismount during teleportation
+            org.bukkit.event.entity.EntityDismountEvent event = new org.bukkit.event.entity.EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity(), !suppressCancellation, dismountReason); // Paper - Force entity dismount during teleportation // EMC - DismountReasonAPI - Add dismountReason
             // Suppress during worldgen
             if (this.valid) {
                 org.bukkit.Bukkit.getPluginManager().callEvent(event);
@@ -3963,6 +3975,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             );
     }
 
+    public org.bukkit.event.entity.EntityDismountEvent.DismountReason dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN; // EMC - DismountReason API
     public final boolean isInvulnerableToBase(DamageSource damageSource) {
         return this.isRemoved()
             || this.invulnerable && !damageSource.is(DamageTypeTags.BYPASSES_INVULNERABILITY) && !damageSource.isCreativePlayer()
@@ -5198,10 +5211,19 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         }
 
         if (this.removalReason.shouldDestroy()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD; // EMC - DismountReason API
             this.stopRiding();
         }
 
         if (this.removalReason != Entity.RemovalReason.UNLOADED_TO_CHUNK) { this.getPassengers().forEach(Entity::stopRiding); } // Paper - rewrite chunk system
+        // EMC start - DismountReason API
+        if (removalReason != RemovalReason.UNLOADED_TO_CHUNK) {
+            this.getPassengers().forEach(entity -> {
+                entity.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; // EMC - Add this
+                entity.stopRiding();
+            }); // Paper - chunk system - don't adjust passenger state when unloading, it's just not safe (and messes with our logic in entity chunk unload)
+        }
+        // EMC end - DismountReason API
         this.levelCallback.onRemove(removalReason);
         this.onRemoval(removalReason);
         // Paper start - Folia schedulers
diff --git a/net/minecraft/world/entity/Leashable.java b/net/minecraft/world/entity/Leashable.java
index 54c7af09f5f64021895074ea5a1578513aa67bf2..f19755bd4dabb9043fa7fd0048a128ebca037bc6 100644
--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -354,6 +354,7 @@ public interface Leashable {
         }
 
         if (entity.isPassenger()) {
+            entity.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.LEASH; // EMC - DismountReason API
             entity.stopRiding();
         }
     }
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 6b1bbedb825e3258470a4ba17ec8e8647125df20..a9cce2ef06e91f1ef33ab2be22ade8a3a7483aa4 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -470,6 +470,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 }
 
                 if (this.isPassenger() && this.getVehicle() != null && this.getVehicle().dismountsUnderwater()) {
+                    this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.WATER; // EMC - DismountReason API
                     this.stopRiding();
                 }
             } else if (this.getAirSupply() < this.getMaxAirSupply()) {
@@ -4585,6 +4586,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     public void startSleeping(BlockPos pos) {
         if (this.isPassenger()) {
+            if (this instanceof net.minecraft.world.entity.player.Player) this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.PLAYER; // EMC - DismountReason API
             this.stopRiding();
         }
 
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 1b84b758a2b16d4680176b816edd57ee5ca3446f..b3e32f96dc1332dccdaeaea0a9a20981b83ddf0d 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -470,6 +470,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     @Override
     public void rideTick() {
         if (!this.level().isClientSide() && this.wantsToStopRiding() && this.isPassenger()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.PLAYER; // EMC - DismountReason API
             this.stopRiding();
             // CraftBukkit start - SPIGOT-7316: no longer passenger, dismount and return
             if (!this.isPassenger()) {
diff --git a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
index 195f041e94a0d4e813c17e94f0fc7318d1a2ea95..c89bd9a9889a0baa5bd0b7fa1c3688e7c31c1d43 100644
--- a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
@@ -51,6 +51,7 @@ public record TeleportRandomlyConsumeEffect(float diameter) implements ConsumeEf
             );
             double d2 = entity.getZ() + (entity.getRandom().nextDouble() - 0.5) * this.diameter;
             if (entity.isPassenger()) {
+                entity.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TELEPORT; // EMC - DismountReason API
                 entity.stopRiding();
             }
 
