From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 7 Aug 2013 19:52:14 -0400
Subject: [PATCH] EntityDismountEvent Dismount Reason API


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index e1e77671f1be5c33bcaa4423d738f14c663dd069..cec1d55b2816918091ff478be79fa35ec7b17b05 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -880,6 +880,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                                         if (!vehicle.isRemoved() && vehicle.hasPassenger(entity)) {
                                             return;
                                         }
+                                        if (vehicle.isRemoved()) { entity.dismountReason =  org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; } // EMC - DismountReason API
 
                                         entity.stopRiding();
                                     }
@@ -1407,6 +1408,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     private void tickPassenger(final Entity vehicle, final Entity entity, final boolean isActive) { // Paper - EAR 2
         if (entity.isRemoved() || entity.getVehicle() != vehicle) {
+            if (entity.isRemoved()) { entity.dismountReason =  org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; } // EMC - DismountReason API
             entity.stopRiding();
         } else if (entity instanceof Player || this.entityTickList.contains(entity)) {
             entity.setOldPosAndRot();
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 466e6cb6fc343b85431d5245f6741ef4fcda0022..d57a4c6806aa82e6dbe1350fa141bab99eb6237a 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1544,6 +1544,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             }
             // CraftBukkit end
             if (!transition.asPassenger()) {
+                this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TELEPORT; // EMC - DismountReason API
                 this.removeVehicle();
             }
 
@@ -2142,7 +2143,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
     public void disconnect() {
         this.disconnected = true;
-        this.ejectPassengers();
+        this.ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason.DISCONNECT); // EMC - DismountReason API'
 
         // Paper start - Workaround vehicle not tracking the passenger disconnection dismount
         if (this.isPassenger() && this.getVehicle() instanceof ServerPlayer) {
@@ -2365,6 +2366,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             this.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.CHANGE_GAME_MODE, mode.getId()));
             if (mode == GameType.SPECTATOR) {
                 this.removeEntitiesOnShoulder();
+                this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.SPECTATE; // EMC - DismountReason API
                 this.stopRiding();
                 this.stopUsingItem();
                 EnchantmentHelper.stopLocationBasedEffects(this);
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 82748af2479b0f8f15e39d1ad1ccd2a0e583bcdd..e66e44087993788af3e6ed6d803bfbf8bf5e45be 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -473,6 +473,7 @@ public abstract class PlayerList {
             Entity vehicle = player.getRootVehicle();
             if (vehicle.hasExactlyOnePlayerPassenger()) {
                 LOGGER.debug("Removing player mount");
+                player.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DISCONNECT; // EMC - DismountReason API
                 player.stopRiding();
                 vehicle.getPassengersAndSelf().forEach(e -> {
                     // Paper start - Fix villager boat exploit
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index a61c3bee62738119311d81482884897e548ff542..d1dc35e36c30e5c26b3139b7d35f7a73a4779c5c 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -879,6 +879,7 @@ public abstract class Entity
         this.computeSpeed();
         this.inBlockState = null;
         if (this.isPassenger() && this.getVehicle().isRemoved()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; // EMC - DismountReason API
             this.stopRiding();
         }
 
@@ -3305,6 +3306,7 @@ public abstract class Entity
                 }
                 // CraftBukkit end
                 if (this.isPassenger()) {
+                    this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TRANSFER; // EMC - DismountReason API
                     this.stopRiding();
                 }
 
@@ -3330,7 +3332,13 @@ public abstract class Entity
     }
 
     public void ejectPassengers() {
+        // EMC start - DismountReason API
+        ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN);
+    }
+    public void ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason reason) {
         for (int i = this.passengers.size() - 1; i >= 0; i--) {
+            this.passengers.get(i).dismountReason = reason;
+            // EMC end - DismountReason API
             this.passengers.get(i).stopRiding();
         }
     }
@@ -3389,6 +3397,10 @@ public abstract class Entity
 
     protected boolean removePassenger(final Entity passenger, final boolean suppressCancellation) { // CraftBukkit
         // Paper end - Force entity dismount during teleportation
+        // EMC start - DismountReason API
+        org.bukkit.event.entity.EntityDismountEvent.DismountReason dismountReason = passenger.dismountReason;
+        passenger.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN;
+        // EMC end - DismountReason API
         if (passenger.getVehicle() == this) {
             throw new IllegalStateException("Use x.stopRiding(y), not y.removePassenger(x)");
         } else {
@@ -3411,7 +3423,7 @@ public abstract class Entity
                 }
             }
 
-            org.bukkit.event.entity.EntityDismountEvent event = new org.bukkit.event.entity.EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity(), !suppressCancellation); // Paper - Force entity dismount during teleportation
+            org.bukkit.event.entity.EntityDismountEvent event = new org.bukkit.event.entity.EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity(), !suppressCancellation, dismountReason); // Paper - Force entity dismount during teleportation // EMC - DismountReasonAPI - Add dismountReason
             // Suppress during worldgen
             if (this.valid) {
                 org.bukkit.Bukkit.getPluginManager().callEvent(event);
@@ -3981,6 +3993,7 @@ public abstract class Entity
             );
     }
 
+    public org.bukkit.event.entity.EntityDismountEvent.DismountReason dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN; // EMC - DismountReason API
     public final boolean isInvulnerableToBase(final DamageSource source) {
         return this.isRemoved()
             || this.invulnerable && !source.is(DamageTypeTags.BYPASSES_INVULNERABILITY) && !source.isCreativePlayer()
@@ -5127,10 +5140,18 @@ public abstract class Entity
         }
 
         if (this.removalReason.shouldDestroy()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD; // EMC - DismountReason API
             this.stopRiding();
         }
 
-        if (this.removalReason != Entity.RemovalReason.UNLOADED_TO_CHUNK) { this.getPassengers().forEach(Entity::stopRiding); } // Paper - rewrite chunk system
+        // EMC start - DismountReason API
+        if (removalReason != RemovalReason.UNLOADED_TO_CHUNK) {
+            this.getPassengers().forEach(entity -> {
+                entity.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; // EMC - Add this
+                entity.stopRiding();
+            }); // Paper - rewrite chunk system
+        }
+        // EMC end - DismountReason API
         this.levelCallback.onRemove(reason);
         this.onRemoval(reason);
         // Paper start - Folia schedulers
diff --git a/net/minecraft/world/entity/Leashable.java b/net/minecraft/world/entity/Leashable.java
index ff401fa35aab040b319789acb245ac300660399c..9575cb4ae77ba0750819704d32603c658d15fd47 100644
--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -360,6 +360,7 @@ public interface Leashable {
         }
 
         if (entity.isPassenger()) {
+            entity.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.LEASH; // EMC - DismountReason API
             entity.stopRiding();
         }
     }
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 346bfad5b0b0f8bf8ff731fbce9376d06b58f42b..63ba1220171e4130e94cfe3f0f089b26448d36f6 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -464,6 +464,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 }
 
                 if (this.isPassenger() && this.getVehicle() != null && this.getVehicle().dismountsUnderwater()) {
+                    this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.WATER; // EMC - DismountReason API
                     this.stopRiding();
                 }
             } else if (this.getAirSupply() < this.getMaxAirSupply()) {
@@ -4646,6 +4647,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     public void startSleeping(final BlockPos bedPosition) {
         if (this.isPassenger()) {
+            if (this instanceof net.minecraft.world.entity.player.Player) this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.PLAYER; // EMC - DismountReason API
             this.stopRiding();
         }
 
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 57ed2a19849d17213c6af00815dc14e68086ff02..7c59abcbb3828a6abe41dc8a82a049a4ae26412f 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -461,6 +461,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     @Override
     public void rideTick() {
         if (!this.level().isClientSide() && this.wantsToStopRiding() && this.isPassenger()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.PLAYER; // EMC - DismountReason API
             this.stopRiding();
             // CraftBukkit start - SPIGOT-7316: no longer passenger, dismount and return
             if (!this.isPassenger()) {
diff --git a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
index fc29fcaf0b8abd56ef8e50ee6a6d95f7ca9377f8..cfe4abdd835e67935a47b045b00fb97db7b8660a 100644
--- a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
@@ -50,6 +50,7 @@ public record TeleportRandomlyConsumeEffect(float diameter) implements ConsumeEf
             );
             double zz = user.getZ() + (user.getRandom().nextDouble() - 0.5) * this.diameter;
             if (user.isPassenger()) {
+                user.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TELEPORT; // EMC - DismountReason API
                 user.stopRiding();
             }
 
