From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 7 Aug 2013 19:52:14 -0400
Subject: [PATCH] EntityDismountEvent Dismount Reason API


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 547e15e8d175ea0dcd98f4dd0a62bdd5a5b43de3..c3215c2fdd0281c02576d5fa80e04763c871f4d2 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -896,6 +896,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                                         if (!vehicle.isRemoved() && vehicle.hasPassenger(entity)) {
                                             return;
                                         }
+                                        if (vehicle.isRemoved()) { entity.dismountReason =  org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; } // EMC - DismountReason API
 
                                         entity.stopRiding();
                                     }
@@ -1422,6 +1423,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     private void tickPassenger(final Entity vehicle, final Entity entity, final boolean isActive) { // Paper - EAR 2
         if (entity.isRemoved() || entity.getVehicle() != vehicle) {
+            if (entity.isRemoved()) { entity.dismountReason =  org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; } // EMC - DismountReason API
             entity.stopRiding();
         } else if (entity instanceof Player || this.entityTickList.contains(entity)) {
             entity.setOldPosAndRot();
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 36a0191679800700f19c786e930435c6eb500f05..7ec4748d15cbf7f910ff4c10d3e7bbecc0508f87 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1510,6 +1510,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         }
         // CraftBukkit end
         if (!transition.asPassenger()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TELEPORT; // EMC - DismountReason API
             this.removeVehicle();
         }
 
@@ -2103,7 +2104,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
     public void disconnect() {
         this.disconnected = true;
-        this.ejectPassengers();
+        this.ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason.DISCONNECT); // EMC - DismountReason API
 
         // Paper start - Workaround vehicle not tracking the passenger disconnection dismount
         if (this.isPassenger() && this.getVehicle() instanceof ServerPlayer) {
@@ -2325,6 +2326,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         this.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.CHANGE_GAME_MODE, mode.getId()));
         if (mode == GameType.SPECTATOR) {
             this.removeEntitiesOnShoulder();
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.SPECTATE; // EMC - DismountReason API
             this.stopRiding();
             this.stopUsingItem();
             EnchantmentHelper.stopLocationBasedEffects(this);
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 393d8111f971fffef607da5023f62c90e592378e..9effe0b5a8b30d06f5ba23a0963018b3b8acd405 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -469,6 +469,7 @@ public abstract class PlayerList {
             Entity vehicle = player.getRootVehicle();
             if (vehicle.hasExactlyOnePlayerPassenger()) {
                 LOGGER.debug("Removing player mount");
+                player.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DISCONNECT; // EMC - DismountReason API
                 player.stopRiding();
                 vehicle.getPassengersAndSelf().forEach(e -> {
                     // Paper start - Fix villager boat exploit
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 1ee88249e6b1faa8dc6217ba1824e5ae01e85942..d7735a227d44f9ed455740c9f337a9fca0c3ae1a 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -885,6 +885,7 @@ public abstract class Entity
         this.computeSpeed();
         this.inBlockState = null;
         if (this.isPassenger() && this.getVehicle().isRemoved()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; // EMC - DismountReason API
             this.stopRiding();
         }
 
@@ -3374,6 +3375,7 @@ public abstract class Entity
             }
             // CraftBukkit end
             if (this.isPassenger()) {
+                this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TRANSFER; // EMC - DismountReason API
                 this.stopRiding();
             }
 
@@ -3398,7 +3400,13 @@ public abstract class Entity
     }
 
     public void ejectPassengers() {
+        // EMC start - DismountReason API
+        ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN);
+    }
+    public void ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason reason) {
         for (int i = this.passengers.size() - 1; i >= 0; i--) {
+            this.passengers.get(i).dismountReason = reason;
+            // EMC end - DismountReason API
             this.passengers.get(i).stopRiding();
         }
     }
@@ -3457,6 +3465,10 @@ public abstract class Entity
 
     protected boolean removePassenger(final Entity passenger, final boolean suppressCancellation) { // CraftBukkit
         // Paper end - Force entity dismount during teleportation
+        // EMC start - DismountReason API
+        org.bukkit.event.entity.EntityDismountEvent.DismountReason dismountReason = passenger.dismountReason;
+        passenger.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN;
+        // EMC end - DismountReason API
         if (passenger.getVehicle() == this) {
             throw new IllegalStateException("Use x.stopRiding(y), not y.removePassenger(x)");
         }
@@ -3479,7 +3491,7 @@ public abstract class Entity
             }
         }
 
-        org.bukkit.event.entity.EntityDismountEvent event = new org.bukkit.event.entity.EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity(), !suppressCancellation); // Paper - Force entity dismount during teleportation
+        org.bukkit.event.entity.EntityDismountEvent event = new org.bukkit.event.entity.EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity(), !suppressCancellation, dismountReason); // Paper - Force entity dismount during teleportation // EMC - DismountReasonAPI - Add dismountReason
         // Suppress during worldgen
         if (this.valid) {
             org.bukkit.Bukkit.getPluginManager().callEvent(event);
@@ -4054,6 +4066,7 @@ public abstract class Entity
             );
     }
 
+    public org.bukkit.event.entity.EntityDismountEvent.DismountReason dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN; // EMC - DismountReason API
     public final boolean isInvulnerableToBase(final DamageSource source) {
         return this.isRemoved()
             || this.invulnerable && !source.is(DamageTypeTags.BYPASSES_INVULNERABILITY) && !source.isCreativePlayer()
@@ -5214,10 +5227,18 @@ public abstract class Entity
         }
 
         if (this.removalReason.shouldDestroy()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD; // EMC - DismountReason API
             this.stopRiding();
         }
 
-        if (this.removalReason != Entity.RemovalReason.UNLOADED_TO_CHUNK) { this.getPassengers().forEach(Entity::stopRiding); } // Paper - rewrite chunk system
+        // EMC start - DismountReason API
+        if (removalReason != RemovalReason.UNLOADED_TO_CHUNK) {
+            this.getPassengers().forEach(entity -> {
+                entity.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; // EMC - Add this
+                entity.stopRiding();
+            }); // Paper - rewrite chunk system
+        }
+        // EMC end - DismountReason API
         this.levelCallback.onRemove(reason);
         this.onRemoval(reason);
         // Paper start - Folia schedulers
diff --git a/net/minecraft/world/entity/Leashable.java b/net/minecraft/world/entity/Leashable.java
index 98034a4c96bf2972316078d3b3a49140921aab50..02b60d5c72ba4b46f4fbd6dd7b4f7a997f6cf07c 100644
--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -357,6 +357,7 @@ public interface Leashable {
         }
 
         if (entity.isPassenger()) {
+            entity.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.LEASH; // EMC - DismountReason API
             entity.stopRiding();
         }
     }
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index a47d15b75355bcd9fbf03b06eb6d7677fdc084e6..faa01c2f771715c3f96a4d6d351a7d2769bd2bcf 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -483,6 +483,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 }
 
                 if (this.isPassenger() && this.getVehicle() != null && this.getVehicle().dismountsUnderwater()) {
+                    this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.WATER; // EMC - DismountReason API
                     this.stopRiding();
                 }
             } else if (this.getAirSupply() < this.getMaxAirSupply()) {
@@ -4663,6 +4664,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     public void startSleeping(final BlockPos bedPosition) {
         if (this.isPassenger()) {
+            if (this instanceof net.minecraft.world.entity.player.Player) this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.PLAYER; // EMC - DismountReason API
             this.stopRiding();
         }
 
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index b033b3449809bc26675960b629c23f2eb789656d..4398d46e2b99cc83f26b0aad7154f6bb9a1b65ad 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -465,6 +465,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     @Override
     public void rideTick() {
         if (!this.level().isClientSide() && this.wantsToStopRiding() && this.isPassenger()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.PLAYER; // EMC - DismountReason API
             this.stopRiding();
             // CraftBukkit start - SPIGOT-7316: no longer passenger, dismount and return
             if (!this.isPassenger()) {
diff --git a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
index 23ba133fea6e44dfd31e564d9544ec43800d426b..9b0eb615015e302ae1c7c40bf2ca7dd08ba7f172 100644
--- a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
@@ -50,6 +50,7 @@ public record TeleportRandomlyConsumeEffect(float diameter) implements ConsumeEf
             );
             double zz = user.getZ() + (user.getRandom().nextDouble() - 0.5) * this.diameter;
             if (user.isPassenger()) {
+                user.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TELEPORT; // EMC - DismountReason API
                 user.stopRiding();
             }
 
