From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 7 Aug 2013 19:52:14 -0400
Subject: [PATCH] EntityDismountEvent Dismount Reason API


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index bab17d21b74f5e4eea478002c5d20a6b7d15fe41..80d3d8cc89d03a2835d0e34a6d6c89beaf8c0744 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -803,6 +803,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
                                         if (!vehicle.isRemoved() && vehicle.hasPassenger(entity)) {
                                             return;
                                         }
+                                        if (vehicle.isRemoved()) { entity.dismountReason =  org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; } // EMC - DismountReason API
 
                                         entity.stopRiding();
                                     }
@@ -1294,6 +1295,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
 
     private void tickPassenger(Entity ridingEntity, Entity passengerEntity, final boolean isActive) { // Paper - EAR 2
         if (passengerEntity.isRemoved() || passengerEntity.getVehicle() != ridingEntity) {
+            if (passengerEntity.isRemoved()) { passengerEntity.dismountReason =  org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; } // EMC - DismountReason API
             passengerEntity.stopRiding();
         } else if (passengerEntity instanceof Player || this.entityTickList.contains(passengerEntity)) {
             passengerEntity.setOldPosAndRot();
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index cef4cd4baf87521fb2b55dc1d35f85ae4729be63..ea8774ff05c488bcbc2ffe6d255161096ed55227 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1395,6 +1395,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             }
             // CraftBukkit end
             if (!teleportTransition.asPassenger()) {
+                this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TELEPORT; // EMC - DismountReason API
                 this.removeVehicle();
             }
 
@@ -1970,10 +1971,11 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
     public void disconnect() {
         this.disconnected = true;
-        this.ejectPassengers();
+        this.ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason.DISCONNECT); // EMC - DismountReason API
 
         // Paper start - Workaround vehicle not tracking the passenger disconnection dismount
         if (this.isPassenger() && this.getVehicle() instanceof ServerPlayer) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DISCONNECT; // EMC - DismountReason API
             this.stopRiding();
         }
         // Paper end - Workaround vehicle not tracking the passenger disconnection dismount
@@ -2171,6 +2173,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             this.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.CHANGE_GAME_MODE, gameMode.getId()));
             if (gameMode == GameType.SPECTATOR) {
                 this.removeEntitiesOnShoulder();
+                this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.SPECTATE; // EMC - DismountReason API
                 this.stopRiding();
                 EnchantmentHelper.stopLocationBasedEffects(this);
             } else {
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index e1fd699b33832c0f01b14023db7cc2b9d6797227..bce32c3816db55a15e20a52cf47b5eafaec07601 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -545,6 +545,7 @@ public abstract class PlayerList {
             Entity rootVehicle = player.getRootVehicle();
             if (rootVehicle.hasExactlyOnePlayerPassenger()) {
                 LOGGER.debug("Removing player mount");
+                player.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DISCONNECT; // EMC - DismountReason API
                 player.stopRiding();
                 rootVehicle.getPassengersAndSelf().forEach(entity -> {
                     // Paper start - Fix villager boat exploit
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 83e920c3eca98e0c771cee3379e906149281fb0e..55dc937a20a88f8229153fc55d07452c676b20d3 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -849,6 +849,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         if (firstTick && this instanceof net.minecraft.world.entity.NeutralMob neutralMob) neutralMob.tickInitialPersistentAnger(level); // Paper - Prevent entity loading causing async lookups
         this.inBlockState = null;
         if (this.isPassenger() && this.getVehicle().isRemoved()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; // EMC - DismountReason API
             this.stopRiding();
         }
 
@@ -3220,6 +3221,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 }
                 // CraftBukkit end
                 if (this.isPassenger()) {
+                    this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TRANSFER; // EMC - DismountReason API
                     this.stopRiding();
                 }
 
@@ -3241,7 +3243,13 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public void ejectPassengers() {
+        // EMC start - DismountReason API
+        ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN);
+    }
+    public void ejectPassengers(org.bukkit.event.entity.EntityDismountEvent.DismountReason reason) {
         for (int i = this.passengers.size() - 1; i >= 0; i--) {
+            this.passengers.get(i).dismountReason = reason;
+            // EMC end - DismountReason API
             this.passengers.get(i).stopRiding();
         }
     }
@@ -3296,6 +3304,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
     protected boolean removePassenger(Entity passenger, boolean suppressCancellation) { // CraftBukkit
         // Paper end - Force entity dismount during teleportation
+        // EMC start - DismountReason API
+        org.bukkit.event.entity.EntityDismountEvent.DismountReason dismountReason = passenger.dismountReason;
+        passenger.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN;
+        // EMC end - DismountReason API
         if (passenger.getVehicle() == this) {
             throw new IllegalStateException("Use x.stopRiding(y), not y.removePassenger(x)");
         } else {
@@ -3318,7 +3330,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 }
             }
 
-            org.bukkit.event.entity.EntityDismountEvent event = new org.bukkit.event.entity.EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity(), !suppressCancellation); // Paper - Force entity dismount during teleportation
+            org.bukkit.event.entity.EntityDismountEvent event = new org.bukkit.event.entity.EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity(), !suppressCancellation, dismountReason); // Paper - Force entity dismount during teleportation // EMC - DismountReasonAPI - Add dismountReason
             // Suppress during worldgen
             if (this.valid) {
                 org.bukkit.Bukkit.getPluginManager().callEvent(event);
@@ -3878,6 +3890,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             );
     }
 
+    public org.bukkit.event.entity.EntityDismountEvent.DismountReason dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.UNKNOWN; // EMC - DismountReason API
     public final boolean isInvulnerableToBase(DamageSource damageSource) {
         return this.isRemoved()
             || this.invulnerable && !damageSource.is(DamageTypeTags.BYPASSES_INVULNERABILITY) && !damageSource.isCreativePlayer()
@@ -5145,10 +5158,19 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
 
         if (this.removalReason.shouldDestroy()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD; // EMC - DismountReason API
             this.stopRiding();
         }
 
         if (this.removalReason != Entity.RemovalReason.UNLOADED_TO_CHUNK) { this.getPassengers().forEach(Entity::stopRiding); } // Paper - rewrite chunk system
+        // EMC start - DismountReason API
+        if (removalReason != RemovalReason.UNLOADED_TO_CHUNK) {
+            this.getPassengers().forEach(entity -> {
+                entity.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.DEAD_VEHICLE; // EMC - Add this
+                entity.stopRiding();
+            }); // Paper - chunk system - don't adjust passenger state when unloading, it's just not safe (and messes with our logic in entity chunk unload)
+        }
+        // EMC end - DismountReason API
         this.levelCallback.onRemove(removalReason);
         this.onRemoval(removalReason);
         // Paper start - Folia schedulers
diff --git a/net/minecraft/world/entity/Leashable.java b/net/minecraft/world/entity/Leashable.java
index b3568d4a1467912f555ce461fa8dcf218c474047..9ea4f8c6fe82f5de7d9a35db63623d26804a7342 100644
--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -355,6 +355,7 @@ public interface Leashable {
         }
 
         if (entity.isPassenger()) {
+            entity.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.LEASH; // EMC - DismountReason API
             entity.stopRiding();
         }
     }
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 516d4f4e5ee0f4bb01968225db0102e8c91a289c..21970c0bd26a25ce3ae50c64f8b75d626d8b1f85 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -467,6 +467,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 }
 
                 if (this.isPassenger() && this.getVehicle() != null && this.getVehicle().dismountsUnderwater()) {
+                    this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.WATER; // EMC - DismountReason API
                     this.stopRiding();
                 }
             } else if (this.getAirSupply() < this.getMaxAirSupply()) {
@@ -4402,6 +4403,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     public void startSleeping(BlockPos pos) {
         if (this.isPassenger()) {
+            if (this instanceof net.minecraft.world.entity.player.Player) this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.PLAYER; // EMC - DismountReason API
             this.stopRiding();
         }
 
diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index c2acafecb2efaf2700de9d4285ce12f11cddd497..9dd7610c30ef33fa55aeb08068b2df2e51a83b1a 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -579,6 +579,7 @@ public abstract class Player extends LivingEntity {
     @Override
     public void rideTick() {
         if (!this.level().isClientSide && this.wantsToStopRiding() && this.isPassenger()) {
+            this.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.PLAYER; // EMC - DismountReason API
             this.stopRiding();
             // CraftBukkit start - SPIGOT-7316: no longer passenger, dismount and return
             if (!this.isPassenger()) {
diff --git a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
index 56dac4f7c4a1e8fe830bb3da807d2ea06bf2fb7c..a38b4cf170cd2518ca3db9bf137c7484db5ac9a2 100644
--- a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
@@ -51,6 +51,7 @@ public record TeleportRandomlyConsumeEffect(float diameter) implements ConsumeEf
             );
             double d2 = entity.getZ() + (entity.getRandom().nextDouble() - 0.5) * this.diameter;
             if (entity.isPassenger()) {
+                entity.dismountReason = org.bukkit.event.entity.EntityDismountEvent.DismountReason.TELEPORT; // EMC - DismountReason API
                 entity.stopRiding();
             }
 
