From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 18 Dec 2014 22:48:46 -0500
Subject: [PATCH] Add BlockBreakNaturally Event

To give reliable control over all blocks dropping to world to restore custom item meta

diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index cec1d55b2816918091ff478be79fa35ec7b17b05..8d98d66725102a7af6f48db48a32439e2c052c80 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1636,6 +1636,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             if (entity instanceof net.minecraft.world.entity.item.ItemEntity item && item.getItem().isEmpty()) return false; // Paper - Prevent empty items from being added
             // Paper start - capture all item additions to the world
             if (captureDrops != null && entity instanceof net.minecraft.world.entity.item.ItemEntity item) {
+                item.droppedPosition = new BlockPos(entity.getBlockX(), entity.getBlockY(), entity.getBlockZ()); // EMC - BlockBreakNaturallyEvent
                 captureDrops.add(item);
                 return true;
             }
diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index 133bda847d538cd83fae06f55bb33badc9c4b4b5..d0eddb9f61a1840c2492de4d6cda3cc14d49871a 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -47,6 +47,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
     private static final short DEFAULT_PICKUP_DELAY = 0;
     public int age = 0;
     public boolean canDespawn = true; // EMC - Item canDespawn API
+    public BlockPos droppedPosition; // EMC - BlockBreakNaturallyEvent
     public int pickupDelay = 0;
     public int health = 5;
     public @Nullable EntityReference<Entity> thrower;
diff --git a/net/minecraft/world/level/block/Block.java b/net/minecraft/world/level/block/Block.java
index 41aab9a526be77b400bed7a74614eceec95345b5..b894818014b9b5af6eef0261a66a907dfd9a0693 100644
--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -461,7 +461,7 @@ public class Block extends BlockBehaviour implements ItemLike {
         double x = pos.getX() + 0.5 + Mth.nextDouble(random, -0.25, 0.25);
         double y = pos.getY() + 0.5 + Mth.nextDouble(random, -0.25, 0.25) - halfHeight;
         double z = pos.getZ() + 0.5 + Mth.nextDouble(random, -0.25, 0.25);
-        popResource(level, () -> new ItemEntity(level, x, y, z, itemStack), itemStack);
+        popResource(level, () -> new ItemEntity(level, x, y, z, itemStack), itemStack, pos); // EMC - BlockBreakNaturallyEvent
     }
 
     public static void popResourceFromFace(final Level level, final BlockPos pos, final Direction face, final ItemStack itemStack) {
@@ -477,17 +477,19 @@ public class Block extends BlockBehaviour implements ItemLike {
         double deltaX = stepX == 0 ? Mth.nextDouble(random, -0.1, 0.1) : stepX * 0.1;
         double deltaY = stepY == 0 ? Mth.nextDouble(random, 0.0, 0.1) : stepY * 0.1 + 0.1;
         double deltaZ = stepZ == 0 ? Mth.nextDouble(random, -0.1, 0.1) : stepZ * 0.1;
-        popResource(level, () -> new ItemEntity(level, x, y, z, itemStack, deltaX, deltaY, deltaZ), itemStack);
+        popResource(level, () -> new ItemEntity(level, x, y, z, itemStack, deltaX, deltaY, deltaZ), itemStack, pos); // EMC - BlockBreakNaturallyEvent
     }
 
-    private static void popResource(final Level level, final Supplier<ItemEntity> entityFactory, final ItemStack itemStack) {
+    private static void popResource(final Level level, final Supplier<ItemEntity> entityFactory, final ItemStack itemStack, BlockPos blockPos) { // EMC - BlockBreakNaturallyEvent
         if (level instanceof ServerLevel serverLevel && !itemStack.isEmpty() && serverLevel.getGameRules().get(GameRules.BLOCK_DROPS)) {
             ItemEntity entity = entityFactory.get();
             entity.setDefaultPickUpDelay();
             // CraftBukkit start
             if (level.captureDrops != null) {
+                entity.droppedPosition = blockPos.immutable(); // EMC - BlockBreakNaturallyEvent
                 level.captureDrops.add(entity);
             } else {
+                new com.empireminecraft.empirecraft.event.block.BlockBreakNaturallyEvent(org.bukkit.craftbukkit.util.CraftLocation.toBukkit(blockPos, level), (org.bukkit.entity.Item) entity.getBukkitEntity()).callEvent(); // EMC - BlockBreakNaturallyEvent
                 level.addFreshEntity(entity);
             }
             // CraftBukkit end
