From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 18 Dec 2014 22:48:46 -0500
Subject: [PATCH] Add BlockBreakNaturally Event

To give reliable control over all blocks dropping to world to restore custom item meta

diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index fce01fd6c4d19f6579f7ea023eaa747a3eb6c4f2..740db97667bb4ff97b6d79509c1f265b99343961 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1619,6 +1619,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
             if (entity instanceof net.minecraft.world.entity.item.ItemEntity itemEntity && itemEntity.getItem().isEmpty()) return false; // Paper - Prevent empty items from being added
             // Paper start - capture all item additions to the world
             if (captureDrops != null && entity instanceof net.minecraft.world.entity.item.ItemEntity) {
+                ((net.minecraft.world.entity.item.ItemEntity) entity).droppedPosition = new BlockPos(entity.getBlockX(), entity.getBlockY(), entity.getBlockZ()); // EMC - BlockBreakNaturallyEvent
                 captureDrops.add((net.minecraft.world.entity.item.ItemEntity) entity);
                 return true;
             }
diff --git a/net/minecraft/world/entity/item/ItemEntity.java b/net/minecraft/world/entity/item/ItemEntity.java
index d406311acdfce6527d7025cc2bed9b8964fdd8aa..5054f74e69094e7b9da763c9d6f207619e6a48cb 100644
--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -47,6 +47,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
     private static final short DEFAULT_PICKUP_DELAY = 0;
     public int age = 0;
     public boolean canDespawn = true; // EMC - Item canDespawn API
+    public BlockPos droppedPosition; // EMC - BlockBreakNaturallyEvent
     public int pickupDelay = 0;
     public int health = 5;
     public @Nullable EntityReference<Entity> thrower;
diff --git a/net/minecraft/world/level/block/Block.java b/net/minecraft/world/level/block/Block.java
index ab08211e0f64a14a335e3cc4479bcf1f82e25cdd..441c569b7fb97ff6def51f788bf03c5eacd86a34 100644
--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -435,7 +435,7 @@ public class Block extends BlockBehaviour implements ItemLike {
         double d1 = pos.getX() + 0.5 + Mth.nextDouble(level.random, -0.25, 0.25);
         double d2 = pos.getY() + 0.5 + Mth.nextDouble(level.random, -0.25, 0.25) - d;
         double d3 = pos.getZ() + 0.5 + Mth.nextDouble(level.random, -0.25, 0.25);
-        popResource(level, () -> new ItemEntity(level, d1, d2, d3, stack), stack);
+        popResource(level, () -> new ItemEntity(level, d1, d2, d3, stack), stack, pos); // EMC - BlockBreakNaturallyEvent
     }
 
     public static void popResourceFromFace(Level level, BlockPos pos, Direction direction, ItemStack stack) {
@@ -450,17 +450,19 @@ public class Block extends BlockBehaviour implements ItemLike {
         double d5 = stepX == 0 ? Mth.nextDouble(level.random, -0.1, 0.1) : stepX * 0.1;
         double d6 = stepY == 0 ? Mth.nextDouble(level.random, 0.0, 0.1) : stepY * 0.1 + 0.1;
         double d7 = stepZ == 0 ? Mth.nextDouble(level.random, -0.1, 0.1) : stepZ * 0.1;
-        popResource(level, () -> new ItemEntity(level, d2, d3, d4, stack, d5, d6, d7), stack);
+        popResource(level, () -> new ItemEntity(level, d2, d3, d4, stack, d5, d6, d7), stack, pos); // EMC - BlockBreakNaturallyEvent
     }
 
-    private static void popResource(Level level, Supplier<ItemEntity> itemEntitySupplier, ItemStack stack) {
+    private static void popResource(Level level, Supplier<ItemEntity> itemEntitySupplier, ItemStack stack, BlockPos blockPos) { // EMC - BlockBreakNaturallyEvent
         if (level instanceof ServerLevel serverLevel && !stack.isEmpty() && serverLevel.getGameRules().get(GameRules.BLOCK_DROPS)) {
             ItemEntity itemEntity = itemEntitySupplier.get();
             itemEntity.setDefaultPickUpDelay();
             // CraftBukkit start
             if (level.captureDrops != null) {
+                itemEntity.droppedPosition = blockPos.immutable(); // EMC - BlockBreakNaturallyEvent
                 level.captureDrops.add(itemEntity);
             } else {
+                new com.empireminecraft.empirecraft.event.block.BlockBreakNaturallyEvent(org.bukkit.craftbukkit.util.CraftLocation.toBukkit(blockPos, level), (org.bukkit.entity.Item) itemEntity.getBukkitEntity()).callEvent(); // EMC - BlockBreakNaturallyEvent
                 level.addFreshEntity(itemEntity);
             }
             // CraftBukkit end
