From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 20 Jun 2018 22:35:39 -0400
Subject: [PATCH] Tweak Explosions


diff --git a/net/minecraft/world/level/ServerExplosion.java b/net/minecraft/world/level/ServerExplosion.java
index 0ef64fcf11d6cd0e25227fe6053ca8cb2125bfff..4340dbb2b18888f09ab2eb7e6151d6948a879ef2 100644
--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -396,6 +396,7 @@ public class ServerExplosion implements Explosion {
         // calculations in one go
         // additional aggressive caching of block retrieval is very significant, as at low power (i.e tnt) most
         // block retrievals are not unique
+        if (this.fire || this.blockInteraction != Explosion.BlockInteraction.KEEP) { // EMC - don't run block processing if keep flag is set
         for (int ray = 0, len = CACHED_RAYS.length; ray < len;) {
             ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache cachedBlock = initialCache;
 
@@ -464,6 +465,7 @@ public class ServerExplosion implements Explosion {
                 currZ += incZ;
             } while (power > 0.0f);
         }
+        } // EMC
 
         return ret;
         // Paper end - collision optimisations
@@ -478,7 +480,7 @@ public class ServerExplosion implements Explosion {
             int y1 = Mth.floor(this.center.y + doubleRadius + 1.0);
             int z0 = Mth.floor(this.center.z - doubleRadius - 1.0);
             int z1 = Mth.floor(this.center.z + doubleRadius + 1.0);
-            List<Entity> list = this.level.getEntities(this.excludeSourceFromDamage ? this.source : null, new AABB(x0, y0, z0, x1, y1, z1), entity -> entity.isAlive() && !entity.isSpectator()); // Paper - Fix lag from explosions processing dead entities, Allow explosions to damage source
+            List<Entity> list = this.level.getEntities(this.excludeSourceFromDamage ? this.source : null, new AABB(x0, y0, z0, x1, y1, z1), entity -> entity.isAlive() && !entity.isSpectator() && !(entity instanceof ItemEntity)); // Paper - Fix lag from explosions processing dead entities, Allow explosions to damage source // EMC - exclude item entities
 
             for (Entity entity : list) { // Paper - used in loop
                 if (!entity.ignoreExplosion(this)) {
