From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 25 Jun 2018 22:55:42 -0400
Subject: [PATCH] EntityPrepForRangedAttack


diff --git a/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
index c59c05c666ec97e8be12c26c53a33cb53d9c34f5..802b78d8661c8beffd66a75c5de7302cde80b0d2 100644
--- a/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
@@ -80,9 +80,18 @@ public class RangedAttackGoal extends Goal {
             this.seeTime = 0;
         }
 
-        if (!(targetDistSqr > this.attackRadiusSqr) && this.seeTime >= 5) {
+        // EMC start - EntityPrepForRangedAttack
+        //if (!(targetDistSqr > this.attackRadiusSqr) && this.seeTime >= 5) {
+        boolean isInRange = targetDistSqr <= (double) this.attackRadiusSqr;
+        boolean readyforAttack = this.seeTime >= 5;
+        com.empireminecraft.empirecraft.event.entity.EntityPrepForRangedAttack event = new com.empireminecraft.empirecraft.event.entity.EntityPrepForRangedAttack((com.destroystokyo.paper.entity.RangedEntity) mob.getBukkitEntity(), this.target.getBukkitLivingEntity(), readyforAttack, !isInRange, this.seeTime);
+        if (!event.callEvent()) {
+            return;
+        }
+        if (event.shouldStop() && !event.shouldMoveNear()) {
             this.mob.getNavigation().stop();
-        } else {
+        } else if (event.shouldMoveNear()) {
+            // EMC end - EntityPrepForRangedAttack
             this.mob.getNavigation().moveTo(this.target, this.speedModifier);
         }
 
diff --git a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
index 13a815c753f01088af53ba63c07ee9293ca0b440..f8c3d025e68ece915602f5437ec0c51d39ba5acf 100644
--- a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -83,10 +83,19 @@ public class RangedBowAttackGoal<T extends Monster & RangedAttackMob> extends Go
                 this.seeTime--;
             }
 
-            if (!(targetDistSqr > this.attackRadiusSqr) && this.seeTime >= 20) {
+            // EMC start - EntityPrepForRangedAttack
+            //if (!(targetDistSqr > this.attackRadiusSqr) && this.seeTime >= 20) {
+            boolean isInRange = targetDistSqr <= (double) this.attackRadiusSqr;
+            boolean readyforAttack = this.seeTime >= 20;
+            com.empireminecraft.empirecraft.event.entity.EntityPrepForRangedAttack event = new com.empireminecraft.empirecraft.event.entity.EntityPrepForRangedAttack((com.destroystokyo.paper.entity.RangedEntity) mob.getBukkitEntity(), target.getBukkitLivingEntity(), readyforAttack, !isInRange, this.seeTime);
+            if (!event.callEvent()) {
+                    return;
+                }
+            if (event.shouldStop() && !event.shouldMoveNear()) {
                 this.mob.getNavigation().stop();
                 this.strafingTime++;
-            } else {
+            } else if (event.shouldMoveNear()){
+                // EMC end - EntityPrepForRangedAttack
                 this.mob.getNavigation().moveTo(target, this.speedModifier);
                 this.strafingTime = -1;
             }
