From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 25 Jun 2018 22:55:42 -0400
Subject: [PATCH] EntityPrepForRangedAttack


diff --git a/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
index d300ba3c540d9006ddfcc5eee99af28d2070ccca..f1fb5eeae6d4d44fb9e9f4f4c407a6353504b245 100644
--- a/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
@@ -78,9 +78,17 @@ public class RangedAttackGoal extends Goal {
             this.seeTime = 0;
         }
 
-        if (!(d > this.attackRadiusSqr) && this.seeTime >= 5) {
+        // EMC start - EntityPrepForRangedAttack
+        boolean isInRange = d <= (double) this.attackRadiusSqr;
+        boolean readyforAttack = this.seeTime >= 5;
+        com.empireminecraft.empirecraft.event.entity.EntityPrepForRangedAttack event = new com.empireminecraft.empirecraft.event.entity.EntityPrepForRangedAttack((com.destroystokyo.paper.entity.RangedEntity) mob.getBukkitEntity(), this.target.getBukkitLivingEntity(), readyforAttack, !isInRange, this.seeTime);
+        if (!event.callEvent()) {
+            return;
+        }
+        if (event.shouldStop() && !event.shouldMoveNear()) {
             this.mob.getNavigation().stop();
-        } else {
+        } else if (event.shouldMoveNear()) {
+            // EMC end - EntityPrepForRangedAttack
             this.mob.getNavigation().moveTo(this.target, this.speedModifier);
         }
 
diff --git a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
index b20fef130dfbfe92aaf64ad467223dce3e9962b5..18155322bcaa04af3dbee72236e4834955598a8c 100644
--- a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -83,10 +83,18 @@ public class RangedBowAttackGoal<T extends Monster & RangedAttackMob> extends Go
                 this.seeTime--;
             }
 
-            if (!(d > this.attackRadiusSqr) && this.seeTime >= 20) {
+            // EMC start - EntityPrepForRangedAttack
+            boolean isInRange = d <= (double) this.attackRadiusSqr;
+            boolean readyforAttack = this.seeTime >= 20;
+            com.empireminecraft.empirecraft.event.entity.EntityPrepForRangedAttack event = new com.empireminecraft.empirecraft.event.entity.EntityPrepForRangedAttack((com.destroystokyo.paper.entity.RangedEntity) mob.getBukkitEntity(), target.getBukkitLivingEntity(), readyforAttack, !isInRange, this.seeTime);
+            if (!event.callEvent()) {
+                    return;
+                }
+            if (event.shouldStop() && !event.shouldMoveNear()) {
                 this.mob.getNavigation().stop();
                 this.strafingTime++;
-            } else {
+            } else if (event.shouldMoveNear()){
+                // EMC end - EntityPrepForRangedAttack
                 this.mob.getNavigation().moveTo(target, this.speedModifier);
                 this.strafingTime = -1;
             }
