From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 6 May 2015 02:29:30 -0400
Subject: [PATCH] Limit worlds structures can generate


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index bc251c4736870e8b3c3e34294188b09b3fbe80b0..b4ce777b61e98eef768869c171b125ee1a3ac61c 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2258,6 +2258,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     public @Nullable BlockPos findNearestMapStructure(
         final TagKey<Structure> structureTag, final BlockPos origin, final int maxSearchRadius, final boolean createReference
     ) {
+        if ("town".equals(getWorld().getName())) { return null; } // EMC
         Optional<HolderSet.Named<Structure>> tag = this.registryAccess().lookupOrThrow(Registries.STRUCTURE).get(structureTag);
         if (tag.isEmpty()) {
             return null;
diff --git a/net/minecraft/world/level/StructureManager.java b/net/minecraft/world/level/StructureManager.java
index a8f00da4d5c7cca9b6948d7c2d5114f84fdacc00..337907895a5df87f1d7fc9de70790fd347943c60 100644
--- a/net/minecraft/world/level/StructureManager.java
+++ b/net/minecraft/world/level/StructureManager.java
@@ -99,7 +99,7 @@ public class StructureManager {
     }
 
     public boolean shouldGenerateStructures() {
-        return this.worldOptions.generateStructures();
+        return this.worldOptions.generateStructures() && !"town".equals(this.level.getMinecraftWorld().getWorld().getName()); // EMC
     }
 
     public StructureStart getStructureAt(final BlockPos blockPos, final Structure structure) {
