From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 6 May 2015 02:29:30 -0400
Subject: [PATCH] Limit worlds structures can generate


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 9a38ae2ee13d6e394782aca593650c0a77ebbad1..f121342f2551f3b9dec02b24bd398c3a96f81384 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2065,6 +2065,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         if (!this.serverLevelData.worldGenOptions().generateStructures()) { // CraftBukkit
             return null;
         } else {
+            if ("town".equals(getWorld().getName())) { return null; } // EMC
             Optional<HolderSet.Named<Structure>> optional = this.registryAccess().lookupOrThrow(Registries.STRUCTURE).get(structureTag);
             if (optional.isEmpty()) {
                 return null;
diff --git a/net/minecraft/world/level/StructureManager.java b/net/minecraft/world/level/StructureManager.java
index 8bc6a6c86cd8db53feefba7508b6031ba67e242e..1dd2dd189f3fc3af6a2832fb449c8e389522d921 100644
--- a/net/minecraft/world/level/StructureManager.java
+++ b/net/minecraft/world/level/StructureManager.java
@@ -102,7 +102,7 @@ public class StructureManager {
     }
 
     public boolean shouldGenerateStructures() {
-        return this.worldOptions.generateStructures();
+        return this.worldOptions.generateStructures() && !"town".equals(this.level.getMinecraftWorld().getWorld().getName()); // EMC
     }
 
     public StructureStart getStructureAt(BlockPos pos, Structure structure) {
