From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 6 May 2015 02:29:30 -0400
Subject: [PATCH] Limit worlds structures can generate


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 04c01363e8cb97ce265a67c9e0f922f89ff187c3..cfb5074e15fb0d16fb33e0af95978c4fd774ce0f 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2102,6 +2102,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         if (!this.serverLevelData.worldGenOptions().generateStructures()) { // CraftBukkit
             return null;
         } else {
+            if ("town".equals(getWorld().getName())) { return null; } // EMC
             Optional<HolderSet.Named<Structure>> optional = this.registryAccess().lookupOrThrow(Registries.STRUCTURE).get(structureTag);
             if (optional.isEmpty()) {
                 return null;
diff --git a/net/minecraft/world/level/StructureManager.java b/net/minecraft/world/level/StructureManager.java
index 3399922d79a713484e16beb6e4e9985c284ddfb5..746942605ff98c5931c500e670a83bc043334136 100644
--- a/net/minecraft/world/level/StructureManager.java
+++ b/net/minecraft/world/level/StructureManager.java
@@ -102,7 +102,7 @@ public class StructureManager {
     }
 
     public boolean shouldGenerateStructures() {
-        return this.worldOptions.generateStructures();
+        return this.worldOptions.generateStructures() && !"town".equals(this.level.getMinecraftWorld().getWorld().getName()); // EMC
     }
 
     public StructureStart getStructureAt(BlockPos pos, Structure structure) {
