From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 29 Jul 2021 02:07:15 -0500
Subject: [PATCH] Capture item drops during block placement


diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 1d08ac66271a441811a157ed6619b14bf58c403f..da2d8035c20c18a7748ad99bcf9e1a0cefc3f524 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -395,10 +395,18 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
         } finally {
             level.captureBlockStates = false;
         }
+            // EMC start - Capture item Drops during placement
+            level.captureDrops = new java.util.ArrayList<>();
+            org.bukkit.block.Block bblock = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos);
+            // EMC end - Capture item Drops during placement
         DataComponentPatch newPatch = this.components.asPatch();
         int newCount = this.getCount();
         this.setCount(previousCount);
         this.restorePatch(previousPatch);
+            // EMC start - Capture item Drops during placement
+            List<ItemEntity> itemsToDrop = level.captureDrops;
+            level.captureDrops = null;
+            // EMC end - Capture item Drops during placement
         if (result.consumesAction() && level.captureTreeGeneration && !level.capturedBlockStates.isEmpty()) {
             level.captureTreeGeneration = false;
             org.bukkit.Location location = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level);
@@ -417,6 +425,7 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
             org.bukkit.Bukkit.getPluginManager().callEvent(fertilizeEvent);
 
             if (!fertilizeEvent.isCancelled()) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDropItemEvent(bblock, bblock.getState(), (ServerPlayer) player, itemsToDrop); // EMC
                 // Change the stack to its new contents if it hasn't been tampered with.
                 if (this.getCount() == previousCount && Objects.equals(this.components.asPatch(), previousPatch)) {
                     this.restorePatch(newPatch);
@@ -457,6 +466,7 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
 
                 SignItem.openSign = null; // SPIGOT-6758 - Reset on early return
             } else {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDropItemEvent(bblock, bblock.getState(), (ServerPlayer) player, itemsToDrop); // EMC
                 // Change the stack to its new contents if it hasn't been tampered with.
                 if (this.getCount() == previousCount && Objects.equals(this.components.asPatch(), previousPatch)) {
                     this.restorePatch(newPatch);
