From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 6 Sep 2021 02:05:16 -0500
Subject: [PATCH] Mitigate ConcurrentModificationException by copying
 entitySection access


diff --git a/net/minecraft/world/level/entity/PersistentEntitySectionManager.java b/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
index 8b5e5757619a36fa3e4720c970a0e41561fa1a75..29d40e0feda011b88dd1472923c939ed128ebae9 100644
--- a/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
@@ -181,17 +181,17 @@ public class PersistentEntitySectionManager<T extends EntityAccess> implements A
             boolean wasTicking = previousStatus.isTicking();
             boolean isTicking = chunkStatus.isTicking();
             if (wasTicking && !isTicking) {
-                section.getEntities().filter(e -> !e.isAlwaysTicking()).forEach(this::stopTicking);
+                section.getEntities().filter(e -> !e.isAlwaysTicking()).toList().forEach(this::stopTicking); // EMC
             }
 
             if (wasAccessible && !isAccessible) {
-                section.getEntities().filter(e -> !e.isAlwaysTicking()).forEach(this::stopTracking);
+                section.getEntities().filter(e -> !e.isAlwaysTicking()).toList().forEach(this::stopTracking); // EMC
             } else if (!wasAccessible && isAccessible) {
-                section.getEntities().filter(e -> !e.isAlwaysTicking()).forEach(this::startTracking);
+                section.getEntities().filter(e -> !e.isAlwaysTicking()).toList().forEach(this::startTracking); // EMC
             }
 
             if (!wasTicking && isTicking) {
-                section.getEntities().filter(e -> !e.isAlwaysTicking()).forEach(this::startTicking);
+                section.getEntities().filter(e -> !e.isAlwaysTicking()).toList().forEach(this::startTicking); // EMC
             }
         });
     }
