From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 11 Sep 2024 21:34:46 -0500
Subject: [PATCH] Add a data converter to migrate from Empirecraft metadata


diff --git a/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java b/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
index c48181291e18bbec351f267ac35ce97718aec24c..eb5049fb1f3c3116033e719f80ce538a214b4f82 100644
--- a/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
+++ b/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
@@ -243,6 +243,7 @@ public final class MCVersionRegistry {
             3939,
             3943,
             3945,
+            3952, // EMC
             4054,
             4055,
             4057,
diff --git a/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java b/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
index 631782c6a1805db6f8a02eb4c42fdf8a7b8ce731..e72ce7ee4708a16b4a760144041d11e9957a4e48 100644
--- a/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
+++ b/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
@@ -319,6 +319,7 @@ public final class MCTypeRegistry {
         V3939.register();
         V3943.register();
         V3945.register();
+        com.empireminecraft.empirecraft.util.EmpireMetadataConverter.register(); // EMC
         // V1.21.2
         V4054.register();
         V4055.register();
