From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 11 Sep 2024 21:34:46 -0500
Subject: [PATCH] Add a data converter to migrate from Empirecraft metadata


diff --git a/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java b/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
index 79ff70b3c6b73827612f874cd2d4116f66a040d3..dd9e27a0523a8e68ae0f89a03feaeec24f34c6c2 100644
--- a/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
+++ b/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
@@ -242,6 +242,7 @@ public final class MCVersionRegistry {
             3939,
             3943,
             3945,
+            3952, // EMC
             4054,
             4055,
             4057,
diff --git a/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java b/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
index 8716f53330b2ace6863816ee6de21954832daf43..f88b45367d81d02883fbd0a42419aa0ff74b868d 100644
--- a/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
+++ b/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
@@ -317,6 +317,7 @@ public final class MCTypeRegistry {
         V3939.register();
         V3943.register();
         V3945.register();
+        com.empireminecraft.empirecraft.util.EmpireMetadataConverter.register(); // EMC
         // V1.21.2
         V4054.register();
         V4055.register();
diff --git a/net/minecraft/world/level/storage/LevelStorageSource.java b/net/minecraft/world/level/storage/LevelStorageSource.java
index 85cddc7ccb9c73e2464a764d61f46f974ce255e3..817b15044505d5286d1e90721e31262a34247a2d 100644
--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -226,7 +226,7 @@ public class LevelStorageSource {
         CompoundTag levelDataTagRaw = readLevelDataTagRaw(levelPath);
         CompoundTag compoundOrEmpty = levelDataTagRaw.getCompoundOrEmpty("Data");
         int dataVersion = NbtUtils.getDataVersion(compoundOrEmpty, -1);
-        Dynamic<?> dynamic = DataFixTypes.LEVEL.updateToCurrentVersion(dataFixer, new Dynamic<>(NbtOps.INSTANCE, compoundOrEmpty), dataVersion);
+        Dynamic<?> dynamic = new Dynamic(NbtOps.INSTANCE, ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.LEVEL, compoundOrEmpty, dataVersion, net.minecraft.SharedConstants.getCurrentVersion().dataVersion().version())); // EMC
         dynamic = dynamic.update("Player", dynamic1 -> new Dynamic(dynamic1.getOps(), ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.PLAYER, (net.minecraft.nbt.CompoundTag)dynamic1.getValue(), dataVersion, ca.spottedleaf.dataconverter.minecraft.util.Version.getCurrentVersion()))); // Paper - replace data conversion system
         return dynamic.update("WorldGenSettings", dynamic1 -> DataFixTypes.WORLD_GEN_SETTINGS.updateToCurrentVersion(dataFixer, dynamic1, dataVersion));
     }
@@ -246,8 +246,7 @@ public class LevelStorageSource {
                 if (readLightweightData(path) instanceof CompoundTag compoundTag) {
                     CompoundTag compoundOrEmpty = compoundTag.getCompoundOrEmpty("Data");
                     int dataVersion = NbtUtils.getDataVersion(compoundOrEmpty, -1);
-                    Dynamic<?> dynamic = DataFixTypes.LEVEL_SUMMARY
-                        .updateToCurrentVersion(this.fixerUpper, new Dynamic<>(NbtOps.INSTANCE, compoundOrEmpty), dataVersion);
+                    Dynamic<?> dynamic = new Dynamic(NbtOps.INSTANCE, ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.LEVEL, compoundOrEmpty, dataVersion, net.minecraft.SharedConstants.getCurrentVersion().dataVersion().version())); // EMC
                     return this.makeLevelSummary(dynamic, levelDirectory, locked);
                 }
 
