From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 11 Sep 2024 21:34:46 -0500
Subject: [PATCH] Add a data converter to migrate from Empirecraft metadata


diff --git a/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java b/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
index 78eefba4cbf90f56e1844d284e7200172df448d5..94b3750f2d81ed7da3b02ebd0571f5c5e6273b12 100644
--- a/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
+++ b/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
@@ -243,6 +243,7 @@ public final class MCVersionRegistry {
             3939,
             3943,
             3945,
+            3952, // EMC
             4054,
             4055,
             4057,
diff --git a/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java b/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
index a102b8dc03721eb7cf5512f7fdf9135fd86d4235..c5185683947ee74e94fcb68f6294f0a4099e67d6 100644
--- a/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
+++ b/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
@@ -318,6 +318,7 @@ public final class MCTypeRegistry {
         V3939.register();
         V3943.register();
         V3945.register();
+        com.empireminecraft.empirecraft.util.EmpireMetadataConverter.register(); // EMC
         // V1.21.2
         V4054.register();
         V4055.register();
diff --git a/net/minecraft/world/level/storage/LevelStorageSource.java b/net/minecraft/world/level/storage/LevelStorageSource.java
index 140630186c3b0324c248cc2a2f31d3b906557b9c..85ba74db0352783585472fdd6bc03237fdc0c583 100644
--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -223,7 +223,7 @@ public class LevelStorageSource {
         CompoundTag levelDataTagRaw = readLevelDataTagRaw(levelPath);
         CompoundTag compoundOrEmpty = levelDataTagRaw.getCompoundOrEmpty("Data");
         int dataVersion = NbtUtils.getDataVersion(compoundOrEmpty);
-        Dynamic<?> dynamic = DataFixTypes.LEVEL.updateToCurrentVersion(dataFixer, new Dynamic<>(NbtOps.INSTANCE, compoundOrEmpty), dataVersion);
+        Dynamic<?> dynamic = new Dynamic(NbtOps.INSTANCE, ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.LEVEL, compoundOrEmpty, dataVersion, net.minecraft.SharedConstants.getCurrentVersion().dataVersion().version())); // EMC
         dynamic = dynamic.update("Player", dynamic1 -> new Dynamic(dynamic1.getOps(), ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.PLAYER, (net.minecraft.nbt.CompoundTag)dynamic1.getValue(), dataVersion, ca.spottedleaf.dataconverter.minecraft.util.Version.getCurrentVersion()))); // Paper - replace data conversion system
         return dynamic.update("WorldGenSettings", dynamic1 -> DataFixTypes.WORLD_GEN_SETTINGS.updateToCurrentVersion(dataFixer, dynamic1, dataVersion));
     }
@@ -243,8 +243,7 @@ public class LevelStorageSource {
                 if (readLightweightData(path) instanceof CompoundTag compoundTag) {
                     CompoundTag compoundOrEmpty = compoundTag.getCompoundOrEmpty("Data");
                     int dataVersion = NbtUtils.getDataVersion(compoundOrEmpty);
-                    Dynamic<?> dynamic = DataFixTypes.LEVEL_SUMMARY
-                        .updateToCurrentVersion(this.fixerUpper, new Dynamic<>(NbtOps.INSTANCE, compoundOrEmpty), dataVersion);
+                    Dynamic<?> dynamic = new Dynamic(NbtOps.INSTANCE, ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.LEVEL, compoundOrEmpty, dataVersion, net.minecraft.SharedConstants.getCurrentVersion().dataVersion().version())); // EMC
                     return this.makeLevelSummary(dynamic, levelDirectory, locked);
                 }
 
