From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 2 Jan 2025 10:24:45 -0600
Subject: [PATCH] Improve loading old block entity names


diff --git a/net/minecraft/world/level/block/entity/BlockEntity.java b/net/minecraft/world/level/block/entity/BlockEntity.java
index 380b1b51ce19f54580a1445820abbea3ce209da8..65e1d649a7d593afe349e2222e9df147df1620e6 100644
--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -334,7 +334,18 @@ public abstract class BlockEntity {
     @Nullable
     public static Component parseCustomNameSafe(String customName, HolderLookup.Provider registries) {
         try {
-            return Component.Serializer.fromJson(customName, registries);
+            // EMC start
+            if (customName.isEmpty()) {
+                return null;
+            }
+            if (customName.startsWith("{") && customName.endsWith("}")) {
+                try {
+                    return Component.Serializer.fromJson(customName, registries);
+                } catch (com.google.gson.JsonParseException ignored) {
+                }
+            }
+            return Component.Serializer.fromJsonLenient(customName, registries);
+            // EMC end
         } catch (Exception var3) {
             LOGGER.warn("Failed to parse custom name from string '{}', discarding", customName, var3);
             return null;
