From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 29 Nov 2025 13:02:59 -0600
Subject: [PATCH] Implement CheckShelfConnectionEvent


diff --git a/net/minecraft/world/level/block/SideChainPartBlock.java b/net/minecraft/world/level/block/SideChainPartBlock.java
index 9f53bee653b4aded96a9d5c7d324517c32feed05..74a92e586c504f08b0f9c6e677836df4c9058ed2 100644
--- a/net/minecraft/world/level/block/SideChainPartBlock.java
+++ b/net/minecraft/world/level/block/SideChainPartBlock.java
@@ -64,13 +64,18 @@ public interface SideChainPartBlock {
                 int i = neighbors.left().isConnectable() ? this.getAllBlocksConnectedTo(level, neighbors.left().pos()).size() : 0;
                 int i1 = neighbors.right().isConnectable() ? this.getAllBlocksConnectedTo(level, neighbors.right().pos()).size() : 0;
                 int i2 = 1;
-                if (this.canConnect(i, i2)) {
+
+                // EMC start - CheckShelfConnectionEvent
+                var leftEvent = new com.empireminecraft.empirecraft.event.block.CheckShelfConnectionEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), org.bukkit.craftbukkit.block.CraftBlock.at(level, neighbors.left().pos()));
+                if (this.canConnect(i, i2) && leftEvent.callEvent()) {
                     sideChainPart = sideChainPart.whenConnectedToTheLeft();
                     neighbors.left().connectToTheRight();
                     i2 += i;
                 }
 
-                if (this.canConnect(i1, i2)) {
+                var rightEvent = new com.empireminecraft.empirecraft.event.block.CheckShelfConnectionEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), org.bukkit.craftbukkit.block.CraftBlock.at(level, neighbors.right().pos()));
+                if (this.canConnect(i1, i2) && rightEvent.callEvent()) {
+                    // EMC end
                     sideChainPart = sideChainPart.whenConnectedToTheRight();
                     neighbors.right().connectToTheLeft();
                 }
