From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 11 Sep 2024 21:34:46 -0500
Subject: [PATCH] Add a data converter to migrate from Empirecraft metadata


diff --git a/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java b/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
index ab0d941248c389fec08a5ca6e45f7f5bf72cae58..34a0473e7184ee2b6ef6e23e8f51923bb159b1ec 100644
--- a/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
+++ b/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
@@ -242,6 +242,7 @@ public final class MCVersionRegistry {
             3939,
             3943,
             3945,
+            3952, // EMC
             4054,
             4055,
             4057,
diff --git a/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java b/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
index 4888b386d0e7f0237c2fe416e9034658cba0f1b0..c3409ad809d96fd36b6f259d9e463c1c27990096 100644
--- a/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
+++ b/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
@@ -316,6 +316,7 @@ public final class MCTypeRegistry {
         V3939.register();
         V3943.register();
         V3945.register();
+        com.empireminecraft.empirecraft.util.EmpireMetadataConverter.register(); // EMC
         // V1.21.2
         V4054.register();
         V4055.register();
diff --git a/net/minecraft/world/level/storage/LevelStorageSource.java b/net/minecraft/world/level/storage/LevelStorageSource.java
index 3b40822ea6ec9783fe3cb8eaba069a8d626d8382..17bdda921620c59b83f09217d90660c07f32e4a7 100644
--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -226,7 +226,7 @@ public class LevelStorageSource {
         CompoundTag levelDataTagRaw = readLevelDataTagRaw(levelPath);
         CompoundTag compoundOrEmpty = levelDataTagRaw.getCompoundOrEmpty("Data");
         int dataVersion = NbtUtils.getDataVersion(compoundOrEmpty, -1);
-        Dynamic<?> dynamic = DataFixTypes.LEVEL.updateToCurrentVersion(dataFixer, new Dynamic<>(NbtOps.INSTANCE, compoundOrEmpty), dataVersion);
+        Dynamic<?> dynamic = new Dynamic(NbtOps.INSTANCE, ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.LEVEL, compoundOrEmpty, dataVersion, net.minecraft.SharedConstants.getCurrentVersion().dataVersion().version())); // EMC
         dynamic = dynamic.update("Player", dynamic1 -> new Dynamic(dynamic1.getOps(), ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.PLAYER, (net.minecraft.nbt.CompoundTag)dynamic1.getValue(), dataVersion, ca.spottedleaf.dataconverter.minecraft.util.Version.getCurrentVersion()))); // Paper - replace data conversion system
         return dynamic.update("WorldGenSettings", dynamic1 -> DataFixTypes.WORLD_GEN_SETTINGS.updateToCurrentVersion(dataFixer, dynamic1, dataVersion));
     }
@@ -246,8 +246,7 @@ public class LevelStorageSource {
                 if (readLightweightData(path) instanceof CompoundTag compoundTag) {
                     CompoundTag compoundOrEmpty = compoundTag.getCompoundOrEmpty("Data");
                     int dataVersion = NbtUtils.getDataVersion(compoundOrEmpty, -1);
-                    Dynamic<?> dynamic = DataFixTypes.LEVEL_SUMMARY
-                        .updateToCurrentVersion(this.fixerUpper, new Dynamic<>(NbtOps.INSTANCE, compoundOrEmpty), dataVersion);
+                    Dynamic<?> dynamic = new Dynamic(NbtOps.INSTANCE, ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.LEVEL, compoundOrEmpty, dataVersion, net.minecraft.SharedConstants.getCurrentVersion().dataVersion().version())); // EMC
                     return this.makeLevelSummary(dynamic, levelDirectory, locked);
                 }
 
