From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 29 Nov 2025 13:02:59 -0600
Subject: [PATCH] Implement CheckShelfConnectionEvent


diff --git a/net/minecraft/world/level/block/SideChainPartBlock.java b/net/minecraft/world/level/block/SideChainPartBlock.java
index b974e28be5803988bda3d8ee4a4eb8230bace8f6..e86158dc813d298fc96b0dbd8c222f1384776f31 100644
--- a/net/minecraft/world/level/block/SideChainPartBlock.java
+++ b/net/minecraft/world/level/block/SideChainPartBlock.java
@@ -66,13 +66,17 @@ public interface SideChainPartBlock {
                 int existingChainOnTheLeft = neighbors.left().isConnectable() ? this.getAllBlocksConnectedTo(level, neighbors.left().pos()).size() : 0;
                 int existingChainOnTheRight = neighbors.right().isConnectable() ? this.getAllBlocksConnectedTo(level, neighbors.right().pos()).size() : 0;
                 int currentChainLength = 1;
-                if (this.canConnect(existingChainOnTheLeft, currentChainLength)) {
+                // EMC start - CheckShelfConnectionEvent
+                var leftEvent = new com.empireminecraft.empirecraft.event.block.CheckShelfConnectionEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), org.bukkit.craftbukkit.block.CraftBlock.at(level, neighbors.left().pos()));
+                if (this.canConnect(existingChainOnTheLeft, currentChainLength) && leftEvent.callEvent()) {
                     newPartForSelf = newPartForSelf.whenConnectedToTheLeft();
                     neighbors.left().connectToTheRight();
                     currentChainLength += existingChainOnTheLeft;
                 }
 
-                if (this.canConnect(existingChainOnTheRight, currentChainLength)) {
+                var rightEvent = new com.empireminecraft.empirecraft.event.block.CheckShelfConnectionEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), org.bukkit.craftbukkit.block.CraftBlock.at(level, neighbors.right().pos()));
+                if (this.canConnect(existingChainOnTheRight, currentChainLength) && rightEvent.callEvent()) {
+                    // EMC end
                     newPartForSelf = newPartForSelf.whenConnectedToTheRight();
                     neighbors.right().connectToTheLeft();
                 }
