From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 29 Nov 2025 20:10:27 -0600
Subject: [PATCH] Implement SpawnPatternBlocks events


diff --git a/net/minecraft/world/level/block/CarvedPumpkinBlock.java b/net/minecraft/world/level/block/CarvedPumpkinBlock.java
index 4ed03c9bb81a903c6ec68efe72845f20b083f3ac..741b6e0a3f41f6f7d270f5ef3d4af15092e31dd2 100644
--- a/net/minecraft/world/level/block/CarvedPumpkinBlock.java
+++ b/net/minecraft/world/level/block/CarvedPumpkinBlock.java
@@ -110,17 +110,34 @@ public class CarvedPumpkinBlock extends HorizontalDirectionalBlock {
         golem.snapTo(spawnPos.getX() + 0.5, spawnPos.getY() + 0.05, spawnPos.getZ() + 0.5, 0.0F, 0.0F);
         // Paper start
         org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason;
+        // EMC start
+        org.bukkit.entity.EntityType entityType = null;
         if (golem.getType() == net.minecraft.world.entity.EntityTypes.SNOW_GOLEM) {
             spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_SNOWMAN;
+            entityType = org.bukkit.entity.EntityType.SNOW_GOLEM;
         } else if (golem.getType() == net.minecraft.world.entity.EntityTypes.COPPER_GOLEM) {
             spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_COPPERGOLEM;
+            entityType = org.bukkit.entity.EntityType.COPPER_GOLEM;
         } else {
             spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM;
+            entityType = org.bukkit.entity.EntityType.IRON_GOLEM;
+        }
+        java.util.List<org.bukkit.craftbukkit.block.CraftBlock> list = new java.util.ArrayList<>();
+        for (int i = 0; i < match.getWidth(); i++) {
+            for (int i1 = 0; i1 < match.getHeight(); i1++) {
+                BlockInWorld block = match.getBlock(i, i1, 0);
+                list.add(org.bukkit.craftbukkit.block.CraftBlock.at(level, block.getPos()));
+            }
+        }
+        if (!(new com.empireminecraft.empirecraft.event.CheckSpawnPatternBlockEvent(java.util.Collections.unmodifiableList(list), entityType).callEvent())) {
+            return;
         }
         if (!level.addFreshEntity(golem, spawnReason)) {
             return;
         }
         // Paper end
+        new com.empireminecraft.empirecraft.event.ClearSpawnPatternBlocksEvent(java.util.Collections.unmodifiableList(list), entityType).callEvent();
+        // EMC end
         clearPatternBlocks(level, match); // Paper - from above
 
         for (ServerPlayer player : level.getEntitiesOfClass(ServerPlayer.class, golem.getBoundingBox().inflate(5.0))) {
diff --git a/net/minecraft/world/level/block/WitherSkullBlock.java b/net/minecraft/world/level/block/WitherSkullBlock.java
index 65da2aae8bbddebe829e9b44b36152fdb76d507d..36c8d85766735596922e506c216d7bf3d1624f88 100644
--- a/net/minecraft/world/level/block/WitherSkullBlock.java
+++ b/net/minecraft/world/level/block/WitherSkullBlock.java
@@ -69,9 +69,22 @@ public class WitherSkullBlock extends SkullBlock {
                         witherBoss.yBodyRot = match.getForwards().getAxis() == Direction.Axis.X ? 0.0F : 90.0F;
                         witherBoss.makeInvulnerable();
                         // CraftBukkit start
+                        // EMC start
+                        java.util.List<org.bukkit.craftbukkit.block.CraftBlock> list = new java.util.ArrayList<>();
+                        for (int i = 0; i < match.getWidth(); i++) {
+                            for (int i1 = 0; i1 < match.getHeight(); i1++) {
+                                BlockInWorld block = match.getBlock(i, i1, 0);
+                                list.add(org.bukkit.craftbukkit.block.CraftBlock.at(level, block.getPos()));
+                            }
+                        }
+                        if (!(new com.empireminecraft.empirecraft.event.CheckSpawnPatternBlockEvent(java.util.Collections.unmodifiableList(list), org.bukkit.entity.EntityType.WITHER).callEvent())) {
+                            return;
+                        }
                         if (!level.addFreshEntity(witherBoss, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_WITHER)) {
                             return;
                         }
+                        new com.empireminecraft.empirecraft.event.ClearSpawnPatternBlocksEvent(java.util.Collections.unmodifiableList(list), org.bukkit.entity.EntityType.WITHER).callEvent();
+                        // EMC end
                         CarvedPumpkinBlock.clearPatternBlocks(level, match); // CraftBukkit - from above
                         // CraftBukkit end
 
