From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 29 Nov 2025 20:10:27 -0600
Subject: [PATCH] Implement SpawnPatternBlocks events


diff --git a/net/minecraft/world/level/block/CarvedPumpkinBlock.java b/net/minecraft/world/level/block/CarvedPumpkinBlock.java
index 5a316fedee1992efd4a21c4678b659c4589d57ec..3b383746f625e8355f7cee39a043fab453a7082d 100644
--- a/net/minecraft/world/level/block/CarvedPumpkinBlock.java
+++ b/net/minecraft/world/level/block/CarvedPumpkinBlock.java
@@ -113,17 +113,34 @@ public class CarvedPumpkinBlock extends HorizontalDirectionalBlock {
             return;
         }
         org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason;
+        // EMC start
+        org.bukkit.entity.EntityType entityType = null;
         if (golem.getType() == net.minecraft.world.entity.EntityTypes.SNOW_GOLEM) {
             spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_SNOWMAN;
+            entityType = org.bukkit.entity.EntityType.SNOW_GOLEM;
         } else if (golem.getType() == net.minecraft.world.entity.EntityTypes.COPPER_GOLEM) {
             spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_COPPERGOLEM;
+            entityType = org.bukkit.entity.EntityType.COPPER_GOLEM;
         } else {
             spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM;
+            entityType = org.bukkit.entity.EntityType.IRON_GOLEM;
+        }
+        java.util.List<org.bukkit.craftbukkit.block.CraftBlock> list = new java.util.ArrayList<>();
+        for (int i = 0; i < match.getWidth(); i++) {
+            for (int i1 = 0; i1 < match.getHeight(); i1++) {
+                BlockInWorld block = match.getBlock(i, i1, 0);
+                list.add(org.bukkit.craftbukkit.block.CraftBlock.at(level, block.getPos()));
+            }
+        }
+        if (!(new com.empireminecraft.empirecraft.event.CheckSpawnPatternBlockEvent(java.util.Collections.unmodifiableList(list), entityType).callEvent())) {
+            return;
         }
         if (!level.addFreshEntity(golem, spawnReason)) {
             return;
         }
         // Paper end
+        new com.empireminecraft.empirecraft.event.ClearSpawnPatternBlocksEvent(java.util.Collections.unmodifiableList(list), entityType).callEvent();
+        // EMC end
         clearPatternBlocks(level, match); // Paper - from above
 
         for (ServerPlayer player : level.getEntitiesOfClass(ServerPlayer.class, golem.getBoundingBox().inflate(5.0))) {
diff --git a/net/minecraft/world/level/block/WitherSkullBlock.java b/net/minecraft/world/level/block/WitherSkullBlock.java
index 2e65881cd2db4dab9967a4d2c597365c31939c42..ac47db3d5ed72e789c5ada9488f1eff54c89ce85 100644
--- a/net/minecraft/world/level/block/WitherSkullBlock.java
+++ b/net/minecraft/world/level/block/WitherSkullBlock.java
@@ -74,9 +74,22 @@ public class WitherSkullBlock extends SkullBlock {
                         }
                         // Paper end
                         // CraftBukkit start
+                        // EMC start
+                        java.util.List<org.bukkit.craftbukkit.block.CraftBlock> list = new java.util.ArrayList<>();
+                        for (int i = 0; i < match.getWidth(); i++) {
+                            for (int i1 = 0; i1 < match.getHeight(); i1++) {
+                                BlockInWorld block = match.getBlock(i, i1, 0);
+                                list.add(org.bukkit.craftbukkit.block.CraftBlock.at(level, block.getPos()));
+                            }
+                        }
+                        if (!(new com.empireminecraft.empirecraft.event.CheckSpawnPatternBlockEvent(java.util.Collections.unmodifiableList(list), org.bukkit.entity.EntityType.WITHER).callEvent())) {
+                            return;
+                        }
                         if (!level.addFreshEntity(witherBoss, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_WITHER)) {
                             return;
                         }
+                        new com.empireminecraft.empirecraft.event.ClearSpawnPatternBlocksEvent(java.util.Collections.unmodifiableList(list), org.bukkit.entity.EntityType.WITHER).callEvent();
+                        // EMC end
                         CarvedPumpkinBlock.clearPatternBlocks(level, match); // CraftBukkit - from above
                         // CraftBukkit end
 
