From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 11 Sep 2024 21:34:46 -0500
Subject: [PATCH] Add a data converter to migrate from Empirecraft metadata


diff --git a/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java b/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
index 344c8c4f3207b6c8b565e5ad6db2470a272b77c3..b33bfde2011cffbc5f9abc0daa5c1b1bbca268f6 100644
--- a/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
+++ b/ca/spottedleaf/dataconverter/minecraft/MCVersionRegistry.java
@@ -240,6 +240,7 @@ public final class MCVersionRegistry {
             3939,
             3943,
             3945,
+            3952, // EMC
             4054,
             4055,
             4057,
diff --git a/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java b/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
index d42bff4fec99eb0b19d132794f4e3306b6dddb0f..09f430695aa53bba761dc8d3017f4cb0ab73e61a 100644
--- a/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
+++ b/ca/spottedleaf/dataconverter/minecraft/datatypes/MCTypeRegistry.java
@@ -308,6 +308,7 @@ public final class MCTypeRegistry {
         V3939.register();
         V3943.register();
         V3945.register();
+        com.empireminecraft.empirecraft.util.EmpireMetadataConverter.register(); // EMC
         // V1.21.2
         V4054.register();
         V4055.register();
diff --git a/net/minecraft/world/level/storage/LevelStorageSource.java b/net/minecraft/world/level/storage/LevelStorageSource.java
index de43e54698125ce9f319d4889dd49f7029fe95e0..6b7444909c2cde80ca7030cd5345183f2b90c236 100644
--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -226,7 +226,7 @@ public class LevelStorageSource {
         CompoundTag levelDataTagRaw = readLevelDataTagRaw(levelPath);
         CompoundTag compound = levelDataTagRaw.getCompound("Data");
         int dataVersion = NbtUtils.getDataVersion(compound, -1);
-        Dynamic<?> dynamic = DataFixTypes.LEVEL.updateToCurrentVersion(dataFixer, new Dynamic<>(NbtOps.INSTANCE, compound), dataVersion);
+        Dynamic<?> dynamic = new Dynamic(NbtOps.INSTANCE, ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.LEVEL, compound, dataVersion, net.minecraft.SharedConstants.getCurrentVersion().getDataVersion().getVersion())); // EMC
         dynamic = dynamic.update("Player", dynamic1 -> new Dynamic(dynamic1.getOps(), ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.PLAYER, (net.minecraft.nbt.CompoundTag)dynamic1.getValue(), dataVersion, net.minecraft.SharedConstants.getCurrentVersion().getDataVersion().getVersion()))); // Paper - replace data conversion system
         return dynamic.update("WorldGenSettings", dynamic1 -> DataFixTypes.WORLD_GEN_SETTINGS.updateToCurrentVersion(dataFixer, dynamic1, dataVersion));
     }
@@ -246,7 +246,7 @@ public class LevelStorageSource {
                 if (readLightweightData(path) instanceof CompoundTag compoundTag) {
                     CompoundTag compound = compoundTag.getCompound("Data");
                     int dataVersion = NbtUtils.getDataVersion(compound, -1);
-                    Dynamic<?> dynamic = DataFixTypes.LEVEL.updateToCurrentVersion(this.fixerUpper, new Dynamic<>(NbtOps.INSTANCE, compound), dataVersion);
+                    Dynamic<?> dynamic = new Dynamic(NbtOps.INSTANCE, ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.LEVEL, compound, dataVersion, net.minecraft.SharedConstants.getCurrentVersion().getDataVersion().getVersion())); // EMC
                     return this.makeLevelSummary(dynamic, levelDirectory, locked);
                 }
 
