From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 30 Dec 2025 12:39:08 -0600
Subject: [PATCH] Remove thrower/owner data on converting old items


diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V2514.java b/ca/spottedleaf/dataconverter/minecraft/versions/V2514.java
index 59200b33a30427ad3a1729ee0cc7c8a0d382d211..078039cbe9b921a11dd41667ae716cf3e71a5799 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V2514.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V2514.java
@@ -159,6 +159,7 @@ public final class V2514 {
 
     static void replaceUUIDMLTag(final MapType data, final String oldPath, final String newPath) {
         final int[] uuid = createUUIDFromLongs(data.getMap(oldPath), "M", "L");
+        data.remove(oldPath); // EMC - always remove old data even if it fails
         if (uuid != null) {
             data.remove(oldPath);
             data.setInts(newPath, uuid);
