From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 30 Dec 2025 12:39:08 -0600
Subject: [PATCH] Remove thrower/owner data on converting old item entities


diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V2514.java b/ca/spottedleaf/dataconverter/minecraft/versions/V2514.java
index 59200b33a30427ad3a1729ee0cc7c8a0d382d211..0b0a860da9a5bd107c0cec853972bd39f41a9a46 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V2514.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V2514.java
@@ -159,8 +159,9 @@ public final class V2514 {
 
     static void replaceUUIDMLTag(final MapType data, final String oldPath, final String newPath) {
         final int[] uuid = createUUIDFromLongs(data.getMap(oldPath), "M", "L");
+        data.remove(oldPath); // EMC - always remove old data even if it fails
         if (uuid != null) {
-            data.remove(oldPath);
+            //data.remove(oldPath); // EMC - move up
             data.setInts(newPath, uuid);
         }
     }
