From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 30 Dec 2025 20:34:17 -0600
Subject: [PATCH] Resync on cancelled nautilus interactions


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 429173a60a1bbc47dfec8865c0fdcdd03c10f86d..ca9d823a3c39c8df01feab79f3cb0414dca0a6cd 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2858,7 +2858,7 @@ public class ServerGamePacketListenerImpl
                                         // Refresh the current entity metadata
                                         target.refreshEntityData(ServerGamePacketListenerImpl.this.player);
                                         // SPIGOT-7136 - Allays
-                                        if (target instanceof net.minecraft.world.entity.animal.allay.Allay || target instanceof net.minecraft.world.entity.animal.equine.AbstractHorse) { // Paper - Fix horse armor desync
+                                        if (target instanceof net.minecraft.world.entity.animal.allay.Allay || target instanceof net.minecraft.world.entity.animal.equine.AbstractHorse || target instanceof net.minecraft.world.entity.animal.nautilus.AbstractNautilus) { // Paper - Fix horse armor desync // EMC - resync nautilus as well
                                             ServerGamePacketListenerImpl.this.send(new net.minecraft.network.protocol.game.ClientboundSetEquipmentPacket(
                                                 target.getId(), java.util.Arrays.stream(net.minecraft.world.entity.EquipmentSlot.values())
                                                 .map((slot) -> com.mojang.datafixers.util.Pair.of(slot, ((LivingEntity) target).getItemBySlot(slot).copy()))
