From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 31 Dec 2025 09:25:49 -0600
Subject: [PATCH] Alter data converter to prevent gamerule conversion to empty
 strings


diff --git a/ca/spottedleaf/dataconverter/minecraft/versions/V4658.java b/ca/spottedleaf/dataconverter/minecraft/versions/V4658.java
index 86e0602e1b8a0e21abda7d0d204ca9a45f314114..67fe4cff2f9f6fa45178c55e4a7ffde1af3bd61b 100644
--- a/ca/spottedleaf/dataconverter/minecraft/versions/V4658.java
+++ b/ca/spottedleaf/dataconverter/minecraft/versions/V4658.java
@@ -14,6 +14,9 @@ public final class V4658 {
     public static void register() {
         MCTypeRegistry.LEVEL.addStructureConverter(new DataConverter<>(VERSION) {
             private static void convertBool(final MapType data, final String src, final String dst) {
+                if (!data.hasKey(src)) {
+                    return;
+                }
                 final String value = data.getString(src, "");
 
                 data.remove(src);
@@ -22,6 +25,9 @@ public final class V4658 {
             }
 
             private static void convertBoolInv(final MapType data, final String src, final String dst) {
+                if (!data.hasKey(src)) {
+                    return;
+                }
                 final String value = data.getString(src, "");
 
                 data.remove(src);
@@ -38,6 +44,9 @@ public final class V4658 {
             }
 
             private static void convertInt(final MapType data, final String src, final String dst, final int min, final int max) {
+                if (!data.hasKey(src)) {
+                    return;
+                }
                 final String value = data.getString(src, "");
 
                 data.remove(src);
