--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -399,6 +_,7 @@
     // Paper end - Folia reagion threading API
 
     static {
+        com.empireminecraft.empirecraft.api.API api = com.empireminecraft.empirecraft.api.CraftEmpireAPI.instance; // EMC - Force load at server start
         ConfigurationSerialization.registerClass(CraftOfflinePlayer.class);
         ConfigurationSerialization.registerClass(CraftPlayerProfile.class);
         ConfigurationSerialization.registerClass(com.destroystokyo.paper.profile.CraftPlayerProfile.class); // Paper
