--- a/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
+++ b/src/main/java/org/bukkit/craftbukkit/scheduler/CraftScheduler.java
@@ -479,7 +_,7 @@
                         task.getTaskId(),
                         task.getOwner().getDescription().getFullName());
                     task.getOwner().getLogger().log(
-                            Level.WARNING,
+                            Level.SEVERE, // EMC - Improve Task Exception Logging
                         logMessage,
                             throwable);
                     org.bukkit.Bukkit.getServer().getPluginManager().callEvent(
