--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -185,6 +_,7 @@
     // CraftBukkit start
     public boolean fauxSleeping;
     public int oldLevel = -1;
+    public float lastAttackPct = 0; // EMC - Human#getAttackPct API
 
     @Override
     public org.bukkit.craftbukkit.entity.CraftHumanEntity getBukkitEntity() {
@@ -1036,6 +_,7 @@
                 DamageSource damageSource = Optional.ofNullable(weaponItem.getItem().getDamageSource(this)).orElse(this.damageSources().playerAttack(this));
                 float f1 = this.getEnchantedDamage(target, f, damageSource) - f;
                 float attackStrengthScale = this.getAttackStrengthScale(0.5F);
+                this.lastAttackPct = attackStrengthScale; // EMC - Human#getAttackPct API
                 f *= 0.2F + attackStrengthScale * attackStrengthScale * 0.8F;
                 f1 *= attackStrengthScale;
                 // this.resetAttackStrengthTicker(); // CraftBukkit - Moved to EntityLiving to reset the cooldown after the damage is dealt
