--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -183,6 +_,7 @@
     // CraftBukkit start
     public boolean fauxSleeping;
     public int oldLevel = -1;
+    public float lastAttackPct = 0; // EMC - Human#getAttackPct API
 
     @Override
     public org.bukkit.craftbukkit.entity.CraftHumanEntity getBukkitEntity() {
@@ -1032,6 +_,7 @@
             ItemStack weaponItem = this.getWeaponItem();
             DamageSource damageSource = this.createAttackSource(weaponItem); final DamageSource dmgSourceFinal = damageSource; // Paper - damage events
             float attackStrengthScale = this.getAttackStrengthScale(0.5F);
+            this.lastAttackPct = attackStrengthScale; // EMC - Human#getAttackPct API
             float f1 = attackStrengthScale * (this.getEnchantedDamage(target, f, damageSource) - f);
             f *= this.baseDamageScaleFactor();
             this.onAttack();
