--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -177,6 +_,7 @@
     // CraftBukkit start
     public boolean fauxSleeping;
     public int oldLevel = -1;
+    public float lastAttackPct = 0; // EMC - Human#getAttackPct API
 
     @Override
     public org.bukkit.craftbukkit.entity.CraftHumanEntity getBukkitEntity() {
@@ -1032,6 +_,7 @@
             ItemStack attackingItemStack = this.getWeaponItem();
             DamageSource damageSource = this.createAttackSource(attackingItemStack); final DamageSource dmgSourceFinal = damageSource; // Paper - damage events
             float attackStrengthScale = this.getAttackStrengthScale(0.5F);
+            this.lastAttackPct = attackStrengthScale; // EMC - Human#getAttackPct API
             float magicBoost = attackStrengthScale * (this.getEnchantedDamage(entity, baseDamage, damageSource) - baseDamage);
             baseDamage *= this.baseDamageScaleFactor();
             this.onAttack(entity); // Paper
