--- a/src/main/java/org/bukkit/event/inventory/FurnaceExtractEvent.java
+++ b/src/main/java/org/bukkit/event/inventory/FurnaceExtractEvent.java
@@ -20,8 +_,12 @@
     private final int itemAmount;
 
     @ApiStatus.Internal
-    public FurnaceExtractEvent(@NotNull Player player, @NotNull Block block, @NotNull ItemStack itemStack, int itemAmount, int exp) {
+    public FurnaceExtractEvent(@org.jetbrains.annotations.Nullable Player player, @NotNull Block block, @org.jetbrains.annotations.Nullable ItemStack itemStack, int itemAmount, int exp, @org.jetbrains.annotations.Nullable ItemStack resultItemStack, int resultAmount) { // EMC - Add Nullable to Player, itemStack, and resultItemStack and add resultAmount
         super(block, exp);
+        // EMC - Add resultType and resultAmount
+        this.resultItemStack = resultItemStack;
+        this.resultAmount = resultAmount;
+        // EMC end - Add resultType and resultAmount
         this.player = player;
         this.itemStack = itemStack;
         this.itemAmount = itemAmount;
@@ -32,7 +_,7 @@
      *
      * @return the relevant player
      */
-    @NotNull
+    @org.jetbrains.annotations.Nullable // EMC - Add Nullable to Player
     public Player getPlayer() {
         return this.player;
     }
@@ -65,4 +_,28 @@
     public int getItemAmount() {
         return this.itemAmount;
     }
+    // EMC start - Add resultType and resultAmount
+    private final ItemStack resultItemStack;
+    private final int resultAmount;
+    /**
+     * Get the Material of the recipe result being retrieved
+     * This is different from {@link #getItemType} for stored result experience
+     *
+     * @return the material of the recipe resul
+     */
+    @org.jetbrains.annotations.Nullable
+    public ItemStack getResultItemStack() {
+        return resultItemStack;
+    }
+
+    /**
+     * Get the recipe result count being retrieved
+     * This is different from {@link #getItemAmount()} for stored result experience
+     *
+     * @return the amount of the recipe result
+     */
+    public int getResultAmount() {
+        return resultAmount;
+    }
+    // EMC end - Add resultType and resultAmount
 }
