--- a/src/main/java/org/bukkit/event/inventory/FurnaceExtractEvent.java
+++ b/src/main/java/org/bukkit/event/inventory/FurnaceExtractEvent.java
@@ -21,8 +_,12 @@
     private final int itemAmount;
 
     @ApiStatus.Internal
-    public FurnaceExtractEvent(@NotNull Player player, @NotNull Block block, @NotNull Material itemType, int itemAmount, int exp) {
+    public FurnaceExtractEvent(@org.jetbrains.annotations.Nullable Player player, @NotNull Block block, @NotNull Material itemType, int itemAmount, int exp, @NotNull Material resultType, int resultAmount) { // EMC - Add Nullable to Player and resultType and resultAmount
         super(block, exp);
+        // EMC - Add resultType and resultAmount
+        this.resultType = resultType;
+        this.resultAmount = resultAmount;
+        // EMC end - Add resultType and resultAmount
         this.player = player;
         if (itemType != null && itemType.isLegacy()) {
             itemType = Bukkit.getUnsafe().fromLegacy(new MaterialData(itemType), true);
@@ -36,7 +_,7 @@
      *
      * @return the relevant player
      */
-    @NotNull
+    @org.jetbrains.annotations.Nullable // EMC - Add Nullable to Player
     public Player getPlayer() {
         return this.player;
     }
@@ -59,4 +_,28 @@
     public int getItemAmount() {
         return this.itemAmount;
     }
+    // EMC start - Add resultType and resultAmount
+    private final Material resultType;
+    private final int resultAmount;
+    /**
+     * Get the Material of the recipe result being retrieved
+     * This is different from {@link #getItemType} for stored result experience
+     *
+     * @return the material of the recipe resul
+     */
+    @org.jetbrains.annotations.Nullable
+    public Material getResultType() {
+        return resultType;
+    }
+
+    /**
+     * Get the recipe result count being retrieved
+     * This is different from {@link #getItemAmount()} for stored result experience
+     *
+     * @return the amount of the recipe result
+     */
+    public int getResultAmount() {
+        return resultAmount;
+    }
+    // EMC end - Add resultType and resultAmount
 }
