From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 3 Dec 2016 22:27:48 -0500
Subject: [PATCH] Debug API's


diff --git a/src/main/java/com/empireminecraft/api/API.java b/src/main/java/com/empireminecraft/api/API.java
index 01b7bda835782d4f6603415785e5c219f0e9f2bc..caa2d3704b8ffe9b3634452e68a01d1872de754d 100644
--- a/src/main/java/com/empireminecraft/api/API.java
+++ b/src/main/java/com/empireminecraft/api/API.java
@@ -23,9 +23,34 @@
 
 package com.empireminecraft.api;
 
+import org.apache.commons.lang.exception.ExceptionUtils;
+import org.bukkit.Bukkit;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 public abstract class API {
 
     public static EAPI_Entity entity;
     public static EAPI_Misc misc;
+
+    @NotNull
+    public static String stack() {
+        return ExceptionUtils.getFullStackTrace(new Throwable());
+    }
+
+    public static void exception(@NotNull Throwable e) {
+        exception(null, e);
+    }
+
+    public static void exception(@Nullable String msg, @NotNull Throwable e) {
+        if (msg != null) {
+            Bukkit.getLogger().severe(msg);
+        }
+        if (e.getMessage() != null) {
+            Bukkit.getLogger().severe(e.getMessage());
+        }
+        for (String line : ExceptionUtils.getFullStackTrace(e).split("\n")) {
+            Bukkit.getLogger().severe(line);
+        }
+    }
 }
