--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -46,6 +_,7 @@
     private static final short DEFAULT_AGE = 0;
     private static final short DEFAULT_PICKUP_DELAY = 0;
     public int age = 0;
+    public boolean canDespawn = true; // EMC - Item canDespawn API
     public int pickupDelay = 0;
     public int health = 5;
     public @Nullable EntityReference<Entity> thrower;
@@ -135,7 +_,7 @@
             this.age++;
         }
 
-        if (!this.level().isClientSide() && this.age >= this.despawnRate) {// Paper - Alternative item-despawn-rate
+        if (canDespawn && !this.level().isClientSide() && this.age >= this.despawnRate) {// Paper - Alternative item-despawn-rate // EMC - Item canDespawn API
             // CraftBukkit start - fire ItemDespawnEvent
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callItemDespawnEvent(this).isCancelled()) {
                 this.age = 0;
@@ -283,7 +_,7 @@
         ItemStack thisItemStack = this.getItem();
         ItemStack otherItemStack = other.getItem();
         if (Objects.equals(this.target, other.target) && areMergable(thisItemStack, otherItemStack)) {
-            if (otherItemStack.getCount() < thisItemStack.getCount()) {
+            if (this.age < other.age) { // EMC - Merge into newer item entity
                 merge(this, thisItemStack, other, otherItemStack);
             } else {
                 merge(other, otherItemStack, this, thisItemStack);
