--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -40,6 +_,7 @@
     private static final int INFINITE_PICKUP_DELAY = 32767;
     private static final int INFINITE_LIFETIME = -32768;
     public int age;
+    public boolean canDespawn = true; // EMC - Item canDespawn API
     public int pickupDelay;
     public int health = 5;
     @Nullable
@@ -135,7 +_,7 @@
             this.age++;
         }
 
-        if (!this.level().isClientSide && this.age >= this.despawnRate) {// Paper - Alternative item-despawn-rate
+        if (canDespawn && !this.level().isClientSide && this.age >= this.despawnRate) {// Paper - Alternative item-despawn-rate // EMC - Item canDespawn API
             // CraftBukkit start - fire ItemDespawnEvent
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callItemDespawnEvent(this).isCancelled()) {
                 this.age = 0;
