--- a/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/arrow/AbstractArrow.java
@@ -555,6 +_,11 @@
         double knockback = this.firedFromWeapon != null && this.level() instanceof ServerLevel serverLevel
             ? EnchantmentHelper.modifyKnockback(serverLevel, this.firedFromWeapon, mob, damageSource, 0.0F)
             : 0.0F;
+        // EMC start - ArrowKnockbackEvent
+        var event = new com.empireminecraft.empirecraft.event.entity.ArrowKnockbackEvent(this.getBukkitEntity(), mob.getBukkitEntity(), knockback);
+        event.callEvent();
+        knockback = event.getKnockback();
+        // EMC end - ArrowKnockbackEvent
         if (knockback > 0.0) {
             double knockbackResistance = Math.max(0.0, 1.0 - mob.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE));
             Vec3 movement = this.getDeltaMovement().multiply(1.0, 0.0, 1.0).normalize().scale(knockback * 0.6 * knockbackResistance);
