--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -392,6 +_,7 @@
 
     public void stop() {
         this.path = null;
+        new com.empireminecraft.empirecraft.event.entity.EntityPathfindEndEvent((org.bukkit.entity.Mob)this.mob.getBukkitEntity()).callEvent(); // EMC - EntityPathfindEndEvent
     }
 
     protected abstract Vec3 getTempMobPos();
