From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 31 Dec 2014 13:32:49 -0500
Subject: [PATCH] Fix ForgeModLoader 1.8 Clients


diff --git a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
index 0d478b7f01ed31ec67609973f8527a19d3a986ce..269a9f0e13676a98b1598a29386183c0793b9703 100644
--- a/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -123,6 +123,11 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                 //if (org.spigotmc.SpigotConfig.bungee) { // Paper - comment out, we check above!
                     connection.isProxied = true; // EMC
                         String[] split = packet.hostName.split("\00");
+                    // EMC - Fix FML
+                    if (split.length == 6 && "FML".equals(split[1])) {
+                        split = new String[] {split[0], split[3], split[4], split[5]};
+                    }
+                    // EMC end
                         if ( ( split.length == 3 || split.length == 4 ) && ( ServerHandshakePacketListenerImpl.BYPASS_HOSTCHECK || ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher( split[1] ).matches() ) ) { // Paper
                             // Paper start - Unix domain socket support
                             java.net.SocketAddress socketAddress = connection.getRemoteAddress();
