From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 17 Nov 2015 01:28:08 -0500
Subject: [PATCH] Fire Vehicle Events for all entities


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 16244921560081575e8f23d84c084a9a189e63cc..223990b27be77bdbae1145d0929ea6a0e5fa3bcd 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2704,9 +2704,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
             // CraftBukkit start
             CraftEntity craft = (CraftEntity) entity.getBukkitEntity().getVehicle();
             Entity orig = craft == null ? null : craft.getHandle();
-            if (this.getBukkitEntity() instanceof Vehicle && entity.getBukkitEntity() instanceof LivingEntity) {
+            if (entity.getBukkitEntity() instanceof LivingEntity) { // EMC
                 VehicleExitEvent event = new VehicleExitEvent(
-                        (Vehicle) this.getBukkitEntity(),
+                        this.getBukkitEntity(), // EMC
                         (LivingEntity) entity.getBukkitEntity(), !suppressCancellation // Paper
                 );
                 // Suppress during worldgen
