--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -147,6 +_,7 @@
                 LOGGER.error("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
                 LOGGER.error("If you do not agree to the above EULA please stop your server and remove this flag immediately.");
             }
+            eulaAgreed = true; // EMC - Empire Minecraft agrees to the Minecraft EULA
             if (!eula.hasAgreedToEULA() && !eulaAgreed) {
                 // Spigot end
                 LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
