--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -148,7 +_,7 @@
             AdvancementHolder advancement = manager.get(id);
             if (advancement == null) {
                 if (!id.getNamespace().equals(Identifier.DEFAULT_NAMESPACE)) return; // CraftBukkit
-                LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", id, this.playerSavePath);
+                //LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", id, this.playerSavePath); // EMC - Undesired log
             } else {
                 this.startProgress(advancement, progress);
                 this.progressChanged.add(advancement);
