From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 26 Jul 2013 21:10:16 -0400
Subject: [PATCH] Add force to spawn reasons


diff --git a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
index 6ff1988092de06f9d751cd40da521c2ed6e2e4bd..d69e08279a191089dad2cae85f02038ca759e66e 100644
--- a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
+++ b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
@@ -221,6 +221,7 @@ public class CreatureSpawnEvent extends EntitySpawnEvent {
          * When a creature is spawned by plugins
          */
         CUSTOM,
+        FORCE, // EMC
         /**
          * When an entity is missing a SpawnReason
          */
