From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 26 Jul 2013 21:10:16 -0400
Subject: [PATCH] Add force to spawn reasons


diff --git a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
index 153fcde1cb0819bca6786685275dfd6f31e2b9a8..92c634c22c44128a57b37a4395ee908008fcacf2 100644
--- a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
+++ b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
@@ -185,6 +185,7 @@ public class CreatureSpawnEvent extends EntitySpawnEvent {
          * When a creature is spawned by plugins
          */
         CUSTOM,
+        FORCE, // EMC
         /**
          * When an entity is missing a SpawnReason
          */
