From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 26 Jul 2013 21:10:16 -0400
Subject: [PATCH] Add force to spawn reasons


diff --git a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
index 9a06487e0f76cd7765e6f900b7458a3cf0aa44e7..01ba53bbfceca84fe9183fbb38597218ee4c252b 100644
--- a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
+++ b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
@@ -202,6 +202,7 @@ public class CreatureSpawnEvent extends EntitySpawnEvent {
          * When a creature is spawned by plugins
          */
         CUSTOM,
+        FORCE, // EMC
         /**
          * When an entity is missing a SpawnReason
          */
