From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 7 Aug 2013 19:39:41 -0400
Subject: [PATCH] Add dismount reason API for EntityDismountEvent


diff --git a/src/main/java/org/spigotmc/event/entity/EntityDismountEvent.java b/src/main/java/org/spigotmc/event/entity/EntityDismountEvent.java
index a7632c8f5cb1bce4be0e456ec34f4a69c5ce80f3..58967e2c12525eb47864afe858efbe024b36dae9 100644
--- a/src/main/java/org/spigotmc/event/entity/EntityDismountEvent.java
+++ b/src/main/java/org/spigotmc/event/entity/EntityDismountEvent.java
@@ -10,7 +10,17 @@ import org.jetbrains.annotations.NotNull;
  * Called when an entity stops riding another entity.
  */
 public class EntityDismountEvent extends EntityEvent implements Cancellable {
+    // EMC start
+    public enum DismountReason {
+        PLAYER, WATER, DEAD, DEAD_VEHICLE, TELEPORT, SPECTATE, LEASH, TRANSFER, UNKNOWN, DISCONNECT, FORCE
+    }
+    public DismountReason reason;
 
+    @NotNull
+    public DismountReason getReason() {
+        return reason;
+    }
+    // EMC end
     private static final HandlerList handlers = new HandlerList();
     private boolean cancelled;
     private final Entity dismounted;
@@ -18,15 +28,16 @@ public class EntityDismountEvent extends EntityEvent implements Cancellable {
 
     public EntityDismountEvent(@NotNull Entity what, @NotNull Entity dismounted) {
         // Paper start
-        this(what, dismounted, true);
+        this(what, dismounted, true, DismountReason.UNKNOWN); // EMC
     }
 
 
-    public EntityDismountEvent(@NotNull Entity what, @NotNull Entity dismounted, boolean isCancellable) {
+    public EntityDismountEvent(@NotNull Entity what, @NotNull Entity dismounted, boolean isCancellable, @NotNull DismountReason reason) { // EMC
         // Paper end
         super( what );
         this.dismounted = dismounted;
         this.isCancellable = isCancellable; // Paper
+        this.reason = reason; // EMC
     }
 
     @NotNull
