From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 7 Aug 2013 19:39:41 -0400
Subject: [PATCH] Add API for vehicle exit event reason


diff --git a/src/main/java/org/bukkit/event/vehicle/VehicleExitEvent.java b/src/main/java/org/bukkit/event/vehicle/VehicleExitEvent.java
index a976c32de6ad5e90b0a96a0f387136ab0f5eb52e..352702c1a0f3d0d2e4d592bee06eaeb44fdbe494 100644
--- a/src/main/java/org/bukkit/event/vehicle/VehicleExitEvent.java
+++ b/src/main/java/org/bukkit/event/vehicle/VehicleExitEvent.java
@@ -10,6 +10,17 @@ import org.jetbrains.annotations.NotNull;
  * Raised when a living entity exits a vehicle.
  */
 public class VehicleExitEvent extends VehicleEvent implements Cancellable {
+    // EMC start
+    public enum DismountReason {
+        PLAYER, WATER, DEAD, DEAD_VEHICLE, TRANSFER, UNKNOWN, DISCONNECT;
+    }
+    public DismountReason reason = DismountReason.UNKNOWN;
+
+    @NotNull
+    public DismountReason getReason() {
+        return reason;
+    }
+    // EMC end
     private static final HandlerList handlers = new HandlerList();
     private boolean cancelled;
     private final LivingEntity exited;
