From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 14 Aug 2014 18:21:11 -0400
Subject: [PATCH] Add Player.getFishingHook() API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index da83b4cbed0be6f693c7cbb1cc032356f12d7883..8a5f16d5488d2eb0b8c0ebdbcb54587fdb49ee90 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1859,6 +1859,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         return net.kyori.adventure.text.event.HoverEvent.showEntity(op.apply(net.kyori.adventure.text.event.HoverEvent.ShowEntity.of(this.getType().getKey(), this.getUniqueId(), this.displayName())));
     }
 
+    /**
+     * Gets the current FishHook a player has attached.
+     * @return current fishing hook or null if the player is not fishing
+     */
+    @Nullable
+    public FishHook getFishingHook();
+
     /**
      * Request that the player's client download and switch resource packs.
      * <p>
