From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 14 Aug 2014 18:21:11 -0400
Subject: [PATCH] Add Player.getFishingHook() API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 6cd0b10d1dc4506cfb1e4db5e1260cb705566cec..9485daf919488baf5f363a61ae6db6c17e56c318 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2375,6 +2375,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         return net.kyori.adventure.text.event.HoverEvent.showEntity(op.apply(net.kyori.adventure.text.event.HoverEvent.ShowEntity.of(this.getType().getKey(), this.getUniqueId(), this.displayName())));
     }
 
+    /**
+     * Gets the current FishHook a player has attached.
+     * @return current fishing hook or null if the player is not fishing
+     */
+    @Nullable
+    public FishHook getFishingHook();
+
     /**
      * Request that the player's client download and switch resource packs.
      * <p>
