From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 14 Aug 2014 18:21:11 -0400
Subject: [PATCH] Add Player.getFishingHook() API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index f5ded992672dde910cd4e1be50c238b142157ff8..0c505d6cc341ec68841f3537ad9e5338801d1423 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2043,6 +2043,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
         return net.kyori.adventure.text.event.HoverEvent.showEntity(op.apply(net.kyori.adventure.text.event.HoverEvent.ShowEntity.of(this.getType().getKey(), this.getUniqueId(), this.displayName())));
     }
 
+    /**
+     * Gets the current FishHook a player has attached.
+     * @return current fishing hook or null if the player is not fishing
+     */
+    @Nullable
+    public FishHook getFishingHook();
+
     /**
      * Request that the player's client download and switch resource packs.
      * <p>
