From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 3 Jun 2015 21:39:34 -0400
Subject: [PATCH] Ignore completion starting with _


diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index b8623575b1c1b565560c2dd6438190716845a652..bac343f7ac7b1c756e7b385bb453ef95420720d4 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -214,7 +214,7 @@ public class SimpleCommandMap implements CommandMap {
 
                 String name = commandEntry.getKey(); // Use the alias, not command name
 
-                if (StringUtil.startsWithIgnoreCase(name, cmdLine)) {
+                if (name.charAt(0) != '_' && StringUtil.startsWithIgnoreCase(name, cmdLine)) { // EMC
                     completions.add(prefix + name);
                 }
             }
