From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 13 Dec 2015 01:19:27 -0500
Subject: [PATCH] Reload Entity API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index a9e455c5b3bbe4edbdb71f86f5c6eebc2f605547..59bc3ec8f6a4ef7e07aa3907151ab46d7dbf33e4 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -677,6 +677,13 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
     Spigot spigot();
     // Spigot end
 
+    /**
+     * Reattach the internal hooks if an entity with the same id is still in the world.
+     *
+     * @return True if the reload is successful.
+     */
+    public boolean reloadEntity(); // EMC
+
     // Paper start
     @NotNull
     @Override
