From 0e6158fc8d20cc92bd697abc7bd138b734b397b8 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 5 May 2015 00:34:04 -0400
Subject: [PATCH] Fix PlayerInteractEvent for Armor Stands

---
 .../org/bukkit/event/player/PlayerInteractAtEntityEvent.java    | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java b/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
index 1075dbb81..e521458a1 100644
--- a/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
@@ -32,6 +32,7 @@ public class PlayerInteractAtEntityEvent extends PlayerInteractEntityEvent {
         return position.clone();
     }
 
+    /* // EMC start
     @NotNull
     @Override
     public HandlerList getHandlers() {
@@ -42,4 +43,5 @@ public class PlayerInteractAtEntityEvent extends PlayerInteractEntityEvent {
     public static HandlerList getHandlerList() {
         return handlers;
     }
+    */ // EMC end
 }
-- 
2.21.0

