From bddd9e7c6ca9ff65ad7fc2c90b14da73dae34923 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 5 May 2015 00:34:04 -0400
Subject: [PATCH] Fix PlayerInteractEvent for Armor Stands

---
 .../org/bukkit/event/player/PlayerInteractAtEntityEvent.java    | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java b/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
index 431715de8..50d724ca1 100644
--- a/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerInteractAtEntityEvent.java
@@ -30,6 +30,7 @@ public class PlayerInteractAtEntityEvent extends PlayerInteractEntityEvent {
         return position.clone();
     }
 
+    /* // EMC start
     @Override
     public HandlerList getHandlers() {
         return handlers;
@@ -38,4 +39,5 @@ public class PlayerInteractAtEntityEvent extends PlayerInteractEntityEvent {
     public static HandlerList getHandlerList() {
         return handlers;
     }
+    */ // EMC end
 }
-- 
2.19.0

