From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 4 Dec 2016 01:14:52 -0500
Subject: [PATCH] BlockBreakNaturallyEvent


diff --git a/src/main/java/com/empireminecraft/customevents/BlockBreakNaturallyEvent.java b/src/main/java/com/empireminecraft/customevents/BlockBreakNaturallyEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..cfd5d73c2e3be9888e51c0e6268695ac95c48586
--- /dev/null
+++ b/src/main/java/com/empireminecraft/customevents/BlockBreakNaturallyEvent.java
@@ -0,0 +1,40 @@
+package com.empireminecraft.customevents;
+
+import org.bukkit.Location;
+import org.bukkit.entity.Item;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.block.BlockEvent;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Collections;
+
+public class BlockBreakNaturallyEvent extends BlockEvent {
+    private final Collection<Item> items;
+
+    public BlockBreakNaturallyEvent(@NotNull Location pos, @NotNull Collection<Item> items) {
+        super(pos.getBlock());
+        this.items = items;
+    }
+
+    public BlockBreakNaturallyEvent(@NotNull Location pos, @NotNull Item item) {
+        this(pos, Collections.singleton(item));
+    }
+
+    @NotNull
+    public Collection<Item> getItems() {
+        return items;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
