From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 4 Dec 2016 01:09:56 -0500
Subject: [PATCH] BatMoveEvent


diff --git a/src/main/java/com/empireminecraft/customevents/BatMoveEvent.java b/src/main/java/com/empireminecraft/customevents/BatMoveEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..0279886f14ae96ed0d27649f31dd226d492e1bb2
--- /dev/null
+++ b/src/main/java/com/empireminecraft/customevents/BatMoveEvent.java
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2015. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ *
+ */
+
+package com.empireminecraft.customevents;
+
+import org.bukkit.Location;
+import org.bukkit.entity.Bat;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Random;
+
+public class BatMoveEvent extends EntityEvent implements Cancellable {
+    private final Bat bat;
+    private Location moveTarget;
+    private static final Random random = new Random();
+
+    public BatMoveEvent(@NotNull Bat bat, @NotNull Location moveTarget) {
+        super(bat);
+        this.bat = bat;
+        this.moveTarget = moveTarget;
+    }
+
+    @NotNull
+    public Bat getBat() {
+        return bat;
+    }
+
+    @NotNull
+    public Location getMoveTarget() {
+        return moveTarget;
+    }
+
+    public void setMoveTarget(@NotNull Location moveTarget) {
+        this.moveTarget = moveTarget;
+    }
+
+    @NotNull
+    private static final HandlerList handlers = new HandlerList();
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+    private boolean cancelled = false;
+
+    @Override
+    public boolean isCancelled() {
+        return cancelled;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+        cancelled = cancel;
+    }
+
+    @NotNull
+    public Location generateLoc() {
+        return generateLoc(bat);
+    }
+
+    @NotNull
+    public static Location generateLoc(@NotNull Bat bat) {
+        Location loc = bat.getLocation();
+        return new Location(bat.getWorld(),
+            loc.getBlockX() + random.nextInt(15) - 7,
+            loc.getBlockY() + random.nextInt(6) - 2,
+            loc.getBlockZ() + random.nextInt(15) - 7
+        );
+    }
+}
