From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 4 Dec 2016 15:02:53 -0500
Subject: [PATCH] Add setRemovedItem to FlowerPotManipulateEvent


diff --git a/src/main/java/io/papermc/paper/event/player/PlayerFlowerPotManipulateEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerFlowerPotManipulateEvent.java
index 03881ae56b3dd68a0aa600382b4689f213ec05f3..c019f06b90cd963a5c3dc51310c3c3eb520249af 100644
--- a/src/main/java/io/papermc/paper/event/player/PlayerFlowerPotManipulateEvent.java
+++ b/src/main/java/io/papermc/paper/event/player/PlayerFlowerPotManipulateEvent.java
@@ -18,7 +18,7 @@ public class PlayerFlowerPotManipulateEvent extends PlayerEvent implements Cance
     private static final HandlerList HANDLER_LIST = new HandlerList();
 
     private final Block flowerpot;
-    private final ItemStack item;
+    private ItemStack item; // EMC - remove final
     private final boolean placing;
 
     private boolean cancelled;
@@ -49,6 +49,16 @@ public class PlayerFlowerPotManipulateEvent extends PlayerEvent implements Cance
     public ItemStack getItem() {
         return this.item;
     }
+    // EMC start
+    /**
+     * Sets a different item to be taken from the flower pot
+     *
+     * @param item the ItemStack to be given when removed from the flowerpot
+     */
+    public void setRemovedItem(ItemStack item) {
+        this.item = item;
+    }
+    // EMC end
 
     /**
      * Gets if the item is being placed into the flowerpot.
