From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 4 Dec 2016 15:02:53 -0500
Subject: [PATCH] Add setRemovedItem to FlowerPotManipulateEvent


diff --git a/src/main/java/io/papermc/paper/event/player/PlayerFlowerPotManipulateEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerFlowerPotManipulateEvent.java
index 4c141f3d8f668cdf9c75865a8e3ecbd012d9e521..4a871ed7d53797cf313b707b6a290d80c760397b 100644
--- a/src/main/java/io/papermc/paper/event/player/PlayerFlowerPotManipulateEvent.java
+++ b/src/main/java/io/papermc/paper/event/player/PlayerFlowerPotManipulateEvent.java
@@ -17,7 +17,7 @@ public class PlayerFlowerPotManipulateEvent extends PlayerEvent implements Cance
     @NotNull
     private final Block flowerpot;
     @NotNull
-    private final ItemStack item;
+    private ItemStack item; // EMC
     private final boolean placing;
 
     private boolean cancel = false;
@@ -59,6 +59,16 @@ public class PlayerFlowerPotManipulateEvent extends PlayerEvent implements Cance
     public ItemStack getItem() {
         return item;
     }
+    // EMC start
+    /**
+     * Sets a different item to be taken from the flower pot
+     *
+     * @param item the ItemStack to be given when removed from the flowerpot
+     */
+    public void setRemovedItem(@NotNull ItemStack item) {
+        this.item = item;
+    }
+    // EMC end
 
     /**
      * Gets if the item is being placed into the flowerpot.
