From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 4 Jul 2013 21:04:26 -0400
Subject: [PATCH] Temporary Metadata API

Entity, World, Chunk, Inventory API

diff --git a/src/main/java/com/empireminecraft/api/meta/TemporaryDataHolder.java b/src/main/java/com/empireminecraft/api/meta/TemporaryDataHolder.java
new file mode 100644
index 0000000000000000000000000000000000000000..d0dcb51dc585cc0e45cf6b7295ae0c0c3fe65749
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/TemporaryDataHolder.java
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.bukkit.NamespacedKey;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Map;
+
+public interface TemporaryDataHolder {
+    @NotNull Map<NamespacedKey, Object> getTemporaryMeta(); // EMC
+
+}
diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index bc8b5bc17706250b8535b1b309134843d2ce2bb1..14649707d71236f01eb99ace6ee469cef29e6906 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -1,6 +1,7 @@
 package org.bukkit;
 
 import java.util.Collection;
+
 import org.bukkit.block.Biome;
 import org.bukkit.block.Block;
 import org.bukkit.block.BlockState;
@@ -20,7 +21,7 @@ import org.jetbrains.annotations.NotNull;
  * then the chunk will only be generated as far as it needs to provide the
  * requested data.
  */
-public interface Chunk extends PersistentDataHolder {
+public interface Chunk extends PersistentDataHolder, com.empireminecraft.api.meta.TemporaryDataHolder {
 
     /**
      * Gets the X-coordinate of this chunk
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index ef32a937e6faf1e8a5d6b1207986715bae5a246c..83fcf0f71ce40f69d6e209d3dbd742d9f5f75a36 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -50,7 +50,7 @@ import org.jetbrains.annotations.Nullable;
 /**
  * Represents a world, which may contain entities, chunks and blocks
  */
-public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient, Metadatable, PersistentDataHolder, Keyed, net.kyori.adventure.audience.ForwardingAudience { // Paper
+public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient, Metadatable, PersistentDataHolder, com.empireminecraft.api.meta.TemporaryDataHolder, Keyed, net.kyori.adventure.audience.ForwardingAudience { // Paper // EMC
 
     // Paper start - void damage configuration
     /**
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index d0ae8a94db20281d3664d74718c65234eb2e5f83..770c468be5bd3dbcd0e9ba5df3c8320f8ee45e9f 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -31,7 +31,7 @@ import org.jetbrains.annotations.Nullable;
  * Not all methods are guaranteed to work/may have side effects when
  * {@link #isInWorld()} is false.
  */
-public interface Entity extends Metadatable, CommandSender, Nameable, PersistentDataHolder, net.kyori.adventure.text.event.HoverEventSource<net.kyori.adventure.text.event.HoverEvent.ShowEntity>, net.kyori.adventure.sound.Sound.Emitter { // Paper
+public interface Entity extends Metadatable, CommandSender, Nameable, PersistentDataHolder, com.empireminecraft.api.meta.TemporaryDataHolder, net.kyori.adventure.text.event.HoverEventSource<net.kyori.adventure.text.event.HoverEvent.ShowEntity>, net.kyori.adventure.sound.Sound.Emitter { // Paper // EMC
 
     /**
      * Gets the entity's current position
diff --git a/src/main/java/org/bukkit/inventory/Inventory.java b/src/main/java/org/bukkit/inventory/Inventory.java
index 0d519813c12d98b28d62e6d01d7ec6e8c2dba3c3..a16a8caed906290a396d278566aa73491c102902 100644
--- a/src/main/java/org/bukkit/inventory/Inventory.java
+++ b/src/main/java/org/bukkit/inventory/Inventory.java
@@ -3,6 +3,7 @@ package org.bukkit.inventory;
 import java.util.HashMap;
 import java.util.List;
 import java.util.ListIterator;
+
 import org.bukkit.Location;
 import org.bukkit.Material;
 import org.bukkit.entity.HumanEntity;
@@ -25,7 +26,7 @@ import org.jetbrains.annotations.Nullable;
  * @see #getContents()
  * @see #getStorageContents()
  */
-public interface Inventory extends Iterable<ItemStack> {
+public interface Inventory extends Iterable<ItemStack>, com.empireminecraft.api.meta.TemporaryDataHolder { // EMC
 
     /**
      * Returns the size of the inventory
