From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 17 Sep 2024 16:49:24 -0500
Subject: [PATCH] Add ItemStack Persistent Data Type


diff --git a/src/main/java/com/empireminecraft/api/meta/ItemStackPersistentDataType.java b/src/main/java/com/empireminecraft/api/meta/ItemStackPersistentDataType.java
new file mode 100644
index 0000000000000000000000000000000000000000..45944bfb93fb9311dc2a2f6e3e92f171b041a7a9
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/ItemStackPersistentDataType.java
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.persistence.PersistentDataAdapterContext;
+import org.bukkit.persistence.PersistentDataType;
+import org.jetbrains.annotations.NotNull;
+
+public class ItemStackPersistentDataType implements PersistentDataType<byte[], ItemStack> {
+    @Override
+    public @NotNull Class<byte[]> getPrimitiveType() {
+        return byte[].class;
+    }
+
+    @Override
+    public @NotNull Class<ItemStack> getComplexType() {
+        return ItemStack.class;
+    }
+
+    @Override
+    public byte @NotNull [] toPrimitive(@NotNull ItemStack complex, @NotNull PersistentDataAdapterContext context) {
+        return complex.serializeAsBytes();
+    }
+
+    @Override
+    public @NotNull ItemStack fromPrimitive(byte @NotNull [] primitive, @NotNull PersistentDataAdapterContext context) {
+        return ItemStack.deserializeBytes(primitive);
+    }
+}
diff --git a/src/main/java/com/empireminecraft/api/meta/TemporaryDataHolder.java b/src/main/java/com/empireminecraft/api/meta/TemporaryDataHolder.java
index d0dcb51dc585cc0e45cf6b7295ae0c0c3fe65749..1f2316536b81f6d0bfd705b15b4e14e9576100f5 100644
--- a/src/main/java/com/empireminecraft/api/meta/TemporaryDataHolder.java
+++ b/src/main/java/com/empireminecraft/api/meta/TemporaryDataHolder.java
@@ -12,6 +12,7 @@ import org.jetbrains.annotations.NotNull;
 import java.util.Map;
 
 public interface TemporaryDataHolder {
-    @NotNull Map<NamespacedKey, Object> getTemporaryMeta(); // EMC
+
+    @NotNull Map<NamespacedKey, Object> getTemporaryMeta();
 
 }
