From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 22 Sep 2024 23:03:26 -0500
Subject: [PATCH] Metadata API structure


diff --git a/src/main/java/com/empireminecraft/api/meta/Meta.java b/src/main/java/com/empireminecraft/api/meta/Meta.java
new file mode 100644
index 0000000000000000000000000000000000000000..3f211056493044ee1755e8a8dc17963292fe38c2
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/Meta.java
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class Meta {
+
+    public static @NotNull <H, K, P, C> C getMeta(@NotNull H holder, @NotNull MetaKey<H, K, C> key, @NotNull C def) {
+        return key.get(holder, def);
+    }
+
+    public static @Nullable <H, K, P, C> C getMeta(@NotNull H holder, @NotNull MetaKey<H, K, C> key) {
+        return key.get(holder);
+    }
+
+    public static <H, K, P, C> boolean hasMeta(@NotNull H holder, @NotNull MetaKey<H, K, C> key) {
+        return key.has(holder);
+    }
+
+    public static <H, K, P, C> @Nullable C setMeta(@NotNull H holder, @NotNull MetaKey<H, K, C> key, @NotNull C value) {
+        return key.set(holder, value);
+    }
+
+    public static <H, K, P, C> @Nullable C removeMeta(@NotNull H holder, @NotNull MetaKey<H, K, C> key) {
+        return key.remove(holder);
+    }
+
+    private Meta() {
+    }
+}
diff --git a/src/main/java/com/empireminecraft/api/meta/MetaKey.java b/src/main/java/com/empireminecraft/api/meta/MetaKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..1273fa1a9f5bc83cb3cb379327f2540f25ed284d
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/MetaKey.java
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.bukkit.NamespacedKey;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public interface MetaKey<H, K, C> {
+
+    @NotNull K key();
+
+    @Nullable C get(@NotNull H holder);
+
+    @NotNull C get(@NotNull H holder, @NotNull C def);
+
+    boolean has(@NotNull H holder);
+
+    @Nullable C set(@NotNull H holder, @NotNull C value);
+
+    @Nullable C remove(@NotNull H holder);
+
+}
diff --git a/src/main/java/com/empireminecraft/api/meta/PersistentListMetaKey.java b/src/main/java/com/empireminecraft/api/meta/PersistentListMetaKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..5f6a3f821250e49445049c63907a404be060c87f
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/PersistentListMetaKey.java
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.persistence.ListPersistentDataType;
+import org.bukkit.persistence.PersistentDataHolder;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+public interface PersistentListMetaKey<H, P, C> extends MetaKey<H, NamespacedKey, List<C>> {
+
+    @NotNull ListPersistentDataType<P, C> type();
+
+}
diff --git a/src/main/java/com/empireminecraft/api/meta/PersistentMetaKey.java b/src/main/java/com/empireminecraft/api/meta/PersistentMetaKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..f0ebc79ed9e0bcd34571dada2d0d9e0b3a31dce4
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/PersistentMetaKey.java
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.persistence.PersistentDataType;
+import org.jetbrains.annotations.NotNull;
+
+public interface PersistentMetaKey<H, P, C> extends MetaKey<H, NamespacedKey, C> {
+
+    @NotNull PersistentDataType<P, C> type();
+
+}
diff --git a/src/main/java/com/empireminecraft/api/meta/TemporaryMetaKey.java b/src/main/java/com/empireminecraft/api/meta/TemporaryMetaKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..5b4636179631f0df0936b9125cfdcd88e5e84003
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/TemporaryMetaKey.java
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.bukkit.NamespacedKey;
+
+public interface TemporaryMetaKey<H, C> extends MetaKey<H, NamespacedKey, C> {
+}
