From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 4 Dec 2016 14:59:41 -0500
Subject: [PATCH] EntityEffectAddedEvent


diff --git a/src/main/java/com/empireminecraft/customevents/EntityEffectAddedEvent.java b/src/main/java/com/empireminecraft/customevents/EntityEffectAddedEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..be1d14026c8bfd063b1fe29737bd57cb04d71293
--- /dev/null
+++ b/src/main/java/com/empireminecraft/customevents/EntityEffectAddedEvent.java
@@ -0,0 +1,54 @@
+package com.empireminecraft.customevents;
+
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.bukkit.potion.PotionEffect;
+import org.jetbrains.annotations.NotNull;
+
+public class EntityEffectAddedEvent extends Event implements Cancellable {
+    private static final HandlerList handlers = new HandlerList();
+    private boolean canceled;
+
+    private final LivingEntity entity;
+    private PotionEffect effect;
+
+    public EntityEffectAddedEvent(@NotNull LivingEntity entity, @NotNull PotionEffect effect) {
+        this.entity = entity;
+        this.effect = effect;
+    }
+
+    @NotNull
+    public PotionEffect getEffect() {
+        return effect;
+    }
+
+    public void setEffect(@NotNull PotionEffect effect) {
+        this.effect = effect;
+    }
+
+    @NotNull
+    public LivingEntity getEntity() {
+        return entity;
+    }
+
+    public boolean isCancelled() {
+        return canceled;
+    }
+
+    public void setCancelled(boolean cancel) {
+        canceled = cancel;
+    }
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+}
