From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 22 Sep 2024 23:03:26 -0500
Subject: [PATCH] Metadata API structure


diff --git a/src/main/java/com/empireminecraft/api/meta/Meta.java b/src/main/java/com/empireminecraft/api/meta/Meta.java
new file mode 100644
index 0000000000000000000000000000000000000000..b3ee57e2d3cb0502944959339a14d3e0dc512336
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/Meta.java
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class Meta {
+
+    public static @NotNull <H, P, C> C getMeta(@NotNull H holder, @NotNull MetaKey<H, C> key, @NotNull C def) {
+        return key.get(holder, def);
+    }
+
+    public static @Nullable <H, P, C> C getMeta(@NotNull H holder, @NotNull MetaKey<H, C> key) {
+        return key.get(holder);
+    }
+
+    public static <H, P, C> boolean hasMeta(@NotNull H holder, @NotNull MetaKey<H, C> key) {
+        return key.has(holder);
+    }
+
+    public static <H, P, C> @Nullable C setMeta(@NotNull H holder, @NotNull MetaKey<H, C> key, @NotNull C value) {
+        return key.set(holder, value);
+    }
+
+    public static <H, P, C> @Nullable C removeMeta(@NotNull H holder, @NotNull MetaKey<H, C> key) {
+        return key.remove(holder);
+    }
+
+    private Meta() {
+    }
+}
diff --git a/src/main/java/com/empireminecraft/api/meta/MetaKey.java b/src/main/java/com/empireminecraft/api/meta/MetaKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..7e29c07fa76313bd87705b8faf4a055eac30604c
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/MetaKey.java
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.bukkit.NamespacedKey;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public interface MetaKey<H, C> {
+
+    @NotNull NamespacedKey key();
+
+    @Nullable C get(@NotNull H holder);
+
+    @NotNull C get(@NotNull H holder, @NotNull C def);
+
+    boolean has(@NotNull H holder);
+
+    @Nullable C set(@NotNull H holder, @NotNull C value);
+
+    @Nullable C remove(@NotNull H holder);
+
+}
diff --git a/src/main/java/com/empireminecraft/api/meta/PersistentListMetaKey.java b/src/main/java/com/empireminecraft/api/meta/PersistentListMetaKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..7087ce8d2c6f72583c6a75fb21af5357bf6337f4
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/PersistentListMetaKey.java
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.bukkit.persistence.ListPersistentDataType;
+import org.bukkit.persistence.PersistentDataHolder;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+public interface PersistentListMetaKey<H, P, C> extends MetaKey<H, List<C>> {
+
+    @NotNull ListPersistentDataType<P, C> type();
+
+}
diff --git a/src/main/java/com/empireminecraft/api/meta/PersistentMetaKey.java b/src/main/java/com/empireminecraft/api/meta/PersistentMetaKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..4274595748f0c0f07d2c440ea853f389957d307b
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/PersistentMetaKey.java
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+import org.bukkit.persistence.PersistentDataType;
+import org.jetbrains.annotations.NotNull;
+
+public interface PersistentMetaKey<H, P, C> extends MetaKey<H, C> {
+
+    @NotNull PersistentDataType<P, C> type();
+
+}
diff --git a/src/main/java/com/empireminecraft/api/meta/TemporaryMetaKey.java b/src/main/java/com/empireminecraft/api/meta/TemporaryMetaKey.java
new file mode 100644
index 0000000000000000000000000000000000000000..732e8f673b73da26b6248dc74ed3ad0eee53c3e2
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/meta/TemporaryMetaKey.java
@@ -0,0 +1,10 @@
+/*
+ * Copyright (c) 2024. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ */
+
+package com.empireminecraft.api.meta;
+
+public interface TemporaryMetaKey<H, C> extends MetaKey<H, C> {
+}
