From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 4 Dec 2016 01:19:40 -0500
Subject: [PATCH] ServerShutdownEvent


diff --git a/src/main/java/com/empireminecraft/customevents/ServerShutdownEvent.java b/src/main/java/com/empireminecraft/customevents/ServerShutdownEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..4d4be1c7df396bbb8ffe5fc1b5281c6f2cf2c576
--- /dev/null
+++ b/src/main/java/com/empireminecraft/customevents/ServerShutdownEvent.java
@@ -0,0 +1,35 @@
+package com.empireminecraft.customevents;
+
+import org.bukkit.Bukkit;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class ServerShutdownEvent extends Event {
+    private String reason;
+    public ServerShutdownEvent(@NotNull String reason) {
+        super(!Bukkit.isPrimaryThread());
+        this.reason = reason;
+    }
+
+    @NotNull
+    public String getReason() {
+        return reason;
+    }
+
+    public void setReason(@NotNull String reason) {
+        this.reason = reason;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
