From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:29:02 -0500
Subject: [PATCH] Add Ranged Attack API


diff --git a/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java b/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
index 09c82908f63233febfe1e07fe756f1c39d23d44f..2e2f5e6ae447c2dab725c5b2201f4cc23b314712 100644
--- a/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
+++ b/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
@@ -2,6 +2,7 @@ package com.destroystokyo.paper.entity;
 
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Mob;
+import org.jetbrains.annotations.Nullable;
 import org.jspecify.annotations.NullMarked;
 
 @NullMarked
@@ -33,4 +34,10 @@ public interface RangedEntity extends Mob {
     default boolean isChargingAttack() {
         return this.isHandRaised();
     }
+
+    // EMC start
+    void setRangedAttackDistance(float range);
+
+    void setRangedAttackSpeed(@Nullable Integer min, @Nullable Integer max);
+    // EMC end
 }
