From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:32:44 -0500
Subject: [PATCH] Add Wolf setAlwaysAngry API


diff --git a/src/main/java/org/bukkit/entity/Wolf.java b/src/main/java/org/bukkit/entity/Wolf.java
index c73489f4b745bc84501ce94f0227b034d9768eae..c870778f0cd43f67f0ad25233c4a99d3980bd779 100644
--- a/src/main/java/org/bukkit/entity/Wolf.java
+++ b/src/main/java/org/bukkit/entity/Wolf.java
@@ -73,6 +73,7 @@ public interface Wolf extends Tameable, Sittable, io.papermc.paper.entity.Collar
      */
     public void setInterested(boolean interested);
 
+    void setAlwaysAngry(boolean alwaysAngry); // EMC
     /**
      * Get the variant of this wolf.
      *
