From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:32:44 -0500
Subject: [PATCH] Add Wolf setAlwaysAngry API


diff --git a/src/main/java/org/bukkit/entity/Wolf.java b/src/main/java/org/bukkit/entity/Wolf.java
index 4b84c04675775e2a606630b00de8afe51665cebc..89ab37a924eba02362f03d4967e2da6032df1adc 100644
--- a/src/main/java/org/bukkit/entity/Wolf.java
+++ b/src/main/java/org/bukkit/entity/Wolf.java
@@ -74,6 +74,7 @@ public interface Wolf extends Tameable, Sittable, io.papermc.paper.entity.Collar
      */
     public void setInterested(boolean interested);
 
+    void setAlwaysAngry(boolean alwaysAngry); // EMC
     /**
      * Get the variant of this wolf.
      *
