From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:38:59 -0500
Subject: [PATCH] Add EnderSignal API


diff --git a/src/main/java/org/bukkit/entity/EnderSignal.java b/src/main/java/org/bukkit/entity/EnderSignal.java
index c8c74651e1b434e5a0fce6d9e8d01754b2d7287a..88b180bfd9243f2bad4d31bab91c348ed8365626 100644
--- a/src/main/java/org/bukkit/entity/EnderSignal.java
+++ b/src/main/java/org/bukkit/entity/EnderSignal.java
@@ -91,4 +91,16 @@ public interface EnderSignal extends Entity {
      * @param timer how long (in ticks) this EnderSignal has been alive.
      */
     public void setDespawnTimer(int timer);
+
+    // EMC start
+    void setExactTarget(boolean exact);
+
+    void setLife(int lifeTime);
+
+    int getLife();
+
+    void setSpeed(double speed);
+
+    double getSpeed();
+    // EMC end
 }
