From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:29:02 -0500
Subject: [PATCH] Add Ranged Attack API


diff --git a/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java b/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
index f2e3233a3d1744e32fb76d3731b9858ef0067e30..1d14c759635e09e2e7e12cedf3415edae13a4360 100644
--- a/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
+++ b/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
@@ -3,6 +3,7 @@ package com.destroystokyo.paper.entity;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Mob;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 public interface RangedEntity extends Mob {
     /**
@@ -28,4 +29,10 @@ public interface RangedEntity extends Mob {
     default boolean isChargingAttack() {
         return isHandRaised();
     }
+
+    // EMC start
+    void setRangedAttackDistance(float range);
+
+    void setRangedAttackSpeed(@Nullable Integer min, @Nullable Integer max);
+    // EMC end
 }
