From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:29:02 -0500
Subject: [PATCH] Add Ranged Attack API


diff --git a/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java b/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
index a087a04c525ea14b197c65063b43ebf097ddc5c2..b591f0f0506198d3e8928747946fc1aa21080698 100644
--- a/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
+++ b/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
@@ -3,6 +3,7 @@ package com.destroystokyo.paper.entity;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Mob;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 public interface RangedEntity extends Mob {
     /**
@@ -32,4 +33,10 @@ public interface RangedEntity extends Mob {
     default boolean isChargingAttack() {
         return isHandRaised();
     }
+
+    // EMC start
+    void setRangedAttackDistance(float range);
+
+    void setRangedAttackSpeed(@Nullable Integer min, @Nullable Integer max);
+    // EMC end
 }
