From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:29:02 -0500
Subject: [PATCH] Add Ranged Attack API


diff --git a/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java b/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
index f2e3233a3d1744e32fb76d3731b9858ef0067e30..1d14c759635e09e2e7e12cedf3415edae13a4360 100644
--- a/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
+++ b/src/main/java/com/destroystokyo/paper/entity/RangedEntity.java
@@ -3,6 +3,7 @@ package com.destroystokyo.paper.entity;
 import org.bukkit.entity.LivingEntity;
 import org.bukkit.entity.Mob;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 public interface RangedEntity extends Mob {
     /**
@@ -28,4 +29,10 @@ public interface RangedEntity extends Mob {
     default boolean isChargingAttack() {
         return isHandRaised();
     }
+
+    // EMC start
+    void setRangedAttackDistance(float range);
+
+    void setRangedAttackSpeed(@Nullable Integer min, @Nullable Integer max);
+    // EMC end
 }
diff --git a/src/main/java/com/empireminecraft/api/EAPI_Entity.java b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
index 14880bb8534c13ee4cfec632d0bb1a52a6057e20..85cc9e71c726c926a5697341794a50b5840cf2b8 100644
--- a/src/main/java/com/empireminecraft/api/EAPI_Entity.java
+++ b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
@@ -143,10 +143,14 @@ public interface EAPI_Entity {
     void setTargetRange(@NotNull Mob mob, @Nullable Double range);
 
     @Deprecated
-    void setRangedAttackDistance(@NotNull RangedEntity ranged, float range);
+    default void setRangedAttackDistance(@NotNull RangedEntity ranged, float range) {
+        ranged.setRangedAttackDistance(range);
+    }
 
     @Deprecated
-    void setRangedAttackSpeed(@NotNull RangedEntity ranged, @Nullable Integer min, @Nullable Integer max);
+    default void setRangedAttackSpeed(@NotNull RangedEntity ranged, @Nullable Integer min, @Nullable Integer max) {
+        ranged.setRangedAttackSpeed(min, max);
+    }
 
     @Deprecated
     void setAlwaysAngry(@NotNull Wolf wolf, boolean alwaysAngry);
