From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 4 Dec 2016 22:30:15 -0500
Subject: [PATCH] Chat API


diff --git a/src/main/java/com/empireminecraft/api/API.java b/src/main/java/com/empireminecraft/api/API.java
index 0aff9298f82ec548ab1be01a5f13282b33b7305b..7eacc2b2d709b0173fb33c8c467b70cb53bc2522 100644
--- a/src/main/java/com/empireminecraft/api/API.java
+++ b/src/main/java/com/empireminecraft/api/API.java
@@ -23,6 +23,7 @@
 
 package com.empireminecraft.api;
 
+import com.empireminecraft.api.chat.EAPI_Chat;
 import com.empireminecraft.api.meta.EAPI_Meta;
 import org.apache.commons.lang.exception.ExceptionUtils;
 import org.bukkit.Bukkit;
@@ -34,6 +35,7 @@ public abstract class API {
     public static EAPI_Entity entity;
     public static EAPI_Misc misc;
     public static EAPI_Meta meta;
+    public static EAPI_Chat chat;
 
     @NotNull
     public static String stack() {
diff --git a/src/main/java/com/empireminecraft/api/chat/ChatBuilder.java b/src/main/java/com/empireminecraft/api/chat/ChatBuilder.java
new file mode 100644
index 0000000000000000000000000000000000000000..4e9aa769119323a8e6ebdfcb16b11c575d142a82
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/chat/ChatBuilder.java
@@ -0,0 +1,30 @@
+package com.empireminecraft.api.chat;
+
+import org.bukkit.ChatColor;
+import org.jetbrains.annotations.NotNull;
+
+public interface ChatBuilder {
+    @NotNull
+    ChatBuilder color(@NotNull ChatColor color);
+
+    @NotNull
+    ChatBuilder color(@NotNull String hex);
+
+    @NotNull
+    ChatBuilder obfuscated();
+
+    @NotNull
+    ChatBuilder bold();
+
+    @NotNull
+    ChatBuilder strikethrough();
+
+    @NotNull
+    ChatBuilder underline();
+
+    @NotNull
+    ChatBuilder italic();
+
+    @NotNull
+    ChatBuilder reset();
+}
diff --git a/src/main/java/com/empireminecraft/api/chat/EAPI_Chat.java b/src/main/java/com/empireminecraft/api/chat/EAPI_Chat.java
new file mode 100644
index 0000000000000000000000000000000000000000..de3a5ccb8c0836a22e53ac8f1e8ffb75c4ae3e55
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/chat/EAPI_Chat.java
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2016 Starlis LLC / Daniel Ennis (Aikar) - MIT License
+ *  
+ *  Permission is hereby granted, free of charge, to any person obtaining
+ *  a copy of this software and associated documentation files (the
+ *  "Software"), to deal in the Software without restriction, including
+ *  without limitation the rights to use, copy, modify, merge, publish,
+ *  distribute, sublicense, and/or sell copies of the Software, and to
+ *  permit persons to whom the Software is furnished to do so, subject to
+ *  the following conditions:
+ *  
+ *  The above copyright notice and this permission notice shall be
+ *  included in all copies or substantial portions of the Software.
+ *  
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ *  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ *  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ *  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ *  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package com.empireminecraft.api.chat;
+
+import org.bukkit.entity.Player;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.UUID;
+
+public interface EAPI_Chat {
+
+    @NotNull
+    FancifulChat createFancyChat();
+
+    //TODO:
+    // FancifulChat fromJson(String json);
+
+    void sendTitleAnimations(@NotNull Player player, int fadeIn, int stay, int fadeOut);
+    void sendTitle(@NotNull Player player, @NotNull FancifulChat title);
+    void sendSubtitle(@NotNull Player player, @NotNull FancifulChat subtitle);
+
+    void sendAction(@NotNull Player player, @NotNull FancifulChat message);
+
+    void sendChat(@NotNull Player player, @NotNull FancifulChat message);
+
+    void sendChat(@Nullable UUID sender, @NotNull Player player, @NotNull FancifulChat message);
+
+    void sendSystem(@NotNull Player player, @NotNull FancifulChat message);
+
+    /* TODO: Add advancement api
+    default void sentToast(Player player, FancifulText title, FancifulText description, ItemStack icon) {
+        sentToast(player, title, description, icon, null, null, 0F, 0F);
+    }
+    void sentToast(Player player, FancifulText title, FancifulText description, ItemStack icon, String frame, String background, Float x, Float y);
+    */
+
+    //todo; Use FancifulPage for book page setting
+}
diff --git a/src/main/java/com/empireminecraft/api/chat/FancifulChat.java b/src/main/java/com/empireminecraft/api/chat/FancifulChat.java
new file mode 100644
index 0000000000000000000000000000000000000000..ed17a0d82ad48ee1d3036283b2355a50b2499eaf
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/chat/FancifulChat.java
@@ -0,0 +1,49 @@
+package com.empireminecraft.api.chat;
+
+import org.bukkit.entity.Entity;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+public interface FancifulChat {
+    @NotNull
+    FancifulChat then(@NotNull String text);
+
+    @NotNull
+    FancifulChat then(@NotNull Object obj);
+
+    //TODO:
+    // String toJson();
+
+    //TODO:
+    // FancifulChat itemTooltip(@NotNull String itemJSON);
+
+    @NotNull
+    FancifulChat tooltip(@NotNull ItemStack itemStack);
+
+    @NotNull
+    FancifulChat tooltip(@NotNull String text);
+
+    @NotNull
+    FancifulChat tooltip(@NotNull String[] lines);
+
+    @NotNull
+    FancifulChat tooltip(@NotNull Entity entity);
+
+    @NotNull
+    FancifulChat link(@NotNull String url);
+
+    @NotNull
+    FancifulChat insert(@NotNull String text);
+
+    @NotNull
+    FancifulChat suggest(@NotNull String command);
+
+    @NotNull
+    FancifulChat command(@NotNull String command);
+
+    @NotNull
+    FancifulChat copy(@NotNull String text);
+
+    @NotNull
+    FancifulChat page(@NotNull String page);
+}
