From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:32:44 -0500
Subject: [PATCH] Add Wolf setAlwaysAngry API


diff --git a/src/main/java/com/empireminecraft/api/EAPI_Entity.java b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
index 85cc9e71c726c926a5697341794a50b5840cf2b8..db1f8c345388e64772960eebaf7b3e23420cd4ba 100644
--- a/src/main/java/com/empireminecraft/api/EAPI_Entity.java
+++ b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
@@ -153,7 +153,9 @@ public interface EAPI_Entity {
     }
 
     @Deprecated
-    void setAlwaysAngry(@NotNull Wolf wolf, boolean alwaysAngry);
+    default void setAlwaysAngry(@NotNull Wolf wolf, boolean alwaysAngry) {
+        wolf.setAlwaysAngry(alwaysAngry);
+    }
 
     @Deprecated
     void setEnderSignalExactTarget(@NotNull EnderSignal enderSignal, boolean exact);
diff --git a/src/main/java/org/bukkit/entity/Wolf.java b/src/main/java/org/bukkit/entity/Wolf.java
index 84db38388bf7a58e66d6cd29620b4fe64b0a897e..4fb97ae6050fbfed255fe98065bf97d597b990c1 100644
--- a/src/main/java/org/bukkit/entity/Wolf.java
+++ b/src/main/java/org/bukkit/entity/Wolf.java
@@ -69,4 +69,6 @@ public interface Wolf extends Tameable, Sittable, io.papermc.paper.entity.Collar
      * @param interested Whether the wolf is interested
      */
     public void setInterested(boolean interested);
+
+    void setAlwaysAngry(boolean alwaysAngry); // EMC
 }
