From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:48:20 -0500
Subject: [PATCH] Add Ghast API


diff --git a/src/main/java/com/empireminecraft/api/EAPI_Entity.java b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
index a3560995f48d16586bcf825de1072b8b17bcd397..7e196935aff7b4ddf1b9e68d89447b84b171b8c6 100644
--- a/src/main/java/com/empireminecraft/api/EAPI_Entity.java
+++ b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
@@ -203,8 +203,12 @@ public interface EAPI_Entity {
     }
 
     @Deprecated
-    int getFireballCooldown(@NotNull Ghast ghast);
+    default int getFireballCooldown(@NotNull Ghast ghast) {
+        return ghast.getFireballCooldown();
+    }
 
     @Deprecated
-    void setFireballCooldown(@NotNull Ghast ghast, int cooldown);
+    default void setFireballCooldown(@NotNull Ghast ghast, int cooldown) {
+        ghast.setFireballCooldown(cooldown);
+    }
 }
diff --git a/src/main/java/org/bukkit/entity/Ghast.java b/src/main/java/org/bukkit/entity/Ghast.java
index 9c618a27d590f186f29c5d9094fc565efd40ca49..cc52ec5b42e7808100002a96a1fa7a3e7cb8f793 100644
--- a/src/main/java/org/bukkit/entity/Ghast.java
+++ b/src/main/java/org/bukkit/entity/Ghast.java
@@ -1,5 +1,7 @@
 package org.bukkit.entity;
 
+import org.jetbrains.annotations.NotNull;
+
 /**
  * Represents a Ghast.
  */
@@ -35,4 +37,10 @@ public interface Ghast extends Flying, Enemy {
      */
     void setExplosionPower(int explosionPower);
     // Paper end
+
+    // EMC start
+    int getFireballCooldown();
+
+    void setFireballCooldown(int cooldown);
+    // EMC end
 }
