From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:41:51 -0500
Subject: [PATCH] Add Blaze API


diff --git a/src/main/java/com/empireminecraft/api/EAPI_Entity.java b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
index e3ad734061790e63211e46b74cd525852b294c84..a3560995f48d16586bcf825de1072b8b17bcd397 100644
--- a/src/main/java/com/empireminecraft/api/EAPI_Entity.java
+++ b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
@@ -183,16 +183,24 @@ public interface EAPI_Entity {
     }
 
     @Deprecated
-    void setFireballs(@NotNull Blaze blaze, int num);
+    default void setFireballs(@NotNull Blaze blaze, int num) {
+        blaze.setFireballs(num);
+    }
 
     @Deprecated
-    int getNumFireballs(@NotNull Blaze blaze);
+    default int getNumFireballs(@NotNull Blaze blaze) {
+        return blaze.getNumFireballs();
+    }
 
     @Deprecated
-    int getTimeBetweenFireballs(@NotNull Blaze blaze);
+    default int getTimeBetweenFireballs(@NotNull Blaze blaze) {
+        return blaze.getTimeBetweenFireballs();
+    }
 
     @Deprecated
-    void setTimeBetweenFireballs(@NotNull Blaze blaze, int timeBetweenFireballs);
+    default void setTimeBetweenFireballs(@NotNull Blaze blaze, int timeBetweenFireballs) {
+        blaze.setTimeBetweenFireballs(timeBetweenFireballs);
+    }
 
     @Deprecated
     int getFireballCooldown(@NotNull Ghast ghast);
diff --git a/src/main/java/org/bukkit/entity/Blaze.java b/src/main/java/org/bukkit/entity/Blaze.java
index 7a5505b7fa84629f81c152fb6da90bdb2392f9df..c77da402fda60c2fa406688b15ec3e4f3b867c95 100644
--- a/src/main/java/org/bukkit/entity/Blaze.java
+++ b/src/main/java/org/bukkit/entity/Blaze.java
@@ -5,4 +5,13 @@ package org.bukkit.entity;
  */
 public interface Blaze extends Monster {
 
+    // EMC start
+    void setFireballs(int num);
+
+    int getNumFireballs();
+
+    int getTimeBetweenFireballs();
+
+    void setTimeBetweenFireballs(int timeBetweenFireballs);
+    // EMC end
 }
