From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:48:20 -0500
Subject: [PATCH] Add Ghast API


diff --git a/src/main/java/org/bukkit/entity/Ghast.java b/src/main/java/org/bukkit/entity/Ghast.java
index 9c618a27d590f186f29c5d9094fc565efd40ca49..cc52ec5b42e7808100002a96a1fa7a3e7cb8f793 100644
--- a/src/main/java/org/bukkit/entity/Ghast.java
+++ b/src/main/java/org/bukkit/entity/Ghast.java
@@ -1,5 +1,7 @@
 package org.bukkit.entity;
 
+import org.jetbrains.annotations.NotNull;
+
 /**
  * Represents a Ghast.
  */
@@ -35,4 +37,10 @@ public interface Ghast extends Flying, Enemy {
      */
     void setExplosionPower(int explosionPower);
     // Paper end
+
+    // EMC start
+    int getFireballCooldown();
+
+    void setFireballCooldown(int cooldown);
+    // EMC end
 }
diff --git a/src/main/java/org/bukkit/entity/Spider.java b/src/main/java/org/bukkit/entity/Spider.java
index f9ee8cc75a6eae3c1fad58b57c70f64b71977c59..802053505bb99753e6a5492c4cb6b5d0ffebc4c0 100644
--- a/src/main/java/org/bukkit/entity/Spider.java
+++ b/src/main/java/org/bukkit/entity/Spider.java
@@ -3,4 +3,5 @@ package org.bukkit.entity;
 /**
  * Represents a Spider.
  */
-public interface Spider extends Monster {}
+public interface Spider extends Monster {
+}
