From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 11:38:59 -0500
Subject: [PATCH] Add EnderSignal API


diff --git a/src/main/java/com/empireminecraft/api/EAPI_Entity.java b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
index db1f8c345388e64772960eebaf7b3e23420cd4ba..e3ad734061790e63211e46b74cd525852b294c84 100644
--- a/src/main/java/com/empireminecraft/api/EAPI_Entity.java
+++ b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
@@ -158,19 +158,29 @@ public interface EAPI_Entity {
     }
 
     @Deprecated
-    void setEnderSignalExactTarget(@NotNull EnderSignal enderSignal, boolean exact);
+    default void setEnderSignalExactTarget(@NotNull EnderSignal enderSignal, boolean exact) {
+        enderSignal.setExactTarget(exact);
+    }
 
     @Deprecated
-    void setEnderSignalLife(@NotNull EnderSignal enderSignal, int lifeTime);
+    default void setEnderSignalLife(@NotNull EnderSignal enderSignal, int lifeTime) {
+        enderSignal.setLife(lifeTime);
+    }
 
     @Deprecated
-    int getEnderSignalLife(@NotNull EnderSignal enderSignal);
+    default int getEnderSignalLife(@NotNull EnderSignal enderSignal) {
+        return enderSignal.getLife();
+    }
 
     @Deprecated
-    void setEnderSignalSpeed(@NotNull EnderSignal enderSignal, double speed);
+    default void setEnderSignalSpeed(@NotNull EnderSignal enderSignal, double speed) {
+        enderSignal.setSpeed(speed);
+    }
 
     @Deprecated
-    double getEnderSignalSpeed(@NotNull EnderSignal enderSignal);
+    default double getEnderSignalSpeed(@NotNull EnderSignal enderSignal) {
+        return enderSignal.getSpeed();
+    }
 
     @Deprecated
     void setFireballs(@NotNull Blaze blaze, int num);
diff --git a/src/main/java/org/bukkit/entity/EnderSignal.java b/src/main/java/org/bukkit/entity/EnderSignal.java
index c8c74651e1b434e5a0fce6d9e8d01754b2d7287a..88b180bfd9243f2bad4d31bab91c348ed8365626 100644
--- a/src/main/java/org/bukkit/entity/EnderSignal.java
+++ b/src/main/java/org/bukkit/entity/EnderSignal.java
@@ -91,4 +91,16 @@ public interface EnderSignal extends Entity {
      * @param timer how long (in ticks) this EnderSignal has been alive.
      */
     public void setDespawnTimer(int timer);
+
+    // EMC start
+    void setExactTarget(boolean exact);
+
+    void setLife(int lifeTime);
+
+    int getLife();
+
+    void setSpeed(double speed);
+
+    double getSpeed();
+    // EMC end
 }
