From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 6 Dec 2016 22:31:40 -0500
Subject: [PATCH] Entity AI API


diff --git a/src/main/java/com/empireminecraft/api/EAPI_Entity.java b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
index 91893d7ff213881e817f9d691fa5dfa4b58cce9d..98c310a673dfd3d9fe4ab406c064f6449dd46f73 100644
--- a/src/main/java/com/empireminecraft/api/EAPI_Entity.java
+++ b/src/main/java/com/empireminecraft/api/EAPI_Entity.java
@@ -23,9 +23,21 @@
 
 package com.empireminecraft.api;
 
+import com.destroystokyo.paper.entity.RangedEntity;
+import org.bukkit.entity.Blaze;
+import org.bukkit.entity.Ghast;
+import org.bukkit.entity.Mob;
 import com.empireminecraft.api.EntityTask.TaskHandler;
+import org.bukkit.entity.Creature;
+import org.bukkit.entity.EnderSignal;
 import org.bukkit.entity.Entity;
+import org.bukkit.entity.Item;
+import org.bukkit.entity.Spider;
+import org.bukkit.entity.Wolf;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Set;
 
 public interface EAPI_Entity {
 
@@ -93,4 +105,69 @@ public interface EAPI_Entity {
 
     @NotNull
     <T extends Entity> EntityTask<T> scheduleTask(@NotNull T entity, int interval, @NotNull EntityTask<T> task);
+
+    boolean isEntityDisabled(@NotNull org.bukkit.entity.Entity entity);
+
+    void setItemCanDespawn(@NotNull Item item, boolean canDespawn);
+
+    /**
+     * Disables Ticking on an Entity
+     *
+     * @param entity
+     * @param disabled
+     */
+    void setDisabledEntity(@NotNull org.bukkit.entity.Entity entity, boolean disabled);
+
+    void setFireProof(@NotNull Mob mob, boolean flag);
+
+    void setPanicProof(@NotNull Mob mob, boolean flag);
+
+    void makeAggressive(@NotNull Creature creature, float range);
+
+    void makeAggressiveDuringDay(@NotNull Spider spider, boolean attack);
+
+    boolean isAggressive(@NotNull Creature creature);
+
+    void makePeaceful(@NotNull Mob mob);
+
+    void addGoalByName(@NotNull Mob mob, @NotNull String goalName, boolean isTargetGoal, int tickRate, @NotNull Class<?>[] argClasses, @NotNull Object... args);
+
+    void removeGoalByName(@NotNull Mob mob, @NotNull String goalName);
+
+    @NotNull
+    Set<String> listGoalNames(@NotNull Mob mob);
+
+    void setEntitySize(@NotNull Entity mob, float width, float height);
+
+    void setTargetRange(@NotNull Mob mob, @Nullable Double range);
+
+    void setRangedAttackDistance(@NotNull RangedEntity ranged, float range);
+
+    void setRangedAttackSpeed(@NotNull RangedEntity ranged, @Nullable Integer min, @Nullable Integer max);
+
+    void setAlwaysAngry(@NotNull Wolf wolf, boolean alwaysAngry);
+
+    void respawnEntity(@NotNull org.bukkit.entity.Entity entity);
+
+    void setEnderSignalExactTarget(@NotNull EnderSignal enderSignal, boolean exact);
+
+    void setEnderSignalLife(@NotNull EnderSignal enderSignal, int lifeTime);
+
+    int getEnderSignalLife(@NotNull EnderSignal enderSignal);
+
+    void setEnderSignalSpeed(@NotNull EnderSignal enderSignal, double speed);
+
+    double getEnderSignalSpeed(@NotNull EnderSignal enderSignal);
+
+    void setFireballs(@NotNull Blaze blaze, int num);
+
+    int getNumFireballs(@NotNull Blaze blaze);
+
+    int getTimeBetweenFireballs(@NotNull Blaze blaze);
+
+    void setTimeBetweenFireballs(@NotNull Blaze blaze, int timeBetweenFireballs);
+
+    int getFireballCooldown(@NotNull Ghast ghast);
+
+    void setFireballCooldown(@NotNull Ghast ghast, int cooldown);
 }
