From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 2 Jul 2023 08:12:49 -0500
Subject: [PATCH] Add Spider API


diff --git a/src/main/java/org/bukkit/entity/Spider.java b/src/main/java/org/bukkit/entity/Spider.java
index 802053505bb99753e6a5492c4cb6b5d0ffebc4c0..4c386e9e8e5bfe10c4f1d7de9417cc381d3acb7c 100644
--- a/src/main/java/org/bukkit/entity/Spider.java
+++ b/src/main/java/org/bukkit/entity/Spider.java
@@ -4,4 +4,5 @@ package org.bukkit.entity;
  * Represents a Spider.
  */
 public interface Spider extends Monster {
+    void makeAggressiveDuringDay(boolean attack); // EMC
 }
