From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jun 2018 02:08:40 -0400
Subject: [PATCH] Human#getAttackPct


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 261475159221ddd528d13991c06be5eaf908030c..a711570f259de4edf7f2143b5246844808741030 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -26,7 +26,7 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
     @Override
     org.bukkit.inventory.@NotNull EntityEquipment getEquipment();
     // Paper end
-
+    float getAttackPct(); // EMC
     /**
      * Returns the name of this player
      *
