From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jun 2018 02:08:40 -0400
Subject: [PATCH] Human#getAttackPct


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 3b204144e2d245098b3dc23b8779f9ed817bb6d1..d4f35f54338282e6946c76e562584d109176607b 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -26,7 +26,7 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
     @Override
     org.bukkit.inventory.@NotNull EntityEquipment getEquipment();
     // Paper end
-
+    float getAttackPct(); // EMC
     /**
      * Returns the name of this player
      *
