From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jun 2018 02:08:40 -0400
Subject: [PATCH] Human#getAttackPct


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index 5dbbeb92aec212d2e849d16320ed9c084b99fc35..4c4de37ead854032deea1e51e00806e084caa5e0 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -25,7 +25,7 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
     @Override
     org.bukkit.inventory.@NotNull EntityEquipment getEquipment();
     // Paper end
-
+    float getAttackPct(); // EMC
     /**
      * Returns the name of this player
      *
