From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jun 2018 02:08:40 -0400
Subject: [PATCH] Human#getAttackPct


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index e8b6cfe7e454c666b4d60b702a3b211dab238830..8f0804c9e616f2c54c29e01459ce2da782ea0072 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -25,7 +25,7 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
     @Override
     org.bukkit.inventory.@NotNull EntityEquipment getEquipment();
     // Paper end
-
+    float getAttackPct(); // EMC
     /**
      * Returns the name of this player
      *
