From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jun 2018 02:08:40 -0400
Subject: [PATCH] Human#getAttackPct


diff --git a/src/main/java/org/bukkit/entity/HumanEntity.java b/src/main/java/org/bukkit/entity/HumanEntity.java
index f17ef44bc02e97ff084e5abfde8d6c6822f00d62..c8310c643fb5fb64d09bd10faed77b0826b98a3c 100644
--- a/src/main/java/org/bukkit/entity/HumanEntity.java
+++ b/src/main/java/org/bukkit/entity/HumanEntity.java
@@ -26,7 +26,7 @@ public interface HumanEntity extends LivingEntity, AnimalTamer, InventoryHolder
     @Override
     org.bukkit.inventory.@NotNull EntityEquipment getEquipment();
     // Paper end
-
+    float getAttackPct(); // EMC
     /**
      * Returns the name of this player
      *
