From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 29 Mar 2019 21:45:09 -0400
Subject: [PATCH] amend ensureServerConversions

pending paper merge, store data version on stacks

diff --git a/src/main/java/org/bukkit/inventory/ItemFactory.java b/src/main/java/org/bukkit/inventory/ItemFactory.java
index 40edff7c93b6bf75de81102326667135b9344666..3609ac117de4762cae2c14c3b1d3ef772bf500fd 100644
--- a/src/main/java/org/bukkit/inventory/ItemFactory.java
+++ b/src/main/java/org/bukkit/inventory/ItemFactory.java
@@ -210,13 +210,18 @@ public interface ItemFactory {
     String getI18NDisplayName(@Nullable ItemStack item);
 
     /**
-     * Minecraft updates are converting simple item stacks into more complex NBT oriented Item Stacks.
-     *
-     * Use this method to to ensure any desired data conversions are processed.
-     * The input itemstack will not be the same as the returned itemstack.
+     * Ensure that this stack is represented by an Internal Item Stack
+     * @param item The stack to ensure is internal
+     * @return A CraftItemStack backed ItemStack with an internal representation
+     * @deprecated Don't use this
+     */
+    @Deprecated
+    @NotNull
+    ItemStack _asServerItemStack(@NotNull ItemStack item);
+    /**
+     * Runs all Minecraft Data Fixer updates on this ItemStack that are needed.
      *
-     * @param item The item to process conversions on
-     * @return A potentially Data Converted ItemStack
+     * @return A new stack with up to date data
      */
     @NotNull
     ItemStack ensureServerConversions(@NotNull ItemStack item);
diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index cf558463352427c1d775f1c7dc116ec851ebfa5b..7ec050c60784482dc44ac315153f5d7f254d46d4 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -31,6 +31,31 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, net.kyor
     private int amount = 0;
     private MaterialData data = null;
     private ItemMeta meta;
+    // Paper start
+    private Integer dataVersion = null;
+
+    /**
+     * Get's the DataVersion of this item stack, usually always the current version.
+     * If you do not know your data version from pre 1.13.2, you may call .setDataVersion(-1)
+     * and call {@link #ensureServerConversions()}.
+     * @return The Data Version of this itemstack.
+     */
+    public int getDataVersion() {
+        return dataVersion != null ? dataVersion : Bukkit.getUnsafe().getDataVersion();
+    }
+
+    /**
+     * Sets the DataVersion of this item stack, in preparation for calling {@link #ensureServerConversions()}
+     * You should not use this method unless you are converting known legacy data.
+     * @param dataVersion The data version. Use -1 for pre 1.13.2 data. Any stack created post 1.13.2 needs to know its data version
+     * @return previous data version
+     */
+    public int setDataVersion(int dataVersion) {
+        int previousDataVersion = this.getDataVersion();
+        this.dataVersion = dataVersion;
+        return previousDataVersion;
+    }
+    // Paper end
 
     @Utility
     protected ItemStack() {}
@@ -526,7 +551,7 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, net.kyor
     public Map<String, Object> serialize() {
         Map<String, Object> result = new LinkedHashMap<String, Object>();
 
-        result.put("v", Bukkit.getUnsafe().getDataVersion()); // Include version to indicate we are using modern material names (or LEGACY prefix)
+        result.put("v", this.getDataVersion()); // Paper
         result.put("type", getType().name());
 
         if (getAmount() != 1) {
@@ -578,6 +603,7 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, net.kyor
         }
 
         ItemStack result = new ItemStack(type, amount, damage);
+        result.dataVersion = version; // Paper
 
         if (args.containsKey("enchantments")) { // Backward compatiblity, @deprecated
             Object raw = args.get("enchantments");
@@ -747,12 +773,9 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, net.kyor
     }
 
     /**
-     * Minecraft updates are converting simple item stacks into more complex NBT oriented Item Stacks.
-     *
-     * Use this method to to ensure any desired data conversions are processed.
-     * The input itemstack will not be the same as the returned itemstack.
+     * Runs all Minecraft Data Fixer updates on this ItemStack that are needed.
      *
-     * @return A potentially Data Converted ItemStack
+     * @return A new stack with up to date data
      */
     @NotNull
     public ItemStack ensureServerConversions() {
