From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 23 Oct 2022 20:42:51 -0500
Subject: [PATCH] Disable the reload command


diff --git a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
index bdfe68b386b5ca2878475e548d3c9a3808fce848..027ccd332ffc47819106d7b620d9944e41845fb1 100644
--- a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
@@ -25,6 +25,13 @@ public class ReloadCommand extends BukkitCommand {
     public boolean execute(@NotNull CommandSender sender, @NotNull String currentAlias, @NotNull String[] args) { // Paper
         if (!testPermission(sender)) return true;
 
+        // EMC start
+        if (true) {
+            Command.broadcastCommandMessage(sender, net.kyori.adventure.text.Component.text("Reloading is disabled.", net.kyori.adventure.text.format.NamedTextColor.RED));
+            return true;
+        }
+        // EMC end
+
         // Paper start - Reload permissions.yml & require confirm
         boolean confirmed = System.getProperty("LetMeReload") != null;
         if (args.length == 1) {
