From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 29 Dec 2020 01:31:22 -0600
Subject: [PATCH] Add FakePlayer api for faking tablist users


diff --git a/src/main/java/com/empireminecraft/api/API.java b/src/main/java/com/empireminecraft/api/API.java
index dedc7a411e1eee1b36ecbf831c7953f0947f80c8..6d7948cf2421b530b205ef05d1a13f13d095586b 100644
--- a/src/main/java/com/empireminecraft/api/API.java
+++ b/src/main/java/com/empireminecraft/api/API.java
@@ -32,4 +32,5 @@ public abstract class API {
     public static EAPI_Misc misc;
     public static EAPI_Meta meta;
     public static EAPI_Chat chat;
+    public static EAPI_Server server;
 }
diff --git a/src/main/java/com/empireminecraft/api/EAPI_Server.java b/src/main/java/com/empireminecraft/api/EAPI_Server.java
new file mode 100644
index 0000000000000000000000000000000000000000..93abf800f656aee2017a411fe3436214f6a0bc54
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/EAPI_Server.java
@@ -0,0 +1,20 @@
+package com.empireminecraft.api;
+
+import com.destroystokyo.paper.profile.PlayerProfile;
+import net.kyori.adventure.text.Component;
+import org.bukkit.GameMode;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Set;
+import java.util.UUID;
+
+public interface EAPI_Server {
+
+    @Nullable
+    FakePlayer getFakePlayer(@NotNull UUID uuid);
+    @NotNull
+    FakePlayer createFakePlayer(@NotNull PlayerProfile profile, int ping, @NotNull GameMode gameMode, @NotNull Component playerListName) throws IllegalArgumentException;
+    @NotNull
+    Set<FakePlayer> getJoinedFakePlayers();
+}
diff --git a/src/main/java/com/empireminecraft/api/FakePlayer.java b/src/main/java/com/empireminecraft/api/FakePlayer.java
new file mode 100644
index 0000000000000000000000000000000000000000..b38aec6573da01a24199b3cabc8fbd84085a0666
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/FakePlayer.java
@@ -0,0 +1,48 @@
+package com.empireminecraft.api;
+
+import com.destroystokyo.paper.profile.PlayerProfile;
+import net.kyori.adventure.text.Component;
+import org.bukkit.GameMode;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.UUID;
+
+public interface FakePlayer {
+
+    @NotNull
+    PlayerProfile getProfile();
+
+    @NotNull
+    Component getPlayerListName();
+
+    int getPing();
+
+    @NotNull
+    GameMode getGameMode();
+
+    @Nullable
+    UUID getUniqueId();
+
+    void join();
+    void quit();
+
+    default void setPing(int ping) {
+        setPing(ping, true);
+    }
+    void setPing(int ping, boolean update);
+
+    default void setGameMode(@NotNull GameMode gameMode) {
+        setGameMode(gameMode, true);
+    }
+    void setGameMode(@NotNull GameMode gameMode, boolean update);
+
+    default void setPlayerListName(@NotNull Component playerListName) {
+        setPlayerListName(playerListName, true);
+    }
+    void setPlayerListName(@NotNull Component playerListName, boolean update);
+
+    void update();
+
+    boolean isJoined();
+}
diff --git a/src/main/java/com/empireminecraft/customevents/FakePlayerEvent.java b/src/main/java/com/empireminecraft/customevents/FakePlayerEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..63771966577fa725105e1f966902acfa1c1cc4ac
--- /dev/null
+++ b/src/main/java/com/empireminecraft/customevents/FakePlayerEvent.java
@@ -0,0 +1,26 @@
+package com.empireminecraft.customevents;
+
+import com.empireminecraft.api.FakePlayer;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Represents a fakeplayer related event
+ */
+public abstract class FakePlayerEvent extends Event {
+    protected FakePlayer fakePlayer;
+
+    public FakePlayerEvent(@NotNull final FakePlayer who) {
+        fakePlayer = who;
+    }
+
+    /**
+     * Returns the fakeplayer involved in this event
+     *
+     * @return Player who is involved in this event
+     */
+    @NotNull
+    public final FakePlayer getFakePlayer() {
+        return fakePlayer;
+    }
+}
diff --git a/src/main/java/com/empireminecraft/customevents/FakePlayerJoinEvent.java b/src/main/java/com/empireminecraft/customevents/FakePlayerJoinEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..d5483977262295d425dad0e5454c12d1c3ae98a5
--- /dev/null
+++ b/src/main/java/com/empireminecraft/customevents/FakePlayerJoinEvent.java
@@ -0,0 +1,23 @@
+package com.empireminecraft.customevents;
+
+import com.empireminecraft.api.FakePlayer;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class FakePlayerJoinEvent extends FakePlayerEvent {
+    public FakePlayerJoinEvent(@NotNull FakePlayer who) {
+        super(who);
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/com/empireminecraft/customevents/FakePlayerQuitEvent.java b/src/main/java/com/empireminecraft/customevents/FakePlayerQuitEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..bfb8d04d55f80103ea797ee744a118257819ba46
--- /dev/null
+++ b/src/main/java/com/empireminecraft/customevents/FakePlayerQuitEvent.java
@@ -0,0 +1,23 @@
+package com.empireminecraft.customevents;
+
+import com.empireminecraft.api.FakePlayer;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class FakePlayerQuitEvent extends FakePlayerEvent {
+    public FakePlayerQuitEvent(@NotNull FakePlayer who) {
+        super(who);
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 5e51bc9176b481ec70fd7e5ac23d04d59f8d1a9a..4bd695c5539f883abdf53b7b3026479abe3f6af1 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1,10 +1,12 @@
 package org.bukkit.entity;
 
 import java.net.InetSocketAddress;
+import java.util.Set;
 import java.util.UUID;
 import com.destroystokyo.paper.ClientOption; // Paper
 import com.destroystokyo.paper.Title; // Paper
 import net.kyori.adventure.text.Component;
+import com.empireminecraft.api.FakePlayer; // EMC
 import org.bukkit.DyeColor;
 import org.bukkit.Effect;
 import org.bukkit.GameMode;
@@ -1411,7 +1413,39 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      *     player
      */
     public boolean canSee(@NotNull Player player);
-
+    // EMC start
+    /**
+     * Checks to see if a fake player has been hidden from this player
+     *
+     * @param fakePlayer FakePlayer to check
+     * @return True if the provided fakePlayer is not being hidden from this
+     *     player
+     */
+    boolean canSee(@NotNull FakePlayer fakePlayer);
+    /**
+     * Hides a fake player from this player
+     *
+     * @param plugin Plugin that wants to hide the fake player
+     * @param fakePlayer FakePlayer to hide
+     */
+    void hideFakePlayer(@NotNull Plugin plugin, @NotNull FakePlayer fakePlayer);
+    /**
+     * Allows this player to see a fake player that was previously hidden. If
+     * another another plugin had hidden the fake player too, then the player will
+     * remain hidden until the other plugin calls this method too.
+     *
+     * @param plugin Plugin that wants to show the fake player
+     * @param fakePlayer FakePlayer to show
+     */
+    void showFakePlayer(@NotNull Plugin plugin, @NotNull FakePlayer fakePlayer);
+    /**
+     * Gets all fake players hidden with {@link #hideFakePlayer(Plugin, FakePlayer)}.
+     *
+     * @return a Set with all hidden fake players
+     */
+    @NotNull
+    Set<FakePlayer> getHiddenFakePlayers();
+    // EMC end
     /**
      * Visually hides an entity from this player.
      *
