From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 16 Mar 2024 17:27:20 -0500
Subject: [PATCH] Add Parrot API


diff --git a/src/main/java/org/bukkit/entity/Parrot.java b/src/main/java/org/bukkit/entity/Parrot.java
index c545a025c2ff5d3a0a4fff4e379277f38acf5c06..f5fe7fe962dead1b15ad77d3cc7b66b15f7c4ad1 100644
--- a/src/main/java/org/bukkit/entity/Parrot.java
+++ b/src/main/java/org/bukkit/entity/Parrot.java
@@ -54,4 +54,7 @@ public interface Parrot extends Tameable, Sittable {
          */
         GRAY;
     }
+    // EMC start
+    void setCanSitOnShoulders(boolean canSitOnShoulders);
+    // EMC end
 }
