From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 22 Mar 2024 22:29:04 -0500
Subject: [PATCH] Re-set name and lore to make itemstacks more consistent
 internally


diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index fba3db234dbed4541e9e95e97eb925e2aae37d21..cc82f4d52b73bba7303621dd8917d191b7065dff 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -573,6 +573,16 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
 
         // EMC start
         result = result.ensureServerConversions(); // Paper
+        if (result.hasItemMeta()) {
+            ItemMeta itemMeta = result.getItemMeta();
+            if (itemMeta.hasDisplayName()) {
+                itemMeta.displayName(itemMeta.displayName());
+            }
+            if (itemMeta.hasLore()) {
+                itemMeta.lore(itemMeta.lore());
+            }
+            result.setItemMeta(itemMeta);
+        }
         if (version <= 1343) {
             result = convertHorse(result);
         }
