From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 22 Mar 2024 22:29:04 -0500
Subject: [PATCH] Re-set name and lore to make itemstacks more consistent
 internally


diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index 3e3fe6e15cc015c81d3e73362e0872a6cd0afa5a..4d4623acdb0f16de3acd48dee2ad0c4b4c71175e 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -591,6 +591,16 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
 
         // EMC start
         result = result.ensureServerConversions(); // Paper
+        if (result.hasItemMeta()) {
+            ItemMeta itemMeta = result.getItemMeta();
+            if (itemMeta.hasDisplayName()) {
+                itemMeta.displayName(itemMeta.displayName());
+            }
+            if (itemMeta.hasLore()) {
+                itemMeta.lore(itemMeta.lore());
+            }
+            result.setItemMeta(itemMeta);
+        }
         if (version <= 1343) {
             result = convertHorse(result);
         }
