From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 4 Jul 2024 07:43:39 -0500
Subject: [PATCH] Incomplete playLaunchSound api for launch projectile event


diff --git a/src/main/java/com/destroystokyo/paper/event/player/PlayerLaunchProjectileEvent.java b/src/main/java/com/destroystokyo/paper/event/player/PlayerLaunchProjectileEvent.java
index a6d483e6be8b8527d7cfd676f6056179e8e9bf33..4697214de02156653ce31f8e926fa1ed98941918 100644
--- a/src/main/java/com/destroystokyo/paper/event/player/PlayerLaunchProjectileEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerLaunchProjectileEvent.java
@@ -25,6 +25,7 @@ public class PlayerLaunchProjectileEvent extends PlayerEvent implements Cancella
     private final Projectile projectile;
     private final ItemStack itemStack;
     private boolean consumeItem = true;
+    private boolean playLaunchSound = true;
 
     private boolean cancelled;
 
@@ -62,6 +63,14 @@ public class PlayerLaunchProjectileEvent extends PlayerEvent implements Cancella
         return this.consumeItem;
     }
 
+    public boolean shouldPlayLaunchSound() {
+        return this.playLaunchSound;
+    }
+
+    public void setShouldPlayLaunchSound(boolean playLaunchSound) {
+        this.playLaunchSound = playLaunchSound;
+    }
+    
     /**
      * Set whether to consume the ItemStack or not
      *
