From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 4 Jul 2024 07:43:39 -0500
Subject: [PATCH] Incomplete playLaunchSound api for launch projectile event


diff --git a/src/main/java/com/destroystokyo/paper/event/player/PlayerLaunchProjectileEvent.java b/src/main/java/com/destroystokyo/paper/event/player/PlayerLaunchProjectileEvent.java
index efd947eb0aa0633891d9c6a8bde66d33e29020d7..bd2fd472f8b73b42c484c433b5630bbc82485891 100644
--- a/src/main/java/com/destroystokyo/paper/event/player/PlayerLaunchProjectileEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerLaunchProjectileEvent.java
@@ -24,6 +24,7 @@ public class PlayerLaunchProjectileEvent extends PlayerEvent implements Cancella
     @NotNull private final Projectile projectile;
     @NotNull private final ItemStack itemStack;
     private boolean consumeItem = true;
+    private boolean playLaunchSound = true;
 
     private boolean cancelled;
 
@@ -63,6 +64,14 @@ public class PlayerLaunchProjectileEvent extends PlayerEvent implements Cancella
         return this.consumeItem;
     }
 
+    public boolean shouldPlayLaunchSound() {
+        return this.playLaunchSound;
+    }
+
+    public void setShouldPlayLaunchSound(boolean playLaunchSound) {
+        this.playLaunchSound = playLaunchSound;
+    }
+    
     /**
      * Set whether to consume the ItemStack or not
      *
