From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 1 Jan 2021 03:32:44 -0500
Subject: [PATCH] Ensure server conversions on ExactChoice ingredients


diff --git a/src/main/java/org/bukkit/inventory/RecipeChoice.java b/src/main/java/org/bukkit/inventory/RecipeChoice.java
index 7a7054bab646afc57b7180828dd3e245aee12248..48432bdfb89fe81c7df849f5e1d597035ef042fd 100644
--- a/src/main/java/org/bukkit/inventory/RecipeChoice.java
+++ b/src/main/java/org/bukkit/inventory/RecipeChoice.java
@@ -7,6 +7,8 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
 import org.bukkit.Material;
 import org.bukkit.Tag;
 import org.jetbrains.annotations.NotNull;
@@ -178,7 +180,7 @@ public interface RecipeChoice extends Predicate<ItemStack>, Cloneable {
                 Preconditions.checkArgument(choice != null, "Cannot have null choice");
             }
 
-            this.choices = new ArrayList<>(choices);
+            this.choices = choices.stream().map(ItemStack::ensureServerConversions).collect(Collectors.toList()); // Paper
         }
 
         @NotNull
