From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 23 Jun 2021 12:24:21 -0500
Subject: [PATCH] Stray interface should extend RangedEntity


diff --git a/src/main/java/org/bukkit/entity/Stray.java b/src/main/java/org/bukkit/entity/Stray.java
index a1427e585b77bb4e37dd11ad42786e909f2d56bb..3238f8d2d4474062020d823c9c7b268450f4cf97 100644
--- a/src/main/java/org/bukkit/entity/Stray.java
+++ b/src/main/java/org/bukkit/entity/Stray.java
@@ -3,4 +3,4 @@ package org.bukkit.entity;
 /**
  * Represents a Stray - variant of {@link AbstractSkeleton}.
  */
-public interface Stray extends AbstractSkeleton { }
+public interface Stray extends AbstractSkeleton, com.destroystokyo.paper.entity.RangedEntity { } // Paper
