From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 22 Mar 2024 22:29:04 -0500
Subject: [PATCH] Re-set name and lore to make itemstacks more consistent
 internally


diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index b261b2da61e280a7a726f6d028a4511176e8d864..bb652cde43868471c6218a2592aa1495c7b80028 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -684,6 +684,16 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, Translat
         if (version <= 1343) {
             convertHorse(result);
         }
+        if (result.hasItemMeta()) {
+            ItemMeta itemMeta = result.getItemMeta();
+            if (itemMeta.hasDisplayName()) {
+                itemMeta.displayName(itemMeta.displayName());
+            }
+            if (itemMeta.hasLore()) {
+                itemMeta.lore(itemMeta.lore());
+            }
+            result.setItemMeta(itemMeta);
+        }
         return result;
         // EMC end
     }
