From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 26 Jun 2023 18:38:01 -0500
Subject: [PATCH] ShulkerBoxDispensedEvent


diff --git a/src/main/java/com/empireminecraft/customevents/ShulkerBoxDispensedEvent.java b/src/main/java/com/empireminecraft/customevents/ShulkerBoxDispensedEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..04d7ab69cf09b172eb831c3edd04d7578ddc8ee2
--- /dev/null
+++ b/src/main/java/com/empireminecraft/customevents/ShulkerBoxDispensedEvent.java
@@ -0,0 +1,29 @@
+package com.empireminecraft.customevents;
+
+import org.bukkit.block.Block;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.block.BlockEvent;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+public class ShulkerBoxDispensedEvent extends BlockEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private final ItemStack item;
+
+    public ShulkerBoxDispensedEvent(@NotNull Block block, @NotNull ItemStack item) {
+        super(block);
+        this.item = item;
+    }
+
+    public @NotNull ItemStack getItem() {
+        return item;
+    }
+
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return handlers;
+    }
+}
