From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 10 Dec 2016 14:00:25 -0500
Subject: [PATCH] MC Log utils


diff --git a/src/main/java/com/empireminecraft/util/MCLog.java b/src/main/java/com/empireminecraft/util/MCLog.java
new file mode 100644
index 0000000000000000000000000000000000000000..e2f2e05d8217beb798213552e835aeeb274e8df8
--- /dev/null
+++ b/src/main/java/com/empireminecraft/util/MCLog.java
@@ -0,0 +1,54 @@
+package com.empireminecraft.util;
+
+import org.apache.commons.lang.exception.ExceptionUtils;
+
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+
+public class MCLog {
+    private static final Logger LOGGER = Logger.getLogger("NMS");
+    private static final Pattern NEWLINE = Pattern.compile("\n");
+
+    private MCLog() {}
+
+
+    public static void log(String message) {
+        info(message);
+    }
+    
+
+    public static void info(String message) {
+        for (String s : NEWLINE.split(message)) {
+            LOGGER.info(s);
+        }
+    }
+
+    public static void warn(String message) {
+        for (String s : NEWLINE.split(message)) {
+            LOGGER.warning(s);
+        }
+    }
+
+    public static void severe(String message) {
+        for (String s : NEWLINE.split(message)) {
+            LOGGER.severe(s);
+        }
+    }
+
+    public static void exception(String msg) {
+        exception(new Throwable(msg));
+    }
+
+    public static void exception(Throwable e) {
+        exception(e.getMessage(), e);
+    }
+
+    public static void exception(String msg, Throwable e) {
+        if (msg != null) {
+            severe(msg);
+        }
+        severe(ExceptionUtils.getFullStackTrace(e));
+    }
+
+
+}
