From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 29 Nov 2012 10:48:07 -0500
Subject: [PATCH] SpawnerInitiateEvent


diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
index f57e1b78204dff661ad5d3ee93a88a00330af2dc..7c74f60944c34f190c1d494c24d76f8f9a29e0dd 100644
--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
+++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
@@ -59,6 +59,25 @@ public abstract class BaseSpawner {
     }
 
     public boolean isNearPlayer(Level world, BlockPos pos) {
+        // EMC start - Add in SpawnerInitiateEvent
+        if (this.nextSpawnData == null) {
+            return false;
+        }
+        Optional<net.minecraft.world.entity.EntityType<?>> type = net.minecraft.world.entity.EntityType.by(this.nextSpawnData.getEntityToSpawn());
+        if (type.isEmpty()) {
+            return false;
+        }
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        org.bukkit.Location loc = new org.bukkit.Location(world.getWorld(), x, y, z);
+
+        if (!(new com.empireminecraft.customevents.SpawnerInitiateEvent(
+                org.bukkit.entity.EntityType.fromName(net.minecraft.world.entity.EntityType.getKey(type.get()).getPath()),
+                world.getWorld(), loc).callEvent())) {
+            return false;
+        }
+        // EMC end
         return world.hasNearbyAlivePlayerThatAffectsSpawning((double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double) this.requiredPlayerRange); // Paper - Affects Spawning API
     }
 
