From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 6 Dec 2016 22:22:14 -0500
Subject: [PATCH] Entity AI API


diff --git a/src/main/java/com/empireminecraft/api/CraftEAPI_Entity.java b/src/main/java/com/empireminecraft/api/CraftEAPI_Entity.java
index 616fe997d4a816e36dc272056bda55c9e67407e1..343bd35a5d63fe733358347357aaf0b92e15a227 100644
--- a/src/main/java/com/empireminecraft/api/CraftEAPI_Entity.java
+++ b/src/main/java/com/empireminecraft/api/CraftEAPI_Entity.java
@@ -23,10 +23,27 @@
 
 package com.empireminecraft.api;
 
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.world.entity.PathfinderMob;
+import net.minecraft.world.entity.ai.goal.Goal;
+import net.minecraft.world.entity.ai.goal.MeleeAttackGoal;
+import net.minecraft.world.entity.ai.goal.WrappedGoal;
+import net.minecraft.world.entity.ai.goal.target.NearestAttackableTargetGoal;
+import net.minecraft.world.entity.item.ItemEntity;
+import net.minecraft.world.entity.player.Player;
+import org.bukkit.craftbukkit.entity.CraftCreature;
 import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.entity.CraftMob;
+import org.bukkit.entity.Creature;
 import org.bukkit.entity.Entity;
+import org.bukkit.entity.Item;
+import org.bukkit.entity.Mob;
 import org.jetbrains.annotations.NotNull;
 
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Pattern;
+
 public class CraftEAPI_Entity implements EAPI_Entity {
 
     @NotNull
@@ -39,4 +56,100 @@ public class CraftEAPI_Entity implements EAPI_Entity {
     public void cancelTasks(@NotNull Entity entity) {
         ((CraftEntity) entity).getHandle().entityTasks.clear();
     }
+
+    @Override
+    public void setItemCanDespawn(@NotNull Item item, boolean canDespawn) {
+        final net.minecraft.world.entity.Entity handle = ((CraftEntity) item).getHandle();
+        ((ItemEntity) handle).canDespawn = canDespawn;
+    }
+
+    @Override
+    public void setFireProof(@NotNull Mob mob, boolean flag) {
+        ((CraftEntity) mob).getHandle().fireProof = flag;
+    }
+
+    @Override
+    public void setPanicProof(@NotNull Mob mob, boolean flag) {
+        ((CraftMob) mob).getHandle().panicProof = flag;
+    }
+
+    @Override
+    public void makeAggressive(@NotNull Creature creature, float range) {
+        PathfinderMob handle = ((CraftCreature) creature).getHandle();
+        if (!handle.markedAggressive) {
+            handle.goalSelector.addGoal(2, new MeleeAttackGoal(handle, 1.0F, false));
+            handle.targetSelector.addGoal(2, new NearestAttackableTargetGoal<>(handle, Player.class, true));
+            setTargetRange(creature, (double) range);
+            handle.markedAggressive = true;
+        }
+    }
+
+    @Override
+    public boolean isAggressive(@NotNull Creature creature) {
+        PathfinderMob entity = ((CraftCreature) creature).getHandle();
+        return entity.markedAggressive;
+    }
+
+    @Override
+    public void makePeaceful(@NotNull Mob mob) {
+        net.minecraft.world.entity.Mob handle = (net.minecraft.world.entity.Mob) ((CraftEntity) mob).getHandle();
+
+        EntityAIApi.getGoalSets(handle).forEach(set -> set.removeIf(o -> EntityAIApi.isHostileGoal(o.getGoal())));
+    }
+
+    @Override
+    public void addGoalByName(@NotNull Mob mob, @NotNull String goalName, boolean isTargetGoal, int tickRate, @NotNull Class<?>[] argClasses, @NotNull Object... args) {
+        net.minecraft.world.entity.Mob handle = (net.minecraft.world.entity.Mob) ((CraftEntity) mob).getHandle();
+        String nms = handle.getClass().getPackage().getName();
+        String className = nms + "." + goalName;
+        try {
+            Goal goal = EntityAIApi.getPathfinderGoal(handle, argClasses, className, args);
+            if (goal == null) {
+                MinecraftServer.LOGGER.error("Could not create goal for " + className);
+                return;
+            }
+            if (isTargetGoal) {
+                handle.targetSelector.addGoal(tickRate, goal);
+            } else {
+                handle.goalSelector.addGoal(tickRate, goal);
+            }
+        } catch (Exception e) {
+            MinecraftServer.LOGGER.error("Error adding goal by name");
+            e.printStackTrace();
+        }
+    }
+
+    private static final Pattern ENTITY_PACKAGE = Pattern.compile("net.minecraft.world.entity.");
+    @Override
+    public void removeGoalByName(@NotNull Mob mob, @NotNull String goalName) {
+        net.minecraft.world.entity.Mob entity = (net.minecraft.world.entity.Mob) ((CraftEntity) mob).getHandle();
+        for (Set<WrappedGoal> set : EntityAIApi.getGoalSets(entity)) {
+            set.removeIf(o -> ENTITY_PACKAGE.matcher(o.getGoal().getClass().getName()).replaceAll("")
+                                            .equalsIgnoreCase(goalName));
+        }
+    }
+    @NotNull
+    @Override
+    public Set<String> listGoalNames(@NotNull Mob mob) {
+        Set<String> goals = new HashSet<>();
+        net.minecraft.world.entity.Mob entity = (net.minecraft.world.entity.Mob) ((CraftEntity) mob).getHandle();
+        for (Set<WrappedGoal> set : EntityAIApi.getGoalSets(entity)) {
+            for (WrappedGoal wrappedGoal : set) {
+                goals.add(ENTITY_PACKAGE.matcher(wrappedGoal.getGoal().getClass().getName()).replaceAll(""));
+            }
+        }
+        return goals;
+    }
+
+    @Override
+    public void setEntitySize(@NotNull Entity entity, float width, float height) {
+        ((CraftEntity) entity).getHandle().setDimensions(width, height);
+    }
+
+    @Override
+    public void setTargetRange(@NotNull Mob mob, Double range) {
+        net.minecraft.world.entity.Mob entity = (net.minecraft.world.entity.Mob) ((CraftEntity) mob).getHandle();
+        entity.targetRange = range;
+    }
+
 }
diff --git a/src/main/java/com/empireminecraft/api/EntityAIApi.java b/src/main/java/com/empireminecraft/api/EntityAIApi.java
new file mode 100644
index 0000000000000000000000000000000000000000..ab934d0ee251cd0d1a5a13bbee70fb8b4024b2cd
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/EntityAIApi.java
@@ -0,0 +1,127 @@
+package com.empireminecraft.api;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.ai.goal.BreakDoorGoal;
+import net.minecraft.world.entity.ai.goal.Goal;
+import net.minecraft.world.entity.ai.goal.MeleeAttackGoal;
+import net.minecraft.world.entity.ai.goal.MoveThroughVillageGoal;
+import net.minecraft.world.entity.ai.goal.OcelotAttackGoal;
+import net.minecraft.world.entity.ai.goal.RangedAttackGoal;
+import net.minecraft.world.entity.ai.goal.RangedBowAttackGoal;
+import net.minecraft.world.entity.ai.goal.WrappedGoal;
+import net.minecraft.world.entity.ai.goal.target.HurtByTargetGoal;
+import net.minecraft.world.entity.ai.goal.target.NearestAttackableTargetGoal;
+import net.minecraft.world.entity.monster.Blaze;
+import net.minecraft.world.level.pathfinder.Node;
+import net.minecraft.world.level.pathfinder.Path;
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Mob;
+import org.bukkit.Location;
+
+import java.lang.reflect.Constructor;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public final class EntityAIApi {
+
+    private EntityAIApi() {
+    }
+
+    public static net.minecraft.world.entity.Mob getMobHandle(Mob entity) {
+        return (net.minecraft.world.entity.Mob) (((CraftEntity) entity).getHandle());
+    }
+
+    public static boolean isHostileGoal(Goal goal) {
+        if (goal instanceof MeleeAttackGoal ||
+            goal instanceof OcelotAttackGoal ||
+            goal instanceof Blaze.BlazeAttackGoal ||
+            goal instanceof RangedAttackGoal ||
+            goal instanceof RangedBowAttackGoal ||
+            goal instanceof MoveThroughVillageGoal ||
+            goal instanceof BreakDoorGoal ||
+            goal instanceof NearestAttackableTargetGoal ||
+            goal instanceof HurtByTargetGoal) {
+            return true;
+        }
+        return false;
+    }
+
+    static Goal getPathfinderGoal(net.minecraft.world.entity.Mob handle, Class<?>[] argClasses, String className, Object[] args) throws Exception {
+        Class<?> aClass = Class.forName(className);
+        CTOR:
+        for (Constructor<?> ctor : aClass.getDeclaredConstructors()) {
+            Class<?>[] parameterTypes = ctor.getParameterTypes();
+            if ((args.length +1) != parameterTypes.length) {
+                continue;
+            }
+            if (!parameterTypes[0].isAssignableFrom(handle.getClass())) {
+                continue;
+            }
+            for (int i = 1; i < parameterTypes.length; i++) {
+                Class<?> pCls = parameterTypes[i];
+                // TODO: Map argClasses from CB counterparts to NMS such as LivingEntity.class => EntityLiving.class
+                if (!pCls.isAssignableFrom(argClasses[i-1])) {
+                    continue CTOR;
+                }
+            }
+            Object[] newArgs = new Object[args.length+1];
+            System.arraycopy(args, 0, newArgs, 1, args.length);
+            newArgs[0] = handle;
+            return (Goal) ctor.newInstance(newArgs);
+        }
+        return null;
+    }
+
+    public static List<Set<WrappedGoal>> getGoalSets(net.minecraft.world.entity.Mob entity) {
+        List<Set<WrappedGoal>> check = new ArrayList<>();
+        check.add(new HashSet<>(entity.goalSelector.lockedFlags.values()));
+        check.add(new HashSet<>(entity.goalSelector.availableGoals));
+        check.add(new HashSet<>(entity.targetSelector.lockedFlags.values()));
+        check.add(new HashSet<>(entity.targetSelector.availableGoals));
+        return check;
+    }
+
+
+    private static Path getPathEntity(Mob entity) {
+        return getMobHandle(entity).getNavigation().getPath();
+    }
+
+    private static Path getPathEntity(Location loc, net.minecraft.world.entity.Mob handle) {
+        final boolean onGround = handle.onGround();
+        handle.setOnGround(true);
+        Path path = null;
+        if (loc != null) {
+            path = handle.getNavigation().createPath(new BlockPos(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), 0);
+        }
+
+        handle.setOnGround(onGround);
+        return path;
+    }
+
+
+    private static Path getPathEntity(LivingEntity target, net.minecraft.world.entity.Mob handle) {
+        final boolean onGround = handle.onGround();
+        handle.setOnGround(true);
+        Path path = null;
+        if (target != null) {
+            path = handle.getNavigation().createPath(((CraftEntity) target).getHandle(), 0);
+        }
+
+        handle.setOnGround(onGround);
+        return path;
+    }
+
+    private static Location getFinalLocation(Mob entity, Path pathEntity) {
+        if (pathEntity == null) {
+            return null;
+        }
+        final Node pathPoint = pathEntity.getEndNode();
+        if (pathPoint == null) {
+            return null;
+        }
+        return new Location(entity.getWorld(), pathPoint.x, pathPoint.y, pathPoint.z);
+    }
+}
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index f2e329790c9a3db686ba037db8c1761d2123f5fc..bf6621e88f52866833fec0e9e8771301f22b3a84 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -357,6 +357,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     protected Object2DoubleMap<TagKey<Fluid>> fluidHeight;
     protected boolean wasEyeInWater;
     private final Set<TagKey<Fluid>> fluidOnEyes;
+    public boolean fireProof = false; // EMC
     public int invulnerableTime;
     protected boolean firstTick;
     protected final SynchedEntityData entityData;
@@ -389,7 +390,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     private final Set<String> tags;
     private final double[] pistonDeltas;
     private long pistonDeltasGameTime;
-    private EntityDimensions dimensions;
+    private EntityDimensions dimensions; public void setDimensions(float width, float height) { dimensions = EntityDimensions.fixed(width, height); } // EMC
     private float eyeHeight;
     public boolean isInPowderSnow;
     public boolean wasInPowderSnow;
@@ -1853,7 +1854,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     public boolean fireImmune() {
-        return this.getType().fireImmune();
+        return this.getType().fireImmune() || this.fireProof; // EMC
     }
 
     public boolean causeFallDamage(float fallDistance, float damageMultiplier, DamageSource damageSource) {
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 30f4f1254fc295442d72d50479e8af635f2fe983..3e4f702ba5d42cd537ad526527fd89dd12684424 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -361,7 +361,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     public static AttributeSupplier.Builder createLivingAttributes() {
-        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS).add(Attributes.MAX_ABSORPTION).add(Attributes.STEP_HEIGHT).add(Attributes.SCALE).add(Attributes.GRAVITY).add(Attributes.SAFE_FALL_DISTANCE).add(Attributes.FALL_DAMAGE_MULTIPLIER).add(Attributes.JUMP_STRENGTH).add(Attributes.OXYGEN_BONUS).add(Attributes.BURNING_TIME).add(Attributes.EXPLOSION_KNOCKBACK_RESISTANCE).add(Attributes.WATER_MOVEMENT_EFFICIENCY).add(Attributes.MOVEMENT_EFFICIENCY).add(Attributes.ATTACK_KNOCKBACK);
+        return AttributeSupplier.builder().add(Attributes.MAX_HEALTH).add(Attributes.KNOCKBACK_RESISTANCE).add(Attributes.MOVEMENT_SPEED).add(Attributes.ARMOR).add(Attributes.ARMOR_TOUGHNESS).add(Attributes.MAX_ABSORPTION).add(Attributes.STEP_HEIGHT).add(Attributes.SCALE).add(Attributes.GRAVITY).add(Attributes.SAFE_FALL_DISTANCE).add(Attributes.FALL_DAMAGE_MULTIPLIER).add(Attributes.JUMP_STRENGTH).add(Attributes.OXYGEN_BONUS).add(Attributes.BURNING_TIME).add(Attributes.EXPLOSION_KNOCKBACK_RESISTANCE).add(Attributes.WATER_MOVEMENT_EFFICIENCY).add(Attributes.MOVEMENT_EFFICIENCY).add(Attributes.ATTACK_KNOCKBACK).add(Attributes.ATTACK_DAMAGE, 2.0D); // EMC - add Attack Damage
     }
 
     @Override
@@ -834,7 +834,19 @@ public abstract class LivingEntity extends Entity implements Attackable {
         }
         // Paper end - Friction API
         if (nbt.contains("attributes", 9) && this.level() != null && !this.level().isClientSide) {
-            this.getAttributes().load(nbt.getList("attributes", 10));
+            // EMC start
+            ListTag nbttaglist = nbt.getList("Attributes", 10);
+            Set<Integer> removeIndices = new HashSet<>();
+            for (int i = 0; i < nbttaglist.size(); ++i) {
+                if ("generic.targetRange".equals(nbttaglist.getCompound(i).getString("Name"))) {
+                    removeIndices.add(i);
+                }
+            }
+            for (int index : removeIndices) {
+                nbttaglist.remove(index);
+            }
+            this.getAttributes().load(nbttaglist);
+            // EMC end
         }
 
         if (nbt.contains("active_effects", 9)) {
@@ -3123,8 +3135,20 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     public boolean doHurtTarget(Entity target) {
-        this.setLastHurtMob(target);
-        return false;
+        // EMC start - make all mobs able to attack
+        boolean flag = target.hurt(this.damageSources().mobAttack(this), (float) ((int) this.getAttribute(Attributes.ATTACK_DAMAGE).getValue()));
+        if (flag) {
+            Level world = this.level();
+
+            if (world instanceof ServerLevel worldserver) {
+                DamageSource damagesource = this.damageSources().mobAttack(this);
+                EnchantmentHelper.doPostAttackEffects(worldserver, target, damagesource);
+            }
+
+            this.setLastHurtMob(target);
+        }
+        return flag;
+        // EMC end
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 763abeea3f14f15c27d600e0bdae44b387687bb4..7bf359cbf08ca91480c74843e6b78c95e2a11f72 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -145,6 +145,8 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
     private Leashable.LeashData leashData;
     private BlockPos restrictCenter;
     private float restrictRadius;
+    public @Nullable Double targetRange; // EMC
+    public boolean panicProof = false; // EMC
 
     public boolean aware = true; // CraftBukkit
 
diff --git a/src/main/java/net/minecraft/world/entity/PathfinderMob.java b/src/main/java/net/minecraft/world/entity/PathfinderMob.java
index ec9e76c548393235dcc6658c29e72e07e5d3510b..777799dcd8d9c455203b94fb6de8688a479ac043 100644
--- a/src/main/java/net/minecraft/world/entity/PathfinderMob.java
+++ b/src/main/java/net/minecraft/world/entity/PathfinderMob.java
@@ -23,6 +23,7 @@ public abstract class PathfinderMob extends Mob {
     }
 
     public BlockPos movingTarget; public BlockPos getMovingTarget() { return movingTarget; } // Paper
+    public boolean markedAggressive = false; // EMC
 
     public float getWalkTargetValue(BlockPos pos) {
         return this.getWalkTargetValue(pos, this.level());
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/FleeSunGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
index 63dde370ba36f8545181a656694a71dbda8bbf9c..77e0f1fc29bf3e7575e4b5bf0a693d1dea8f087b 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/FleeSunGoal.java
@@ -29,6 +29,7 @@ public class FleeSunGoal extends Goal {
         return this.mob.getTarget() == null
             && this.level.isDay()
             && this.mob.isOnFire()
+            && !this.mob.panicProof
             && this.level.canSeeSky(this.mob.blockPosition())
             && this.mob.getItemBySlot(EquipmentSlot.HEAD).isEmpty()
             && this.setWantedPos();
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/GoalSelector.java b/src/main/java/net/minecraft/world/entity/ai/goal/GoalSelector.java
index 9bdbf3e9453bc3ce96d52d04b8cde0d05f7356d8..184054bfc6655e290c1109900662c6a1658c20ad 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -22,8 +22,8 @@ public class GoalSelector {
             return false;
         }
     };
-    private final Map<Goal.Flag, WrappedGoal> lockedFlags = new EnumMap<>(Goal.Flag.class);
-    private final Set<WrappedGoal> availableGoals = new ObjectLinkedOpenHashSet<>();
+    public final Map<Goal.Flag, WrappedGoal> lockedFlags = new EnumMap<>(Goal.Flag.class); // EMC - make public
+    public final Set<WrappedGoal> availableGoals = new ObjectLinkedOpenHashSet<>(); // EMC - make public
     private final Supplier<ProfilerFiller> profiler;
     private static final Goal.Flag[] GOAL_FLAG_VALUES = Goal.Flag.values(); // Paper - remove streams from pathfindergoalselector
     private final ca.spottedleaf.moonrise.common.set.OptimizedSmallEnumSet<net.minecraft.world.entity.ai.goal.Goal.Flag> goalTypes = new ca.spottedleaf.moonrise.common.set.OptimizedSmallEnumSet<>(Goal.Flag.class); // Paper - remove streams from pathfindergoalselector
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/PanicGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/PanicGoal.java
index 20d4e3fb00a9eaac8484e95329bd64a368872b59..c0d52981a8bb05990da959bc1b0248e7234fdb8d 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/PanicGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/PanicGoal.java
@@ -41,6 +41,7 @@ public class PanicGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (this.mob.panicProof) { return false; } // EMC
         if (!this.shouldPanic()) {
             return false;
         } else {
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
index f09d7e5052ae0166a8f3363d4eff1c000af07a96..f115fe09c72e5f55e4fee5273f54e215a8c6d938 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -80,6 +80,7 @@ public abstract class TargetGoal extends Goal {
     }
 
     protected double getFollowDistance() {
+        if (this.mob.targetRange != null) { return this.mob.targetRange; } // EMC
         return this.mob.getAttributeValue(Attributes.FOLLOW_RANGE);
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index 4423973d4d9a2c3879d98d1d4c8b8c117c677ac5..defc6fbb4c2835eb1bf5300196441ce0d346208f 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -51,6 +51,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
     private static final int INFINITE_PICKUP_DELAY = 32767;
     private static final int INFINITE_LIFETIME = -32768;
     public int age;
+    public boolean canDespawn = true; // EMC
     public int pickupDelay;
     public int health;
     @Nullable
@@ -225,7 +226,7 @@ public class ItemEntity extends Entity implements TraceableEntity {
                 }
             }
 
-            if (!this.level().isClientSide && this.age >= this.despawnRate) { // Spigot // Paper - Alternative item-despawn-rate
+            if (canDespawn && !this.level().isClientSide && this.age >= this.despawnRate) { // Spigot // Paper - Alternative item-despawn-rate // EMC - add canDespawn
                 // CraftBukkit start - fire ItemDespawnEvent
                 if (CraftEventFactory.callItemDespawnEvent(this).isCancelled()) {
                     this.age = 0;
diff --git a/src/main/java/net/minecraft/world/entity/monster/Blaze.java b/src/main/java/net/minecraft/world/entity/monster/Blaze.java
index 61004bb35b0edcc4578b8a9c1b280096466ba279..fc5eb680a8b1ff027c631176407d5c03a48528fd 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Blaze.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Blaze.java
@@ -152,7 +152,7 @@ public class Blaze extends Monster {
         this.entityData.set(DATA_FLAGS_ID, b);
     }
 
-    static class BlazeAttackGoal extends Goal {
+    public static class BlazeAttackGoal extends Goal { // EMC
         private final Blaze blaze;
         private int attackStep;
         private int attackTime;
