From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 12:14:27 -0500
Subject: [PATCH] RangedEntity-API


diff --git a/src/main/java/com/destroystokyo/paper/entity/CraftRangedEntity.java b/src/main/java/com/destroystokyo/paper/entity/CraftRangedEntity.java
index d7a8eb1b8f24ed2741ae9dae62d3f6146f273e1d..46558a49e62c0e1e10b49b0d9d919ca54b5c1a32 100644
--- a/src/main/java/com/destroystokyo/paper/entity/CraftRangedEntity.java
+++ b/src/main/java/com/destroystokyo/paper/entity/CraftRangedEntity.java
@@ -1,6 +1,11 @@
 package com.destroystokyo.paper.entity;
 
+import com.empireminecraft.api.API;
+import com.empireminecraft.api.EntityAIApi;
 import net.minecraft.world.entity.Mob;
+import net.minecraft.world.entity.ai.goal.Goal;
+import net.minecraft.world.entity.ai.goal.RangedAttackGoal;
+import net.minecraft.world.entity.ai.goal.RangedBowAttackGoal;
 import net.minecraft.world.entity.monster.RangedAttackMob;
 import org.bukkit.craftbukkit.entity.CraftLivingEntity;
 import org.bukkit.entity.LivingEntity;
@@ -17,4 +22,41 @@ public interface CraftRangedEntity<T extends Mob & RangedAttackMob> extends Rang
     default void setChargingAttack(boolean raiseHands) {
         getHandle().setAggressive(raiseHands);
     }
+    // EMC start
+    @Override
+    default void setRangedAttackDistance(float range) {
+        API.entity.setTargetRange(this, (double) range);
+        net.minecraft.world.entity.Mob entity = getHandle();
+        EntityAIApi.getGoalSets(entity).forEach(set -> set.forEach(o -> {
+            Goal goal = o.getGoal();
+            if (goal instanceof RangedAttackGoal) {
+                ((RangedAttackGoal) goal).setAttackRadius(range);
+            } else if (goal instanceof RangedBowAttackGoal) {
+                ((RangedBowAttackGoal<?>) goal).setAttackRadius(range);
+            }
+        }));
+    }
+
+    @Override
+    default void setRangedAttackSpeed(Integer min, Integer max) {
+        if (min == null && max == null) {
+            return;
+        }
+        net.minecraft.world.entity.Mob entity = getHandle();
+        EntityAIApi.getGoalSets(entity).forEach(set -> set.forEach(o -> {
+            Goal goal = o.getGoal();
+            if (goal instanceof RangedAttackGoal) {
+                if (min != null) {
+                    ((RangedAttackGoal) goal).setMinInterval(min);
+                }
+                if (max != null) {
+                    ((RangedAttackGoal) goal).setMaxInterval(max);
+                }
+            } else if (goal instanceof RangedBowAttackGoal) {
+                int speed = min != null && max != null ? Math.min(min, max) : (min != null ? min : max);
+                ((RangedBowAttackGoal<?>) goal).setMinAttackInterval(speed);
+            }
+        }));
+    }
+    // EMC end
 }
