From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 12:15:01 -0500
Subject: [PATCH] Wolf-API


diff --git a/src/main/java/net/minecraft/world/entity/animal/Wolf.java b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
index eecb7511582e5e316b71fa4a4734881424be5ca7..1736e969defb2b480a17d3e7ccf67c3ffc611f65 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Wolf.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Wolf.java
@@ -463,6 +463,13 @@ public class Wolf extends TamableAnimal implements NeutralMob {
         this.entityData.set(Wolf.DATA_REMAINING_ANGER_TIME, angerTime);
     }
 
+    // EMC start
+    public boolean alwaysAngry = false;
+    @Override
+    public boolean isAngry() {
+        return alwaysAngry || NeutralMob.super.isAngry();
+    }
+    // EMC end
     @Override
     public void startPersistentAngerTimer() {
         this.setRemainingPersistentAngerTime(Wolf.PERSISTENT_ANGER_TIME.sample(this.random));
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftWolf.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftWolf.java
index e16459c9cfcac790edd6d912750d32c68387cbbc..92d69e922a658d1d3432cea21efb651ad6100d08 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftWolf.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftWolf.java
@@ -57,4 +57,14 @@ public class CraftWolf extends CraftTameableAnimal implements Wolf {
     public void setInterested(boolean flag) {
         this.getHandle().setIsInterested(flag);
     }
+    // EMC start
+    @Override
+    public void setAlwaysAngry(boolean alwaysAngry) {
+        net.minecraft.world.entity.animal.Wolf entity = getHandle();
+        entity.alwaysAngry = alwaysAngry;
+        if (alwaysAngry) {
+            entity.startPersistentAngerTimer();
+        }
+    }
+    // EMC end
 }
