From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 12:15:47 -0500
Subject: [PATCH] EnderSignal-API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderSignal.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderSignal.java
index cd0692963a80c098c93afdaf040e76c18ddcb101..a532da9433fcb27054df43acf642e73f311643ee 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderSignal.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderSignal.java
@@ -79,4 +79,30 @@ public class CraftEnderSignal extends CraftEntity implements EnderSignal {
     public void setDespawnTimer(int time) {
         this.getHandle().life = time;
     }
+    // EMC start
+    @Override
+    public void setExactTarget(boolean exact) {
+        getHandle().exactTarget = exact;
+    }
+
+    @Override
+    public void setSpeed(double speed) {
+        getHandle().speed = speed / 10000D;
+    }
+
+    @Override
+    public double getSpeed() {
+        return getHandle().speed * 10000D;
+    }
+
+    @Override
+    public void setLife(int lifeTime) {
+        getHandle().lifeTime = lifeTime;
+    }
+
+    @Override
+    public int getLife() {
+        return getHandle().lifeTime;
+    }
+    // EMC end
 }
