From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 12:16:13 -0500
Subject: [PATCH] Ghast-API


diff --git a/src/main/java/net/minecraft/world/entity/monster/Ghast.java b/src/main/java/net/minecraft/world/entity/monster/Ghast.java
index a836a902bebf318ceabaed4e98fab1141b46a28b..ef238e2e9c559b1906d07f67dc424eeb97b4552c 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Ghast.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Ghast.java
@@ -36,6 +36,7 @@ public class Ghast extends FlyingMob implements Enemy {
 
     private static final EntityDataAccessor<Boolean> DATA_IS_CHARGING = SynchedEntityData.defineId(Ghast.class, EntityDataSerializers.BOOLEAN);
     private int explosionPower = 1;
+    public int fireballCooldown = 40; // EMC
 
     public Ghast(EntityType<? extends Ghast> type, Level world) {
         super(type, world);
@@ -342,7 +343,7 @@ public class Ghast extends FlyingMob implements Enemy {
                         entitylargefireball.bukkitYield = entitylargefireball.explosionPower = this.ghast.getExplosionPower();
                         entitylargefireball.setPos(this.ghast.getX() + vec3d.x * 4.0D, this.ghast.getY(0.5D) + 0.5D, entitylargefireball.getZ() + vec3d.z * 4.0D);
                         world.addFreshEntity(entitylargefireball);
-                        this.chargeTime = -40;
+                        this.chargeTime = -ghast.fireballCooldown; // EMC
                     }
                 } else if (this.chargeTime > 0) {
                     --this.chargeTime;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftGhast.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftGhast.java
index 97fa4e1e70203194bd939618b2fad92665af6d59..d1d9180ef0c304983335ba15c7d5b1d7383fdc70 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftGhast.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftGhast.java
@@ -41,4 +41,15 @@ public class CraftGhast extends CraftFlying implements Ghast, CraftEnemy {
         this.getHandle().setExplosionPower(explosionPower);
     }
     // Paper end
+    // EMC start
+    @Override
+    public int getFireballCooldown() {
+        return getHandle().fireballCooldown;
+    }
+
+    @Override
+    public void setFireballCooldown(int cooldown) {
+        getHandle().fireballCooldown = cooldown;
+    }
+    // EMC end
 }
