From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 12:16:04 -0500
Subject: [PATCH] Blaze-API


diff --git a/src/main/java/net/minecraft/world/entity/monster/Blaze.java b/src/main/java/net/minecraft/world/entity/monster/Blaze.java
index 58c2b8b8bfd5a40259aa6252243884d14c183ef2..923a09a01d72e1a855bb4675ebc60ad76b4088bb 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Blaze.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Blaze.java
@@ -30,6 +30,10 @@ public class Blaze extends Monster {
     private int nextHeightOffsetChangeTick;
     private static final EntityDataAccessor<Byte> DATA_FLAGS_ID = SynchedEntityData.defineId(Blaze.class, EntityDataSerializers.BYTE);
 
+    // EMC start
+    public int numFireballs = 4;
+    public int timeBetweenFireballs = 6;
+    // EMC end
     public Blaze(EntityType<? extends Blaze> type, Level world) {
         super(type, world);
         this.setPathfindingMalus(BlockPathTypes.WATER, -1.0F);
@@ -152,7 +156,7 @@ public class Blaze extends Monster {
         this.entityData.set(DATA_FLAGS_ID, b);
     }
 
-    static class BlazeAttackGoal extends Goal {
+    public static class BlazeAttackGoal extends Goal { // EMC - make public
         private final Blaze blaze;
         private int attackStep;
         private int attackTime;
@@ -218,8 +222,8 @@ public class Blaze extends Monster {
                         if (this.attackStep == 1) {
                             this.attackTime = 60;
                             this.blaze.setCharged(true);
-                        } else if (this.attackStep <= 4) {
-                            this.attackTime = 6;
+                        } else if (this.attackStep <= this.blaze.numFireballs) { // EMC
+                            this.attackTime = this.blaze.timeBetweenFireballs; // EMC
                         } else {
                             this.attackTime = 100;
                             this.attackStep = 0;
@@ -241,7 +245,9 @@ public class Blaze extends Monster {
                                     this.blaze.getRandom().triangle(g, 2.297 * h)
                                 );
                                 smallFireball.setPos(smallFireball.getX(), this.blaze.getY(0.5) + 0.5, smallFireball.getZ());
+                                if (new com.empireminecraft.customevents.BlazeLaunchFireballEvent((org.bukkit.entity.Blaze) this.blaze.getBukkitEntity(), (org.bukkit.entity.SmallFireball) smallFireball.getBukkitEntity()).callEvent()) { // EMC
                                 this.blaze.level().addFreshEntity(smallFireball);
+                                } else { smallFireball.kill(); }// EMC
                             }
                         }
                     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftBlaze.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftBlaze.java
index a4c9c73691300880777483b0beb17e1bd6779d06..ff93f26790d1e52e5c0296fe5205e363c53f1541 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftBlaze.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftBlaze.java
@@ -17,4 +17,25 @@ public class CraftBlaze extends CraftMonster implements Blaze {
     public String toString() {
         return "CraftBlaze";
     }
+    // EMC start
+    @Override
+    public void setFireballs(int num) {
+        getHandle().numFireballs = num;
+    }
+
+    @Override
+    public int getNumFireballs() {
+        return getHandle().numFireballs;
+    }
+
+    @Override
+    public int getTimeBetweenFireballs() {
+        return getHandle().timeBetweenFireballs;
+    }
+
+    @Override
+    public void setTimeBetweenFireballs(int timeBetweenFireballs) {
+        getHandle().timeBetweenFireballs = timeBetweenFireballs;
+    }
+    // EMC end
 }
