From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 2 Jul 2023 08:52:42 -0500
Subject: [PATCH] Spider-API


diff --git a/src/main/java/net/minecraft/world/entity/monster/Spider.java b/src/main/java/net/minecraft/world/entity/monster/Spider.java
index a30fb47559eb74b7fe634678e63a85e7e2cad9a4..fa8cabf7f309bf63232cfa264de4002d0859c688 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Spider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Spider.java
@@ -184,6 +184,7 @@ public class Spider extends Monster {
         return vehicle.getBbWidth() <= this.getBbWidth() ? new Vec3(0.0D, 0.3125D * (double) this.getScale(), 0.0D) : super.getVehicleAttachmentPoint(vehicle);
     }
 
+    public boolean attackDuringDay = false; // EMC
     private static class SpiderAttackGoal extends MeleeAttackGoal {
 
         public SpiderAttackGoal(Spider spider) {
@@ -198,6 +199,7 @@ public class Spider extends Monster {
         @Override
         public boolean canContinueToUse() {
             float f = this.mob.getLightLevelDependentMagicValue();
+            if (((Spider) this.mob).attackDuringDay) { f = 0.0F; } // EMC
 
             if (f >= 0.5F && this.mob.getRandom().nextInt(100) == 0) {
                 this.mob.setTarget((LivingEntity) null);
@@ -217,6 +219,7 @@ public class Spider extends Monster {
         @Override
         public boolean canUse() {
             float f = this.mob.getLightLevelDependentMagicValue();
+            if (((Spider) this.mob).attackDuringDay) { f = 0.0F; } // EMC
 
             return f >= 0.5F ? false : super.canUse();
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
index b4afc37c21fc478df44fca7ec3fbc33d337dc6b7..eb3f232db076a49ba3c303fa42173496c02f777e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
@@ -18,4 +18,10 @@ public class CraftSpider extends CraftMonster implements Spider {
     public String toString() {
         return "CraftSpider";
     }
+    // EMC start
+    @Override
+    public void makeAggressiveDuringDay(boolean attack) {
+        getHandle().attackDuringDay = attack;
+    }
+    // EMC end
 }
