From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 12:16:04 -0500
Subject: [PATCH] Blaze-API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftBlaze.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftBlaze.java
index 84cd915955188492696e5948fbb1c4eca0cca039..8a28ef7ab039f86da343f644d3677ff9d89ea7e5 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftBlaze.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftBlaze.java
@@ -23,4 +23,25 @@ public class CraftBlaze extends CraftMonster implements Blaze {
     public EntityType getType() {
         return EntityType.BLAZE;
     }
+    // EMC start
+    @Override
+    public void setFireballs(int num) {
+        getHandle().numFireballs = num;
+    }
+
+    @Override
+    public int getNumFireballs() {
+        return getHandle().numFireballs;
+    }
+
+    @Override
+    public int getTimeBetweenFireballs() {
+        return getHandle().timeBetweenFireballs;
+    }
+
+    @Override
+    public void setTimeBetweenFireballs(int timeBetweenFireballs) {
+        getHandle().timeBetweenFireballs = timeBetweenFireballs;
+    }
+    // EMC end
 }
