From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 1 Apr 2023 12:16:13 -0500
Subject: [PATCH] Ghast-API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftGhast.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftGhast.java
index 29508a44221b83f5c31ee986dab1bf66af84a405..c46bcdbf69a902f2f93275031875bd2a15e3f39e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftGhast.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftGhast.java
@@ -24,14 +24,6 @@ public class CraftGhast extends CraftFlying implements Ghast, CraftEnemy {
     public EntityType getType() {
         return EntityType.GHAST;
     }
-    // EMC start
-    public int getFireballCooldown() {
-        return getHandle().fireballCooldown;
-    }
-    public void setFireballCooldown(int cooldown) {
-        getHandle().fireballCooldown = cooldown;
-    }
-    // EMC end
 
     @Override
     public boolean isCharging() {
@@ -55,4 +47,15 @@ public class CraftGhast extends CraftFlying implements Ghast, CraftEnemy {
         this.getHandle().setExplosionPower(explosionPower);
     }
     // Paper end
+    // EMC start
+    @Override
+    public int getFireballCooldown() {
+        return getHandle().fireballCooldown;
+    }
+
+    @Override
+    public void setFireballCooldown(int cooldown) {
+        getHandle().fireballCooldown = cooldown;
+    }
+    // EMC end
 }
