From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 2 Jul 2023 08:52:42 -0500
Subject: [PATCH] Spider-API


diff --git a/src/main/java/net/minecraft/world/entity/monster/Spider.java b/src/main/java/net/minecraft/world/entity/monster/Spider.java
index 71b5a9c97a13f703073c0122742ff9e8a0e49df2..6ecbdc0db8ec5eb5a8c9c94182842ac97b6ebe29 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Spider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Spider.java
@@ -199,6 +199,7 @@ public class Spider extends Monster {
         return vehicle.getBbWidth() <= this.getBbWidth() ? -0.3125F : 0.0F;
     }
 
+    public boolean attackDuringDay = false; // EMC
     private static class SpiderAttackGoal extends MeleeAttackGoal {
 
         public SpiderAttackGoal(Spider spider) {
@@ -213,6 +214,7 @@ public class Spider extends Monster {
         @Override
         public boolean canContinueToUse() {
             float f = this.mob.getLightLevelDependentMagicValue();
+            if (((Spider) this.mob).attackDuringDay) { f = 0.0F; } // EMC
 
             if (f >= 0.5F && this.mob.getRandom().nextInt(100) == 0) {
                 this.mob.setTarget((LivingEntity) null);
@@ -232,6 +234,7 @@ public class Spider extends Monster {
         @Override
         public boolean canUse() {
             float f = this.mob.getLightLevelDependentMagicValue();
+            if (((Spider) this.mob).attackDuringDay) { f = 0.0F; } // EMC
 
             return f >= 0.5F ? false : super.canUse();
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
index 3da93b408a864e1663291ed22aaccfa6beba1280..b445bf0e072cf310de24cd8f6d348c37108cf15c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
@@ -18,4 +18,10 @@ public class CraftSpider extends CraftMonster implements Spider {
     public String toString() {
         return "CraftSpider";
     }
+    // EMC start
+    @Override
+    public void makeAggressiveDuringDay(boolean attack) {
+        getHandle().attackDuringDay = attack;
+    }
+    // EMC end
 }
