From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 2 Jul 2023 08:52:42 -0500
Subject: [PATCH] Spider-API


diff --git a/src/main/java/net/minecraft/world/entity/monster/Spider.java b/src/main/java/net/minecraft/world/entity/monster/Spider.java
index 05432184077752b1d0cb764a5e39ed875748b2d6..0351ce8e1f5fb17d14e74662b40b219ba24b98f7 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Spider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Spider.java
@@ -192,6 +192,7 @@ public class Spider extends Monster {
         return 0.65F;
     }
 
+    public boolean attackDuringDay = false; // EMC
     private static class SpiderAttackGoal extends MeleeAttackGoal {
 
         public SpiderAttackGoal(Spider spider) {
@@ -206,6 +207,7 @@ public class Spider extends Monster {
         @Override
         public boolean canContinueToUse() {
             float f = this.mob.getLightLevelDependentMagicValue();
+            if (((Spider) this.mob).attackDuringDay) { f = 0.0F; } // EMC
 
             if (f >= 0.5F && this.mob.getRandom().nextInt(100) == 0) {
                 this.mob.setTarget((LivingEntity) null);
@@ -230,6 +232,7 @@ public class Spider extends Monster {
         @Override
         public boolean canUse() {
             float f = this.mob.getLightLevelDependentMagicValue();
+            if (((Spider) this.mob).attackDuringDay) { f = 0.0F; } // EMC
 
             return f >= 0.5F ? false : super.canUse();
         }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
index 994d8d815aa110bd6ac9939a4ce78f1bbad2662f..b77cc772a94ba83b8e7321fe06d881d3de560025 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftSpider.java
@@ -24,4 +24,10 @@ public class CraftSpider extends CraftMonster implements Spider {
     public EntityType getType() {
         return EntityType.SPIDER;
     }
+    // EMC start
+    @Override
+    public void makeAggressiveDuringDay(boolean attack) {
+        getHandle().attackDuringDay = attack;
+    }
+    // EMC end
 }
