From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 10 Jan 2014 22:05:59 -0500
Subject: [PATCH] Disable Snooper

What purpose does it provide...

diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 66c99cf8ebcfff889ee1242fe10bfb2028506c21..8d58e743de1cf8d9b6ce87e608b3065a409786f8 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1441,11 +1441,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         //} // Paper
 
         this.profiler.push("snooper");
-        if (((DedicatedServer) this).getProperties().snooperEnabled && !this.snooper.isStarted() && this.tickCount > 100) { // Spigot
+        if (false && ((DedicatedServer) this).getProperties().snooperEnabled && !this.snooper.isStarted() && this.tickCount > 100) { // Spigot // EMC - disable snooper
             this.snooper.start();
         }
 
-        if (((DedicatedServer) this).getProperties().snooperEnabled && this.tickCount % 6000 == 0) { // Spigot
+        if (false && ((DedicatedServer) this).getProperties().snooperEnabled && this.tickCount % 6000 == 0) { // Spigot // EMC - disable snooper
             this.snooper.prepare();
         }
 
