From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 10 Jan 2014 22:05:59 -0500
Subject: [PATCH] Disable Snooper

What purpose does it provide...

diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index fbc9ba1277788580b08a03cd61a2e94349b01ebf..717af7dff30a7614060eceaec0dbc97d37357c6c 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1443,11 +1443,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         //} // Paper
 
         this.profiler.push("snooper");
-        if (((DedicatedServer) this).getProperties().snooperEnabled && !this.snooper.isStarted() && this.tickCount > 100) { // Spigot
+        if (false && ((DedicatedServer) this).getProperties().snooperEnabled && !this.snooper.isStarted() && this.tickCount > 100) { // Spigot // EMC - disable snooper
             this.snooper.start();
         }
 
-        if (((DedicatedServer) this).getProperties().snooperEnabled && this.tickCount % 6000 == 0) { // Spigot
+        if (false && ((DedicatedServer) this).getProperties().snooperEnabled && this.tickCount % 6000 == 0) { // Spigot // EMC - disable snooper
             this.snooper.prepare();
         }
 
