From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 10 Jan 2014 22:05:59 -0500
Subject: [PATCH] Disable Snooper

What purpose does it provide...

diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 651892092347bc0467424c8bdcc304b2532e40a5..1728d9c275c4688097472200d953d9e7470b1da1 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1517,11 +1517,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         //} // Paper
 
         this.profiler.push("snooper");
-        if (((DedicatedServer) this).getProperties().snooperEnabled && !this.snooper.isStarted() && this.tickCount > 100) { // Spigot
+        if (false && ((DedicatedServer) this).getProperties().snooperEnabled && !this.snooper.isStarted() && this.tickCount > 100) { // Spigot // EMC - disable snooper
             this.snooper.start();
         }
 
-        if (((DedicatedServer) this).getProperties().snooperEnabled && this.tickCount % 6000 == 0) { // Spigot
+        if (false && ((DedicatedServer) this).getProperties().snooperEnabled && this.tickCount % 6000 == 0) { // Spigot // EMC - disable snooper
             this.snooper.prepare();
         }
 
