From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 9 Sep 2013 19:41:07 -0400
Subject: [PATCH] Better hopper events


diff --git a/src/main/java/com/empireminecraft/api/HopperEvents.java b/src/main/java/com/empireminecraft/api/HopperEvents.java
new file mode 100644
index 0000000000000000000000000000000000000000..b5f09c53b182dd8d7a59dda9b0ebecc05a545f6c
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/HopperEvents.java
@@ -0,0 +1,22 @@
+package com.empireminecraft.api;
+
+import com.empireminecraft.customevents.HopperDrainEvent;
+import com.empireminecraft.customevents.HopperFillEvent;
+import org.bukkit.Bukkit;
+import org.bukkit.inventory.Inventory;
+
+public final class HopperEvents {
+    private HopperEvents() {
+    }
+
+    public static boolean drain(Inventory hopper, Inventory source) {
+        HopperDrainEvent event = new HopperDrainEvent(hopper, source);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isCancelled();
+    }
+    public static boolean fill(Inventory hopper, Inventory target)  {
+        HopperFillEvent event = new HopperFillEvent(hopper, target);
+        Bukkit.getPluginManager().callEvent(event);
+        return event.isCancelled();
+    }
+}
diff --git a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 757edf74751dc7183454656fda9cecc4eb601e4c..092892696e5130f9cf4b7590153cec2c1f6fa95a 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -439,6 +439,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
             if (HopperBlockEntity.isFullContainer(iinventory1, enumdirection)) {
                 return false;
+            } else if (com.empireminecraft.api.HopperEvents.fill(hopper.getOwner(false).getInventory(), getInventory(iinventory1))) { return false; // EMC
             } else {
                 // Paper start - replace logic; MAKE SURE TO CHECK FOR DIFFS ON UPDATES
                 return hopperPush(world, iinventory1, enumdirection, hopper);
@@ -521,6 +522,11 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         Container iinventory = HopperBlockEntity.getSourceContainer(world, hopper);
 
         if (iinventory != null) {
+            // EMC start
+            if (com.empireminecraft.api.HopperEvents.drain(new CraftInventory(hopper), getInventory(iinventory))) {
+                return false;
+            }
+            // EMC end
             Direction enumdirection = Direction.DOWN;
 
             // Paper start - Perf: Optimize Hoppers
