From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 21 Feb 2014 00:50:46 -0500
Subject: [PATCH] Add EntityEffectAddedEvent


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 4b4d2c11e4e0517e549fd92df76bc25a6a49da89..8748627cf9e0269be70870ef7e67686e93b08a4e 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -131,6 +131,7 @@ import org.bukkit.Location;
 import org.bukkit.craftbukkit.attribute.CraftAttributeMap;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.potion.CraftPotionUtil;
 import org.bukkit.entity.Player;
 import org.bukkit.event.entity.ArrowBodyCountChangeEvent;
 import org.bukkit.event.entity.EntityDamageEvent;
@@ -1108,7 +1109,13 @@ public abstract class LivingEntity extends Entity {
             return true;
         }
         // CraftBukkit end
-
+        // EMC start
+        com.empireminecraft.customevents.EntityEffectAddedEvent addedEvent = new com.empireminecraft.customevents.EntityEffectAddedEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), CraftPotionUtil.toBukkit(mobeffect));
+        if (!addedEvent.callEvent()) {
+            return false;
+        }
+        mobeffect = CraftPotionUtil.fromBukkit(addedEvent.getEffect());
+        // EMC end
         if (!this.canBeAffected(mobeffect)) {
             return false;
         } else {
