From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 21 Feb 2014 00:50:46 -0500
Subject: [PATCH] Add EntityEffectAddedEvent


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 3a99ed36474df2f7ccf5c51383208d2c878fb3d7..013b76e7a4d308c56a29e408baa7c5e7f03a2f39 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -128,6 +128,7 @@ import org.bukkit.Location;
 import org.bukkit.craftbukkit.attribute.CraftAttributeMap;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.potion.CraftPotionUtil;
 import org.bukkit.entity.Player;
 import org.bukkit.event.entity.ArrowBodyCountChangeEvent;
 import org.bukkit.event.entity.EntityDamageEvent;
@@ -1103,7 +1104,13 @@ public abstract class LivingEntity extends Entity {
             return true;
         }
         // CraftBukkit end
-
+        // EMC start
+        com.empireminecraft.customevents.EntityEffectAddedEvent addedEvent = new com.empireminecraft.customevents.EntityEffectAddedEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), CraftPotionUtil.toBukkit(mobeffect));
+        if (!addedEvent.callEvent()) {
+            return false;
+        }
+        mobeffect = CraftPotionUtil.fromBukkit(addedEvent.getEffect());
+        // EMC end
         if (!this.canBeAffected(mobeffect)) {
             return false;
         } else {
