From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 21 Feb 2014 00:50:46 -0500
Subject: [PATCH] Add EntityEffectAddedEvent


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index d165cbc8a2fcd10fefc9d5374501a77a5901d49a..5176eb3d299ea461a67471f22142a6c79d55d0f0 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -130,6 +130,7 @@ import org.bukkit.Location;
 import org.bukkit.craftbukkit.attribute.CraftAttributeMap;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.potion.CraftPotionUtil;
 import org.bukkit.entity.Player;
 import org.bukkit.event.entity.ArrowBodyCountChangeEvent;
 import org.bukkit.event.entity.EntityDamageEvent;
@@ -1106,7 +1107,13 @@ public abstract class LivingEntity extends Entity {
             return true;
         }
         // CraftBukkit end
-
+        // EMC start
+        com.empireminecraft.customevents.EntityEffectAddedEvent addedEvent = new com.empireminecraft.customevents.EntityEffectAddedEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), CraftPotionUtil.toBukkit(mobeffect));
+        if (!addedEvent.callEvent()) {
+            return false;
+        }
+        mobeffect = CraftPotionUtil.fromBukkit(addedEvent.getEffect());
+        // EMC end
         if (!this.canBeAffected(mobeffect)) {
             return false;
         } else {
