From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 21 Feb 2014 00:50:46 -0500
Subject: [PATCH] Add EntityEffectAddedEvent


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 4e63f9ef452b4d881360c345f554c03e5c401b22..334990e9a981126b5aedc8d8029f05dbf1c72bc1 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -145,6 +145,7 @@ import org.bukkit.Location;
 import org.bukkit.craftbukkit.attribute.CraftAttributeMap;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.potion.CraftPotionUtil;
 import org.bukkit.entity.Player;
 import org.bukkit.event.entity.ArrowBodyCountChangeEvent;
 import org.bukkit.event.entity.EntityDamageEvent;
@@ -1165,7 +1166,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
             return true;
         }
         // CraftBukkit end
-
+        // EMC start
+        com.empireminecraft.customevents.EntityEffectAddedEvent addedEvent = new com.empireminecraft.customevents.EntityEffectAddedEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), CraftPotionUtil.toBukkit(mobeffect));
+        if (!addedEvent.callEvent()) {
+            return false;
+        }
+        mobeffect = CraftPotionUtil.fromBukkit(addedEvent.getEffect());
+        // EMC end
         if (!this.canBeAffected(mobeffect)) {
             return false;
         } else {
