From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 14 Apr 2014 20:44:38 -0400
Subject: [PATCH] Disable UUID conversion - Already done


diff --git a/src/main/java/net/minecraft/server/players/OldUsersConverter.java b/src/main/java/net/minecraft/server/players/OldUsersConverter.java
index 5445cb5910ec63408dc4379eec5e12d305182527..489d4835b259399511eacde9cf7d0352fb202cef 100644
--- a/src/main/java/net/minecraft/server/players/OldUsersConverter.java
+++ b/src/main/java/net/minecraft/server/players/OldUsersConverter.java
@@ -434,6 +434,7 @@ public class OldUsersConverter {
     }
 
     public static boolean serverReadyAfterUserconversion(MinecraftServer server) {
+        if (true) return true; // EMC - Disable UUID conversion
         boolean flag = OldUsersConverter.areOldUserlistsRemoved();
 
         flag = flag && OldUsersConverter.areOldPlayersConverted(server);
