From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 16 Mar 2014 20:44:46 -0400
Subject: [PATCH] EntityKnockbackEvent

Control knockback power of entity attacks

diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 923f84f4a2953a15e47665d9c4868fc5d29e6d8e..fffd2a730abefbd52432dea93438e65190ff8a5f 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1571,7 +1571,7 @@ public abstract class Mob extends LivingEntity {
 
         if (target instanceof LivingEntity) {
             f += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
-            f1 += (float) EnchantmentHelper.getKnockbackBonus(this);
+            f1 += (float) EnchantmentHelper.getKnockbackBonus(this, (LivingEntity) target); // EMC
         }
 
         int i = EnchantmentHelper.getFireAspect(this);
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 2dc2b1bf2f6c9a8a80e1d43f7d32cbab79d332f2..46c3acaa8c35a67a7aaaa91e0f67efd475297921 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1230,7 +1230,7 @@ public abstract class Player extends LivingEntity {
                     boolean flag = f2 > 0.9F;
                     boolean flag1 = false;
                     byte b0 = 0;
-                    int i = b0 + EnchantmentHelper.getKnockbackBonus(this);
+                    int i = b0 + EnchantmentHelper.getKnockbackBonus(this, target instanceof LivingEntity ? (LivingEntity) target : null); // EMC
 
                     if (this.isSprinting() && flag) {
                         sendSoundEffect(this, this.getX(), this.getY(), this.getZ(), SoundEvents.PLAYER_ATTACK_KNOCKBACK, this.getSoundSource(), 1.0F, 1.0F); // Paper - send while respecting visibility
diff --git a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
index 3a09ce6d0ea51436adcae4719d3f28d1868c283c..88eff4a6bfdb304ed46c63fbaa3fe7697d7f8f26 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
@@ -207,7 +207,15 @@ public class EnchantmentHelper {
     }
 
     public static int getKnockbackBonus(LivingEntity entity) {
-        return getEnchantmentLevel(Enchantments.KNOCKBACK, entity);
+        // EMC start - add target
+        return getKnockbackBonus(entity, null);
+    }
+    public static int getKnockbackBonus(LivingEntity entity, @Nullable LivingEntity target) {
+        int level = getEnchantmentLevel(Enchantments.KNOCKBACK, entity);
+        com.empireminecraft.customevents.EntityKnockbackEvent event = new com.empireminecraft.customevents.EntityKnockbackEvent((org.bukkit.entity.LivingEntity) entity.getBukkitEntity(), target != null ? (org.bukkit.entity.LivingEntity) target.getBukkitEntity() : null, level);
+        event.callEvent();
+        return event.getLevel();
+        // EMC end
     }
 
     public static int getFireAspect(LivingEntity entity) {
