From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 14 Apr 2014 20:44:38 -0400
Subject: [PATCH] Disable UUID conversion - Already done


diff --git a/src/main/java/net/minecraft/server/players/OldUsersConverter.java b/src/main/java/net/minecraft/server/players/OldUsersConverter.java
index 6aacc724c8c8d6fbe3067226989039ca9cee4929..586c7384529e918d69d78ad247148c70eaf0eb97 100644
--- a/src/main/java/net/minecraft/server/players/OldUsersConverter.java
+++ b/src/main/java/net/minecraft/server/players/OldUsersConverter.java
@@ -433,6 +433,7 @@ public class OldUsersConverter {
     }
 
     public static boolean serverReadyAfterUserconversion(MinecraftServer server) {
+        if (true) return true; // EMC - Disable UUID conversion
         boolean flag = OldUsersConverter.areOldUserlistsRemoved();
 
         flag = flag && OldUsersConverter.areOldPlayersConverted(server);
