From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 14 Apr 2014 20:44:38 -0400
Subject: [PATCH] Disable UUID conversion - Already done

---
 .../minecraft/server/players/NameReferencingFileConverter.java   | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/main/java/net/minecraft/server/players/NameReferencingFileConverter.java b/src/main/java/net/minecraft/server/players/NameReferencingFileConverter.java
index 8c1f328ca1ba12ed63ec7bd7efad54ff633ba802..e0bd6c242b712e7df776e774626ff302ba03767a 100644
--- a/src/main/java/net/minecraft/server/players/NameReferencingFileConverter.java
+++ b/src/main/java/net/minecraft/server/players/NameReferencingFileConverter.java
@@ -434,6 +434,7 @@ public class NameReferencingFileConverter {
     }
 
     public static boolean e(MinecraftServer minecraftserver) {
+        if (true) return true; // EMC - Disable UUID conversion
         boolean flag = b();
 
         flag = flag && f(minecraftserver);
