From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 14 Apr 2014 20:44:38 -0400
Subject: [PATCH] Disable UUID conversion - Already done

---
 .../java/net/minecraft/server/NameReferencingFileConverter.java  | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/main/java/net/minecraft/server/NameReferencingFileConverter.java b/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
index c5957c2d6c54b076ebe7f9a432e30551d2b6febc..7d748f5fc078fa91421ae49935766084061ad507 100644
--- a/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
+++ b/src/main/java/net/minecraft/server/NameReferencingFileConverter.java
@@ -424,6 +424,7 @@ public class NameReferencingFileConverter {
     }
 
     public static boolean e(MinecraftServer minecraftserver) {
+        if (true) return true; // EMC - Disable UUID conversion
         boolean flag = b();
 
         flag = flag && f(minecraftserver);
