From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 21 Feb 2014 00:50:46 -0500
Subject: [PATCH] Add EntityEffectAddedEvent


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 96fe9a1b42c5343516fda3677fc76dfd68c10b8f..e088e914a3713028660745d91b08a0d3c2aea2b6 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -131,6 +131,7 @@ import org.bukkit.Location;
 import org.bukkit.craftbukkit.attribute.CraftAttributeMap;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.potion.CraftPotionUtil;
 import org.bukkit.entity.Player;
 import org.bukkit.event.entity.ArrowBodyCountChangeEvent;
 import org.bukkit.event.entity.EntityDamageEvent;
@@ -1159,7 +1160,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
             return true;
         }
         // CraftBukkit end
-
+        // EMC start
+        com.empireminecraft.customevents.EntityEffectAddedEvent addedEvent = new com.empireminecraft.customevents.EntityEffectAddedEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), CraftPotionUtil.toBukkit(mobeffect));
+        if (!addedEvent.callEvent()) {
+            return false;
+        }
+        mobeffect = CraftPotionUtil.fromBukkit(addedEvent.getEffect());
+        // EMC end
         if (!this.canBeAffected(mobeffect)) {
             return false;
         } else {
