From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 16 Mar 2014 20:44:46 -0400
Subject: [PATCH] EntityKnockbackEvent

Control knockback power of entity attacks

diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index b41521246e272d36992fba4cf63eac29f37f058b..04da9ed1194ce371f746d9f2bf6e365daf5b1df7 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1571,7 +1571,7 @@ public abstract class Mob extends LivingEntity {
 
         if (target instanceof LivingEntity) {
             f += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
-            f1 += (float) EnchantmentHelper.getKnockbackBonus(this);
+            f1 += (float) EnchantmentHelper.getKnockbackBonus(this, (LivingEntity) target); // EMC
         }
 
         int i = EnchantmentHelper.getFireAspect(this);
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 04d07bd687862921d51be7daa4e9eb3d2fe0c0c0..c38f20cbc3f79f76a184ae4d74995d9814980823 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1226,7 +1226,7 @@ public abstract class Player extends LivingEntity {
                     boolean flag = f2 > 0.9F;
                     boolean flag1 = false;
                     byte b0 = 0;
-                    int i = b0 + EnchantmentHelper.getKnockbackBonus(this);
+                    int i = b0 + EnchantmentHelper.getKnockbackBonus(this, target instanceof LivingEntity ? (LivingEntity) target : null); // EMC
 
                     if (this.isSprinting() && flag) {
                         sendSoundEffect(this, this.getX(), this.getY(), this.getZ(), SoundEvents.PLAYER_ATTACK_KNOCKBACK, this.getSoundSource(), 1.0F, 1.0F); // Paper - send while respecting visibility
diff --git a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
index 3a09ce6d0ea51436adcae4719d3f28d1868c283c..88eff4a6bfdb304ed46c63fbaa3fe7697d7f8f26 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
@@ -207,7 +207,15 @@ public class EnchantmentHelper {
     }
 
     public static int getKnockbackBonus(LivingEntity entity) {
-        return getEnchantmentLevel(Enchantments.KNOCKBACK, entity);
+        // EMC start - add target
+        return getKnockbackBonus(entity, null);
+    }
+    public static int getKnockbackBonus(LivingEntity entity, @Nullable LivingEntity target) {
+        int level = getEnchantmentLevel(Enchantments.KNOCKBACK, entity);
+        com.empireminecraft.customevents.EntityKnockbackEvent event = new com.empireminecraft.customevents.EntityKnockbackEvent((org.bukkit.entity.LivingEntity) entity.getBukkitEntity(), target != null ? (org.bukkit.entity.LivingEntity) target.getBukkitEntity() : null, level);
+        event.callEvent();
+        return event.getLevel();
+        // EMC end
     }
 
     public static int getFireAspect(LivingEntity entity) {
