From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 16 Mar 2014 20:44:46 -0400
Subject: [PATCH] EntityKnockbackEvent

Control knockback power of entity attacks

diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 74fbd5a375e875763f7aad9eabf5b38365cff1aa..28efd9bd344341b46bda8b00c2ccd8acdab48b80 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1950,11 +1950,16 @@ public abstract class LivingEntity extends Entity implements Attackable {
         float f = (float) this.getAttributeValue(Attributes.ATTACK_KNOCKBACK);
         Level world = this.level();
 
+        // EMC start
         if (world instanceof ServerLevel worldserver) {
-            return EnchantmentHelper.modifyKnockback(worldserver, this.getWeaponItem(), target, damageSource, f);
-        } else {
-            return f;
-        }
+            f = EnchantmentHelper.modifyKnockback(worldserver, this.getWeaponItem(), target, damageSource, f);
+        }// else {
+        //    return f;
+        //}
+        com.empireminecraft.customevents.EntityKnockbackEvent event = new com.empireminecraft.customevents.EntityKnockbackEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), (org.bukkit.entity.LivingEntity) target.getBukkitEntity(), f);
+        event.callEvent();
+        return event.getLevel();
+        // EMC end
     }
 
     protected void dropFromLootTable(DamageSource damageSource, boolean causedByPlayer) {
