From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 28 Feb 2014 22:25:04 -0500
Subject: [PATCH] Fire EntityTargetLivingEntityEvent for
 GoalNearestAttackableTarget

If cancelled, move to next in list

diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
index 7f4fb6ad4b3b3da52a111b0c58499f27d8443124..f04ddefeef0151a7f05327249c8c8844cc60bc24 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/target/NearestAttackableTargetGoal.java
@@ -57,9 +57,20 @@ public class NearestAttackableTargetGoal<T extends LivingEntity> extends TargetG
 
     protected void findTarget() {
         if (this.targetType != Player.class && this.targetType != ServerPlayer.class) {
-            this.target = this.mob.level.getNearestEntity(this.mob.level.getEntitiesOfClass(this.targetType, this.getTargetSearchArea(this.getFollowDistance()), (entityliving) -> {
-                return true;
-            }), this.targetConditions, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ());
+            // EMC start
+            java.util.List<T> list = this.mob.level.getNearbyEntities(this.targetType, this.targetConditions, this.mob, this.getTargetSearchArea(this.getFollowDistance()));
+            list.sort(java.util.Comparator.comparingDouble(this.mob::distanceToSqr));
+            for (T entity : list) {
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(
+                    this.mob, entity,
+                    org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER
+                ).isCancelled()) {
+                    continue;
+                }
+                this.target = entity;
+                break;
+            }
+            // EMC end
         } else {
             this.target = this.mob.level.getNearestPlayer(this.targetConditions, this.mob, this.mob.getX(), this.mob.getEyeY(), this.mob.getZ());
         }
