From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 14 Apr 2014 20:44:38 -0400
Subject: [PATCH] Disable UUID conversion - Already done


diff --git a/src/main/java/net/minecraft/server/players/OldUsersConverter.java b/src/main/java/net/minecraft/server/players/OldUsersConverter.java
index 45d4638d568ea2aee805aa1b0542533019e5870d..8a9b8b18eebc6c03e6cdcd00dfdc87c1715ffc2b 100644
--- a/src/main/java/net/minecraft/server/players/OldUsersConverter.java
+++ b/src/main/java/net/minecraft/server/players/OldUsersConverter.java
@@ -427,6 +427,7 @@ public class OldUsersConverter {
     }
 
     public static boolean serverReadyAfterUserconversion(MinecraftServer server) {
+        if (true) return true; // EMC - Disable UUID conversion
         boolean flag = OldUsersConverter.areOldUserlistsRemoved();
 
         flag = flag && OldUsersConverter.areOldPlayersConverted(server);
