From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 16 Mar 2014 20:44:46 -0400
Subject: [PATCH] EntityKnockbackEvent

Control knockback power of entity attacks

diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 71212aa216bdcd0d53741fb17009c39f76cf5a3d..b36a62fec33d52ec986c9fbc2d3c3cefbbb47613 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1545,7 +1545,7 @@ public abstract class Mob extends LivingEntity {
 
         if (target instanceof LivingEntity) {
             f += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
-            f1 += (float) EnchantmentHelper.getKnockbackBonus((LivingEntity) this);
+            f1 += (float) EnchantmentHelper.getKnockbackBonus((LivingEntity) this, (LivingEntity) target); // EMC
         }
 
         int i = EnchantmentHelper.getFireAspect(this);
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 936f8e9d677c050d1ea4ae5f59b38d3ec45eb591..417f2b16c519c9330b2b08ed3cc3ed7a60d73dc2 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1205,7 +1205,7 @@ public abstract class Player extends LivingEntity {
                     boolean flag = f2 > 0.9F;
                     boolean flag1 = false;
                     byte b0 = 0;
-                    int i = b0 + EnchantmentHelper.getKnockbackBonus((LivingEntity) this);
+                    int i = b0 + EnchantmentHelper.getKnockbackBonus((LivingEntity) this, target instanceof LivingEntity ? (LivingEntity) target : null); // EMC
 
                     if (this.isSprinting() && flag) {
                         sendSoundEffect(this, this.getX(), this.getY(), this.getZ(), SoundEvents.PLAYER_ATTACK_KNOCKBACK, this.getSoundSource(), 1.0F, 1.0F); // Paper - send while respecting visibility
diff --git a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
index d439e8ce87bf7da03683a336941c7673b8b166e4..1577d830768ac32959981b8265e2faf31449464d 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
@@ -207,7 +207,16 @@ public class EnchantmentHelper {
     }
 
     public static int getKnockbackBonus(LivingEntity entity) {
-        return getEnchantmentLevel(Enchantments.KNOCKBACK, entity);
+        // EMC start - rename and add target
+        return getKnockbackBonus(entity, null);
+    }
+
+    public static int getKnockbackBonus(LivingEntity entity, @Nullable LivingEntity target) { // EMC - OBF HELPER
+        int level = getEnchantmentLevel(Enchantments.KNOCKBACK, entity);
+        com.empireminecraft.customevents.EntityKnockbackEvent event = new com.empireminecraft.customevents.EntityKnockbackEvent((org.bukkit.entity.LivingEntity) entity.getBukkitEntity(), target != null ? (org.bukkit.entity.LivingEntity) target.getBukkitEntity() : null, level);
+        event.callEvent();
+        return event.getLevel();
+        // EMC end
     }
 
     public static int getFireAspect(LivingEntity entity) {
