From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 14 Aug 2014 18:23:33 -0400
Subject: [PATCH] Add Player.getFishingHook() API and kill hook after rod
 damage event


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index f5886a88fd98ede5e85a91eccccb05ac33eb40e2..1a0a89ebab374de5a327f22d90025e5351e8e466 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -555,8 +555,8 @@ public class FishingHook extends Projectile {
             }
             // CraftBukkit end
 
-            this.discard();
-            return i;
+            //this.discard(); // EMC - move to after damage event in EntityFishingRod
+            return i == 0 ? -1 : i; // EMC - use -1 to represent "0 but kill entity"
         } else {
             return 0;
         }
diff --git a/src/main/java/net/minecraft/world/item/FishingRodItem.java b/src/main/java/net/minecraft/world/item/FishingRodItem.java
index f63db099188bd9bcc46960ded440177a26634925..39b3e11a5c6b398d6c97d3a54687a77c3a552cc7 100644
--- a/src/main/java/net/minecraft/world/item/FishingRodItem.java
+++ b/src/main/java/net/minecraft/world/item/FishingRodItem.java
@@ -27,9 +27,10 @@ public class FishingRodItem extends Item implements Vanishable {
         if (user.fishing != null) {
             if (!world.isClientSide) {
                 i = user.fishing.retrieve(itemstack);
-                itemstack.hurtAndBreak(i, user, (entityhuman1) -> {
+                itemstack.hurtAndBreak(Math.max(0, i), user, (entityhuman1) -> { // EMC - Ignore -1 return using Math.max
                     entityhuman1.broadcastBreakEvent(hand);
                 });
+                if (i != 0) { user.fishing.discard(); } // EMC - removed EntityFishingHook, and 0 = cancelled
             }
 
             world.playSound((Player) null, user.getX(), user.getY(), user.getZ(), SoundEvents.FISHING_BOBBER_RETRIEVE, SoundSource.NEUTRAL, 1.0F, 0.4F / (world.getRandom().nextFloat() * 0.4F + 0.8F));
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index abcce6bd0b32d0188751393a6f56f2e82530912c..dc73fc1be7a888bbab43230f78f5410ed4c48ecd 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2380,6 +2380,14 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
 
     // Paper start
+    @Override
+    public org.bukkit.entity.FishHook getFishingHook() {
+        if (getHandle().fishing == null) {
+            return null;
+        }
+        return (org.bukkit.entity.FishHook) getHandle().fishing.getBukkitEntity();
+    }
+
     public void setAffectsSpawning(boolean affects) {
         this.getHandle().affectsSpawning = affects;
     }
