From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 12 Mar 2014 23:10:04 -0400
Subject: [PATCH] LivingEntityArmorProtectEvent

To control if armor should protect entity, how much if so, and how much item damage to give.

diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index e088e914a3713028660745d91b08a0d3c2aea2b6..57c8ae457bafd03dbf7522d489ffb6fb16f15873 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.entity;
 
 import com.destroystokyo.paper.event.player.PlayerArmorChangeEvent; // Paper
+import com.empireminecraft.customevents.LivingEntityArmorProtectEvent;
 import com.google.common.base.Objects;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
@@ -2115,7 +2116,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     public int getArmorValue() {
-        return Mth.floor(this.getAttributeValue(Attributes.ARMOR));
+        // EMC start
+        final double value = this.getAttributeValue(Attributes.ARMOR);
+        LivingEntityArmorProtectEvent event = new LivingEntityArmorProtectEvent(this.getBukkitLivingEntity(), value);
+        event.callEvent();
+        return Mth.floor(event.getArmorValue());
+        // EMC end
     }
 
     protected void hurtArmor(DamageSource source, float amount) {}
