From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 12 Mar 2014 23:10:04 -0400
Subject: [PATCH] LivingEntityArmorProtectEvent

To control if armor should protect entity, how much if so, and how much item damage to give.

diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 261d040df1f23448914815bc0cfabe6cc7a5f9be..3bbaac70c31534e0f018039327d881f19d745892 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.entity;
 
+import com.empireminecraft.customevents.LivingEntityArmorProtectEvent;
 import com.google.common.base.Objects;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
@@ -2127,7 +2128,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     public int getArmorValue() {
-        return Mth.floor(this.getAttributeValue(Attributes.ARMOR));
+        // EMC start
+        final double value = this.getAttributeValue(Attributes.ARMOR);
+        LivingEntityArmorProtectEvent event = new LivingEntityArmorProtectEvent(this.getBukkitLivingEntity(), value);
+        event.callEvent();
+        return Mth.floor(event.getArmorValue());
+        // EMC end
     }
 
     protected void hurtArmor(DamageSource source, float amount) {}
