From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 16 Mar 2014 20:44:46 -0400
Subject: [PATCH] EntityKnockbackEvent

Control knockback power of entity attacks

diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 5e3b1d69905e20497ad0647e9369304cf6d3b20b..a28d71f6c071ab2d2ce8accade6747ab0869a659 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1667,7 +1667,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
 
         if (target instanceof LivingEntity) {
             f += EnchantmentHelper.getDamageBonus(this.getMainHandItem(), ((LivingEntity) target).getMobType());
-            f1 += (float) EnchantmentHelper.getKnockbackBonus(this);
+            f1 += (float) EnchantmentHelper.getKnockbackBonus(this, (LivingEntity) target); // EMC
         }
 
         int i = EnchantmentHelper.getFireAspect(this);
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 4fad2f5da03c26e0260798bac5cc9ae7bdac13c1..b669f6ba72fedf5983cf00b59b20e015efdea53a 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1269,7 +1269,7 @@ public abstract class Player extends LivingEntity {
                     boolean flag = f2 > 0.9F;
                     boolean flag1 = false;
                     byte b0 = 0;
-                    int i = b0 + EnchantmentHelper.getKnockbackBonus(this);
+                    int i = b0 + EnchantmentHelper.getKnockbackBonus(this, target instanceof LivingEntity ? (LivingEntity) target : null); // EMC
 
                     if (this.isSprinting() && flag) {
                         sendSoundEffect(this, this.getX(), this.getY(), this.getZ(), SoundEvents.PLAYER_ATTACK_KNOCKBACK, this.getSoundSource(), 1.0F, 1.0F); // Paper - send while respecting visibility
diff --git a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
index 496a9b9b095bd322fba8229a5d47e2a0107aeb96..c111f1f72288ca3929ce6064eacc9bf17263a804 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentHelper.java
@@ -199,7 +199,15 @@ public class EnchantmentHelper {
     }
 
     public static int getKnockbackBonus(LivingEntity entity) {
-        return getEnchantmentLevel(Enchantments.KNOCKBACK, entity);
+        // EMC start - add target
+        return getKnockbackBonus(entity, null);
+    }
+    public static int getKnockbackBonus(LivingEntity entity, @Nullable LivingEntity target) {
+        int level = getEnchantmentLevel(Enchantments.KNOCKBACK, entity);
+        com.empireminecraft.customevents.EntityKnockbackEvent event = new com.empireminecraft.customevents.EntityKnockbackEvent((org.bukkit.entity.LivingEntity) entity.getBukkitEntity(), target != null ? (org.bukkit.entity.LivingEntity) target.getBukkitEntity() : null, level);
+        event.callEvent();
+        return event.getLevel();
+        // EMC end
     }
 
     public static int getFireAspect(LivingEntity entity) {
