From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 14 Aug 2014 18:23:33 -0400
Subject: [PATCH] Add Player.getFishingHook() API and kill hook after rod
 damage event


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index f958ce70efb64d4077fae60a790a0df6786ad0ea..05e3a7abdcc4090c05266d83ce72703c59850d54 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -557,8 +557,8 @@ public class FishingHook extends Projectile {
             }
             // CraftBukkit end
 
-            this.discard();
-            return i;
+            //this.discard(); // EMC - move to after damage event in EntityFishingRod
+            return i == 0 ? -1 : i; // EMC - use -1 to represent "0 but kill entity"
         } else {
             return 0;
         }
diff --git a/src/main/java/net/minecraft/world/item/FishingRodItem.java b/src/main/java/net/minecraft/world/item/FishingRodItem.java
index 7a4ae2d69f530182f44a31eebd1c2e44ada5e44b..c573e757d9e7f57c0b3d52903130ec85b18de9c2 100644
--- a/src/main/java/net/minecraft/world/item/FishingRodItem.java
+++ b/src/main/java/net/minecraft/world/item/FishingRodItem.java
@@ -28,9 +28,10 @@ public class FishingRodItem extends Item implements Vanishable {
         if (user.fishing != null) {
             if (!world.isClientSide) {
                 i = user.fishing.retrieve(itemstack);
-                itemstack.hurtAndBreak(i, user, (entityhuman1) -> {
+                itemstack.hurtAndBreak(Math.max(0, i), user, (entityhuman1) -> { // EMC - Ignore -1 return using Math.max
                     entityhuman1.broadcastBreakEvent(hand);
                 });
+                if (i != 0) { user.fishing.discard(); } // EMC - removed EntityFishingHook, and 0 = cancelled
             }
 
             world.playSound((Player) null, user.getX(), user.getY(), user.getZ(), SoundEvents.FISHING_BOBBER_RETRIEVE, SoundSource.NEUTRAL, 1.0F, 0.4F / (world.getRandom().nextFloat() * 0.4F + 0.8F));
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 06b4135d3a684ef31c969b23d2349a6fef8a905f..a5f53ee75524ad8840785dc180d8a7b502d6b2e6 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2253,6 +2253,14 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
 
     // Paper start
+    @Override
+    public org.bukkit.entity.FishHook getFishingHook() {
+        if (getHandle().fishing == null) {
+            return null;
+        }
+        return (org.bukkit.entity.FishHook) getHandle().fishing.getBukkitEntity();
+    }
+
     public void setAffectsSpawning(boolean affects) {
         this.getHandle().affectsSpawning = affects;
     }
