From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 25 Apr 2014 00:38:10 -0400
Subject: [PATCH] Cancel Vehicle/Passenger if spawn is cancelled

Otherwise lots of chickens

diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 1410779cc6f12b7ddbc23757b503b9a4c3fdbb56..df6021ccec964f97717f42d0ff1ea814b919f89b 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -707,6 +707,16 @@ public class CraftEventFactory {
 
         CreatureSpawnEvent event = new CreatureSpawnEvent(entity, spawnReason);
         craftServer.getPluginManager().callEvent(event);
+        // EMC start
+        if (event.isCancelled() && spawnReason != SpawnReason.FORCE) {
+            if (entity.getVehicle() != null) {
+                entity.getVehicle().remove();
+            }
+            for (org.bukkit.entity.Entity passenger : entity.getPassengers()) {
+                passenger.remove();
+            }
+        }
+        // EMC end
         return event;
     }
 
