From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 6 Nov 2014 21:36:14 -0500
Subject: [PATCH] Mark Withers as Persistent

We don't want them despawning due to player range

diff --git a/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java b/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
index e81e8f050bd9df34b6a64c741428503b434f03a3..6991ba7a27cee3282ac4762e5005d0e0893ed5ea 100644
--- a/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -588,6 +588,7 @@ public class WitherBoss extends Monster implements PowerableMob, RangedAttackMob
         return this.getHealth() <= this.getMaxHealth() / 2.0F;
     }
 
+    @Override public boolean removeWhenFarAway(double distanceSquared) { return false; } // EMC
     @Override
     public MobType getMobType() {
         return MobType.UNDEAD;
