From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 6 Nov 2014 21:36:14 -0500
Subject: [PATCH] Mark Withers as Persistent

We don't want them despawning due to player range

diff --git a/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java b/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
index 563a028018421cbafdeff398175a32d6830134af..0bf896d6dfb67e655771a6f64ecf56c31fff8e78 100644
--- a/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -595,6 +595,7 @@ public class WitherBoss extends Monster implements PowerableMob, RangedAttackMob
         return this.getHealth() <= this.getMaxHealth() / 2.0F;
     }
 
+    @Override public boolean removeWhenFarAway(double distanceSquared) { return false; } // EMC
     @Override
     public MobType getMobType() {
         return MobType.UNDEAD;
