From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 14 Aug 2014 18:23:33 -0400
Subject: [PATCH] Add Player.getFishingHook() API and kill hook after rod
 damage event


diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index 852a4edde291bf368b2396e3c94ab402e3c66622..7680bf7fb8a23f6da3339b6e0322dae67d7c55dc 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -558,8 +558,8 @@ public class FishingHook extends Projectile {
             }
             // CraftBukkit end
 
-            this.discard();
-            return i;
+            //this.discard(); // EMC - move to after damage event in EntityFishingRod
+            return i == 0 ? -1 : i; // EMC - use -1 to represent "0 but kill entity"
         } else {
             return 0;
         }
diff --git a/src/main/java/net/minecraft/world/item/FishingRodItem.java b/src/main/java/net/minecraft/world/item/FishingRodItem.java
index d0e8729c76c13ea2469e3b1010d7199a94622f86..2f4f9f394e473971e50f00aa41e5e23eb4ea07f9 100644
--- a/src/main/java/net/minecraft/world/item/FishingRodItem.java
+++ b/src/main/java/net/minecraft/world/item/FishingRodItem.java
@@ -28,9 +28,10 @@ public class FishingRodItem extends Item implements Vanishable {
         if (user.fishing != null) {
             if (!world.isClientSide) {
                 i = user.fishing.retrieve(itemstack);
-                itemstack.hurtAndBreak(i, user, (entityhuman1) -> {
+                itemstack.hurtAndBreak(Math.max(0, i), user, (entityhuman1) -> { // EMC - Ignore -1 return using Math.max
                     entityhuman1.broadcastBreakEvent(hand);
                 });
+                if (i != 0) { user.fishing.discard(); } // EMC - removed EntityFishingHook, and 0 = cancelled
             }
 
             world.playSound((Player) null, user.getX(), user.getY(), user.getZ(), SoundEvents.FISHING_BOBBER_RETRIEVE, SoundSource.NEUTRAL, 1.0F, 0.4F / (world.getRandom().nextFloat() * 0.4F + 0.8F));
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 8c8698f6e17562651fe790dc19825e428d5d0352..e2e1942f8f7953252b5e8ded9188f81a3407d6bd 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2175,6 +2175,14 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     }
 
     // Paper start
+    @Override
+    public org.bukkit.entity.FishHook getFishingHook() {
+        if (getHandle().fishing == null) {
+            return null;
+        }
+        return (org.bukkit.entity.FishHook) getHandle().fishing.getBukkitEntity();
+    }
+
     public void setAffectsSpawning(boolean affects) {
         this.getHandle().affectsSpawning = affects;
     }
