From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 19 Oct 2014 02:32:26 -0400
Subject: [PATCH] Exterminate the chickeneer - EMC-610

Wipe out chickens when the Jockey is killed.

diff --git a/src/main/java/net/minecraft/world/entity/animal/Chicken.java b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
index 9fe34da92c26185af241e6d5be90e695b03cfa99..ab284a3fac22723d55e828e7113a8927c66b8773 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Chicken.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
@@ -166,6 +166,12 @@ public class Chicken extends Animal {
         }
 
     }
+    // EMC start
+    @Override
+    public boolean save(CompoundTag nbt) {
+        return (isVehicle() || !isChickenJockey()) && super.save(nbt);
+    }
+    // EMC end
 
     @Override
     public void addAdditionalSaveData(CompoundTag nbt) {
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index 5e6af78f639bfc403e49089775e285de1c266edb..77098107d5f81dc630bd6120020883147ef2fdb6 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -598,6 +598,7 @@ public class Zombie extends Monster {
         super.dropCustomDeathLoot(source, lootingMultiplier, allowDrops);
         Entity entity = source.getEntity();
 
+        if (getVehicle() instanceof Chicken) {((LivingEntity) getVehicle()).die(source);} // EMC
         if (entity instanceof Creeper) {
             Creeper entitycreeper = (Creeper) entity;
 
