From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 19 Oct 2014 02:32:26 -0400
Subject: [PATCH] Exterminate the chickeneer - EMC-610

Wipe out chickens when the Jockey is killed.

diff --git a/src/main/java/net/minecraft/world/entity/animal/Chicken.java b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
index cece07e000a4355e88aef7848fd264191521484e..9ea43e23744743431a298f2afd7066cbe7b255bf 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Chicken.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
@@ -166,6 +166,12 @@ public class Chicken extends Animal {
         }
 
     }
+    // EMC start
+    @Override
+    public boolean save(CompoundTag nbt) {
+        return (isVehicle() || !isChickenJockey()) && super.save(nbt);
+    }
+    // EMC end
 
     @Override
     public void addAdditionalSaveData(CompoundTag nbt) {
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index 11fb677d998b7ec82baab17322aceb78c6890fce..467a02aa68dca4713265d8d8e31f7fd6f7777e61 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -598,6 +598,7 @@ public class Zombie extends Monster {
         super.dropCustomDeathLoot(source, lootingMultiplier, allowDrops);
         Entity entity = source.getEntity();
 
+        if (getVehicle() instanceof Chicken) {((LivingEntity) getVehicle()).die(source);} // EMC
         if (entity instanceof Creeper) {
             Creeper entitycreeper = (Creeper) entity;
 
