From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 18 Dec 2014 22:48:46 -0500
Subject: [PATCH] Add BlockBreakNaturally Event

To give reliable control over all blocks dropping to world to restore custom item meta

diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 9b64cfa0d8527a6539ca1a95fb56c2ccf46ece79..b6de8331a371e5b900d54f5a2acb4986d55dc67a 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1405,6 +1405,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
         } else {
             // Paper start - capture all item additions to the world
             if (captureDrops != null && entity instanceof net.minecraft.world.entity.item.ItemEntity) {
+                ((net.minecraft.world.entity.item.ItemEntity) entity).droppedPosition = new BlockPos(entity.getX(), entity.getY(), entity.getZ()); // EMC
                 captureDrops.add((net.minecraft.world.entity.item.ItemEntity) entity);
                 return true;
             }
diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index 2c8570381e4e84edb29fe54b7541760c26d451f7..7658c4c2b09ca51686c5a400e8aea85d99463ade 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -46,6 +46,7 @@ public class ItemEntity extends Entity {
     private static final int INFINITE_LIFETIME = -32768;
     public int age;
     public boolean canDespawn = true; // EMC
+    public BlockPos droppedPosition; // EMC
     public int pickupDelay;
     private int health;
     @Nullable
diff --git a/src/main/java/net/minecraft/world/level/block/Block.java b/src/main/java/net/minecraft/world/level/block/Block.java
index ab5b9f00123e2ede2931ffc520684e482aac49b4..417eeaf6131b6e8a5ad039619fcd418aa7c1fd7c 100644
--- a/src/main/java/net/minecraft/world/level/block/Block.java
+++ b/src/main/java/net/minecraft/world/level/block/Block.java
@@ -21,6 +21,7 @@ import net.minecraft.core.Registry;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.network.chat.TranslatableComponent;
+import net.minecraft.server.MCUtil;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.BlockTags;
@@ -367,7 +368,7 @@ public class Block extends BlockBehaviour implements ItemLike {
 
         Block.popResource(world, () -> {
             return new ItemEntity(world, d0, d1, d2, stack);
-        }, stack);
+        }, stack, pos);
     }
 
     public static void popResourceFromFace(Level world, BlockPos pos, Direction direction, ItemStack stack) {
@@ -387,18 +388,20 @@ public class Block extends BlockBehaviour implements ItemLike {
 
         Block.popResource(world, () -> {
             return new ItemEntity(world, d0, d1, d2, stack, d3, d4, d5);
-        }, stack);
+        }, stack, pos);
     }
 
-    private static void popResource(Level world, Supplier<ItemEntity> itemEntitySupplier, ItemStack stack) {
+    private static void popResource(Level world, Supplier<ItemEntity> itemEntitySupplier, ItemStack stack, BlockPos blockPos) { // EMC - add blockPos
         if (!world.isClientSide && !stack.isEmpty() && world.getGameRules().getBoolean(GameRules.RULE_DOBLOCKDROPS)) {
             ItemEntity entityitem = (ItemEntity) itemEntitySupplier.get();
 
             entityitem.setDefaultPickUpDelay();
             // CraftBukkit start
             if (world.captureDrops != null) {
+                entityitem.droppedPosition = blockPos.immutable(); // EMC
                 world.captureDrops.add(entityitem);
             } else {
+                new com.empireminecraft.customevents.BlockBreakNaturallyEvent(MCUtil.toLocation(world, blockPos), (org.bukkit.entity.Item) entityitem.getBukkitEntity()).callEvent(); // EMC
                 world.addFreshEntity(entityitem);
             }
             // CraftBukkit end
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 538bee399a928b117aaa968a0d8b3bf450591235..5d0d006cd28934e8fb597e511fe162c3ceb0b43d 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -405,6 +405,16 @@ public class CraftEventFactory {
         Bukkit.getPluginManager().callEvent(event);
 
         if (!event.isCancelled()) {
+            // EMC start
+            com.google.common.collect.Multimap<org.bukkit.Location, org.bukkit.entity.Item> drops = com.google.common.collect.ArrayListMultimap.create();
+            for (Item item : list) {
+                ItemEntity entityItem = (ItemEntity) ((CraftEntity) item).getHandle();
+                drops.put(net.minecraft.server.MCUtil.toLocation(entityItem.level, entityItem.droppedPosition), (org.bukkit.entity.Item) entityItem.getBukkitEntity());
+            }
+            for (org.bukkit.Location location : drops.keySet()) {
+                new com.empireminecraft.customevents.BlockBreakNaturallyEvent(location, drops.get(location)).callEvent();
+            }
+            // EMC end
             // Paper start
             for (Item bukkit : list) {
                 if (!bukkit.isValid()) {
