From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 19 Oct 2014 02:32:26 -0400
Subject: [PATCH] Exterminate the chickeneer - EMC-610

Wipe out chickens when the Jockey is killed.

diff --git a/src/main/java/net/minecraft/world/entity/animal/Chicken.java b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
index 8460bed561c09a647f6e0209f7c5448e5a42b281..7b4b9191fceb3187cf37eccb5be5944d8f2f3489 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Chicken.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Chicken.java
@@ -166,6 +166,12 @@ public class Chicken extends Animal {
         }
 
     }
+    // EMC start
+    @Override
+    public boolean save(CompoundTag nbt) {
+        return (isVehicle() || !isChickenJockey()) && super.save(nbt);
+    }
+    // EMC end
 
     @Override
     public void addAdditionalSaveData(CompoundTag nbt) {
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index 92c1f34a32aa92ac8cbfed25ba4336e24c7761ce..a512776b41cf9d8417b7aa84940125eefb57213c 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -593,6 +593,7 @@ public class Zombie extends Monster {
         super.dropCustomDeathLoot(source, lootingMultiplier, allowDrops);
         Entity entity = source.getEntity();
 
+        if (getVehicle() instanceof Chicken) {((LivingEntity) getVehicle()).die(source);} // EMC
         if (entity instanceof Creeper) {
             Creeper entitycreeper = (Creeper) entity;
 
