From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 6 Nov 2014 21:36:14 -0500
Subject: [PATCH] Mark Withers as Persistent

We don't want them despawning due to player range

diff --git a/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java b/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
index 1c8f6863b976cfcb559de9b3e3cf9292831166ee..39a1f25de2cb615d867db6b4c69e3bd62ce6f0f7 100644
--- a/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/src/main/java/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -585,6 +585,7 @@ public class WitherBoss extends Monster implements PowerableMob, RangedAttackMob
         return this.getHealth() <= this.getMaxHealth() / 2.0F;
     }
 
+    @Override public boolean removeWhenFarAway(double distanceSquared) { return false; } // EMC
     @Override
     public MobType getMobType() {
         return MobType.UNDEAD;
