From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 30 Apr 2015 00:46:26 -0400
Subject: [PATCH] Don't require flag for sign conversion

Also reverts part of Paper's old sign conversion patch

diff --git a/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
index 6c96c3dde4e1b1b1b6952841edb41006e73737e9..297f34666971b1a78bb8199a752e4def4f1de5b2 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -25,7 +25,7 @@ import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 
 public class SignBlockEntity extends BlockEntity implements CommandSource { // CraftBukkit - implements
-    private static final boolean CONVERT_LEGACY_SIGNS = Boolean.getBoolean("convertLegacySigns"); // Paper
+    private static final boolean CONVERT_LEGACY_SIGNS = true; // EMC
 
     public static final int LINES = 4;
     private static final String[] RAW_TEXT_FIELD_NAMES = new String[]{"Text1", "Text2", "Text3", "Text4"};
@@ -85,7 +85,7 @@ public class SignBlockEntity extends BlockEntity implements CommandSource { // C
         // CraftBukkit start - Add an option to convert signs correctly
         // This is done with a flag instead of all the time because
         // we have no way to tell whether a sign is from 1.7.10 or 1.8
-        boolean oldSign = Boolean.getBoolean("convertLegacySigns") && !nbt.getBoolean("Bukkit.isConverted");
+        boolean oldSign = !nbt.getBoolean("Bukkit.isConverted"); // EMC - don't use flag
         // CraftBukkit end
 
         for (int i = 0; i < 4; ++i) {
