From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 18 Dec 2014 23:03:44 -0500
Subject: [PATCH] Use FlowerPotManipulateEvent item for flowerpot removal


diff --git a/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java b/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
index 7194565ad1a910fe300d8664a30b35c4eff18cb7..35cb09b63c51fdd0639a28b3fc9e084967b9434b 100644
--- a/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
@@ -101,6 +101,7 @@ public class FlowerPotBlock extends Block {
                 return InteractionResult.PASS;
             }
             // Paper end - Add PlayerFlowerPotManipulateEvent
+            itemStack = ItemStack.fromBukkitCopy(event.getItem()); // EMC
             if (!player.addItem(itemStack)) {
                 player.drop(itemStack, false);
             }
