From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 18 Dec 2014 23:03:44 -0500
Subject: [PATCH] Add FlowerPotRemoveEvent


diff --git a/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java b/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
index db05c1ea847d60ad45d33cd798cb34ad3f5cfd75..9e00e5df8ce6ee43175e27edebc1d3e499850aac 100644
--- a/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
@@ -80,6 +80,12 @@ public class FlowerPotBlock extends Block {
                 }
             } else {
                 ItemStack itemStack2 = new ItemStack(this.content);
+                // EMC Start
+                com.empireminecraft.customevents.FlowerPotRemoveEvent removeEvent = new com.empireminecraft.customevents.FlowerPotRemoveEvent(
+                    player1, net.minecraft.server.MCUtil.toLocation(world, pos), itemStack2.getBukkitStack());
+                removeEvent.callEvent();
+                itemStack2 = ItemStack.fromBukkitCopy(removeEvent.getItem());
+                // EMC End
                 if (itemStack.isEmpty()) {
                     player.setItemInHand(hand, itemStack2);
                 } else if (!player.addItem(itemStack2)) {
