From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 18 Dec 2014 23:03:44 -0500
Subject: [PATCH] Use FlowerPotManipulateEvent item for flowerpot removal


diff --git a/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java b/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
index b229b21b4a9b953983e743f345e46e4235e2809d..c96337e3f0de54f4af73ef9fe2a881b654f91371 100644
--- a/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
@@ -78,7 +78,7 @@ public class FlowerPotBlock extends Block {
                     itemStack.shrink(1);
                 }
             } else {
-                ItemStack itemStack2 = new ItemStack(this.content);
+                ItemStack itemStack2 = ItemStack.fromBukkitCopy(event.getItem()); // EMC
                 if (itemStack.isEmpty()) {
                     player.setItemInHand(hand, itemStack2);
                 } else if (!player.addItem(itemStack2)) {
