From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 8 May 2015 19:56:21 -0400
Subject: [PATCH] ServerShutdownEvent and move disablePlugins


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 98eeb58ff725d231f7d8dbeb90393dea186cbd79..751050b3ccce8c3376baaa3180744aa2f529f43f 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -245,6 +245,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     protected final Proxy proxy;
     private boolean onlineMode;
     private boolean preventProxyConnections;
+    public String shutdownReason = "Server Shutting Down"; // EMC
     private boolean pvp;
     private boolean allowFlight;
     @Nullable
@@ -943,8 +944,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         MinecraftTimings.stopServer(); // Paper
         // CraftBukkit start
         if (this.server != null) {
-            this.server.disablePlugins();
-            this.server.waitForAsyncTasksShutdown(); // Paper
+            //this.server.disablePlugins(); // EMC - Moved below after everything has unloaded so plugins can get events
+            //this.server.waitForAsyncTasksShutdown(); // Paper // EMC - moved down
+            new com.empireminecraft.customevents.ServerShutdownEvent(shutdownReason).callEvent(); // EMC
         }
         // CraftBukkit end
         if (this.getConnection() != null) {
@@ -973,6 +975,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         this.saveAllChunks(false, true, false, true); // Paper - rewrite chunk system - move closing into here
 
+        if (this.server != null) { this.server.disablePlugins(); this.server.waitForAsyncTasksShutdown(); } // EMC
         this.isSaving = false;
         this.resources.close();
 
