From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 18 Dec 2014 23:03:44 -0500
Subject: [PATCH] Use FlowerPotManipulateEvent item for flowerpot removal


diff --git a/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java b/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
index fbe399fa2c27ba283e4a51f9a2559baf86fdff11..12451f3b525cdb40724fb563d24ced0ff6f54466 100644
--- a/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FlowerPotBlock.java
@@ -88,7 +88,7 @@ public class FlowerPotBlock extends Block {
                     itemStack.shrink(1);
                 }
             } else {
-                ItemStack itemStack2 = new ItemStack(this.potted);
+                ItemStack itemStack2 = ItemStack.fromBukkitCopy(event.getItem()); // EMC
                 if (itemStack.isEmpty()) {
                     player.setItemInHand(hand, itemStack2);
                 } else if (!player.addItem(itemStack2)) {
