From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 8 May 2015 19:56:21 -0400
Subject: [PATCH] ServerShutdownEvent and move disablePlugins


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 696d075ca2883f3c37e35f983c4d020e5db89d16..35d4a2f9a2ab452e9c594ab30484336761d5ca6d 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -252,6 +252,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     protected final Proxy proxy;
     private boolean onlineMode;
     private boolean preventProxyConnections;
+    public String shutdownReason = "Server Shutting Down"; // EMC
     private boolean pvp;
     private boolean allowFlight;
     private net.kyori.adventure.text.Component motd; // Paper - Adventure
@@ -1042,8 +1043,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.server.spark.disable(); // Paper - spark
         // CraftBukkit start
         if (this.server != null) {
-            this.server.disablePlugins();
-            this.server.waitForAsyncTasksShutdown(); // Paper - Wait for Async Tasks during shutdown
+            //this.server.disablePlugins();
+            //this.server.waitForAsyncTasksShutdown(); // Paper - Wait for Async Tasks during shutdown
+            new com.empireminecraft.customevents.ServerShutdownEvent(shutdownReason).callEvent(); // EMC
         }
         // CraftBukkit end
         if (io.papermc.paper.plugin.PluginInitializerManager.instance().pluginRemapper != null) io.papermc.paper.plugin.PluginInitializerManager.instance().pluginRemapper.shutdown(); // Paper - Plugin remapping
@@ -1087,6 +1089,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         this.saveAllChunks(false, true, true, true); // Paper - rewrite chunk system
 
+        if (this.server != null) { this.server.disablePlugins(); this.server.waitForAsyncTasksShutdown(); } // EMC
         this.isSaving = false;
         this.resources.close();
 
