From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 8 May 2015 19:56:21 -0400
Subject: [PATCH] ServerShutdownEvent and move disablePlugins


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index aa3b8d70d3c792ef4ecc4bc808dda1a19570c241..91967488aaebddd85e7b30586d76b443fbc5c2ee 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -238,6 +238,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     protected final Proxy proxy;
     private boolean onlineMode;
     private boolean preventProxyConnections;
+    public String shutdownReason = "Server Shutting Down"; // EMC
     private boolean pvp;
     private boolean allowFlight;
     @Nullable
@@ -970,8 +971,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         MinecraftTimings.stopServer(); // Paper
         // CraftBukkit start
         if (this.server != null) {
-            this.server.disablePlugins();
-            this.server.waitForAsyncTasksShutdown(); // Paper
+            //this.server.disablePlugins(); // EMC - Moved below after everything has unloaded so plugins can get events
+            //this.server.waitForAsyncTasksShutdown(); // Paper // EMC - moved down
+            new com.empireminecraft.customevents.ServerShutdownEvent(shutdownReason).callEvent(); // EMC
         }
         // CraftBukkit end
         if (this.getConnection() != null) {
@@ -1037,6 +1039,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 }
             }
         }
+        if (server != null) { this.server.disablePlugins(); this.server.waitForAsyncTasksShutdown(); } // EMC
 
         this.isSaving = false;
         this.resources.close();
