From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 8 May 2015 19:56:21 -0400
Subject: [PATCH] ServerShutdownEvent and move disablePlugins


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 89d1748cdddcd4ddb299d2efe0e93a03e848b8a2..12cb800b7df438bb7c35f1a0311568d4428e7c2a 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -246,6 +246,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     protected final Proxy proxy;
     private boolean onlineMode;
     private boolean preventProxyConnections;
+    public String shutdownReason = "Server Shutting Down"; // EMC
     private boolean pvp;
     private boolean allowFlight;
     private net.kyori.adventure.text.Component motd; // Paper - Adventure
@@ -939,8 +940,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         MinecraftTimings.stopServer(); // Paper
         // CraftBukkit start
         if (this.server != null) {
-            this.server.disablePlugins();
-            this.server.waitForAsyncTasksShutdown(); // Paper
+            //this.server.disablePlugins(); // EMC - Moved below after everything has unloaded so plugins can get events
+            //this.server.waitForAsyncTasksShutdown(); // Paper // EMC - moved down
+            new com.empireminecraft.customevents.ServerShutdownEvent(shutdownReason).callEvent(); // EMC
         }
         // CraftBukkit end
         if (this.getConnection() != null) {
@@ -969,6 +971,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         this.saveAllChunks(false, true, false, true); // Paper - rewrite chunk system - move closing into here
 
+        if (this.server != null) { this.server.disablePlugins(); this.server.waitForAsyncTasksShutdown(); } // EMC
         this.isSaving = false;
         this.resources.close();
 
