From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 15 May 2015 21:21:19 -0400
Subject: [PATCH] Remove "Copy of" from Written Books


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 6c49c41d2ee05af140277dc79eb1cacf6f4070e8..b6578ff0979cfc218a3c540f6cb678941187e80f 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -285,6 +285,7 @@ public final class ItemStack {
             this.tag = nbttagcompound.getCompound("tag").copy();
             this.processEnchantOrder(this.tag); // Paper
             this.processText(); // Paper - Update itemstack legacy name and lore
+            if (item == Items.WRITTEN_BOOK && tag.contains("generation")) { tag.remove("generation"); } // EMC
             this.getItem().verifyTagAfterLoad(this.tag);
         }
 
diff --git a/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java b/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
index 5a3b8cb9e275f747c82cf13a2a911a7654de8fc6..0821772c9e3f3c5c74324294c8b88c115f91e6f1 100644
--- a/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
+++ b/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
@@ -68,7 +68,7 @@ public class BookCloningRecipe extends CustomRecipe {
         if (!itemStack.isEmpty() && itemStack.hasTag() && i >= 1 && WrittenBookItem.getGeneration(itemStack) < 2) {
             ItemStack itemStack3 = new ItemStack(Items.WRITTEN_BOOK, i);
             CompoundTag compoundTag = itemStack.getTag().copy();
-            compoundTag.putInt("generation", WrittenBookItem.getGeneration(itemStack) + 1);
+            //compoundTag.putInt("generation", WrittenBookItem.getGeneration(itemStack) + 1); // EMC
             itemStack3.setTag(compoundTag);
             return itemStack3;
         } else {
