From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 15 May 2015 21:21:19 -0400
Subject: [PATCH] Remove "Copy of" from Written Books


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 47e0b76b9165ce94783f05d071bf8aa8c94ddc1d..5c10a5a2d096b2a8830736fb81596650eb6ab31b 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -262,6 +262,7 @@ public final class ItemStack {
             // CraftBukkit end
             this.processEnchantOrder(this.tag); // Paper
             this.processText(); // Paper
+            if (item == Items.WRITTEN_BOOK && tag.contains("generation")) { tag.remove("generation"); } // EMC
             this.getItem().verifyTagAfterLoad(this.tag);
         }
 
diff --git a/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java b/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
index c0e35cdbab4468afb2e0092e1d30e74d694d54e4..3701ca10da4fe4dc0a53c4f0065ec0c8c6bdf67b 100644
--- a/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
+++ b/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
@@ -69,7 +69,7 @@ public class BookCloningRecipe extends CustomRecipe {
         if (!itemStack.isEmpty() && itemStack.hasTag() && i >= 1 && WrittenBookItem.getGeneration(itemStack) < 2) {
             ItemStack itemStack3 = new ItemStack(Items.WRITTEN_BOOK, i);
             CompoundTag compoundTag = itemStack.getTag().copy();
-            compoundTag.putInt("generation", WrittenBookItem.getGeneration(itemStack) + 1);
+            //compoundTag.putInt("generation", WrittenBookItem.getGeneration(itemStack) + 1); // EMC
             itemStack3.setTag(compoundTag);
             return itemStack3;
         } else {
