From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 15 May 2015 21:21:19 -0400
Subject: [PATCH] Remove "Copy of" from Written Books


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 19b9c9b83d1a0974fcde32153dc5f54dcbd6c91b..ac4719e37f77b190e8cbc22155a8a7ebe215536e 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -256,6 +256,7 @@ public final class ItemStack {
             // CraftBukkit end
             this.processEnchantOrder(this.tag); // Paper
             this.processText(); // Paper
+            if (item == Items.WRITTEN_BOOK && tag.contains("generation")) { tag.remove("generation"); } // EMC
             this.getItem().verifyTagAfterLoad(this.tag);
         }
 
diff --git a/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java b/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
index 25cd2b0c6af8e7fa455069b2a618b110df1f82c9..2ab06446c45deb47a23cf95225a2fa56c4a038a5 100644
--- a/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
+++ b/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
@@ -68,7 +68,7 @@ public class BookCloningRecipe extends CustomRecipe {
         if (!itemStack.isEmpty() && itemStack.hasTag() && i >= 1 && WrittenBookItem.getGeneration(itemStack) < 2) {
             ItemStack itemStack3 = new ItemStack(Items.WRITTEN_BOOK, i);
             CompoundTag compoundTag = itemStack.getTag().copy();
-            compoundTag.putInt("generation", WrittenBookItem.getGeneration(itemStack) + 1);
+            //compoundTag.putInt("generation", WrittenBookItem.getGeneration(itemStack) + 1); // EMC
             itemStack3.setTag(compoundTag);
             return itemStack3;
         } else {
