From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Fri, 15 May 2015 21:21:19 -0400
Subject: [PATCH] Remove "Copy of" from Written Books


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 419ae4cb98b1ce430876125fe33676f08c993eb7..844fd0b0550f5064970cb67fc2d10470ac16879d 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -265,6 +265,7 @@ public final class ItemStack {
             // CraftBukkit end
             this.processEnchantOrder(this.tag); // Paper
             this.processText(); // Paper
+            if (item == Items.WRITTEN_BOOK && tag.contains("generation")) { tag.remove("generation"); } // EMC
             this.getItem().verifyTagAfterLoad(this.tag);
         }
 
diff --git a/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java b/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
index 818fa56beb2f14030980f3662b28a6e67e39c69f..42c26f54fa8ddf5d0438445898f26033df1a3bc5 100644
--- a/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
+++ b/src/main/java/net/minecraft/world/item/crafting/BookCloningRecipe.java
@@ -69,7 +69,7 @@ public class BookCloningRecipe extends CustomRecipe {
         if (!itemStack.isEmpty() && itemStack.hasTag() && i >= 1 && WrittenBookItem.getGeneration(itemStack) < 2) {
             ItemStack itemStack3 = new ItemStack(Items.WRITTEN_BOOK, i);
             CompoundTag compoundTag = itemStack.getTag().copy();
-            compoundTag.putInt("generation", WrittenBookItem.getGeneration(itemStack) + 1);
+            //compoundTag.putInt("generation", WrittenBookItem.getGeneration(itemStack) + 1); // EMC
             itemStack3.setTag(compoundTag);
             return itemStack3;
         } else {
