From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 14 Jan 2016 00:49:14 -0500
Subject: [PATCH] Do not drop items from Give command


diff --git a/src/main/java/net/minecraft/server/commands/GiveCommand.java b/src/main/java/net/minecraft/server/commands/GiveCommand.java
index 0d9de4c61c7b26a6ff37c12fde629161fd0c3d5a..18fd23dc8221c8274428ec449a6e2207777756da 100644
--- a/src/main/java/net/minecraft/server/commands/GiveCommand.java
+++ b/src/main/java/net/minecraft/server/commands/GiveCommand.java
@@ -58,6 +58,7 @@ public class GiveCommand {
                     l -= i1;
                     ItemStack itemstack1 = item.createItemStack(i1, false);
                     boolean flag = entityplayer.getInventory().add(itemstack1);
+                    if (true) { continue; } // EMC - never drop items
                     ItemEntity entityitem;
 
                     if (flag && itemstack1.isEmpty()) {
