From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 8 Jun 2015 23:55:20 -0400
Subject: [PATCH] SendSignEvent


diff --git a/src/main/java/com/empireminecraft/api/SendSignEventImpl.java b/src/main/java/com/empireminecraft/api/SendSignEventImpl.java
new file mode 100644
index 0000000000000000000000000000000000000000..4538f2d2e46dcfb749d0078696c524764968492d
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/SendSignEventImpl.java
@@ -0,0 +1,44 @@
+package com.empireminecraft.api;
+
+import com.empireminecraft.customevents.SendSignEvent;
+import net.minecraft.core.BlockPos;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.TextComponent;
+import net.minecraft.world.level.Level;
+import org.bukkit.block.Block;
+import org.bukkit.craftbukkit.util.CraftChatMessage;
+import org.jetbrains.annotations.NotNull;
+
+public class SendSignEventImpl extends SendSignEvent {
+    public final Component[] lines;
+    private final String[] slines;
+    private final Block block;
+
+    public SendSignEventImpl(Level level, BlockPos position, Component[] lines) {
+        this.block = level.getWorld().getBlockAt(position.getX(), position.getY(), position.getZ());
+        this.lines = new Component[lines.length];
+        this.slines = new String[lines.length];
+        for (int i = 0; i < lines.length; i++) {
+            this.lines[i] = lines[i];
+            slines[i] = CraftChatMessage.fromComponent(lines[i]);
+        }
+    }
+
+    @NotNull
+    public Block getBlock() {
+        return block;
+    }
+
+    @NotNull
+    public String[] getLines() {
+        return slines;
+    }
+
+    public void setLine(int i, @NotNull String line) {
+        slines[i] = line;
+        lines[i] = new TextComponent("");
+        for (Component comp : CraftChatMessage.fromString(line)) {
+            lines[i].getSiblings().add(comp);
+        }
+    }
+}
diff --git a/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
index 36501794370efc839f9b013f016492c86a970ff7..74ef57e18fb09e372bd27e138a249d45d84bec47 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.level.block.entity;
 
+import com.empireminecraft.api.SendSignEventImpl;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import java.util.UUID;
 import java.util.function.Function;
@@ -53,14 +54,19 @@ public class SignBlockEntity extends BlockEntity implements CommandSource { // C
 
     @Override
     public CompoundTag save(CompoundTag nbt) {
+        // EMC start
+        return save(nbt, this.messages, this.filteredMessages);
+    }
+    public CompoundTag save(CompoundTag nbt, Component[] lines, Component[] filteredMessages) {
+        // EMC end
         super.save(nbt);
 
         for (int i = 0; i < 4; ++i) {
-            Component ichatbasecomponent = this.messages[i];
+            Component ichatbasecomponent = lines[i]; // EMC
             String s = Component.Serializer.toJson(ichatbasecomponent);
 
             nbt.putString(SignBlockEntity.RAW_TEXT_FIELD_NAMES[i], s);
-            Component ichatbasecomponent1 = this.filteredMessages[i];
+            Component ichatbasecomponent1 = filteredMessages[i]; // EMC
 
             if (!ichatbasecomponent1.equals(ichatbasecomponent)) {
                 nbt.putString(SignBlockEntity.FILTERED_TEXT_FIELD_NAMES[i], Component.Serializer.toJson(ichatbasecomponent1));
@@ -189,6 +195,13 @@ public class SignBlockEntity extends BlockEntity implements CommandSource { // C
 
     @Override
     public CompoundTag getUpdateTag() {
+        // EMC start
+        if (this.level != null) {
+            SendSignEventImpl event = new SendSignEventImpl(this.level, this.getBlockPos(), this.messages);
+            event.callEvent();
+            return this.save(new CompoundTag(), event.lines, event.lines);
+        }
+        // EMC end
         return this.save(new CompoundTag());
     }
 
