From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 25 Jun 2015 21:00:09 -0400
Subject: [PATCH] ServerReloadEvent


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 708be121f76405e135cf59f2413e75ce44cc1a0a..5c70cd49a5c0da0ece6c8bd7eb7547f3c6272b25 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -847,6 +847,7 @@ public final class CraftServer implements Server {
     @Override
     public void reload() {
         org.spigotmc.WatchdogThread.hasStarted = false; // Paper - Disable watchdog early timeout on reload
+        new com.empireminecraft.customevents.ServerReloadEvent().callEvent(); // EMC
         this.reloadCount++;
         this.configuration = YamlConfiguration.loadConfiguration(this.getConfigFile());
         this.commandsConfiguration = YamlConfiguration.loadConfiguration(this.getCommandsConfigFile());
