From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 21 Dec 2015 13:33:00 -0500
Subject: [PATCH] SnowmanThrowSnowballEvent


diff --git a/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java b/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java
index 96bfba9d3ccdbf95b8eea4038bf42e6b1430181d..e4f5f55ab9708a2525768e47a7aea692118b03a0 100644
--- a/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java
@@ -138,10 +138,27 @@ public class SnowGolem extends AbstractGolem implements Shearable, RangedAttackM
         double d4 = Math.sqrt(d1 * d1 + d3 * d3) * 0.20000000298023224D;
 
         entitysnowball.shoot(d1, d2 + d4, d3, 1.6F, 12.0F);
+        if (!fireEvent(this, entitysnowball, target)) { return; } // EMC
         this.playSound(SoundEvents.SNOW_GOLEM_SHOOT, 1.0F, 0.4F / (this.getRandom().nextFloat() * 0.4F + 0.8F));
         this.level().addFreshEntity(entitysnowball);
     }
 
+    // EMC start
+    public static boolean fireEvent(SnowGolem snowman, Snowball snowball, LivingEntity target) {
+        org.bukkit.entity.Snowman bukkitSnowman = (org.bukkit.entity.Snowman) snowman.getBukkitEntity();
+        org.bukkit.entity.Snowball bukkitSnowball = (org.bukkit.entity.Snowball) snowball.getBukkitEntity();
+        snowball.projectileSource = bukkitSnowman;
+        if (!(new org.bukkit.event.entity.ProjectileLaunchEvent(bukkitSnowball).callEvent())) {
+            return false;
+        }
+        com.empireminecraft.customevents.SnowmanThrowSnowballEvent event = new com.empireminecraft.customevents.SnowmanThrowSnowballEvent(
+            bukkitSnowman,
+            bukkitSnowball,
+            target.getBukkitLivingEntity()
+        );
+        return event.callEvent();
+    }
+    // EMC end
     @Override
     protected float getStandingEyeHeight(Pose pose, EntityDimensions dimensions) {
         return 1.7F;
