From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 17 Nov 2015 01:28:08 -0500
Subject: [PATCH] Fire Vehicle Events for all entities


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index bfb45ce4897191e50738e42c643a405528a5962c..9103e852a6d7a30b6336500719b49d2980e7b2be 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2703,9 +2703,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
             // CraftBukkit start
             CraftEntity craft = (CraftEntity) entity.getBukkitEntity().getVehicle();
             Entity orig = craft == null ? null : craft.getHandle();
-            if (this.getBukkitEntity() instanceof Vehicle && entity.getBukkitEntity() instanceof LivingEntity) {
+            if (entity.getBukkitEntity() instanceof LivingEntity) { // EMC
                 VehicleExitEvent event = new VehicleExitEvent(
-                        (Vehicle) this.getBukkitEntity(),
+                        this.getBukkitEntity(), // EMC
                         (LivingEntity) entity.getBukkitEntity(), !suppressCancellation // Paper
                 );
                 event.reason = reason; // EMC
