From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 14 Jan 2016 00:49:14 -0500
Subject: [PATCH] Do not drop items from Give command


diff --git a/src/main/java/net/minecraft/server/commands/GiveCommand.java b/src/main/java/net/minecraft/server/commands/GiveCommand.java
index d601d287e94a59ff93b8a83a44dac02544d211df..831d882826d50fae437a882edd2a3bfc5fde88d9 100644
--- a/src/main/java/net/minecraft/server/commands/GiveCommand.java
+++ b/src/main/java/net/minecraft/server/commands/GiveCommand.java
@@ -57,6 +57,7 @@ public class GiveCommand {
                     l -= i1;
                     ItemStack itemstack1 = item.createItemStack(i1, false);
                     boolean flag = entityplayer.getInventory().add(itemstack1);
+                    if (true) { continue; } // EMC - never drop items
                     ItemEntity entityitem;
 
                     if (flag && itemstack1.isEmpty()) {
