From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 14 Jan 2016 00:49:14 -0500
Subject: [PATCH] Do not drop items from Give command


diff --git a/src/main/java/net/minecraft/server/commands/GiveCommand.java b/src/main/java/net/minecraft/server/commands/GiveCommand.java
index a60a84d5df0b05cf35668df8c5f08b811bc42a9f..cad16980917fb03c9a1405e018261838521629be 100644
--- a/src/main/java/net/minecraft/server/commands/GiveCommand.java
+++ b/src/main/java/net/minecraft/server/commands/GiveCommand.java
@@ -57,6 +57,7 @@ public class GiveCommand {
                     boolean flag = entityplayer.getInventory().add(itemstack);
                     ItemEntity entityitem;
 
+                    if (true) { continue; } // EMC - never drop items
                     if (flag && itemstack.isEmpty()) {
                         itemstack.setCount(1);
                         entityitem = entityplayer.drop(itemstack, false, false, false); // SPIGOT-2942: Add boolean to call event
