From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 16 Nov 2015 23:23:03 -0500
Subject: [PATCH] Implement BatMoveEvent

Control where the bat goes

diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index bd68c1eed703e1c3e3fd1a480c89f1920faaf229..9ac5626f053166cae636dddb51362f3b5d1270b8 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -43,6 +43,7 @@ import net.minecraft.world.entity.ai.goal.GoalSelector;
 import net.minecraft.world.entity.ai.navigation.GroundPathNavigation;
 import net.minecraft.world.entity.ai.navigation.PathNavigation;
 import net.minecraft.world.entity.ai.sensing.Sensing;
+import net.minecraft.world.entity.ambient.Bat;
 import net.minecraft.world.entity.decoration.HangingEntity;
 import net.minecraft.world.entity.decoration.LeashFenceKnotEntity;
 import net.minecraft.world.entity.item.ItemEntity;
@@ -215,7 +216,7 @@ public abstract class Mob extends LivingEntity {
     // Paper end
 
     public MoveControl getMoveControl() {
-        if (this.isPassenger() && this.getVehicle() instanceof Mob) {
+        if (this.isPassenger() && this.getVehicle() instanceof Mob && !(this instanceof Bat)) { // EMC
             Mob entityinsentient = (Mob) this.getVehicle();
 
             return entityinsentient.getMoveControl();
diff --git a/src/main/java/net/minecraft/world/entity/ambient/Bat.java b/src/main/java/net/minecraft/world/entity/ambient/Bat.java
index 153194d937d210e2e4fd8864e4a3c000f85d7e2e..9d082d1cd548f316bea77b1031e5bd4bbb45a4ed 100644
--- a/src/main/java/net/minecraft/world/entity/ambient/Bat.java
+++ b/src/main/java/net/minecraft/world/entity/ambient/Bat.java
@@ -164,7 +164,7 @@ public class Bat extends AmbientCreature {
             }
 
             if (this.targetPosition == null || this.random.nextInt(30) == 0 || this.targetPosition.closerThan((Position) this.position(), 2.0D)) {
-                this.targetPosition = new BlockPos(this.getX() + (double) this.random.nextInt(7) - (double) this.random.nextInt(7), this.getY() + (double) this.random.nextInt(6) - 2.0D, this.getZ() + (double) this.random.nextInt(7) - (double) this.random.nextInt(7));
+                this.targetPosition = getBatMoveTarget(this); // EMC
             }
 
             double d0 = (double) this.targetPosition.getX() + 0.5D - this.getX();
@@ -266,4 +266,16 @@ public class Bat extends AmbientCreature {
     protected float getStandingEyeHeight(Pose pose, EntityDimensions dimensions) {
         return dimensions.height / 2.0F;
     }
+    // EMC start
+    public static BlockPos getBatMoveTarget(Bat bat) {
+        org.bukkit.entity.Bat bukkitBat = (org.bukkit.entity.Bat) bat.getBukkitEntity();
+        org.bukkit.Location loc = com.empireminecraft.customevents.BatMoveEvent.generateLoc(bukkitBat);
+        com.empireminecraft.customevents.BatMoveEvent event = new com.empireminecraft.customevents.BatMoveEvent(bukkitBat, loc);
+        if (!event.callEvent()) {
+            return new BlockPos(bat.getX(), bat.getY(), bat.getZ());
+        }
+        org.bukkit.Location target = event.getMoveTarget();
+        return new BlockPos(target.getBlockX(), target.getBlockY(), target.getBlockZ());
+    }
+    // EMC end
 }
