From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 15 Jun 2016 22:06:57 -0400
Subject: [PATCH] Don't trigger Lootable Refresh for non player interaction


diff --git a/src/main/java/com/destroystokyo/paper/loottable/PaperLootableInventoryData.java b/src/main/java/com/destroystokyo/paper/loottable/PaperLootableInventoryData.java
index cd61276a45894a02cbefc41a63c27e2cf6361d1e..521c398430351a93363159b7d068a2fa9cbd1cad 100644
--- a/src/main/java/com/destroystokyo/paper/loottable/PaperLootableInventoryData.java
+++ b/src/main/java/com/destroystokyo/paper/loottable/PaperLootableInventoryData.java
@@ -49,6 +49,7 @@ public class PaperLootableInventoryData {
             return false;
         }
 
+        if (player == null) return false; // EMC - never fill for non-player
         // ALWAYS process the first fill or if the feature is disabled
         if (this.lastFill == -1 || !holderInterface.paperConfig(lootTableHolder).lootables.autoReplenish) {
             return true;
