From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 17 Nov 2015 01:28:08 -0500
Subject: [PATCH] Fire Vehicle Events for all entities


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index bcdd74dc137f07563a16d6b354863e77381a8a4b..614c0f7ec6491843d572dc6b03b8cfce74d44f49 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2455,9 +2455,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
             // CraftBukkit start
             CraftEntity craft = (CraftEntity) entity.getBukkitEntity().getVehicle();
             Entity orig = craft == null ? null : craft.getHandle();
-            if (this.getBukkitEntity() instanceof Vehicle && entity.getBukkitEntity() instanceof LivingEntity) {
+            if (entity.getBukkitEntity() instanceof LivingEntity) { // EMC
                 VehicleExitEvent event = new VehicleExitEvent(
-                        (Vehicle) this.getBukkitEntity(),
+                        this.getBukkitEntity(), // EMC
                         (LivingEntity) entity.getBukkitEntity(), !suppressCancellation // Paper
                 );
                 event.reason = reason; // EMC
