From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 30 Jun 2016 22:31:06 -0400
Subject: [PATCH] Don't use Players scoreboard team for pets


diff --git a/src/main/java/net/minecraft/world/entity/TamableAnimal.java b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
index 9aee0569447d351729c26eedbe24d5defe620162..44c47f5891af001010b9902eba643c77f8db7693 100644
--- a/src/main/java/net/minecraft/world/entity/TamableAnimal.java
+++ b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
@@ -214,6 +214,8 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         return true;
     }
 
+    // EMC start - don't link to owners team
+    /*
     @Override
     public PlayerTeam getTeam() {
         if (this.isTame()) {
@@ -225,7 +227,8 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         }
 
         return super.getTeam();
-    }
+    }*/
+    // EMC end
 
     @Override
     public boolean isAlliedTo(Entity other) {
