From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 13 Dec 2015 01:19:17 -0500
Subject: [PATCH] Reload Entity API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index c472715997cea1aa42b1543fb355ff24159d025a..fd17cb26e00de7502c9932193231f5b7682de93d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1251,4 +1251,14 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return getHandle().isTicking();
     }
     // Paper end
+    // EMC start
+    public boolean reloadEntity() {
+        Entity entity = this.entity.level.getEntities().get(this.entity.getUUID());
+        if (entity != null) {
+            this.entity = entity;
+            return true;
+        }
+        return false;
+    }
+    // EMC end
 }
