From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 23 Apr 2016 20:03:27 -0400
Subject: [PATCH] Give a Dragon Head instead of egg Every Dragon Kill


diff --git a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
index e9eb32469a5c03f7a3677ef50fd4541c1ed662ad..92782b8858aa0e59061758501739cfa68b7d6b0a 100644
--- a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -376,7 +376,7 @@ public class EndDragonFight {
             // Paper start - DragonEggFormEvent
             BlockPos eggPosition = this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.END_PODIUM_LOCATION);
             org.bukkit.craftbukkit.block.CraftBlockState eggState = org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(this.level, eggPosition);
-            eggState.setData(Blocks.DRAGON_EGG.defaultBlockState());
+            eggState.setData(Blocks.DRAGON_HEAD.defaultBlockState()); // EMC
             io.papermc.paper.event.block.DragonEggFormEvent eggEvent = new io.papermc.paper.event.block.DragonEggFormEvent(org.bukkit.craftbukkit.block.CraftBlock.at(this.level, eggPosition), eggState,
                 new org.bukkit.craftbukkit.boss.CraftDragonBattle(this));
             // Paper end - DragonEggFormEvent
