From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jan 2016 03:42:03 -0500
Subject: [PATCH] Player Death fires PlayerDropItem event


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index e0456a2606efa2e4099f8706c3b45ea2b16a0145..f88a8877975b17a742b30dcd8a6a5ed57e72699f 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -953,7 +953,7 @@ public class CraftEventFactory {
         for (org.bukkit.inventory.ItemStack stack : event.getDrops()) {
             if (stack == null || stack.getType() == Material.AIR) continue;
 
-            world.dropItem(entity.getLocation(), stack);
+            victim.drop(CraftItemStack.asNMSCopy(stack), true, false); // EMC
         }
 
         return event;
