From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 8 Jun 2016 21:04:13 -0400
Subject: [PATCH] Allow damage to shields for all damage


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 51279781d26cb770bec55706ce40f4615f40723b..f9243dc31fe7a0ff615f76fc4ce64319df67ba54 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1031,7 +1031,7 @@ public abstract class Player extends LivingEntity {
                 this.awardStat(Stats.ITEM_USED.get(this.useItem.getItem()));
             }
 
-            if (amount >= 3.0F) {
+            if (amount >= 0.0F) { // EMC - shields always can take damage
                 int i = 1 + Mth.floor(amount);
                 InteractionHand enumhand = this.getUsedItemHand();
 
