From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 29 Dec 2016 23:27:50 -0500
Subject: [PATCH] Make CraftMetaItem not crash on HideFlags error


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/SerializableMeta.java b/src/main/java/org/bukkit/craftbukkit/inventory/SerializableMeta.java
index dcf02bd0f7f4c67f5ab98003cc932b960704eef1..c03dd40deac1a89b409039ece5c97396fe6ab841 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/SerializableMeta.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/SerializableMeta.java
@@ -134,7 +134,13 @@ public final class SerializableMeta implements ConfigurationSerializable {
             }
             return null;
         }
-        throw new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz);
+        // EMC start
+        if (!nullable) {
+            throw new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz);
+        }
+        new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz).printStackTrace();
+        return null;
+        // EMC end
     }
 
     // Paper start - General ItemMeta Fixes
