From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 11 Feb 2016 23:13:57 -0500
Subject: [PATCH] Shutdown server if Empire does not load


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index cb9f4239734b73cb22912fad6b3d8c29c69c1cea..6b5cf5a97f7eb814cc1a1f93e720783e363b9308 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -735,6 +735,13 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Paper end
 
         this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
+        // EMC start
+        org.bukkit.plugin.Plugin empire = server.getPluginManager().getPlugin("Empire"); // EMC
+        if (empire == null || !empire.isEnabled()) {
+            System.err.println("CRITICAL - Empire did not load. ABORT ABORT ABORT!");
+            org.bukkit.Bukkit.getServer().shutdown();
+        }
+        // EMC end
         this.server.getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.STARTUP));
         this.connection.acceptConnections();
     }
