From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 11 Feb 2016 23:13:57 -0500
Subject: [PATCH] Shutdown server if Empire does not load


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 8d4c8de2849138701b4ba6f0957b8ae32bae652e..1faa177742a133c1bf59f2dfe39b7d3e9744350a 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -637,6 +637,13 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Paper end
 
         this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
+        // EMC start
+        org.bukkit.plugin.Plugin empire = server.getPluginManager().getPlugin("Empire"); // EMC
+        if (empire == null || !empire.isEnabled()) {
+            System.err.println("CRITICAL - Empire did not load. ABORT ABORT ABORT!");
+            org.bukkit.Bukkit.getServer().shutdown();
+        }
+        // EMC end
         this.server.getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.STARTUP));
         this.connection.acceptConnections();
     }
