From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 8 Jun 2016 21:04:13 -0400
Subject: [PATCH] Allow damage to shields for all damage


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 46c3acaa8c35a67a7aaaa91e0f67efd475297921..97f76924d16a45fe75d623f6c00ac48366a6a6b5 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1005,7 +1005,7 @@ public abstract class Player extends LivingEntity {
                 this.awardStat(Stats.ITEM_USED.get(this.useItem.getItem()));
             }
 
-            if (amount >= 3.0F) {
+            if (amount >= 0.0F) { // EMC - shields always can take damage
                 int i = 1 + Mth.floor(amount);
                 InteractionHand enumhand = this.getUsedItemHand();
 
