From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 23 Apr 2016 20:03:27 -0400
Subject: [PATCH] Give a Dragon Head instead of egg Every Dragon Kill


diff --git a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
index e2cd77d9de9709aa50f4b6febabe21bfcf94dc5d..0cf165feaa3b0cb478075f84e0c262b975159a99 100644
--- a/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/src/main/java/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -360,7 +360,7 @@ public class EndDragonFight {
             // Paper start - DragonEggFormEvent
             BlockPos eggPosition = this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.getLocation(this.origin));
             org.bukkit.craftbukkit.block.CraftBlockState eggState = org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(this.level, eggPosition);
-            eggState.setData(Blocks.DRAGON_EGG.defaultBlockState());
+            eggState.setData(Blocks.DRAGON_HEAD.defaultBlockState()); // EMC
             io.papermc.paper.event.block.DragonEggFormEvent eggEvent = new io.papermc.paper.event.block.DragonEggFormEvent(org.bukkit.craftbukkit.block.CraftBlock.at(this.level, eggPosition), eggState,
                 new org.bukkit.craftbukkit.boss.CraftDragonBattle(this));
             // Paper end - DragonEggFormEvent
