From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 30 Jun 2016 22:31:06 -0400
Subject: [PATCH] Don't use Players scoreboard team for pets


diff --git a/src/main/java/net/minecraft/world/entity/TamableAnimal.java b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
index 1282911b97292cde30dff83de756bc91bf319cb2..ad51a8eaf6c178662d9a980194ae90a0589f89cf 100644
--- a/src/main/java/net/minecraft/world/entity/TamableAnimal.java
+++ b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
@@ -172,6 +172,8 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         return true;
     }
 
+    // EMC start - don't link to owners team
+    /*
     @Override
     public Team getTeam() {
         if (this.isTame()) {
@@ -182,7 +184,8 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         }
 
         return super.getTeam();
-    }
+    }*/
+    // EMC end
 
     @Override
     public boolean isAlliedTo(Entity other) {
