From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 30 Jun 2016 22:31:06 -0400
Subject: [PATCH] Don't use Players scoreboard team for pets


diff --git a/src/main/java/net/minecraft/world/entity/TamableAnimal.java b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
index acc25fb309568864dd7b53ad6a7a3ee6ff18e82a..176218c98238b5c1a0baa5148af3a389a81952f4 100644
--- a/src/main/java/net/minecraft/world/entity/TamableAnimal.java
+++ b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
@@ -177,6 +177,8 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         return true;
     }
 
+    // EMC start - don't link to owners team
+    /*
     @Override
     public Team getTeam() {
         if (this.isTame()) {
@@ -187,7 +189,8 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         }
 
         return super.getTeam();
-    }
+    }*/
+    // EMC end
 
     @Override
     public boolean isAlliedTo(Entity other) {
