From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 30 Jun 2016 22:31:06 -0400
Subject: [PATCH] Don't use Players scoreboard team for pets


diff --git a/src/main/java/net/minecraft/world/entity/TamableAnimal.java b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
index cf5cd0e49c4966e7b636a8ad4dec7a12806c6a4d..fa119a9c59cede4fd728656f89a156a99d25fdaa 100644
--- a/src/main/java/net/minecraft/world/entity/TamableAnimal.java
+++ b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
@@ -170,6 +170,8 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         return true;
     }
 
+    // EMC start - don't link to owners team
+    /*
     @Override
     public PlayerTeam getTeam() {
         if (this.isTame()) {
@@ -180,7 +182,8 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         }
 
         return super.getTeam();
-    }
+    }*/
+    // EMC end
 
     @Override
     public boolean isAlliedTo(Entity other) {
