From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 11 Feb 2016 23:13:57 -0500
Subject: [PATCH] Shutdown server if Empire does not load


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 1daedd9c4cddbb0511b5b5315a3ed8820dde8704..5eba1a6ef2dc80375239d719fd89f28ed320566f 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -642,6 +642,13 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Paper end
 
         this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
+        // EMC start
+        org.bukkit.plugin.Plugin empire = server.getPluginManager().getPlugin("Empire"); // EMC
+        if (empire == null || !empire.isEnabled()) {
+            System.err.println("CRITICAL - Empire did not load. ABORT ABORT ABORT!");
+            org.bukkit.Bukkit.getServer().shutdown();
+        }
+        // EMC end
         this.server.getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.STARTUP));
         this.connection.acceptConnections();
         // CraftBukkit end
