From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 11 Feb 2016 23:13:57 -0500
Subject: [PATCH] Shutdown server if Empire does not load


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 4b216dae62c395140f8821a2b1b4352c27853597..ebb36394899c8b73410069ddd47edf2784d61988 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -638,6 +638,13 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Paper end
 
         this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
+        // EMC start
+        org.bukkit.plugin.Plugin empire = server.getPluginManager().getPlugin("Empire"); // EMC
+        if (empire == null || !empire.isEnabled()) {
+            System.err.println("CRITICAL - Empire did not load. ABORT ABORT ABORT!");
+            org.bukkit.Bukkit.getServer().shutdown();
+        }
+        // EMC end
         this.server.getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.STARTUP));
         this.connection.acceptConnections();
         // CraftBukkit end
