From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 23 Jun 2016 23:24:20 -0400
Subject: [PATCH] Do not return item to player on cancel drop item event

we will control this in plugin code

diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index cb92e69e3bf462873aee75f80e607c56d80bb40a..2b33d9c401c67070067f99aa435061af64cc0eda 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -780,6 +780,7 @@ public abstract class Player extends LivingEntity {
             this.level().getCraftServer().getPluginManager().callEvent(event);
 
             if (event.isCancelled()) {
+                if (true) { return null; } // EMC
                 org.bukkit.inventory.ItemStack cur = player.getInventory().getItemInHand();
                 if (flag1 && (cur == null || cur.getAmount() == 0)) {
                     // The complete stack was dropped
