From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 29 Dec 2016 23:27:50 -0500
Subject: [PATCH] Make CraftMetaItem not crash on HideFlags error


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
index 4d6aa15ca797b88bf21383ca33bdd0b65d3f1912..6185a8c14aa459f51197cf375490dea6a5df01d2 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -235,7 +235,12 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
                 }
                 return null;
             }
-            throw new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz);
+            if (nullable) {
+                new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz).printStackTrace();
+            } else {
+                throw new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz);
+            }
+            return null;
         }
     }
 
