From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 20 Jun 2018 22:35:39 -0400
Subject: [PATCH] Tweak Explosions


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index ad57bf49476192dea6a7367cbd0ad3f11e142e1b..ebe169acb5baa68fcffd95a67e5df27ff5c3950b 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -437,6 +437,7 @@ public class Explosion {
         int i;
         int j;
 
+        if (this.fire || this.blockInteraction != Explosion.BlockInteraction.KEEP) { // EMC - don't run block processing if keep flag is set
         // Paper start - collision optimisations
         for (int ray = 0, len = CACHED_RAYS.length; ray < len;) {
             ca.spottedleaf.moonrise.patches.collisions.ExplosionBlockCache cachedBlock = initialCache;
@@ -507,6 +508,7 @@ public class Explosion {
                         } while (power > 0.0f);
                         // Paper end - collision optimisations
                     }
+        } // EMC
 
         // Paper - optimise collisions
         float f2 = this.radius * 2.0F;
@@ -517,7 +519,7 @@ public class Explosion {
         int i1 = Mth.floor(this.y + (double) f2 + 1.0D);
         int j1 = Mth.floor(this.z - (double) f2 - 1.0D);
         int k1 = Mth.floor(this.z + (double) f2 + 1.0D);
-        List<Entity> list = this.level.getEntities(this.source, new AABB((double) i, (double) l, (double) j1, (double) j, (double) i1, (double) k1), (com.google.common.base.Predicate<Entity>) entity -> entity.isAlive() && !entity.isSpectator()); // Paper - Fix lag from explosions processing dead entities
+        List<Entity> list = this.level.getEntities(this.source, new AABB((double) i, (double) l, (double) j1, (double) j, (double) i1, (double) k1), (com.google.common.base.Predicate<Entity>) entity -> entity.isAlive() && !entity.isSpectator() && !(entity instanceof ItemEntity)); // Paper - Fix lag from explosions processing dead entitiess // EMC - exclude item entities
         Vec3 vec3d = new Vec3(this.x, this.y, this.z);
         Iterator iterator = list.iterator();
 
@@ -636,7 +638,7 @@ public class Explosion {
         if (particles) {
             ParticleOptions particleparam;
 
-            if (this.radius >= 2.0F && flag1) {
+            if (this.radius >= 2.0F /*&& flag1*/) { // EMC - don't care about block breaks flag for animation
                 particleparam = this.largeExplosionParticles;
             } else {
                 particleparam = this.smallExplosionParticles;
