From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 8 Jun 2016 21:04:13 -0400
Subject: [PATCH] Allow damage to shields for all damage


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 417f2b16c519c9330b2b08ed3cc3ed7a60d73dc2..cd52b8b59aeb807b2057cf92aca3e7c64ddb9b65 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -984,7 +984,7 @@ public abstract class Player extends LivingEntity {
                 this.awardStat(Stats.ITEM_USED.get(this.useItem.getItem()));
             }
 
-            if (amount >= 3.0F) {
+            if (amount >= 0.0F) { // EMC - shields always can take damage
                 int i = 1 + Mth.floor(amount);
                 InteractionHand enumhand = this.getUsedItemHand();
 
