From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 23 Jun 2016 23:24:20 -0400
Subject: [PATCH] Do not return item to player on cancel drop item event

we will control this in plugin code

diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 2a6a39d4e2d80ac8650a97f9c388b52d131598c0..a3dc5794e30071af5ac7241ab8f9d8759df15416 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -751,6 +751,7 @@ public abstract class Player extends LivingEntity {
             this.level.getCraftServer().getPluginManager().callEvent(event);
 
             if (event.isCancelled() && !alwaysSucceed) { // Paper - Fix duplicating /give items on item drop cancel
+                if (true) { return null; } // EMC
                 org.bukkit.inventory.ItemStack cur = player.getInventory().getItemInHand();
                 if (retainOwnership && (cur == null || cur.getAmount() == 0)) {
                     // The complete stack was dropped
