From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 25 Jun 2018 23:16:51 -0400
Subject: [PATCH] EntityPathfindEndEvent


diff --git a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 2e9991e6b3c05584002744a2ee2579b1dba218b2..b020880db970aabfef64e8cc50fe8def2594740b 100644
--- a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -381,6 +381,7 @@ public abstract class PathNavigation {
 
     public void stop() {
         this.path = null;
+        new com.destroystokyo.paper.event.entity.EntityPathfindEndEvent((org.bukkit.entity.Mob)this.mob.getBukkitEntity()).callEvent(); // Paper
     }
 
     protected abstract Vec3 getTempMobPos();
