From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 30 Jun 2016 22:31:06 -0400
Subject: [PATCH] Don't use Players scoreboard team for pets


diff --git a/src/main/java/net/minecraft/world/entity/TamableAnimal.java b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
index 0fc4a3e196ecdf16250d8b62136b2a4fe1ab6f84..d422f1165b8079db6b5c354b06d5e12a2dba0f5e 100644
--- a/src/main/java/net/minecraft/world/entity/TamableAnimal.java
+++ b/src/main/java/net/minecraft/world/entity/TamableAnimal.java
@@ -172,6 +172,8 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         return true;
     }
 
+    // EMC start - don't link to owners team
+    /*
     @Override
     public Team getTeam() {
         if (this.isTame()) {
@@ -182,7 +184,8 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
         }
 
         return super.getTeam();
-    }
+    }*/
+    // EMC end
 
     @Override
     public boolean isAlliedTo(Entity other) {
