From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 29 Dec 2016 23:27:50 -0500
Subject: [PATCH] Make CraftMetaItem not crash on HideFlags error


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
index 94cefa1ebd37664f6db40e199a052d40ffcef03f..d730b149bb217690b7d6dea5ce385754a3a8019e 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -234,7 +234,12 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
                 }
                 return null;
             }
-            throw new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz);
+            if (nullable) {
+                new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz).printStackTrace();
+            } else {
+                throw new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz);
+            }
+            return null;
         }
     }
 
