From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 7 Nov 2017 00:01:04 -0500
Subject: [PATCH] Don't use snapshots for TileEntity::getOwner


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
index 79a84b4f8de6c255769dc2c274f3e64af5920102..78c9abc5f6d1774c6adcffa00272ade5943e6d60 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -250,7 +250,7 @@ public abstract class BlockEntity {
     // CraftBukkit start - add method
     public InventoryHolder getOwner() {
         // Paper start
-        return getOwner(true);
+        return getOwner(false); // EMC - don't use snapshots
     }
     public InventoryHolder getOwner(boolean useSnapshot) {
         // Paper end
