From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Tue, 13 Dec 2016 00:40:10 -0500
Subject: [PATCH] Villager Treasure Map Events / Changes


diff --git a/src/main/java/com/empireminecraft/api/CraftEAPI_Misc.java b/src/main/java/com/empireminecraft/api/CraftEAPI_Misc.java
index fa8cbd5a38a3969ce9c8e0fc2cd36566ea252a2c..071082fb2012920859e1c4d04b792b830954088e 100644
--- a/src/main/java/com/empireminecraft/api/CraftEAPI_Misc.java
+++ b/src/main/java/com/empireminecraft/api/CraftEAPI_Misc.java
@@ -23,5 +23,34 @@
 
 package com.empireminecraft.api;
 
+import net.minecraft.core.BlockPos;
+import net.minecraft.network.chat.TranslatableComponent;
+import net.minecraft.server.MCUtil;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.MapItem;
+import net.minecraft.world.level.saveddata.maps.MapDecoration;
+import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.jetbrains.annotations.NotNull;
+
 public class CraftEAPI_Misc implements EAPI_Misc {
+
+    @Override
+    @NotNull
+    public org.bukkit.inventory.ItemStack createTreasureMap(@NotNull String mapType, int iconType, @NotNull Location loc) {
+        MapDecoration.Type[] values = MapDecoration.Type.values();
+        if (iconType >= values.length) {
+            throw new IllegalArgumentException("Invalid Icon Type");
+        }
+        ServerLevel level = MCUtil.getNMSWorld(loc.getWorld());
+        BlockPos pos = MCUtil.toBlockPosition(loc);
+        ItemStack itemstack = MapItem.create(level, pos.getX(), pos.getZ(), (byte) 2, true, true);
+        MapItem.renderBiomePreviewMap(level, itemstack);
+
+        MapItemSavedData.addTargetDecoration(itemstack, pos, "+", values[iconType]);
+        itemstack.setHoverName(new TranslatableComponent("filled_map." + mapType.toLowerCase()));
+        return CraftItemStack.asCraftMirror(itemstack);
+    }
 }
diff --git a/src/main/java/net/minecraft/world/entity/npc/VillagerTrades.java b/src/main/java/net/minecraft/world/entity/npc/VillagerTrades.java
index 90d77153d818534b1cb53ce80ea5409a709a2384..395a1d5c210aee5e0b83f89a583cd0fa56186811 100644
--- a/src/main/java/net/minecraft/world/entity/npc/VillagerTrades.java
+++ b/src/main/java/net/minecraft/world/entity/npc/VillagerTrades.java
@@ -369,10 +369,10 @@ public class VillagerTrades {
         private final int villagerXp;
 
         public TreasureMapForEmeralds(int price, StructureFeature<?> feature, MapDecoration.Type iconType, int maxUses, int experience) {
-            this.emeraldCost = price;
+            this.emeraldCost = Math.min(64 , 35 + price); // EMC - boost cost due to modified behavior - each map is unique, when its not in vanilla
             this.destination = feature;
             this.destinationType = iconType;
-            this.maxUses = maxUses;
+            this.maxUses = 2; // EMC - max 2 purchase before lock
             this.villagerXp = experience;
         }
 
@@ -384,12 +384,13 @@ public class VillagerTrades {
             } else {
                 ServerLevel serverLevel = (ServerLevel)entity.level;
                 if (!serverLevel.paperConfig.enableTreasureMaps) return null; // Paper
-                BlockPos blockPos = serverLevel.findNearestMapFeature(this.destination, entity.blockPosition(), 100, !serverLevel.paperConfig.treasureMapsAlreadyDiscovered); // Paper
-                if (blockPos != null) {
-                    ItemStack itemStack = MapItem.create(serverLevel, blockPos.getX(), blockPos.getZ(), (byte)2, true, true);
-                    MapItem.renderBiomePreviewMap(serverLevel, itemStack);
-                    MapItemSavedData.addTargetDecoration(itemStack, blockPos, "+", this.destinationType);
-                    itemStack.setHoverName(new TranslatableComponent("filled_map." + this.destination.getFeatureName().toLowerCase(Locale.ROOT)));
+                // EMC start - refactor treasure maps to be pending - ABSOLUTELY DEPENDS ON PLUGIN CODE!!!!!!!!!!!
+                //BlockPosition blockposition = worldserver.a(this.b, entity.getChunkCoordinates(), 100, !worldserver.paperConfig.treasureMapsAlreadyDiscovered); // Paper
+                com.empireminecraft.customevents.CreatePendingTreasureMapEvent event = new com.empireminecraft.customevents.CreatePendingTreasureMapEvent(this.destination.getFeatureName(), destinationType.ordinal());
+                event.callEvent();
+                if (event.getItem() != null) {
+                    ItemStack itemStack = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    // EMC end
                     return new MerchantOffer(new ItemStack(Items.EMERALD, this.emeraldCost), new ItemStack(Items.COMPASS), itemStack, this.maxUses, this.villagerXp, 0.2F);
                 } else {
                     return null;
