From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jun 2018 02:09:37 -0400
Subject: [PATCH] Human#getAttackPct


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 07dc8ebf5288c8cc863bde643d160f80563869ee..64b884515e145ab1ae500cd5a788113d08cdf1db 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -193,6 +193,7 @@ public abstract class Player extends LivingEntity {
     // CraftBukkit start
     public boolean fauxSleeping;
     public int oldLevel = -1;
+    public float lastAttackPct = 0; // EMC
 
     @Override
     public CraftHumanEntity getBukkitEntity() {
@@ -1267,6 +1268,7 @@ public abstract class Player extends LivingEntity {
                 }
 
                 float f2 = this.getAttackStrengthScale(0.5F);
+                this.lastAttackPct = f2; // EMC
 
                 f *= 0.2F + f2 * f2 * 0.8F;
                 f1 *= f2;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index e779dcc4982ff51e4d450265fd61bc26e8e74d3a..c2a7f60f219e0025bd51237bcc4cf55cfc37480e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -80,6 +80,7 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         this.inventory = new CraftInventoryPlayer(entity.getInventory());
         this.enderChest = new CraftInventory(entity.getEnderChestInventory());
     }
+    @Override public float getAttackPct() { return getHandle().lastAttackPct; } // EMC
 
     @Override
     public PlayerInventory getInventory() {
