From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jun 2018 02:09:37 -0400
Subject: [PATCH] Human#getAttackPct


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 3a7d0851ca7cf63f3e82081ff0e4ff5b3d470d13..6a22ae7a8a36d04fe4c08a62edd2ca2e7343c4c1 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -192,6 +192,7 @@ public abstract class Player extends LivingEntity {
     // CraftBukkit start
     public boolean fauxSleeping;
     public int oldLevel = -1;
+    public float lastAttackPct = 0; // EMC
 
     @Override
     public CraftHumanEntity getBukkitEntity() {
@@ -1266,6 +1267,7 @@ public abstract class Player extends LivingEntity {
                 }
 
                 float f2 = this.getAttackStrengthScale(0.5F);
+                this.lastAttackPct = f2; // EMC
 
                 f *= 0.2F + f2 * f2 * 0.8F;
                 f1 *= f2;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 73517e2cc98e9cda6e577976232a6c9314226ece..feb2411b230d47889c7700a677c4a410119ca43a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -77,6 +77,7 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         this.inventory = new CraftInventoryPlayer(entity.getInventory());
         this.enderChest = new CraftInventory(entity.getEnderChestInventory());
     }
+    @Override public float getAttackPct() { return getHandle().lastAttackPct; } // EMC
 
     @Override
     public PlayerInventory getInventory() {
