From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jun 2018 02:09:37 -0400
Subject: [PATCH] Human#getAttackPct


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 2b33d9c401c67070067f99aa435061af64cc0eda..52014b6afc6209a9b86af1614af12c7889f8f92e 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -193,6 +193,7 @@ public abstract class Player extends LivingEntity {
     // CraftBukkit start
     public boolean fauxSleeping;
     public int oldLevel = -1;
+    public float lastAttackPct = 0; // EMC
 
     @Override
     public CraftHumanEntity getBukkitEntity() {
@@ -1270,6 +1271,7 @@ public abstract class Player extends LivingEntity {
                 }
 
                 float f2 = this.getAttackStrengthScale(0.5F);
+                this.lastAttackPct = f2; // EMC
 
                 f *= 0.2F + f2 * f2 * 0.8F;
                 f1 *= f2;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 017e97c1618b8ee4640b36a0ec1b07026047bfc3..ad75716d39043bfd594a8c8b3452e7685d51aebc 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -81,6 +81,7 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         this.inventory = new CraftInventoryPlayer(entity.getInventory());
         this.enderChest = new CraftInventory(entity.getEnderChestInventory());
     }
+    @Override public float getAttackPct() { return getHandle().lastAttackPct; } // EMC
 
     @Override
     public PlayerInventory getInventory() {
