From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 14 Jul 2019 13:50:53 -0500
Subject: [PATCH] ConduitNewTargetEvent and remove target for cancelled damage
 events


diff --git a/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
index 73e532dc998e5701c1a73da846da3d3a79871b81..c97d49dc1e72f5a526d259c22b1f8f58bb0b7517 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
@@ -27,6 +27,7 @@ import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.block.CraftBlock;
 
 public class ConduitBlockEntity extends BlockEntity {
 
@@ -241,6 +242,14 @@ public class ConduitBlockEntity extends BlockEntity {
                 return entityliving1 instanceof Enemy && entityliving1.isInWaterOrRain();
             });
 
+            // EMC start
+            List<org.bukkit.entity.LivingEntity> bukkitEntities = Lists.newArrayListWithCapacity(list1.size());
+            for (LivingEntity entity : list1) {
+                bukkitEntities.add(entity.getBukkitLivingEntity());
+            }
+            new com.empireminecraft.customevents.ConduitNewTargetEvent(bukkitEntities, CraftBlock.at(world, blockposition)).callEvent();
+            list1.removeIf(entityLiving -> !bukkitEntities.contains(entityLiving.getBukkitLivingEntity()));
+            // EMC end
             if (!list1.isEmpty()) {
                 tileentityconduit.destroyTarget = (LivingEntity) list1.get(world.random.nextInt(list1.size()));
             }
@@ -252,7 +261,7 @@ public class ConduitBlockEntity extends BlockEntity {
         if (damageTarget && tileentityconduit.destroyTarget != null) {
             if (tileentityconduit.destroyTarget.hurt(world.damageSources().magic().directBlock(world, blockposition), 4.0F)) {
                 world.playSound(null, tileentityconduit.destroyTarget.getX(), tileentityconduit.destroyTarget.getY(), tileentityconduit.destroyTarget.getZ(), SoundEvents.CONDUIT_ATTACK_TARGET, SoundSource.BLOCKS, 1.0F, 1.0F);
-            }
+            } else { tileentityconduit.destroyTarget = null; } // EMC
             // CraftBukkit end
         }
 
