From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 29 Nov 2019 02:48:17 -0600
Subject: [PATCH] Add the PlayerThrowTridentEvent


diff --git a/src/main/java/net/minecraft/world/item/TridentItem.java b/src/main/java/net/minecraft/world/item/TridentItem.java
index f1b2d388a1a40a1d909a2e726f32d6c15e1eb0eb..58df0631f039c1913b226e85ecafd384863ee905 100644
--- a/src/main/java/net/minecraft/world/item/TridentItem.java
+++ b/src/main/java/net/minecraft/world/item/TridentItem.java
@@ -97,6 +97,7 @@ public class TridentItem extends Item implements ProjectileItem {
                                     return;
                                 }
 
+                                (new com.destroystokyo.paper.event.player.PlayerThrowTridentEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack), entitythrowntrident.getBukkitEntity())).callEvent(); // Paper
                                 if (event.shouldConsume()) { // Paper - PlayerLaunchProjectileEvent
                                 stack.hurtAndBreak(1, entityhuman, LivingEntity.getSlotForHand(user.getUsedItemHand()));
                                 } // Paper - PlayerLaunchProjectileEvent
