From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 25 Jun 2018 22:30:08 -0400
Subject: [PATCH] EntityRandomStrollEvent


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
index 676e1580849e8942eb41c7a3a80f464138b22764..1bdd18999db13f8072f3396f0211bbbba4880c0d 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
@@ -2,6 +2,8 @@ package net.minecraft.world.entity.ai.goal;
 
 import java.util.EnumSet;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.util.MCUtil;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.entity.ai.util.DefaultRandomPos;
 import net.minecraft.world.phys.Vec3;
@@ -73,7 +75,15 @@ public class RandomStrollGoal extends Goal {
 
     @Override
     public void start() {
-        this.mob.getNavigation().moveTo(this.wantedX, this.wantedY, this.wantedZ, this.speedModifier);
+        // Paper start
+        org.bukkit.Location loc = MCUtil.toLocation(mob.level(), wantedX, wantedY, wantedZ);
+        com.destroystokyo.paper.event.entity.EntityRandomStrollEvent event = new com.destroystokyo.paper.event.entity.EntityRandomStrollEvent(mob.getBukkitCreature(), loc, speedModifier);
+        if (!event.callEvent()) {
+            return;
+        }
+        loc = event.getLocation();
+        this.mob.getNavigation().moveTo(loc.getX(), loc.getY(), loc.getZ(), event.getSpeed());
+        // Paper end
     }
 
     @Override
