From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 25 Jun 2018 22:55:42 -0400
Subject: [PATCH] EntityPrepForRangedAttack


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
index 1ad065ac927a5c092532491ba7439b93342b1137..4ad85d3a333459c7fe4bed9bf9dc44c504fff563 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/RangedAttackGoal.java
@@ -2,6 +2,8 @@ package net.minecraft.world.entity.ai.goal;
 
 import java.util.EnumSet;
 import javax.annotation.Nullable;
+
+import com.destroystokyo.paper.event.entity.EntityPrepForRangedAttack;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Mob;
@@ -77,9 +79,17 @@ public class RangedAttackGoal extends Goal {
             this.seeTime = 0;
         }
 
-        if (!(d > (double)this.attackRadiusSqr) && this.seeTime >= 5) {
+        // EMC start
+        boolean isInRange = d <= (double) this.attackRadiusSqr;
+        boolean readyforAttack = this.seeTime >= 5;
+        EntityPrepForRangedAttack event = new EntityPrepForRangedAttack((com.destroystokyo.paper.entity.RangedEntity) mob.getBukkitEntity(), this.target.getBukkitLivingEntity(), readyforAttack, !isInRange, this.seeTime);
+        if (!event.callEvent()) {
+            return;
+        }
+        if (event.shouldStop() && !event.shouldMoveNear()) {
             this.mob.getNavigation().stop();
-        } else {
+        } else if (event.shouldMoveNear()) {
+            // EMC end
             this.mob.getNavigation().moveTo(this.target, this.speedModifier);
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
index 9796c98c2b7380e980abb00d798ba154d166cac7..8d554b7a0ee3fbdfa698ffcdb890a857953fee52 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.entity.ai.goal;
 
+import com.destroystokyo.paper.event.entity.EntityPrepForRangedAttack;
 import java.util.EnumSet;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Mob;
@@ -83,10 +84,18 @@ public class RangedBowAttackGoal<T extends Monster & RangedAttackMob> extends Go
                 this.seeTime--;
             }
 
-            if (!(d > (double)this.attackRadiusSqr) && this.seeTime >= 20) {
+            // EMC start
+            boolean isInRange = d <= (double) this.attackRadiusSqr;
+            boolean readyforAttack = this.seeTime >= 20;
+            EntityPrepForRangedAttack event = new EntityPrepForRangedAttack((com.destroystokyo.paper.entity.RangedEntity) mob.getBukkitEntity(), livingEntity.getBukkitLivingEntity(), readyforAttack, !isInRange, this.seeTime);
+            if (!event.callEvent()) {
+                return;
+            }
+            if (event.shouldStop() && !event.shouldMoveNear()) {
                 this.mob.getNavigation().stop();
                 this.strafingTime++;
-            } else {
+            } else if (event.shouldMoveNear()){
+                // EMC end {
                 this.mob.getNavigation().moveTo(livingEntity, this.speedModifier);
                 this.strafingTime = -1;
             }
