From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 26 Jan 2020 23:28:04 -0600
Subject: [PATCH] Use lenient json to component to resolve old WE schematic
 issue


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
index cffeaff83ec5037c21ff33ac32287017c1263b67..574a3c7173a9ea24a542d5a489b2df3f9c920d1b 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -378,7 +378,7 @@ public abstract class BlockEntity {
     @Nullable
     public static Component parseCustomNameSafe(String json, HolderLookup.Provider registryLookup) {
         try {
-            return Component.Serializer.fromJson(json, registryLookup);
+            return Component.Serializer.fromJsonLenient(json, registryLookup); // EMC
         } catch (Exception exception) {
             BlockEntity.LOGGER.warn("Failed to parse custom name from string '{}', discarding", json, exception);
             return null;
