From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 25 Jun 2018 23:16:51 -0400
Subject: [PATCH] EntityPathfindEndEvent


diff --git a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index a04fb58c29b462e20ed0b702cef478e3985b7cd2..f2b4b37805277cd7b14458538e3e80ab09136c0f 100644
--- a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -379,6 +379,7 @@ public abstract class PathNavigation {
 
     public void stop() {
         this.path = null;
+        new com.destroystokyo.paper.event.entity.EntityPathfindEndEvent((org.bukkit.entity.Mob)this.mob.getBukkitEntity()).callEvent(); // Paper
     }
 
     protected abstract Vec3 getTempMobPos();
