From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 26 Jan 2020 23:28:04 -0600
Subject: [PATCH] Use lenient json to component to resolve old WE schematic
 issue


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c0041c47ccaaa55b8b4592385cf20168e70c9e08..fe5f937b8a5dc1f5348eb249fbb85fbaab19ddfb 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2616,7 +2616,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                         String s = nbt.getString("CustomName");
 
                         try {
-                            this.setCustomName(Component.Serializer.fromJson(s, this.registryAccess()));
+                            this.setCustomName(Component.Serializer.fromJsonLenient(s, this.registryAccess())); // EMC
                         } catch (Exception exception) {
                             Entity.LOGGER.warn("Failed to parse entity custom name {}", s, exception);
                         }
