From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jun 2018 02:09:37 -0400
Subject: [PATCH] Human#getAttackPct


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 9a7690ba561e17a9f97d38a603290b46c0fc2209..85c691969ad7e828aa3ca2acc8e3ae415fbf0ce4 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -191,6 +191,7 @@ public abstract class Player extends LivingEntity {
     // CraftBukkit start
     public boolean fauxSleeping;
     public int oldLevel = -1;
+    public float lastAttackPct = 0; // EMC
 
     @Override
     public CraftHumanEntity getBukkitEntity() {
@@ -1263,6 +1264,7 @@ public abstract class Player extends LivingEntity {
                 }
 
                 float f2 = this.getAttackStrengthScale(0.5F);
+                this.lastAttackPct = f2; // EMC
 
                 f *= 0.2F + f2 * f2 * 0.8F;
                 f1 *= f2;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 1b008e5217c5bbf566a213abb92e1c7c43a3a7c2..10a2859eb131b99e1dd3e2cffe7695c8be74e0cb 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -81,6 +81,7 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         this.inventory = new CraftInventoryPlayer(entity.getInventory());
         this.enderChest = new CraftInventory(entity.getEnderChestInventory());
     }
+    @Override public float getAttackPct() { return getHandle().lastAttackPct; } // EMC
 
     @Override
     public PlayerInventory getInventory() {
