From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 4 Jul 2018 15:49:34 -0400
Subject: [PATCH] Fire PlayerKickEvent during server shutdown


diff --git a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 231150bac0ae61e9722c2cdfd70d6f7d254681e4..eff1164ac6ffe0fd8f5fc4b28a7ccfe70520a8a7 100644
--- a/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -312,11 +312,11 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
 
         PlayerKickEvent event = new PlayerKickEvent(this.player.getBukkitEntity(), reason, leaveMessage, cause); // Paper - adventure
 
-        if (this.cserver.getServer().isRunning()) {
+        if (true) { // EMC - don't care if its async either, fire it!
             this.cserver.getPluginManager().callEvent(event);
         }
 
-        if (event.isCancelled()) {
+        if (!cserver.getServer().hasStopped() && event.isCancelled()) { // EMC
             // Do not kick the player
             return;
         }
