From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 6 May 2015 02:29:30 -0400
Subject: [PATCH] Limit worlds structures can generate


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index c08106d955ef73d101ed84e96d84353e05026cb6..c795dc526578b676893fca941cf8d1bfa6d423ea 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1932,6 +1932,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         if (!this.serverLevelData.worldGenOptions().generateStructures()) { // CraftBukkit
             return null;
         } else {
+            if ("town".equals(getWorld().getName())) { return null; } // EMC
             Optional<HolderSet.Named<Structure>> optional = this.registryAccess().registryOrThrow(Registries.STRUCTURE).getTag(structureTag);
 
             if (optional.isEmpty()) {
diff --git a/src/main/java/net/minecraft/world/level/StructureManager.java b/src/main/java/net/minecraft/world/level/StructureManager.java
index 54972cce2314eff774250101df43a9b7074e9604..ada7a8ddce83dde912b5ffd9237dbd20a17af084 100644
--- a/src/main/java/net/minecraft/world/level/StructureManager.java
+++ b/src/main/java/net/minecraft/world/level/StructureManager.java
@@ -99,7 +99,7 @@ public class StructureManager {
     }
 
     public boolean shouldGenerateStructures() {
-        return this.worldOptions.generateStructures();
+        return this.worldOptions.generateStructures() && !"town".equals(this.level.getMinecraftWorld().getWorld().getName()); // EMC
     }
 
     public StructureStart getStructureAt(BlockPos pos, Structure structure) {
