From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 4 Jul 2018 15:49:34 -0400
Subject: [PATCH] Fire PlayerKickEvent during server shutdown


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 14e34aeb85bb7fc48300e0617f5a4aad564e2e6e..1776751754a777944d1b48b964cfa60d69caf420 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -516,11 +516,11 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
 
         PlayerKickEvent event = new PlayerKickEvent(this.player.getBukkitEntity(), reason, leaveMessage, cause); // Paper - Adventure & kick event reason
 
-        if (this.cserver.getServer().isRunning()) {
+        if (true) { // EMC - don't care if its async either, fire it!
             this.cserver.getPluginManager().callEvent(event);
         }
 
-        if (event.isCancelled()) {
+        if (!cserver.getServer().hasStopped() && event.isCancelled()) { // EMC
             // Do not kick the player
             return;
         }
