From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 17 Jun 2018 22:16:25 -0400
Subject: [PATCH] Allow negative Amplifier values


diff --git a/src/main/java/net/minecraft/world/effect/MobEffectInstance.java b/src/main/java/net/minecraft/world/effect/MobEffectInstance.java
index 544b3711bc7538609f64ad30995ef927f562a36f..f7cdd9824c28673c6bda3add6d5283b5fe6cd653 100644
--- a/src/main/java/net/minecraft/world/effect/MobEffectInstance.java
+++ b/src/main/java/net/minecraft/world/effect/MobEffectInstance.java
@@ -253,7 +253,7 @@ public class MobEffectInstance implements Comparable<MobEffectInstance> {
             mobEffectInstance = loadSpecifiedEffect(type, nbt.getCompound("HiddenEffect"));
         }
 
-        return new MobEffectInstance(type, j, i < 0 ? 0 : i, bl, bl2, bl3, mobEffectInstance);
+        return new MobEffectInstance(type, j, i, bl, bl2, bl3, mobEffectInstance); // EMC - always use Amplifier, dont check >0
     }
 
     public void setNoCounter(boolean permanent) {
