From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 25 Jun 2018 23:16:51 -0400
Subject: [PATCH] EntityPathfindEndEvent


diff --git a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 10505f32b71b723ed8dbfd9e1348a1691c9f7f18..4fea2b4794e234abe3f9e7dcab36176d4e6dce55 100644
--- a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -365,6 +365,7 @@ public abstract class PathNavigation {
 
     public void stop() {
         this.path = null;
+        new com.destroystokyo.paper.event.entity.EntityPathfindEndEvent((org.bukkit.entity.Mob)this.mob.getBukkitEntity()).callEvent(); // Paper
     }
 
     protected abstract Vec3 getTempMobPos();
