From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 29 Nov 2019 02:48:17 -0600
Subject: [PATCH] Add the PlayerThrowTridentEvent


diff --git a/src/main/java/net/minecraft/world/item/TridentItem.java b/src/main/java/net/minecraft/world/item/TridentItem.java
index fa876ddf54780728e7f3ecfe02aa8a16b8ef6f8d..a06745650b9b3b45dd68749546316c547ab0e802 100644
--- a/src/main/java/net/minecraft/world/item/TridentItem.java
+++ b/src/main/java/net/minecraft/world/item/TridentItem.java
@@ -92,7 +92,7 @@ public class TridentItem extends Item implements Vanishable {
                                 }
                                 return;
                             }
-
+                            (new com.destroystokyo.paper.event.player.PlayerThrowTridentEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack), entitythrowntrident.getBukkitEntity())).callEvent(); // Paper
                             if (event.shouldConsume()) { // Paper - PlayerLaunchProjectileEvent
                             stack.hurtAndBreak(1, entityhuman, (entityhuman1) -> {
                                 entityhuman1.broadcastBreakEvent(user.getUsedItemHand());
