From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 30 Mar 2019 00:45:12 -0400
Subject: [PATCH] Convert Item Stacks


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 8ee2d91346b7043ae348d6fcfe08f7e320ac4620..2c000376698b7738ef3a430114338ee4f61256c7 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -242,6 +242,7 @@ public final class ItemStack {
         this.item = null;
     }
 
+    private static final boolean DEBUG_CONVERSIONS = Boolean.getBoolean("debug.conversions"); // Paper
     // Called to run this stack through the data converter to handle older storage methods and serialized items
     public void convertStack(int version) {
         if (0 < version && version < CraftMagicNumbers.INSTANCE.getDataVersion()) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
index a366e5e985eb148d516a70b9bad575d5e1de50bd..7893e8836813689e0ec9b7c89b9cea3a6233ead8 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
@@ -11,9 +11,11 @@ import org.bukkit.Color;
 import org.bukkit.Material;
 import org.bukkit.configuration.serialization.ConfigurationSerialization;
 import org.bukkit.craftbukkit.util.CraftLegacy;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
 import org.bukkit.inventory.ItemFactory;
 import org.bukkit.inventory.ItemStack;
 import org.bukkit.inventory.meta.ItemMeta;
+import org.jetbrains.annotations.NotNull;
 
 public final class CraftItemFactory implements ItemFactory {
     static final Color DEFAULT_LEATHER_COLOR = Color.fromRGB(0xA06540);
@@ -467,9 +469,25 @@ public final class CraftItemFactory implements ItemFactory {
     }
 
     // Paper start
+    @NotNull
+    public ItemStack _asServerItemStack(@NotNull ItemStack item) {
+        if (item instanceof CraftItemStack) {
+            return item;
+        } else {
+            CraftItemStack newStack = CraftItemStack.asCraftMirror(CraftItemStack.asNMSCopy(item));
+            newStack.setDataVersion(item.getDataVersion());
+            return newStack;
+        }
+    }
     @Override
     public ItemStack ensureServerConversions(ItemStack item) {
-        return CraftItemStack.asCraftMirror(CraftItemStack.asNMSCopy(item));
+        CraftItemStack newItem = (CraftItemStack) _asServerItemStack(item);
+        net.minecraft.world.item.ItemStack handle = newItem.handle;
+        if (handle != null && handle.getTag() != null) {
+            handle.convertStack(item.getDataVersion());
+        }
+        newItem.setDataVersion(CraftMagicNumbers.INSTANCE.getDataVersion());
+        return newItem;
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 8657736cdfe2d5d2a1c0851ca54dd26ce17e0a83..4f20ee630be86bdd68c09f680f794a12af7182d2 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -40,7 +40,12 @@ public final class CraftItemStack extends ItemStack {
 
         net.minecraft.world.item.ItemStack stack = new net.minecraft.world.item.ItemStack(item, original.getAmount());
         if (original.hasItemMeta()) {
-            CraftItemStack.setItemMeta(stack, original.getItemMeta());
+            // Paper start
+            ItemMeta itemMeta = original.getItemMeta();
+            itemMeta.setVersion(original.getDataVersion());
+            CraftItemStack.setItemMeta(stack, itemMeta);
+            stack.convertStack(original.getDataVersion());
+            // Paper end
         }
         return stack;
     }
@@ -277,7 +282,11 @@ public final class CraftItemStack extends ItemStack {
 
     @Override
     public ItemMeta getItemMeta() {
-        return CraftItemStack.getItemMeta(this.handle);
+        // Paper start
+        ItemMeta itemMeta = CraftItemStack.getItemMeta(handle);
+        ((CraftMetaItem) itemMeta).setVersion(this.getDataVersion());
+        return itemMeta;
+        // Paper end
     }
     // Paper start
     public static void applyMetaToItem(net.minecraft.world.item.ItemStack itemStack, ItemMeta meta) {
@@ -628,7 +637,7 @@ public final class CraftItemStack extends ItemStack {
         item.setTag(tag);
 
         ((CraftMetaItem) itemMeta).applyToItem(tag);
-        item.convertStack(((CraftMetaItem) itemMeta).getVersion());
+        //item.convertStack(((CraftMetaItem) itemMeta).getVersion()); // Paper
         // SpigotCraft#463 this is required now by the Vanilla client, so mimic ItemStack constructor in ensuring it
         if (item.getItem() != null && item.getItem().canBeDepleted()) {
             item.setDamageValue(item.getDamageValue());
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
index 7f5b26b9cbfd211832a231ace3c12a90f6ecdc8e..4ce999a19f0ec938c196324fd0cb287e37c11bdd 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -318,6 +318,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             return;
         }
 
+        this.version = meta.version; // Paper - set before all the code below that might use it...
         this.displayName = meta.displayName;
         this.locName = meta.locName;
 
@@ -357,19 +358,22 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             this.deserializeInternal(this.internalTag, meta);
         }
 
-        this.version = meta.version;
+        //this.version = meta.version; // Paper - move up
     }
 
     CraftMetaItem(CompoundTag tag) {
+        boolean isLegacy = version <= 1343; // Paper
         if (tag.contains(DISPLAY.NBT)) {
             CompoundTag display = tag.getCompound(DISPLAY.NBT);
 
             if (display.contains(NAME.NBT)) {
-                this.displayName = limit( display.getString(NAME.NBT), io.papermc.paper.configuration.GlobalConfiguration.get().itemValidation.displayName ); // Spigot // Paper - make configurable
+                if (isLegacy) this.displayName = CraftChatMessage.fromStringOrNullToJSON(limit( display.getString(NAME.NBT), io.papermc.paper.configuration.GlobalConfiguration.get().itemValidation.displayName )); // Paper - add isLegacy
+                else this.displayName = limit( display.getString(NAME.NBT), io.papermc.paper.configuration.GlobalConfiguration.get().itemValidation.displayName ); // Spigot // Paper - make configurable // Paper - add isLegacy
             }
 
             if (display.contains(LOCNAME.NBT)) {
-                this.locName = limit( display.getString(LOCNAME.NBT), 8192 ); // Spigot
+                if (isLegacy) { this.locName = CraftChatMessage.fromStringOrNullToJSON(limit( display.getString(LOCNAME.NBT), 8192 )); } // Paper - add isLegacy
+                else this.locName = limit( display.getString(LOCNAME.NBT), 8192 ); // Spigot // Paper - make configurable // Paper - add isLegacy
             }
 
             if (display.contains(LORE.NBT)) {
