From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 30 Mar 2019 00:45:12 -0400
Subject: [PATCH] Convert Item Stacks


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index baee8a6de2438aa3ffa0111928601c4a83b547c8..e0b8c3afb230776d532adc9c7b64aa65fe946a7f 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -242,6 +242,7 @@ public final class ItemStack {
         this.item = null;
     }
 
+    private static final boolean DEBUG_CONVERSIONS = Boolean.getBoolean("debug.conversions"); // Paper
     // Called to run this stack through the data converter to handle older storage methods and serialized items
     public void convertStack(int version) {
         if (0 < version && version < CraftMagicNumbers.INSTANCE.getDataVersion()) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
index 8aa561a0ebd386553e00b0d3499103a0dfbd7697..936032e58abc64430e9643e9991bfed5d05118bd 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
@@ -18,6 +18,7 @@ import org.bukkit.entity.EntityType;
 import org.bukkit.inventory.ItemFactory;
 import org.bukkit.inventory.ItemStack;
 import org.bukkit.inventory.meta.ItemMeta;
+import org.jetbrains.annotations.NotNull;
 
 public final class CraftItemFactory implements ItemFactory {
     static final Color DEFAULT_LEATHER_COLOR = Color.fromRGB(0xA06540);
@@ -488,9 +489,25 @@ public final class CraftItemFactory implements ItemFactory {
     // Paper end - Adventure
 
     // Paper start - ensure server conversions API
+    @NotNull
+    public ItemStack _asServerItemStack(@NotNull ItemStack item) {
+        if (item instanceof CraftItemStack) {
+            return item;
+        } else {
+            CraftItemStack newStack = CraftItemStack.asCraftMirror(CraftItemStack.asNMSCopy(item));
+            newStack.setDataVersion(item.getDataVersion());
+            return newStack;
+        }
+    }
     @Override
     public ItemStack ensureServerConversions(ItemStack item) {
-        return CraftItemStack.asCraftMirror(CraftItemStack.asNMSCopy(item));
+        CraftItemStack newItem = (CraftItemStack) _asServerItemStack(item);
+        net.minecraft.world.item.ItemStack handle = newItem.handle;
+        if (handle != null && handle.getTag() != null) {
+            handle.convertStack(item.getDataVersion());
+        }
+        newItem.setDataVersion(CraftMagicNumbers.INSTANCE.getDataVersion());
+        return newItem;
     }
     // Paper end - ensure server conversions API
 
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index f62ef038af2dfe7525e72c8b8e30a3fa9ce2b451..b2650ac19504840200e37aa62132f09a47ca7c49 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -40,7 +40,12 @@ public final class CraftItemStack extends ItemStack {
 
         net.minecraft.world.item.ItemStack stack = new net.minecraft.world.item.ItemStack(item, original.getAmount());
         if (original.hasItemMeta()) {
-            CraftItemStack.setItemMeta(stack, original.getItemMeta());
+            // Paper start
+            ItemMeta itemMeta = original.getItemMeta();
+            itemMeta.setVersion(original.getDataVersion());
+            CraftItemStack.setItemMeta(stack, itemMeta);
+            stack.convertStack(original.getDataVersion());
+            // Paper end
         }
         return stack;
     }
@@ -277,7 +282,11 @@ public final class CraftItemStack extends ItemStack {
 
     @Override
     public ItemMeta getItemMeta() {
-        return CraftItemStack.getItemMeta(this.handle);
+        // Paper start
+        ItemMeta itemMeta = CraftItemStack.getItemMeta(handle);
+        ((CraftMetaItem) itemMeta).setVersion(this.getDataVersion());
+        return itemMeta;
+        // Paper end
     }
     // Paper start
     public static void applyMetaToItem(net.minecraft.world.item.ItemStack itemStack, ItemMeta meta) {
@@ -629,7 +638,7 @@ public final class CraftItemStack extends ItemStack {
         item.setTag(tag);
 
         ((CraftMetaItem) itemMeta).applyToItem(tag);
-        item.convertStack(((CraftMetaItem) itemMeta).getVersion());
+        //item.convertStack(((CraftMetaItem) itemMeta).getVersion()); // Paper
         // SpigotCraft#463 this is required now by the Vanilla client, so mimic ItemStack constructor in ensuring it
         if (item.getItem() != null && item.getItem().canBeDepleted()) {
             item.setDamageValue(item.getDamageValue());
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
index 0436d45f18ea6097986b4379c56719e2b49026de..ebc8caed054661f656bf47264e3562ab30a7fe69 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
@@ -318,6 +318,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             return;
         }
 
+        this.version = meta.version; // Paper - set before all the code below that might use it...
         this.displayName = meta.displayName;
         this.locName = meta.locName;
 
@@ -357,19 +358,22 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
             this.deserializeInternal(this.internalTag, meta);
         }
 
-        this.version = meta.version;
+        //this.version = meta.version; // Paper - move up
     }
 
     CraftMetaItem(CompoundTag tag) {
+        boolean isLegacy = version <= 1343; // Paper
         if (tag.contains(DISPLAY.NBT)) {
             CompoundTag display = tag.getCompound(DISPLAY.NBT);
 
             if (display.contains(NAME.NBT)) {
-                this.displayName = limit( display.getString(NAME.NBT), io.papermc.paper.configuration.GlobalConfiguration.get().itemValidation.displayName ); // Spigot // Paper - make configurable
+                if (isLegacy) this.displayName = CraftChatMessage.fromStringOrNullToJSON(limit( display.getString(NAME.NBT), io.papermc.paper.configuration.GlobalConfiguration.get().itemValidation.displayName )); // Paper - add isLegacy
+                else this.displayName = limit( display.getString(NAME.NBT), io.papermc.paper.configuration.GlobalConfiguration.get().itemValidation.displayName ); // Spigot // Paper - make configurable // Paper - add isLegacy
             }
 
             if (display.contains(LOCNAME.NBT)) {
-                this.locName = limit( display.getString(LOCNAME.NBT), 8192 ); // Spigot
+                if (isLegacy) { this.locName = CraftChatMessage.fromStringOrNullToJSON(limit( display.getString(LOCNAME.NBT), 8192 )); } // Paper - add isLegacy
+                else this.locName = limit( display.getString(LOCNAME.NBT), 8192 ); // Spigot // Paper - make configurable // Paper - add isLegacy
             }
 
             if (display.contains(LORE.NBT)) {
