From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 8 Jul 2018 23:21:47 -0400
Subject: [PATCH] Detect async meta


diff --git a/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java b/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
index 1633a1af2e310165624821792a2a9b68e7ad2134..903ed15c1f926cbf282dc3a4ea89e40e9bc2d653 100644
--- a/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
+++ b/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
@@ -448,12 +448,14 @@ public final class MetaApiAccessor {
     @NotNull
     private static PersistentMetaMap getMetaMapFromCompound(@NotNull CompoundTag cmp) {
         PersistentMetaMap map = new PersistentMetaMap();
+        map.allowAsync = true;
         for (Entry<String, Tag> entry : cmp.tags.entrySet()) {
             final Object metaData = getObjectFromNbt(entry.getValue());
             if (metaData != null) {
                 map.put(entry.getKey(), metaData);
             }
         }
+        map.allowAsync = false;
         return map;
     }
 
