From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 8 Jul 2018 23:21:47 -0400
Subject: [PATCH] Detect async meta


diff --git a/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java b/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
index 6cc205b9a4059d5126a2e8a656c88a7edeaf4c8f..1d3427b7c44bdd6da731dbadafa18efaf2cde48b 100644
--- a/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
+++ b/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
@@ -437,12 +437,14 @@ public final class MetaApiAccessor {
     @NotNull
     private static PersistentMetaMap getMetaMapFromCompound(@NotNull CompoundTag cmp) {
         PersistentMetaMap map = new PersistentMetaMap();
+        map.allowAsync = true;
         for (Entry<String, Tag> entry : cmp.tags.entrySet()) {
             final Object metaData = getObjectFromNbt(entry.getValue());
             if (metaData != null) {
                 map.put(entry.getKey(), metaData);
             }
         }
+        map.allowAsync = false;
         return map;
     }
 
