From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 25 Jun 2018 22:30:08 -0400
Subject: [PATCH] EntityRandomStrollEvent

---
 .../ai/goal/PathfinderGoalRandomStroll.java   | 22 ++++++++++++++-----
 1 file changed, 16 insertions(+), 6 deletions(-)

diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/PathfinderGoalRandomStroll.java b/src/main/java/net/minecraft/world/entity/ai/goal/PathfinderGoalRandomStroll.java
index 6e766d19e53047a4fc5d7e628b8cc7dc9b8062f8..f87f70bb2099d73d7e644c9023e9bb440353167f 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/PathfinderGoalRandomStroll.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/PathfinderGoalRandomStroll.java
@@ -2,17 +2,18 @@ package net.minecraft.world.entity.ai.goal;
 
 import java.util.EnumSet;
 import javax.annotation.Nullable;
+import net.minecraft.server.MCUtil;
 import net.minecraft.world.entity.EntityCreature;
 import net.minecraft.world.entity.ai.util.RandomPositionGenerator;
 import net.minecraft.world.phys.Vec3D;
 
 public class PathfinderGoalRandomStroll extends PathfinderGoal {
 
-    protected final EntityCreature a;
-    protected double b;
-    protected double c;
-    protected double d;
-    protected final double e;
+    protected final EntityCreature a; EntityCreature getEntity() { return a; } // Paper - OBFHELPER
+    protected double b; double getX() { return b; } // Paper - OBFHELPER
+    protected double c; double getY() { return c; } // Paper - OBFHELPER
+    protected double d; double getZ() { return d; } // Paper - OBFHELPER
+    protected final double e;double getSpeed() { return e; } // Paper - OBFHELPER
     protected int f;
     protected boolean g;
     private boolean h;
@@ -74,7 +75,16 @@ public class PathfinderGoalRandomStroll extends PathfinderGoal {
 
     @Override
     public void c() {
-        this.a.getNavigation().a(this.b, this.c, this.d, this.e);
+        // Paper start
+        org.bukkit.Location loc = MCUtil.toLocation(getEntity().getWorld(), getX(), getY(), getZ());
+        com.destroystokyo.paper.event.entity.EntityRandomStrollEvent event = new com.destroystokyo.paper.event.entity.EntityRandomStrollEvent(getEntity().getBukkitCreature(), loc, getSpeed());
+
+        if (!event.callEvent()) {
+            return;
+        }
+        loc = event.getLocation();
+        getEntity().getNavigation().a(loc.getX(), loc.getY(), loc.getZ(), event.getSpeed());
+        // Paper end
     }
 
     @Override
