From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 25 Jun 2018 23:16:51 -0400
Subject: [PATCH] EntityPathfindEndEvent


diff --git a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index e605daac0c90f5d0b9315d1499938feb0e478d0e..119d62ae54ed2bbae99747eac012ef92477f7404 100644
--- a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -356,6 +356,7 @@ public abstract class PathNavigation {
     }
 
     public void stop() {
+        new com.destroystokyo.paper.event.entity.EntityPathfindEndEvent((org.bukkit.entity.Mob)this.mob.getBukkitEntity()).callEvent(); // Paper
         this.path = null;
     }
 
