From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 8 Jul 2018 23:21:47 -0400
Subject: [PATCH] Detect async meta


diff --git a/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java b/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
index b99b1ce0c78b3c3e7fbaf4a8b6a3533d13e28c71..4693bb22249bc5b91b387d9631f3df089f5b7e9c 100644
--- a/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
+++ b/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
@@ -443,12 +443,14 @@ public final class MetaApiAccessor {
     @NotNull
     private static PersistentMetaMap getMetaMapFromCompound(@NotNull CompoundTag cmp) {
         PersistentMetaMap map = new PersistentMetaMap();
+        map.allowAsync = true;
         for (Entry<String, Tag> entry : cmp.tags.entrySet()) {
             final Object metaData = getObjectFromNbt(entry.getValue());
             if (metaData != null) {
                 map.put(entry.getKey(), metaData);
             }
         }
+        map.allowAsync = false;
         return map;
     }
 
