From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 29 Nov 2019 02:48:17 -0600
Subject: [PATCH] Add the PlayerThrowTridentEvent


diff --git a/src/main/java/net/minecraft/world/item/TridentItem.java b/src/main/java/net/minecraft/world/item/TridentItem.java
index 998758be827efbcb7693ed36ab1dffc0ef0369bf..5854513c7461d1b2c58ce699608152a315546541 100644
--- a/src/main/java/net/minecraft/world/item/TridentItem.java
+++ b/src/main/java/net/minecraft/world/item/TridentItem.java
@@ -89,6 +89,7 @@ public class TridentItem extends Item implements Vanishable {
                                 }
                                 return;
                             }
+                            (new com.destroystokyo.paper.event.player.PlayerThrowTridentEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack), entitythrowntrident.getBukkitEntity())).callEvent(); // Paper
 
                             stack.hurtAndBreak(1, entityhuman, (entityhuman1) -> {
                                 entityhuman1.broadcastBreakEvent(user.getUsedItemHand());
