From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 14 Jul 2019 13:50:53 -0500
Subject: [PATCH] ConduitNewTargetEvent and remove target for cancelled damage
 events


diff --git a/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
index 963a596154091b79ca139af6274aa323518ad1ad..24284ef4757a080d9fd18e34a72bfdfd6f1038cb 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
@@ -226,6 +226,14 @@ public class ConduitBlockEntity extends BlockEntity {
                 return entityliving1 instanceof Enemy && entityliving1.isInWaterOrRain();
             });
 
+            // EMC start
+            List<org.bukkit.entity.LivingEntity> bukkitEntities = Lists.newArrayListWithCapacity(list1.size());
+            for (LivingEntity entity : list1) {
+                bukkitEntities.add(entity.getBukkitLivingEntity());
+            }
+            new com.empireminecraft.customevents.ConduitNewTargetEvent(bukkitEntities, CraftBlock.at(world, pos)).callEvent();
+            list1.removeIf(entityLiving -> !bukkitEntities.contains(entityLiving.getBukkitLivingEntity()));
+            // EMC end
             if (!list1.isEmpty()) {
                 blockEntity.destroyTarget = (LivingEntity) list1.get(world.random.nextInt(list1.size()));
             }
@@ -238,7 +246,7 @@ public class ConduitBlockEntity extends BlockEntity {
             CraftEventFactory.blockDamage = CraftBlock.at(world, pos);
             if (blockEntity.destroyTarget.hurt(world.damageSources().magic(), 4.0F)) {
                 world.playSound((Player) null, blockEntity.destroyTarget.getX(), blockEntity.destroyTarget.getY(), blockEntity.destroyTarget.getZ(), SoundEvents.CONDUIT_ATTACK_TARGET, SoundSource.BLOCKS, 1.0F, 1.0F);
-            }
+            } else { blockEntity.destroyTarget = null; } // EMC
             CraftEventFactory.blockDamage = null;
             // CraftBukkit end
         }
