From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 4 Jul 2018 15:49:34 -0400
Subject: [PATCH] Fire PlayerKickEvent during server shutdown


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index cc2cbfc53ea10ef74849383ad3ddfb8748f3b3a8..b1821dfe483496fa373005b59b9add276b8da39b 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -441,11 +441,11 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
 
         PlayerKickEvent event = new PlayerKickEvent(this.player.getBukkitEntity(), reason, leaveMessage, cause); // Paper - Adventure & kick event reason
 
-        if (this.cserver.getServer().isRunning()) {
+        if (true) { // EMC - don't care if its async either, fire it!
             this.cserver.getPluginManager().callEvent(event);
         }
 
-        if (event.isCancelled()) {
+        if (!cserver.getServer().hasStopped() && event.isCancelled()) { // EMC
             // Do not kick the player
             return;
         }
