From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 29 Jul 2021 02:07:15 -0500
Subject: [PATCH] Capture item drops during block placement


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index eec0ec43590be7e8ae5b530a7404c98b5e23cb53..e95886bac529b294eb2a9e543d26809d8ea9840e 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -435,10 +435,19 @@ public final class ItemStack implements DataComponentHolder {
             } finally {
                 world.captureBlockStates = false;
             }
+            // Paper start
+            world.captureDrops = new java.util.ArrayList<>();
+            org.bukkit.block.Block bblock = CraftBlock.at(world, blockposition);
+            org.bukkit.block.BlockState state = bblock.getState();
+            // Paper end
             DataComponentPatch newData = this.components.asPatch();
             int newCount = this.getCount();
             this.setCount(oldCount);
             this.restorePatch(oldData);
+            // Paper start
+            List<ItemEntity> itemsToDrop = world.captureDrops;
+            world.captureDrops = null;
+            // Paper end
             if (enuminteractionresult.consumesAction() && world.captureTreeGeneration && world.capturedBlockStates.size() > 0) {
                 world.captureTreeGeneration = false;
                 Location location = CraftLocation.toBukkit(blockposition, world.getWorld());
@@ -458,6 +467,7 @@ public final class ItemStack implements DataComponentHolder {
                 org.bukkit.Bukkit.getPluginManager().callEvent(fertilizeEvent);
 
                 if (!fertilizeEvent.isCancelled()) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDropItemEvent(bblock, state, (ServerPlayer) entityhuman, itemsToDrop); // Paper
                     // Change the stack to its new contents if it hasn't been tampered with.
                     if (this.getCount() == oldCount && Objects.equals(this.components.asPatch(), oldData)) {
                         this.restorePatch(newData);
@@ -510,6 +520,7 @@ public final class ItemStack implements DataComponentHolder {
                     // Paper end - Don't resync blocks
                     SignItem.openSign = null; // SPIGOT-6758 - Reset on early return
                 } else {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDropItemEvent(bblock, state, (ServerPlayer) entityhuman, itemsToDrop); // Paper
                     // Change the stack to its new contents if it hasn't been tampered with.
                     if (this.getCount() == oldCount && Objects.equals(this.components.asPatch(), oldData)) {
                         this.restorePatch(newData);
