From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 26 Jan 2020 23:28:04 -0600
Subject: [PATCH] Use lenient json to component to resolve old WE schematic
 issue


diff --git a/src/main/java/net/minecraft/server/MCUtil.java b/src/main/java/net/minecraft/server/MCUtil.java
index 13082cd8de1a79a3b2fac6055bdaa163dbc7897b..28ee48b4e41ba59f34e57d8040eadaf9937f8a44 100644
--- a/src/main/java/net/minecraft/server/MCUtil.java
+++ b/src/main/java/net/minecraft/server/MCUtil.java
@@ -549,7 +549,7 @@ public final class MCUtil {
         }
         String string = compound.getString(key);
         try {
-            return Component.Serializer.fromJson(string);
+            return Component.Serializer.fromJsonLenient(string); // EMC
         } catch (com.google.gson.JsonParseException e) {
             org.bukkit.Bukkit.getLogger().warning("Unable to parse " + key + " from " + compound +": " + e.getMessage());
         }
