From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 26 Jan 2020 23:28:04 -0600
Subject: [PATCH] Use lenient json to component to resolve old WE schematic
 issue


diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index d02546b18cb689724887b4e85e8d32a18828a4ad..c71dadc87e4364fef6afafa5dd8b7105847cce8b 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -656,7 +656,7 @@ public final class MCUtil {
         }
         String string = compound.getString(key);
         try {
-            return net.minecraft.network.chat.Component.Serializer.fromJson(string);
+            return net.minecraft.network.chat.Component.Serializer.fromJsonLenient(string); // EMC
         } catch (com.google.gson.JsonParseException e) {
             org.bukkit.Bukkit.getLogger().warning("Unable to parse " + key + " from " + compound +": " + e.getMessage());
         }
