From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 8 Jul 2018 23:21:47 -0400
Subject: [PATCH] Detect async meta


diff --git a/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java b/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
index 40748a486766aff09e37e1b58a77e4744597f66d..17300b0ed4a42fbb772479d1e362490ac661ec3c 100644
--- a/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
+++ b/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
@@ -440,12 +440,14 @@ public final class MetaApiAccessor {
     @NotNull
     private static PersistentMetaMap getMetaMapFromCompound(@NotNull CompoundTag cmp) {
         PersistentMetaMap map = new PersistentMetaMap();
+        map.allowAsync = true;
         for (Entry<String, Tag> entry : cmp.tags.entrySet()) {
             final Object metaData = getObjectFromNbt(entry.getValue());
             if (metaData != null) {
                 map.put(entry.getKey(), metaData);
             }
         }
+        map.allowAsync = false;
         return map;
     }
 
