From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 26 Jan 2020 23:28:04 -0600
Subject: [PATCH] Use lenient json to component to resolve old WE schematic
 issue


diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index 8240bb085b619f257f8c0a25775e0b15068e440f..5d77ff026271d142e0365544e3d07bfd6a48576d 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -660,7 +660,7 @@ public final class MCUtil {
         }
         String string = compound.getString(key);
         try {
-            return net.minecraft.network.chat.Component.Serializer.fromJson(string);
+            return net.minecraft.network.chat.Component.Serializer.fromJsonLenient(string); // EMC
         } catch (com.google.gson.JsonParseException e) {
             org.bukkit.Bukkit.getLogger().warning("Unable to parse " + key + " from " + compound +": " + e.getMessage());
         }
