From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 29 Nov 2019 02:48:17 -0600
Subject: [PATCH] Add the PlayerThrowTridentEvent


diff --git a/src/main/java/net/minecraft/world/item/TridentItem.java b/src/main/java/net/minecraft/world/item/TridentItem.java
index 9365f886a23a71c41091b22d46896ff18a5a0635..a69cf7ff0800760b04dc9ca9db13956c96ebd5ff 100644
--- a/src/main/java/net/minecraft/world/item/TridentItem.java
+++ b/src/main/java/net/minecraft/world/item/TridentItem.java
@@ -92,6 +92,7 @@ public class TridentItem extends Item implements Vanishable {
                                 }
                                 return;
                             }
+                            (new com.destroystokyo.paper.event.player.PlayerThrowTridentEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack), entitythrowntrident.getBukkitEntity())).callEvent(); // Paper
                             if (event.shouldConsume()) { // Paper
                             stack.hurtAndBreak(1, entityhuman, (entityhuman1) -> {
                                 entityhuman1.broadcastBreakEvent(user.getUsedItemHand());
