From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 26 Jan 2020 23:28:04 -0600
Subject: [PATCH] Use lenient json to component to resolve old WE schematic
 issue


diff --git a/src/main/java/io/papermc/paper/util/MCUtil.java b/src/main/java/io/papermc/paper/util/MCUtil.java
index 6efb8b10f17c70b05128039376d254e6beda3841..3cbceb7071f693adecf35302c08edd5d8d0fab2e 100644
--- a/src/main/java/io/papermc/paper/util/MCUtil.java
+++ b/src/main/java/io/papermc/paper/util/MCUtil.java
@@ -543,7 +543,7 @@ public final class MCUtil {
         }
         String string = compound.getString(key);
         try {
-            return Component.Serializer.fromJson(string);
+            return Component.Serializer.fromJsonLenient(string); // EMC
         } catch (com.google.gson.JsonParseException e) {
             org.bukkit.Bukkit.getLogger().warning("Unable to parse " + key + " from " + compound +": " + e.getMessage());
         }
