From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 17 Mar 2020 18:59:20 -0500
Subject: [PATCH] Properly handle entityTags on SpawnEggs


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
index 1c4b10a2c30d27d6ec94b739d580dd8f179978a7..6dd38f3ed8d8f44cf00a0731e2554edcb978d993 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
@@ -2,14 +2,20 @@ package org.bukkit.craftbukkit.inventory;
 
 import com.google.common.collect.ImmutableMap.Builder;
 import java.util.Map;
+
+import com.mojang.serialization.Dynamic;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.NbtOps;
 import net.minecraft.nbt.Tag;
 import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.util.datafix.fixes.References;
 import org.bukkit.Material;
 import org.bukkit.configuration.serialization.DelegateDeserialization;
 import org.bukkit.craftbukkit.inventory.CraftMetaItem.ItemMetaKey;
 import org.bukkit.craftbukkit.inventory.CraftMetaItem.SerializableMeta;
 import org.bukkit.craftbukkit.util.CraftLegacy;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
 import org.bukkit.entity.EntityType;
 import org.bukkit.inventory.meta.SpawnEggMeta;
 import org.bukkit.material.MaterialData;
@@ -42,6 +48,16 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
 
         if (tag.contains(ENTITY_TAG.NBT)) {
             this.entityTag = tag.getCompound(ENTITY_TAG.NBT).copy();
+            // EMC start
+            if (!this.entityTag.isEmpty()) {
+                String entityType = entityTag.getString(ENTITY_ID.NBT);
+                this.spawnedType = EntityType.fromName(entityType);
+                entityTag.remove(ENTITY_ID.NBT);
+                if (entityTag != null && entityTag.isEmpty()) {
+                    entityTag = null;
+                }
+            }
+            // EMC end
         }
     }
 
@@ -77,7 +93,13 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             }
 
             // Tag still has some other data, lets try our luck with a conversion
-            if (!this.entityTag.isEmpty()) {
+            if (this.spawnedType == null && this.entityTag.contains(ENTITY_ID.NBT) && !this.entityTag.isEmpty()) { // EMC - only try converting when there is an id and spawnedType is not found already
+                // EMC start - try updating and catch update exceptions
+                try {
+                    entityTag = (CompoundTag) MinecraftServer.getServer().fixerUpper.update(References.ENTITY, new Dynamic(NbtOps.INSTANCE, entityTag), -1, CraftMagicNumbers.INSTANCE.getDataVersion()).getValue();
+                } catch (IllegalArgumentException ignored) {
+                }
+                // EMC end
                 // SPIGOT-4128: This is hopeless until we start versioning stacks. RIP data.
                 // entityTag = (NBTTagCompound) MinecraftServer.getServer().dataConverterManager.update(DataConverterTypes.ENTITY, new Dynamic(DynamicOpsNBT.a, entityTag), -1, CraftMagicNumbers.DATA_VERSION).getValue();
             }
@@ -85,7 +107,15 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             // See if we can read a converted ID tag
             if (this.entityTag.contains(ENTITY_ID.NBT)) {
                 this.spawnedType = EntityType.fromName(new ResourceLocation(this.entityTag.getString(ENTITY_ID.NBT)).getPath());
+                // EMC start
+                if (spawnedType != null) {
+                    entityTag.remove(ENTITY_ID.NBT);
+                }
+            }
+            if (entityTag.isEmpty()) {
+                entityTag = null;
             }
+            // EMC end
         }
     }
 
@@ -236,7 +266,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             CraftMetaSpawnEgg that = (CraftMetaSpawnEgg) meta;
 
             return this.hasSpawnedType() ? that.hasSpawnedType() && this.spawnedType.equals(that.spawnedType) : !that.hasSpawnedType()
-                    && this.entityTag != null ? that.entityTag != null && this.entityTag.equals(that.entityTag) : this.entityTag == null;
+                    && (this.entityTag != null ? that.entityTag != null && this.entityTag.equals(that.entityTag) : that.entityTag == null); // EMC - fix logic
         }
         return true;
     }
@@ -291,6 +321,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             if (this.entityTag != null) {
                 // Remove ID tag as it is now in the material
                 this.entityTag.remove(ENTITY_ID.NBT);
+                if (entityTag.isEmpty()) { entityTag = null; } // EMC
             }
 
             return CraftLegacy.fromLegacy(new MaterialData(Material.LEGACY_MONSTER_EGG, (byte) this.spawnedType.getTypeId()));
