From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 26 Jan 2020 23:28:04 -0600
Subject: [PATCH] Use lenient json to component to resolve old WE schematic
 issue


diff --git a/src/main/java/net/minecraft/server/MCUtil.java b/src/main/java/net/minecraft/server/MCUtil.java
index f9701ef4162868894a81f0d1ac9b58d00d9a2d65..8be18c9e49b699b55f3f4cd34e2e08760c0344b8 100644
--- a/src/main/java/net/minecraft/server/MCUtil.java
+++ b/src/main/java/net/minecraft/server/MCUtil.java
@@ -539,7 +539,7 @@ public final class MCUtil {
         }
         String string = compound.getString(key);
         try {
-            return Component.Serializer.fromJson(string);
+            return Component.Serializer.fromJsonLenient(string); // EMC
         } catch (com.google.gson.JsonParseException e) {
             org.bukkit.Bukkit.getLogger().warning("Unable to parse " + key + " from " + compound +": " + e.getMessage());
         }
