From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 7 Mar 2020 10:32:04 -0600
Subject: [PATCH] EntityAngerEvent


diff --git a/src/main/java/net/minecraft/world/entity/NeutralMob.java b/src/main/java/net/minecraft/world/entity/NeutralMob.java
index 7dac62b6370dae3ad6d098857c3136d5acf2bd74..bda3330538de9f3c37fb69fd6c8b656f126ef148 100644
--- a/src/main/java/net/minecraft/world/entity/NeutralMob.java
+++ b/src/main/java/net/minecraft/world/entity/NeutralMob.java
@@ -3,6 +3,8 @@ package net.minecraft.world.entity;
 import java.util.Objects;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import com.destroystokyo.paper.event.entity.EntityAngerEvent;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.player.Player;
@@ -59,8 +61,15 @@ public interface NeutralMob {
             this.stopBeingAngry();
         } else {
             if (entityliving != null && !Objects.equals(uuid, entityliving.getUUID())) {
-                this.setPersistentAngerTarget(entityliving.getUUID());
-                this.startPersistentAngerTimer();
+                // Paper start
+                if (this instanceof Entity) {
+                    EntityAngerEvent event = new EntityAngerEvent(((Entity)this).getBukkitEntity(), entityliving.getBukkitLivingEntity());
+                    if (event.callEvent()) {
+                        this.setPersistentAngerTarget(entityliving.getUUID());
+                        this.startPersistentAngerTimer();
+                    }
+                }
+                // Paper end
             }
 
             if (this.getRemainingPersistentAngerTime() > 0 && (entityliving == null || entityliving.getType() != EntityType.PLAYER || !angerPersistent)) {
