From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 31 Mar 2023 23:22:24 -0500
Subject: [PATCH] Revert mysql dependency version for now...


diff --git a/build.gradle.kts b/build.gradle.kts
index 9d76aadc67c11d71e2cf685a45fbd5da7626590e..066937b4976c1a2c9a5a385710772b86b6cc32c9 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -35,7 +35,7 @@ dependencies {
     implementation("org.spongepowered:configurate-yaml:4.2.0-SNAPSHOT") // Paper - config files
     implementation("commons-lang:commons-lang:2.6")
     runtimeOnly("org.xerial:sqlite-jdbc:3.46.0.0")
-    runtimeOnly("com.mysql:mysql-connector-j:8.4.0")
+    runtimeOnly("mysql:mysql-connector-java:8.0.29") // EMC
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
     // Paper start - Use Velocity cipher
     implementation("com.velocitypowered:velocity-native:3.3.0-SNAPSHOT") {
