From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 17 Jan 2020 21:32:10 -0600
Subject: [PATCH] Add the BlockHarvestBeehiveEvent


diff --git a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 0d34026a70c72661a9ba6b319690370e589714cc..e16cd0246ffebe6af34329f0f7bc656099b11416 100644
--- a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -9,6 +9,7 @@ import net.minecraft.core.BlockPos;
 import net.minecraft.core.BlockSource;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Position;
+import net.minecraft.server.MCUtil;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvents;
@@ -974,6 +975,11 @@ public interface DispenseItemBehavior {
                 if (iblockdata.is((Tag) BlockTags.BEEHIVES, (blockbase_blockdata) -> {
                     return blockbase_blockdata.hasProperty(BeehiveBlock.HONEY_LEVEL);
                 }) && (Integer) iblockdata.getValue(BeehiveBlock.HONEY_LEVEL) >= 5) {
+                    // EMC start
+                    if (!(new com.empireminecraft.customevents.BlockHarvestBeehiveEvent(bukkitBlock, MCUtil.toBukkitBlock(worldserver, blockposition), craftItem).callEvent())) {
+                        return stack;
+                    }
+                    // EMC end
                     ((BeehiveBlock) iblockdata.getBlock()).releaseBeesAndResetHoneyLevel((Level) worldserver, iblockdata, blockposition, (Player) null, BeehiveBlockEntity.BeeReleaseStatus.BEE_RELEASED);
                     this.setSuccess(true);
                     return this.a(pointer, stack, new ItemStack(Items.HONEY_BOTTLE));
