From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 16 Jun 2022 09:13:17 -0500
Subject: [PATCH] EntityTriggerSculkCatalystEvent


diff --git a/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
index a74732902c0494c67e6acf2fc04581ff9c46b832..2f8cb89a4acea634817a98cb07f86c89c1b9ebdf 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -104,6 +104,12 @@ public class SculkCatalystBlockEntity extends BlockEntity implements GameEventLi
 
                 if (entity instanceof LivingEntity) {
                     LivingEntity entityliving = (LivingEntity) entity;
+                    // EMC start
+                    net.minecraft.core.Position pos = this.positionSource.getPosition(world).orElse(null);
+                    if (pos != null && !(new com.empireminecraft.customevents.EntityTriggerSculkCatalystEvent(entityliving.getBukkitLivingEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, BlockPos.containing(pos))).callEvent())) {
+                        return false;
+                    }
+                    // EMC end
 
                     if (!entityliving.wasExperienceConsumed()) {
                         DamageSource damagesource = entityliving.getLastDamageSource();
