From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 6 May 2015 02:29:30 -0400
Subject: [PATCH] Limit worlds structures can generate


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 519faa36efe54944d11ec87cf9e5c5fabb16f738..54cef1445a59eb6febf7dda65854624cce2befc2 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1914,6 +1914,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
         if (!this.serverLevelData.worldGenSettings().generateStructures()) { // CraftBukkit
             return null;
         } else {
+            if ("town".equals(getWorld().getName())) { return null; } // EMC
             Optional<HolderSet.Named<Structure>> optional = this.registryAccess().registryOrThrow(Registry.STRUCTURE_REGISTRY).getTag(structureTag);
 
             if (optional.isEmpty()) {
diff --git a/src/main/java/net/minecraft/world/level/StructureManager.java b/src/main/java/net/minecraft/world/level/StructureManager.java
index 64ca546a905d69a795bbdafb4f9860b31cd863d1..c7ad8c6cfb531f283f872b5a4519e35e894440bc 100644
--- a/src/main/java/net/minecraft/world/level/StructureManager.java
+++ b/src/main/java/net/minecraft/world/level/StructureManager.java
@@ -93,7 +93,7 @@ public class StructureManager {
     }
 
     public boolean shouldGenerateStructures() {
-        return this.worldGenSettings.generateStructures();
+        return this.worldGenSettings.generateStructures() && !"town".equals(this.level.getMinecraftWorld().getWorld().getName()); // EMC
     }
 
     public StructureStart getStructureAt(BlockPos pos, Structure structure) {
diff --git a/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java b/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java
index 00cb9dafc711607f28529ea9afbcdb492b1b2595..82f19b85568b254c5e484bfa155d4b37bcf21bcc 100644
--- a/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java
+++ b/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java
@@ -71,7 +71,7 @@ public class ChunkStatus {
     });
     public static final ChunkStatus STRUCTURE_STARTS = ChunkStatus.register("structure_starts", ChunkStatus.EMPTY, 0, ChunkStatus.PRE_FEATURES, ChunkStatus.ChunkType.PROTOCHUNK, (chunkstatus, executor, worldserver, chunkgenerator, structuretemplatemanager, lightenginethreaded, function, list, ichunkaccess, flag) -> {
         if (!ichunkaccess.getStatus().isOrAfter(chunkstatus)) {
-            if (worldserver.serverLevelData.worldGenSettings().generateStructures()) { // CraftBukkit
+            if (worldserver.serverLevelData.worldGenSettings().generateStructures() && !"town".equals(worldserver.getWorld().getName())) { // CraftBukkit // EMC
                 chunkgenerator.createStructures(worldserver.registryAccess(), worldserver.getChunkSource().randomState(), worldserver.structureManager(), ichunkaccess, structuretemplatemanager, worldserver.getSeed());
             }
 
