From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 6 Jun 2020 21:17:28 -0500
Subject: [PATCH] Raid total waves API


diff --git a/src/main/java/net/minecraft/world/entity/raid/Raid.java b/src/main/java/net/minecraft/world/entity/raid/Raid.java
index f7399737548483905f3b5c08a03876b0da54b714..3c196ad8a4ddb6a4524119ebfefe7fd978d2bc5d 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raid.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raid.java
@@ -103,7 +103,8 @@ public class Raid {
     private int postRaidTicks;
     private int raidCooldownTicks;
     private final RandomSource random;
-    public final int numGroups;
+    public int numGroups; // EMC - make non-final
+    public boolean firstTrigger = false; // EMC
     private Raid.RaidStatus status;
     private int celebrationTicks;
     private Optional<BlockPos> waveSpawnPos;
@@ -120,6 +121,7 @@ public class Raid {
         this.center = pos;
         this.numGroups = this.getNumGroups(world.getDifficulty());
         this.status = Raid.RaidStatus.ONGOING;
+        this.firstTrigger = true; // EMC
     }
 
     public Raid(ServerLevel world, CompoundTag nbt) {
diff --git a/src/main/java/net/minecraft/world/entity/raid/Raids.java b/src/main/java/net/minecraft/world/entity/raid/Raids.java
index 41457c9f27b18fa2734a6cca297ec5186470e82f..c4865a43fcaaab36abf2ab14197f22836b52c8a0 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raids.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raids.java
@@ -135,6 +135,7 @@ public class Raids extends SavedData {
                         return null;
                     }
 
+                    raid.firstTrigger = false; // EMC
                     if (!this.raidMap.containsKey(raid.getId())) {
                         this.raidMap.put(raid.getId(), raid);
                     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftRaid.java b/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
index 92672558a64d5b7085a075dd2d274e0dc3d3ef4c..28808126fa43e7472c36856dc12e20250b0878e5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
@@ -80,6 +80,13 @@ public final class CraftRaid implements Raid {
         return handle.numGroups;
     }
 
+    // EMC start
+    @Override
+    public void setTotalWaves(int total) {
+        Preconditions.checkArgument(1 <= total && total <= 7, "Total waves out of bounds 1 <= %s <= 7", total);
+        handle.numGroups = total;
+    }
+    // EMC end
     @Override
     public float getTotalHealth() {
         return this.handle.getHealthOfLivingRaiders();
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index ec3784cc3c906e87a36b69216927db1b3ccdb489..f685f6e4c7b5c48cc0af44640b14080c1a30238d 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1911,7 +1911,7 @@ public class CraftEventFactory {
      * Raid events
      */
     public static boolean callRaidTriggerEvent(Raid raid, ServerPlayer player) {
-        RaidTriggerEvent event = new RaidTriggerEvent(new CraftRaid(raid), raid.getLevel().getWorld(), player.getBukkitEntity());
+        RaidTriggerEvent event = new RaidTriggerEvent(new CraftRaid(raid), raid.getLevel().getWorld(), player.getBukkitEntity(), raid.firstTrigger); // EMC - add raid.firstTrigger
         Bukkit.getPluginManager().callEvent(event);
         return !event.isCancelled();
     }
