From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 18 Mar 2020 21:54:38 -0500
Subject: [PATCH] Add getPlayerCause to LightningStrike entities


diff --git a/src/main/java/net/minecraft/world/entity/LightningBolt.java b/src/main/java/net/minecraft/world/entity/LightningBolt.java
index 359c78b4f98c38c81af27a2c4cf0939da51e5a0b..c3c1dc51a283db5b230d842a062dd6ed14a931ff 100644
--- a/src/main/java/net/minecraft/world/entity/LightningBolt.java
+++ b/src/main/java/net/minecraft/world/entity/LightningBolt.java
@@ -44,7 +44,7 @@ public class LightningBolt extends Entity {
     public int flashes;
     public boolean visualOnly;
     @Nullable
-    private ServerPlayer cause;
+    public ServerPlayer cause; // EMC - make public
     private final Set<Entity> hitEntities = Sets.newHashSet();
     private int blocksSetOnFire;
     public boolean isSilent = false; // Spigot
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLightningStrike.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLightningStrike.java
index e515e819774bfb31ec03f05a5502921e66f2b0e2..03814f400b15de2ff6a88c5538a9202e2c7d1486 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLightningStrike.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLightningStrike.java
@@ -79,4 +79,11 @@ public class CraftLightningStrike extends CraftEntity implements LightningStrike
         this.getHandle().setCause(causingPlayer == null ? null : ((CraftPlayer) causingPlayer).getHandle());
     }
     // Paper end
+    // EMC start
+    @Override
+    public org.bukkit.entity.Player getPlayerCause() {
+        net.minecraft.server.level.ServerPlayer serverPlayer = ((LightningBolt) super.getHandle()).cause;
+        return serverPlayer != null ? serverPlayer.getBukkitEntity() : null;
+    }
+    // EMC end
 }
