From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 19 Dec 2020 03:59:09 -0500
Subject: [PATCH] Quick fix the rcon log issue for now


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 9142527592bc163b8952d687983e69df82aa4f56..1813d97953ee0a40a697ae61ebffeef058869054 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -122,6 +122,7 @@ public class Main {
 
             Path path1 = Paths.get("eula.txt");
             Eula eula = new Eula(path1);
+            PaperLogFilters.configureFilter(); // Paper
 
             if (optionset.has("initSettings")) { // CraftBukkit
                 // CraftBukkit start - SPIGOT-5761: Create bukkit.yml and commands.yml if not present
diff --git a/src/main/java/net/minecraft/server/PaperLogFilters.java b/src/main/java/net/minecraft/server/PaperLogFilters.java
new file mode 100644
index 0000000000000000000000000000000000000000..cd518c157d25c7b268b04d0a25c073eaf2c2792b
--- /dev/null
+++ b/src/main/java/net/minecraft/server/PaperLogFilters.java
@@ -0,0 +1,24 @@
+package net.minecraft.server;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.filter.AbstractFilter;
+
+public final class PaperLogFilters {
+    private PaperLogFilters() {
+    }
+
+    public static void configureFilter() {
+        final org.apache.logging.log4j.core.Logger logger = ((org.apache.logging.log4j.core.Logger) LogManager.getRootLogger());
+        logger.addFilter(new AbstractFilter() {
+
+            @Override
+            public Result filter(LogEvent logEvent) {
+                String msg = logEvent.getMessage().getFormattedMessage();
+                boolean match = msg.contains("RCON Client");
+                return match ? Result.DENY : Result.ACCEPT;
+            }
+        });
+
+    }
+}
