From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 6 Jun 2020 21:17:28 -0500
Subject: [PATCH] Raid total waves API


diff --git a/src/main/java/net/minecraft/world/entity/raid/Raid.java b/src/main/java/net/minecraft/world/entity/raid/Raid.java
index f50abfd058e3adf1cf8c0f4468a60bc04e8541e7..03f74912cce42e88de7a7c80fdc37430a2269a32 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raid.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raid.java
@@ -103,7 +103,8 @@ public class Raid {
     private int postRaidTicks;
     private int raidCooldownTicks;
     private final Random random;
-    public final int numGroups;
+    public int numGroups; // EMC - make non-final
+    public boolean firstTrigger = false; // EMC
     private Raid.RaidStatus status;
     private int celebrationTicks;
     private Optional<BlockPos> waveSpawnPos;
@@ -120,6 +121,7 @@ public class Raid {
         this.center = pos;
         this.numGroups = this.getNumGroups(world.getDifficulty());
         this.status = Raid.RaidStatus.ONGOING;
+        this.firstTrigger = true; // EMC
     }
 
     public Raid(ServerLevel world, CompoundTag nbt) {
diff --git a/src/main/java/net/minecraft/world/entity/raid/Raids.java b/src/main/java/net/minecraft/world/entity/raid/Raids.java
index 17b24f57cc5d63f99eb999cfe3e2362758fe0f4f..8b7ed7ab9c9c9c44d8c1b2ad1a8db3e081dcf8a9 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raids.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raids.java
@@ -133,6 +133,7 @@ public class Raids extends SavedData {
                         return null;
                     }
 
+                    raid.firstTrigger = false; // EMC
                     if (!this.raidMap.containsKey(raid.getId())) {
                         this.raidMap.put(raid.getId(), raid);
                     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftRaid.java b/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
index 7a5bfc724e42100b7582836ab8c30feed1a3d221..d18305d37926ee61fddbe20f6afcbe902e41b32f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
@@ -79,6 +79,13 @@ public final class CraftRaid implements Raid {
         return handle.numGroups;
     }
 
+    // EMC start
+    @Override
+    public void setTotalWaves(int total) {
+        Preconditions.checkArgument(1 <= total && total <= 7, "Total waves out of bounds 1 <= %s <= 7", total);
+        handle.numGroups = total;
+    }
+    // EMC end
     @Override
     public float getTotalHealth() {
         return this.handle.getHealthOfLivingRaiders();
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 488f0f91048a25cf9e452f2c66e0992e38f24701..677f382223dd2e0a584eba6318b99eae3abc2712 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1840,7 +1840,7 @@ public class CraftEventFactory {
      * Raid events
      */
     public static boolean callRaidTriggerEvent(Raid raid, ServerPlayer player) {
-        RaidTriggerEvent event = new RaidTriggerEvent(new CraftRaid(raid), raid.getLevel().getWorld(), player.getBukkitEntity());
+        RaidTriggerEvent event = new RaidTriggerEvent(new CraftRaid(raid), raid.getLevel().getWorld(), player.getBukkitEntity(), raid.firstTrigger); // EMC - add raid.firstTrigger
         Bukkit.getPluginManager().callEvent(event);
         return !event.isCancelled();
     }
