From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 27 May 2023 16:03:16 -0500
Subject: [PATCH] PortalFrameCreateBlockCheckEvent


diff --git a/src/main/java/net/minecraft/world/level/portal/PortalForcer.java b/src/main/java/net/minecraft/world/level/portal/PortalForcer.java
index 49f7ba292b82bac1643cc07aa576f3c37b8e8ab3..a7daef2e529d3eec0926fef0f67a2899dde8f734 100644
--- a/src/main/java/net/minecraft/world/level/portal/PortalForcer.java
+++ b/src/main/java/net/minecraft/world/level/portal/PortalForcer.java
@@ -243,6 +243,11 @@ public class PortalForcer {
                 if (k >= 0 && !this.canPortalReplaceBlock(temp)) {
                     return false;
                 }
+                // EMC start
+                if (!(new com.empireminecraft.customevents.PortalFrameCreateBlockCheckEvent(org.bukkit.craftbukkit.block.CraftBlock.at(this.level, temp))).callEvent()) {
+                    return false;
+                }
+                // EMC end
             }
         }
 
