From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 6 May 2015 02:29:30 -0400
Subject: [PATCH] Limit worlds structures can generate


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index e1b61eaceeaa58c4a2e15e83e9dd1d209c48ec33..534c6479314aa5dc080a983438e85179c94e5ffe 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1813,6 +1813,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     }
     public @Nullable BlockPos findNearestMapFeature(HolderSet<ConfiguredStructureFeature<?, ?>> holderSet, BlockPos pos, int radius, boolean skipExistingChunks) {
         {
+            if ("town".equals(getWorld().getName()) ) { return null; } // EMC
             {
                 Pair<BlockPos, Holder<ConfiguredStructureFeature<?, ?>>> pair = this.getChunkSource().getGenerator().findNearestMapFeature(this, holderSet, pos, radius, skipExistingChunks);
                 // Paper end
diff --git a/src/main/java/net/minecraft/world/level/StructureFeatureManager.java b/src/main/java/net/minecraft/world/level/StructureFeatureManager.java
index fa6eaf7041b6a8b17c64322307839dbd358ad3ab..12c092e466c9d187644269b273698b9137f72c77 100644
--- a/src/main/java/net/minecraft/world/level/StructureFeatureManager.java
+++ b/src/main/java/net/minecraft/world/level/StructureFeatureManager.java
@@ -94,7 +94,7 @@ public class StructureFeatureManager {
     }
 
     public boolean shouldGenerateFeatures() {
-        return this.worldGenSettings.generateFeatures();
+        return this.worldGenSettings.generateFeatures() && !"town".equals(this.level.getMinecraftWorld().getWorld().getName()); // EMC
     }
 
     public StructureStart getStructureAt(BlockPos pos, ConfiguredStructureFeature<?, ?> configuredStructureFeature) {
diff --git a/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java b/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java
index 954e141e5617d6d52e2f3008c25fe9e2fe2f0f9a..1d70218e20e81905fdd079d47e89d8ff4b746449 100644
--- a/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java
+++ b/src/main/java/net/minecraft/world/level/chunk/ChunkStatus.java
@@ -47,7 +47,7 @@ public class ChunkStatus {
     });
     public static final ChunkStatus STRUCTURE_STARTS = ChunkStatus.register("structure_starts", ChunkStatus.EMPTY, 0, ChunkStatus.PRE_FEATURES, ChunkStatus.ChunkType.PROTOCHUNK, (chunkstatus, executor, worldserver, chunkgenerator, definedstructuremanager, lightenginethreaded, function, list, ichunkaccess, flag) -> {
         if (!ichunkaccess.getStatus().isOrAfter(chunkstatus)) {
-            if (worldserver.serverLevelData.worldGenSettings().generateFeatures()) { // CraftBukkit
+            if (worldserver.serverLevelData.worldGenSettings().generateFeatures() && !"town".equals(worldserver.getWorld().getName())) { // CraftBukkit // EMC
                 chunkgenerator.createStructures(worldserver.registryAccess(), worldserver.structureFeatureManager(), ichunkaccess, definedstructuremanager, worldserver.getSeed());
             }
 
