From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 17 Jan 2020 21:32:10 -0600
Subject: [PATCH] Add the BlockHarvestBeehiveEvent

---
 .../net/minecraft/core/dispenser/IDispenseBehavior.java     | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/main/java/net/minecraft/core/dispenser/IDispenseBehavior.java b/src/main/java/net/minecraft/core/dispenser/IDispenseBehavior.java
index 93093c05da53e5ddc59fac179081af2e0893706d..e6b7b2a8a5913ccdabd4cc0ce357ea77d75e6a16 100644
--- a/src/main/java/net/minecraft/core/dispenser/IDispenseBehavior.java
+++ b/src/main/java/net/minecraft/core/dispenser/IDispenseBehavior.java
@@ -8,6 +8,7 @@ import net.minecraft.core.BlockPosition;
 import net.minecraft.core.EnumDirection;
 import net.minecraft.core.IPosition;
 import net.minecraft.core.ISourceBlock;
+import net.minecraft.server.MCUtil;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.EntityPlayer;
 import net.minecraft.server.level.WorldServer;
@@ -954,6 +955,11 @@ public interface IDispenseBehavior {
                 if (iblockdata.a((Tag) TagsBlock.BEEHIVES, (blockbase_blockdata) -> {
                     return blockbase_blockdata.b(BlockBeehive.b);
                 }) && (Integer) iblockdata.get(BlockBeehive.b) >= 5) {
+                    // EMC start
+                    if (!(new com.empireminecraft.customevents.BlockHarvestBeehiveEvent(bukkitBlock, MCUtil.toBukkitBlock(worldserver, blockposition), craftItem).callEvent())) {
+                        return itemstack;
+                    }
+                    // EMC end
                     ((BlockBeehive) iblockdata.getBlock()).a(worldserver, iblockdata, blockposition, (EntityHuman) null, TileEntityBeehive.ReleaseStatus.BEE_RELEASED);
                     this.a(true);
                     return this.a(isourceblock, itemstack, new ItemStack(Items.HONEY_BOTTLE));
