From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 2 Jun 2021 09:16:37 -0500
Subject: [PATCH] Fire FurnaceExtractEvent for furnace breakings with null
 players

temp

diff --git a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index 89d06253b00604114e543ebbe12a9993ae95dc41..a4bf7f5b79b188be176794f4152ab2455bd62dd9 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -650,14 +650,14 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
             worldserver.getRecipeManager().byKey((ResourceLocation) entry.getKey()).ifPresent((recipeholder) -> {
                 if (!(recipeholder.value() instanceof AbstractCookingRecipe)) return; // Paper - don't process non-cooking recipes
                 list.add(recipeholder);
-                AbstractFurnaceBlockEntity.createExperience(worldserver, vec3d, entry.getIntValue(), ((AbstractCookingRecipe) recipeholder.value()).getExperience(), blockposition, entityplayer, itemstack, amount); // CraftBukkit
+                AbstractFurnaceBlockEntity.createExperience(worldserver, vec3d, entry.getIntValue(), ((AbstractCookingRecipe) recipeholder.value()).getExperience(), blockposition, entityplayer, itemstack, amount, ((AbstractCookingRecipe) recipeholder.value()).getResultItem(RegistryAccess.EMPTY)); // CraftBukkit // EMC - add result item
             });
         }
 
         return list;
     }
 
-    private static void createExperience(ServerLevel worldserver, Vec3 vec3d, int i, float f, BlockPos blockposition, net.minecraft.world.entity.player.Player entityhuman, ItemStack itemstack, int amount) { // CraftBukkit
+    private static void createExperience(ServerLevel worldserver, Vec3 vec3d, int i, float f, BlockPos blockposition, net.minecraft.world.entity.player.Player entityhuman, ItemStack itemstack, int amount, ItemStack result) { // CraftBukkit // EMC - add result item
         int j = Mth.floor((float) i * f);
         float f1 = Mth.frac((float) i * f);
 
@@ -667,8 +667,8 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
 
         // CraftBukkit start - fire FurnaceExtractEvent / BlockExpEvent
         BlockExpEvent event;
-        if (amount != 0) {
-            event = new FurnaceExtractEvent((Player) entityhuman.getBukkitEntity(), CraftBlock.at(worldserver, blockposition), CraftItemType.minecraftToBukkit(itemstack.getItem()), amount, j);
+        if (true || amount != 0) { // EMC
+            event = new FurnaceExtractEvent(entityhuman == null ? null : (Player) entityhuman.getBukkitEntity(), CraftBlock.at(worldserver, blockposition), itemstack == null ? null : CraftItemType.minecraftToBukkit(itemstack.getItem()), amount, j, CraftItemType.minecraftToBukkit(result.getItem()), i); // EMC - add == null for entityhuman and itemstack and add resultType/resultAmount
         } else {
             event = new BlockExpEvent(CraftBlock.at(worldserver, blockposition), j);
         }
