From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 17 Mar 2020 18:59:20 -0500
Subject: [PATCH] Properly remove empty entityTags from SpawnEggs


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
index 8077fab7cb85e8da027a08a72dfac4879b4daf43..79cd30bd60cb0e0e57a762e0e37b7e5ad334d9dd 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
@@ -42,6 +42,18 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
 
         if (tag.contains(ENTITY_TAG.NBT)) {
             this.entityTag = tag.getCompound(ENTITY_TAG.NBT);
+            // EMC start
+            if (this.entityTag != null && !this.entityTag.isEmpty()) {
+                String entityType = entityTag.getString(ENTITY_ID.NBT);
+                if (entityType != null) {
+                    this.spawnedType = EntityType.fromName(entityType);
+                    entityTag.remove(ENTITY_ID.NBT);
+                }
+                if (entityTag != null && entityTag.isEmpty()) {
+                    entityTag = null;
+                }
+            }
+            // EMC end
         }
     }
 
@@ -85,7 +97,15 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             // See if we can read a converted ID tag
             if (this.entityTag.contains(ENTITY_ID.NBT)) {
                 this.spawnedType = EntityType.fromName(new ResourceLocation(this.entityTag.getString(ENTITY_ID.NBT)).getPath());
+                // EMC start
+                if (spawnedType != null) {
+                    entityTag.remove(ENTITY_ID.NBT);
+                }
+            }
+            if (entityTag.isEmpty()) {
+                entityTag = null;
             }
+            // EMC end
         }
     }
 
@@ -232,7 +252,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             CraftMetaSpawnEgg that = (CraftMetaSpawnEgg) meta;
 
             return this.hasSpawnedType() ? that.hasSpawnedType() && this.spawnedType.equals(that.spawnedType) : !that.hasSpawnedType()
-                    && this.entityTag != null ? that.entityTag != null && this.entityTag.equals(that.entityTag) : this.entityTag == null;
+                    && (this.entityTag != null ? that.entityTag != null && this.entityTag.equals(that.entityTag) : that.entityTag == null); // EMC - fix logic
         }
         return true;
     }
@@ -287,6 +307,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             if (this.entityTag != null) {
                 // Remove ID tag as it is now in the material
                 this.entityTag.remove(ENTITY_ID.NBT);
+                if (entityTag.isEmpty()) { entityTag = null; } // EMC
             }
 
             return CraftLegacy.fromLegacy(new MaterialData(Material.LEGACY_MONSTER_EGG, (byte) this.spawnedType.getTypeId()));
