From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 7 Mar 2020 10:32:04 -0600
Subject: [PATCH] EntityAngerEvent

---
 .../minecraft/world/entity/IEntityAngerable.java    | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/src/main/java/net/minecraft/world/entity/IEntityAngerable.java b/src/main/java/net/minecraft/world/entity/IEntityAngerable.java
index 3e390ba135f358500ee742678ef2f69acb29fa40..7786f380b1315cf048b5202d73153c276c9fce77 100644
--- a/src/main/java/net/minecraft/world/entity/IEntityAngerable.java
+++ b/src/main/java/net/minecraft/world/entity/IEntityAngerable.java
@@ -1,5 +1,7 @@
 package net.minecraft.world.entity;
 
+import com.destroystokyo.paper.event.entity.EntityAngerEvent;
+
 import java.util.Objects;
 import java.util.UUID;
 import javax.annotation.Nullable;
@@ -61,8 +63,15 @@ public interface IEntityAngerable {
             this.pacify();
         } else {
             if (entityliving != null && !Objects.equals(uuid, entityliving.getUniqueID())) {
-                this.setAngerTarget(entityliving.getUniqueID());
-                this.anger();
+                // Paper start
+                if (this instanceof Entity) {
+                    EntityAngerEvent event = new EntityAngerEvent(((Entity)this).getBukkitEntity(), entityliving.getBukkitLivingEntity());
+                    if (event.callEvent()) {
+                        this.setAngerTarget(entityliving.getUniqueID());
+                        this.anger();
+                    }
+                }
+                // Paper end
             }
 
             if (this.getAnger() > 0 && (entityliving == null || entityliving.getEntityType() != EntityTypes.PLAYER || !flag)) {
