From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 6 Jun 2020 21:17:28 -0500
Subject: [PATCH] Raid total waves API


diff --git a/src/main/java/net/minecraft/world/entity/raid/Raid.java b/src/main/java/net/minecraft/world/entity/raid/Raid.java
index 7131226de05bc57830f7a68ba545ebfd19d33a59..cdc732556ab9251ca8a0bdd0e61ad0f2151a8d47 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raid.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raid.java
@@ -103,7 +103,8 @@ public class Raid {
     private int postRaidTicks;
     private int raidCooldownTicks;
     private final Random random;
-    public final int numGroups;
+    public int numGroups; // EMC - make non-final
+    public boolean firstTrigger = false; // EMC
     private Raid.RaidStatus status;
     private int celebrationTicks;
     private Optional<BlockPos> waveSpawnPos;
@@ -120,6 +121,7 @@ public class Raid {
         this.center = pos;
         this.numGroups = this.getNumGroups(world.getDifficulty());
         this.status = Raid.RaidStatus.ONGOING;
+        this.firstTrigger = true; // EMC
     }
 
     public Raid(ServerLevel world, CompoundTag nbt) {
diff --git a/src/main/java/net/minecraft/world/entity/raid/Raids.java b/src/main/java/net/minecraft/world/entity/raid/Raids.java
index e8dd1fdf1a3b697a2e040424df49ac6527b68312..4d9c0e644d7b7d881bae8947cabbae07c55ee963 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raids.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raids.java
@@ -132,6 +132,7 @@ public class Raids extends SavedData {
                         return null;
                     }
 
+                    raid.firstTrigger = false; // EMC
                     if (!this.raidMap.containsKey(raid.getId())) {
                         this.raidMap.put(raid.getId(), raid);
                     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftRaid.java b/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
index 7a5bfc724e42100b7582836ab8c30feed1a3d221..d18305d37926ee61fddbe20f6afcbe902e41b32f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftRaid.java
@@ -79,6 +79,13 @@ public final class CraftRaid implements Raid {
         return handle.numGroups;
     }
 
+    // EMC start
+    @Override
+    public void setTotalWaves(int total) {
+        Preconditions.checkArgument(1 <= total && total <= 7, "Total waves out of bounds 1 <= %s <= 7", total);
+        handle.numGroups = total;
+    }
+    // EMC end
     @Override
     public float getTotalHealth() {
         return this.handle.getHealthOfLivingRaiders();
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 7c1428e6376c77df5b705a59087040a6f1ad18a7..801f6353f4d90da028986f1b600b175a61d94c9b 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1832,7 +1832,7 @@ public class CraftEventFactory {
      * Raid events
      */
     public static boolean callRaidTriggerEvent(Raid raid, ServerPlayer player) {
-        RaidTriggerEvent event = new RaidTriggerEvent(new CraftRaid(raid), raid.getLevel().getWorld(), player.getBukkitEntity());
+        RaidTriggerEvent event = new RaidTriggerEvent(new CraftRaid(raid), raid.getLevel().getWorld(), player.getBukkitEntity(), raid.firstTrigger); // EMC - add raid.firstTrigger
         Bukkit.getPluginManager().callEvent(event);
         return !event.isCancelled();
     }
