From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 10 Jul 2021 23:31:59 -0500
Subject: [PATCH] VegetationPatchPlaceEvent


diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java b/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
index f431b659c89a299a778d12a2875b4bfb71418613..a9ef938b22fdfa7b573f67836aba590b28b6f68a 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
@@ -4,6 +4,8 @@ import com.mojang.serialization.Codec;
 import java.util.HashSet;
 import java.util.Set;
 import java.util.function.Predicate;
+
+import io.papermc.paper.util.MCUtil;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.util.RandomSource;
@@ -12,6 +14,8 @@ import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.configurations.VegetationPatchConfiguration;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.block.CraftBlock;
 
 public class VegetationPatchFeature extends Feature<VegetationPatchConfiguration> {
     public VegetationPatchFeature(Codec<VegetationPatchConfiguration> configCodec) {
@@ -67,6 +71,15 @@ public class VegetationPatchFeature extends Feature<VegetationPatchConfiguration
                     if (world.isEmptyBlock(mutableBlockPos) && blockState.isFaceSturdy(world, mutableBlockPos2, config.surface.getDirection().getOpposite())) {
                         int l = config.depth.sample(random) + (config.extraBottomBlockChance > 0.0F && random.nextFloat() < config.extraBottomBlockChance ? 1 : 0);
                         BlockPos blockPos = mutableBlockPos2.immutable();
+                        // EMC start
+                        if (Bukkit.isPrimaryThread()) {
+                            org.bukkit.block.Block origin = CraftBlock.at(world.getLevel(), pos);
+                            org.bukkit.block.Block place = CraftBlock.at(world.getLevel(), blockPos);
+                            if (!(new com.empireminecraft.customevents.VegetationPatchPlaceEvent(place, origin).callEvent())) {
+                                continue;
+                            }
+                        }
+                        // EMC end
                         boolean bl6 = this.placeGround(world, config, replaceable, random, mutableBlockPos2, l);
                         if (bl6) {
                             set.add(blockPos);
@@ -82,6 +95,15 @@ public class VegetationPatchFeature extends Feature<VegetationPatchConfiguration
     protected void distributeVegetation(FeaturePlaceContext<VegetationPatchConfiguration> context, WorldGenLevel world, VegetationPatchConfiguration config, RandomSource random, Set<BlockPos> positions, int radiusX, int radiusZ) {
         for(BlockPos blockPos : positions) {
             if (config.vegetationChance > 0.0F && random.nextFloat() < config.vegetationChance) {
+                // EMC start
+                if (Bukkit.isPrimaryThread()) {
+                    org.bukkit.block.Block origin = CraftBlock.at(context.level().getLevel(), context.origin());
+                    org.bukkit.block.Block place = CraftBlock.at(context.level().getLevel(), blockPos);
+                    if (!(new com.empireminecraft.customevents.VegetationPatchPlaceEvent(place, origin).callEvent())) {
+                        continue;
+                    }
+                }
+                // EMC end
                 this.placeVegetation(world, config, context.chunkGenerator(), random, blockPos);
             }
         }
