From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 17 Mar 2020 18:59:20 -0500
Subject: [PATCH] Properly handle entityTags on SpawnEggs


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
index 8077fab7cb85e8da027a08a72dfac4879b4daf43..b4c64b9d3d32a30543754ede8f18540cee483be0 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
@@ -1,5 +1,7 @@
 package org.bukkit.craftbukkit.inventory;
 
+import com.empireminecraft.DataConverters;
+import com.empireminecraft.DataConverters.LegacyType;
 import com.google.common.collect.ImmutableMap.Builder;
 import java.util.Map;
 import net.minecraft.nbt.CompoundTag;
@@ -42,6 +44,16 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
 
         if (tag.contains(ENTITY_TAG.NBT)) {
             this.entityTag = tag.getCompound(ENTITY_TAG.NBT);
+            // EMC start
+            if (!this.entityTag.isEmpty()) {
+                String entityType = entityTag.getString(ENTITY_ID.NBT);
+                this.spawnedType = EntityType.fromName(entityType);
+                entityTag.remove(ENTITY_ID.NBT);
+                if (entityTag != null && entityTag.isEmpty()) {
+                    entityTag = null;
+                }
+            }
+            // EMC end
         }
     }
 
@@ -77,7 +89,8 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             }
 
             // Tag still has some other data, lets try our luck with a conversion
-            if (!this.entityTag.isEmpty()) {
+            if (this.spawnedType == null && this.entityTag.contains(ENTITY_ID.NBT) && !this.entityTag.isEmpty()) { // EMC - only try converting when there is an id and spawnedType is not found already
+                entityTag = DataConverters.convert(LegacyType.ENTITY, this.entityTag); // EMC
                 // SPIGOT-4128: This is hopeless until we start versioning stacks. RIP data.
                 // entityTag = (NBTTagCompound) MinecraftServer.getServer().dataConverterManager.update(DataConverterTypes.ENTITY, new Dynamic(DynamicOpsNBT.a, entityTag), -1, CraftMagicNumbers.DATA_VERSION).getValue();
             }
@@ -85,7 +98,15 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             // See if we can read a converted ID tag
             if (this.entityTag.contains(ENTITY_ID.NBT)) {
                 this.spawnedType = EntityType.fromName(new ResourceLocation(this.entityTag.getString(ENTITY_ID.NBT)).getPath());
+                // EMC start
+                if (spawnedType != null) {
+                    entityTag.remove(ENTITY_ID.NBT);
+                }
+            }
+            if (entityTag.isEmpty()) {
+                entityTag = null;
             }
+            // EMC end
         }
     }
 
@@ -232,7 +253,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             CraftMetaSpawnEgg that = (CraftMetaSpawnEgg) meta;
 
             return this.hasSpawnedType() ? that.hasSpawnedType() && this.spawnedType.equals(that.spawnedType) : !that.hasSpawnedType()
-                    && this.entityTag != null ? that.entityTag != null && this.entityTag.equals(that.entityTag) : this.entityTag == null;
+                    && (this.entityTag != null ? that.entityTag != null && this.entityTag.equals(that.entityTag) : that.entityTag == null); // EMC - fix logic
         }
         return true;
     }
@@ -287,6 +308,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             if (this.entityTag != null) {
                 // Remove ID tag as it is now in the material
                 this.entityTag.remove(ENTITY_ID.NBT);
+                if (entityTag.isEmpty()) { entityTag = null; } // EMC
             }
 
             return CraftLegacy.fromLegacy(new MaterialData(Material.LEGACY_MONSTER_EGG, (byte) this.spawnedType.getTypeId()));
