From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 5 Jul 2021 22:01:39 -0500
Subject: [PATCH] DisperWaxBlockEvent


diff --git a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
index b34da7ee89779ae868b7b896e9aa89f7c6d26886..63e03f19ee1cde9ad238b3b8d0196012c3da6259 100644
--- a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -1134,6 +1134,13 @@ public interface DispenseItemBehavior {
                 Optional<BlockState> optional = HoneycombItem.getWaxed(iblockdata);
 
                 if (optional.isPresent()) {
+                    // EMC start
+                    org.bukkit.block.Block block = worldserver.getWorld().getBlockAt(pointer.getPos().getX(), pointer.getPos().getY(), pointer.getPos().getZ());
+                    CraftItemStack craftItem = CraftItemStack.asCraftMirror(stack);
+                    if (!(new com.empireminecraft.customevents.DispenserWaxBlockEvent(MCUtil.toBukkitBlock(worldserver, blockposition), block, craftItem).callEvent())) {
+                        return stack;
+                    }
+                    // EMC end
                     worldserver.setBlockAndUpdate(blockposition, (BlockState) optional.get());
                     worldserver.levelEvent(3003, blockposition, 0);
                     stack.shrink(1);
