From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 29 Jul 2021 02:07:15 -0500
Subject: [PATCH] Capture item drops during block placement


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index c658b950f47393fe1946db275506ea52a884ffc9..fd93e10ee9177f25b61ce1f719336ffc38aabc56 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -397,10 +397,19 @@ public final class ItemStack {
             } finally {
                 world.captureBlockStates = false;
             }
+            // Paper start
+            world.captureDrops = new java.util.ArrayList<>();
+            org.bukkit.block.Block bblock = CraftBlock.at(world, blockposition);
+            org.bukkit.block.BlockState state = bblock.getState();
+            // Paper end
             CompoundTag newData = this.getTagClone();
             int newCount = this.getCount();
             this.setCount(oldCount);
             this.setTagClone(oldData);
+            // Paper start
+            List<ItemEntity> itemsToDrop = world.captureDrops;
+            world.captureDrops = null;
+            // Paper end
             if (enuminteractionresult.consumesAction() && world.captureTreeGeneration && world.capturedBlockStates.size() > 0) {
                 world.captureTreeGeneration = false;
                 Location location = CraftLocation.toBukkit(blockposition, world.getWorld());
@@ -420,6 +429,7 @@ public final class ItemStack {
                 org.bukkit.Bukkit.getPluginManager().callEvent(fertilizeEvent);
 
                 if (!fertilizeEvent.isCancelled()) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDropItemEvent(bblock, state, (ServerPlayer) entityhuman, itemsToDrop); // Paper
                     // Change the stack to its new contents if it hasn't been tampered with.
                     if (this.getCount() == oldCount && Objects.equals(this.tag, oldData)) {
                         this.setTag(newData);
@@ -470,6 +480,7 @@ public final class ItemStack {
                     // Paper end - Don't resync blocks
                     SignItem.openSign = null; // SPIGOT-6758 - Reset on early return
                 } else {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDropItemEvent(bblock, state, (ServerPlayer) entityhuman, itemsToDrop); // Paper
                     // Change the stack to its new contents if it hasn't been tampered with.
                     if (this.getCount() == oldCount && Objects.equals(this.tag, oldData)) {
                         this.setTag(newData);
