From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 2 Nov 2024 18:41:19 -0500
Subject: [PATCH] Add ProjectileUseAmmoEvent


diff --git a/src/main/java/net/minecraft/world/item/ProjectileWeaponItem.java b/src/main/java/net/minecraft/world/item/ProjectileWeaponItem.java
index 32dd0b13a0819f597d8a93c6bc3a155781067544..c9b233c42de176755cec61eecfef87682c84c357 100644
--- a/src/main/java/net/minecraft/world/item/ProjectileWeaponItem.java
+++ b/src/main/java/net/minecraft/world/item/ProjectileWeaponItem.java
@@ -4,6 +4,7 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
+
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.ItemTags;
@@ -166,8 +167,11 @@ public abstract class ProjectileWeaponItem extends Item {
 
             i = 0;
         }
-
-        int j = i;
+        // EMC start
+        com.empireminecraft.customevents.ProjectileUseAmmoEvent event = new com.empireminecraft.customevents.ProjectileUseAmmoEvent(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(projectileStack), shooter.getBukkitEntity(), i);
+        event.callEvent();
+        int j = event.getAmmoCount();
+        // EMC end
 
         if (j > projectileStack.getCount()) {
             return ItemStack.EMPTY;
