From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 3 Nov 2024 10:16:01 -0600
Subject: [PATCH] Add ArrowKnockbackEvent


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index 9ca29b3d4bf8bca5f51f3644e12fcbec2cb5d35e..050cd725c5b6252bf4f5ee37fe0ff84f3fb53a60 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -520,14 +520,18 @@ public abstract class AbstractArrow extends Projectile {
                     ServerLevel worldserver = (ServerLevel) world;
 
                     f = EnchantmentHelper.modifyKnockback(worldserver, this.firedFromWeapon, target, source, 0.0F);
+                    //TODO:
                     break label18;
                 }
             }
 
             f = 0.0F;
         }
-
-        double d0 = (double) f;
+        // EMC start
+        com.empireminecraft.customevents.ArrowKnockbackEvent event = new com.empireminecraft.customevents.ArrowKnockbackEvent(this.getBukkitEntity(), target.getBukkitEntity(), f);
+        event.callEvent();
+        double d0 = event.getKnockback();
+        // EMC end
 
         if (d0 > 0.0D) {
             double d1 = Math.max(0.0D, 1.0D - target.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE));
