From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 7 Aug 2021 11:36:09 -0500
Subject: [PATCH] Change vanilla structure error to info log level


diff --git a/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java b/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java
index 92fd8e437b57c9df06c0287006604b85a1b7c048..876a8cf091c91127f00a4ce1e0b969b7a66ba23e 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java
@@ -58,7 +58,7 @@ public class PoolElementStructurePiece extends StructurePiece {
         DynamicOps<Tag> dynamicOps = RegistryOps.create(NbtOps.INSTANCE, context.registryAccess());
         this.element = StructurePoolElement.CODEC
             .parse(dynamicOps, nbt.getCompound("pool_element"))
-            .resultOrPartial(LOGGER::error)
+            .resultOrPartial(LOGGER::info) // EMC
             .orElseThrow(() -> new IllegalStateException("Invalid pool element found"));
         this.rotation = Rotation.valueOf(nbt.getString("rotation"));
         this.boundingBox = this.element.getBoundingBox(this.structureTemplateManager, this.position, this.rotation);
