From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 29 Dec 2020 01:32:45 -0600
Subject: [PATCH] Fake Player API


diff --git a/src/main/java/com/empireminecraft/api/CraftEAPI_Server.java b/src/main/java/com/empireminecraft/api/CraftEAPI_Server.java
new file mode 100644
index 0000000000000000000000000000000000000000..8dd7b8ddcddb2ad2f66a1e76eca6dfd5b5516fde
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/CraftEAPI_Server.java
@@ -0,0 +1,35 @@
+package com.empireminecraft.api;
+
+import com.destroystokyo.paper.profile.PlayerProfile;
+import com.google.common.collect.Maps;
+import net.kyori.adventure.text.Component;
+import org.bukkit.GameMode;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+
+public class CraftEAPI_Server implements EAPI_Server {
+    public final Map<UUID, FakePlayer> fakePlayerMap = Maps.newHashMap();
+
+    @Nullable
+    @Override
+    public FakePlayer getFakePlayer(@NotNull UUID uuid) {
+        return fakePlayerMap.get(uuid);
+    }
+
+    @NotNull
+    @Override
+    public FakePlayer createFakePlayer(@NotNull PlayerProfile profile, int ping, @NotNull GameMode gameMode, @NotNull Component playerListName) throws IllegalArgumentException {
+        return new CraftFakePlayer(profile, ping, gameMode, playerListName);
+    }
+
+    @NotNull
+    @Override
+    public Set<FakePlayer> getJoinedFakePlayers() {
+        return Set.copyOf(fakePlayerMap.values());
+    }
+
+}
diff --git a/src/main/java/com/empireminecraft/api/CraftEmpireAPI.java b/src/main/java/com/empireminecraft/api/CraftEmpireAPI.java
index b794befb4e5cfe8055c1b5745e52a0f01fc573d1..33757baf4510ab54dd6d73d4dbd1d46ed27b6ae8 100644
--- a/src/main/java/com/empireminecraft/api/CraftEmpireAPI.java
+++ b/src/main/java/com/empireminecraft/api/CraftEmpireAPI.java
@@ -36,6 +36,10 @@ public final class CraftEmpireAPI extends API {
         misc = new CraftEAPI_Misc();
         meta = new CraftEAPI_Meta();
         chat = new CraftEAPI_Chat();
+        server = new CraftEAPI_Server();
     }
 
+    public static CraftEAPI_Server getServer() {
+        return (CraftEAPI_Server) server;
+    }
 }
diff --git a/src/main/java/com/empireminecraft/api/CraftFakePlayer.java b/src/main/java/com/empireminecraft/api/CraftFakePlayer.java
new file mode 100644
index 0000000000000000000000000000000000000000..25f7506f7673eb336834323944474bcc66ad0d92
--- /dev/null
+++ b/src/main/java/com/empireminecraft/api/CraftFakePlayer.java
@@ -0,0 +1,150 @@
+package com.empireminecraft.api;
+
+import com.destroystokyo.paper.profile.PlayerProfile;
+import com.empireminecraft.customevents.FakePlayerJoinEvent;
+import com.empireminecraft.customevents.FakePlayerQuitEvent;
+import net.kyori.adventure.text.Component;
+import net.minecraft.network.protocol.game.ClientboundPlayerInfoPacket;
+import net.minecraft.network.protocol.game.ClientboundPlayerInfoPacket.Action;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.GameMode;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.UUID;
+
+public class CraftFakePlayer implements FakePlayer {
+    private final PlayerProfile profile;
+    // Use uuid field to ensure the PlayerProfile is not altered to change the UUID.
+    private final UUID uuid;
+    public int ping;
+    public GameMode gameMode;
+    public Component playerListName;
+    public boolean sentListPacket;
+
+    public CraftFakePlayer(@NotNull PlayerProfile profile, int ping, @NotNull GameMode gameMode, @NotNull Component playerListName) throws IllegalArgumentException {
+        this.profile = profile;
+        if (profile.getId() == null) {
+            throw new IllegalArgumentException("PlayerProfile does not contain a valid id");
+        }
+        this.uuid = profile.getId();
+        this.ping = ping;
+        this.gameMode = gameMode;
+        this.playerListName = playerListName;
+    }
+
+    @NotNull
+    @Override
+    public PlayerProfile getProfile() {
+        return profile;
+    }
+
+    @NotNull
+    @Override
+    public Component getPlayerListName() {
+        return playerListName;
+    }
+
+    @Override
+    public int getPing() {
+        return ping;
+    }
+
+    @NotNull
+    @Override
+    public GameMode getGameMode() {
+        return gameMode;
+    }
+
+    @Nullable
+    @Override
+    public UUID getUniqueId() {
+        return uuid;
+    }
+
+    @Override
+    public void join() {
+        UUID uuid = this.getUniqueId();
+        if (this.isJoined()) {
+            return;
+        }
+        new FakePlayerJoinEvent(this).callEvent();
+        CraftEmpireAPI.getServer().fakePlayerMap.put(uuid, this);
+
+        ClientboundPlayerInfoPacket packet = new ClientboundPlayerInfoPacket(Action.ADD_PLAYER, this);
+        for (ServerPlayer entityPlayer : MinecraftServer.getServer().getPlayerList().players) {
+            if (entityPlayer.getBukkitEntity().canSee(this)) {
+                entityPlayer.connection.send(packet);
+            }
+        }
+    }
+
+    @Override
+    public void quit() {
+        UUID uuid = this.getUniqueId();
+        if (!this.isJoined()) {
+            return;
+        }
+        new FakePlayerQuitEvent(this).callEvent();
+        CraftEmpireAPI.getServer().fakePlayerMap.remove(uuid);
+
+
+        ClientboundPlayerInfoPacket packet = new ClientboundPlayerInfoPacket(Action.REMOVE_PLAYER, this);
+        for (ServerPlayer entityPlayer : MinecraftServer.getServer().getPlayerList().players) {
+            if (entityPlayer.getBukkitEntity().canSee(this)) {
+                entityPlayer.connection.send(packet);
+            } else {
+                entityPlayer.getBukkitEntity().hiddenFakePlayers.remove(uuid);
+            }
+        }
+    }
+
+    @Override
+    public void setPing(int ping, boolean update) {
+        this.ping = ping;
+        if (update) {
+            this.sendUpdatePackets(Action.UPDATE_LATENCY);
+        }
+    }
+
+    @Override
+    public void setGameMode(@NotNull GameMode gameMode, boolean update) {
+        this.gameMode = gameMode;
+        if (update) {
+            this.sendUpdatePackets(Action.UPDATE_GAME_MODE);
+        }
+    }
+
+    @Override
+    public void setPlayerListName(@NotNull Component playerListName, boolean update) {
+        this.playerListName = playerListName;
+        if (update) {
+            this.sendUpdatePackets(Action.UPDATE_DISPLAY_NAME);
+        }
+    }
+
+    @Override
+    public void update() {
+        this.sendUpdatePackets(Action.UPDATE_LATENCY);
+        this.sendUpdatePackets(Action.UPDATE_GAME_MODE);
+        this.sendUpdatePackets(Action.UPDATE_DISPLAY_NAME);
+    }
+
+    private void sendUpdatePackets(@NotNull Action action) {
+        if (this.isJoined()) {
+            ClientboundPlayerInfoPacket packet = new ClientboundPlayerInfoPacket(action, this);
+            for (ServerPlayer entityPlayer : MinecraftServer.getServer().getPlayerList().players) {
+                if (entityPlayer.getBukkitEntity().canSee(this)) {
+                    entityPlayer.connection.send(packet);
+                }
+            }
+        }
+    }
+
+    @Override
+    public boolean isJoined() {
+        return CraftEmpireAPI.getServer().fakePlayerMap.containsKey(this.getUniqueId());
+    }
+
+}
diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerInfoPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerInfoPacket.java
index 420cc7ab5f7245f31dd5b333eacf4e306fc6e7c1..298f860fb7535637637a05cfaec7e4d918723f5a 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerInfoPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundPlayerInfoPacket.java
@@ -1,5 +1,6 @@
 package net.minecraft.network.protocol.game;
 
+import com.empireminecraft.api.FakePlayer;
 import com.google.common.base.MoreObjects;
 import com.google.common.collect.Lists;
 import com.mojang.authlib.GameProfile;
@@ -8,6 +9,8 @@ import com.mojang.authlib.properties.PropertyMap;
 import java.util.Collection;
 import java.util.List;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.adventure.PaperAdventure;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.protocol.Packet;
@@ -18,6 +21,21 @@ public class ClientboundPlayerInfoPacket implements Packet<ClientGamePacketListe
     private final ClientboundPlayerInfoPacket.Action action;
     private final List<ClientboundPlayerInfoPacket.PlayerUpdate> entries;
 
+    // EMC start
+    public ClientboundPlayerInfoPacket(ClientboundPlayerInfoPacket.Action enumplayerinfoaction, FakePlayer... fakePlayers) {
+        this.action = enumplayerinfoaction;
+        this.entries = Lists.newArrayListWithCapacity(fakePlayers.length);
+        for (FakePlayer fakePlayer : fakePlayers) {
+            if (fakePlayer.getUniqueId() != null) {
+                this.entries.add(new ClientboundPlayerInfoPacket.PlayerUpdate(
+                    com.destroystokyo.paper.profile.CraftPlayerProfile.asAuthlib(fakePlayer.getProfile()),
+                    fakePlayer.getPing(),
+                    GameType.byId(fakePlayer.getGameMode().getValue()),
+                    PaperAdventure.asVanilla(fakePlayer.getPlayerListName())));
+            }
+        }
+    }
+    // EMC end
     public ClientboundPlayerInfoPacket(ClientboundPlayerInfoPacket.Action action, ServerPlayer... players) {
         this.action = action;
         this.entries = Lists.newArrayListWithCapacity(players.length);
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index e945d7c92dc65019b3153cdc7498d59eb0fa78ba..31ccda37a5c1926ea834a19d992578e17335749b 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1,6 +1,8 @@
 package net.minecraft.server.players;
 
 import co.aikar.timings.MinecraftTimings;
+import com.empireminecraft.api.CraftEmpireAPI;
+import com.empireminecraft.api.FakePlayer;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
@@ -377,6 +379,13 @@ public abstract class PlayerList {
         }
         // CraftBukkit end
 
+        // EMC start
+        for (FakePlayer fakePlayer : CraftEmpireAPI.getServer().fakePlayerMap.values()) {
+            if (player.getBukkitEntity().canSee(fakePlayer)) {
+                player.connection.send(new ClientboundPlayerInfoPacket(ClientboundPlayerInfoPacket.Action.ADD_PLAYER, fakePlayer));
+            }
+        }
+        // EMC end
         // CraftBukkit start - sendAll above replaced with this loop
         ClientboundPlayerInfoPacket packet = new ClientboundPlayerInfoPacket(ClientboundPlayerInfoPacket.Action.ADD_PLAYER, player);
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 01c8b6cf2dfb93d21856587e275c7e2c45137d08..99e4b605111ce4397569b63fe1a1f90f14d1e132 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1,6 +1,8 @@
 package org.bukkit.craftbukkit.entity;
 
 import com.destroystokyo.paper.Title;
+import com.empireminecraft.api.CraftEmpireAPI;
+import com.empireminecraft.api.FakePlayer;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.io.BaseEncoding;
@@ -1605,6 +1607,73 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         return !this.hiddenEntities.containsKey(entity.getUniqueId());
     }
 
+    // EMC start
+    public final Map<UUID, Set<WeakReference<Plugin>>> hiddenFakePlayers = new HashMap<>();
+    @Override
+    public boolean canSee(FakePlayer fakePlayer) {
+        return !hiddenFakePlayers.containsKey(fakePlayer.getUniqueId());
+    }
+    @Override
+    public void hideFakePlayer(Plugin plugin, FakePlayer fakePlayer) {
+        if (getHandle().connection == null) {
+            return;
+        }
+        UUID fakeUUID = fakePlayer.getUniqueId();
+        if  (fakeUUID.equals(this.getUniqueId())) {
+            return;
+        }
+
+        Set<WeakReference<Plugin>> hidingPlugins = hiddenFakePlayers.get(fakeUUID);
+        if (hidingPlugins != null) {
+            // Some plugins are already hiding the player. Just mark that this
+            // plugin wants the player hidden too and end.
+            hidingPlugins.add(getPluginWeakReference(plugin));
+            return;
+        }
+        hidingPlugins = new HashSet<>();
+        hidingPlugins.add(getPluginWeakReference(plugin));
+        hiddenFakePlayers.put(fakeUUID, hidingPlugins);
+
+        if (CraftEmpireAPI.getServer().fakePlayerMap.containsKey(fakeUUID)) {
+            getHandle().connection.send(new ClientboundPlayerInfoPacket(ClientboundPlayerInfoPacket.Action.REMOVE_PLAYER, fakePlayer));
+        }
+    }
+    @Override
+    public void showFakePlayer(Plugin plugin, FakePlayer fakePlayer) {
+        if (getHandle().connection == null) {
+            return;
+        }
+        UUID fakeUUID = fakePlayer.getUniqueId();
+        if  (fakeUUID.equals(this.getUniqueId())) {
+            return;
+        }
+
+        Set<WeakReference<Plugin>> hidingPlugins = hiddenFakePlayers.get(fakeUUID);
+        if (hidingPlugins == null) {
+            return; // Player isn't hidden
+        }
+        hidingPlugins.remove(getPluginWeakReference(plugin));
+        if (!hidingPlugins.isEmpty()) {
+            return; // Some other plugins still want the player hidden
+        }
+        hiddenFakePlayers.remove(fakeUUID);
+        if (CraftEmpireAPI.getServer().fakePlayerMap.containsKey(fakeUUID)) {
+            getHandle().connection.send(new ClientboundPlayerInfoPacket(ClientboundPlayerInfoPacket.Action.ADD_PLAYER, fakePlayer));
+        }
+    }
+    @Override
+    public Set<FakePlayer> getHiddenFakePlayers() {
+        Set<FakePlayer> ret = new HashSet<>();
+        for (UUID uuid : hiddenFakePlayers.keySet()) {
+            FakePlayer fakePlayer = CraftEmpireAPI.server.getFakePlayer(uuid);
+            if (fakePlayer != null) {
+                ret.add(fakePlayer);
+            }
+        }
+
+        return java.util.Collections.unmodifiableSet(ret);
+    }
+    // EMC end
     @Override
     public Map<String, Object> serialize() {
         Map<String, Object> result = new LinkedHashMap<String, Object>();
