From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 2 Jun 2021 09:16:37 -0500
Subject: [PATCH] Fire FurnaceExtractEvent for furnace breakings with null
 players

temp

diff --git a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index a3f073066f6e2eea8964461ad2b0409ade202f35..4cddfed1142ed2cea512dde6d0e6854aef036279 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -634,14 +634,14 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
             worldserver.getRecipeManager().byKey((ResourceLocation) entry.getKey()).ifPresent((irecipe) -> {
                 if (!(irecipe instanceof AbstractCookingRecipe)) return; // Paper - don't process non-cooking recipes
                 list.add(irecipe);
-                AbstractFurnaceBlockEntity.createExperience(worldserver, vec3d, entry.getIntValue(), ((AbstractCookingRecipe) irecipe).getExperience(), blockposition, entityplayer, itemstack, amount); // CraftBukkit
+                AbstractFurnaceBlockEntity.createExperience(worldserver, vec3d, entry.getIntValue(), ((AbstractCookingRecipe) irecipe).getExperience(), blockposition, entityplayer, itemstack, amount, irecipe.getResultItem(RegistryAccess.EMPTY)); // CraftBukkit // EMC - add result item
             });
         }
 
         return list;
     }
 
-    private static void createExperience(ServerLevel worldserver, Vec3 vec3d, int i, float f, BlockPos blockposition, net.minecraft.world.entity.player.Player entityhuman, ItemStack itemstack, int amount) { // CraftBukkit
+    private static void createExperience(ServerLevel worldserver, Vec3 vec3d, int i, float f, BlockPos blockposition, net.minecraft.world.entity.player.Player entityhuman, ItemStack itemstack, int amount, ItemStack result) { // CraftBukkit // EMC - add result item
         int j = Mth.floor((float) i * f);
         float f1 = Mth.frac((float) i * f);
 
@@ -651,8 +651,8 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
 
         // CraftBukkit start - fire FurnaceExtractEvent / BlockExpEvent
         BlockExpEvent event;
-        if (amount != 0) {
-            event = new FurnaceExtractEvent((Player) entityhuman.getBukkitEntity(), CraftBlock.at(worldserver, blockposition), org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(itemstack.getItem()), amount, j);
+        if (true || amount != 0) { // EMC
+            event = new FurnaceExtractEvent(entityhuman == null ? null : (Player) entityhuman.getBukkitEntity(), CraftBlock.at(worldserver, blockposition), itemstack == null ? null : org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(itemstack.getItem()), amount, j, org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(result.getItem()), i); // EMC - add == null for entityhuman and itemstack and add resultType/resultAmount
         } else {
             event = new BlockExpEvent(CraftBlock.at(worldserver, blockposition), j);
         }
