From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 15 Dec 2024 04:49:05 -0600
Subject: [PATCH] Suppress some log warnings


diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java b/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java
index 69992ebc999ea3ff9e47e4e049bcc514c01150ca..08fa5e6c3e57c68181ace91f6565bbac6fa5d2dd 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -134,7 +134,7 @@ public class AttributeMap {
                     if (attributeInstance != null) {
                         attributeInstance.load(compoundTag);
                     }
-                }, () -> LOGGER.warn("Ignoring unknown attribute '{}'", resourceLocation));
+                }, () -> LOGGER.debug("Ignoring unknown attribute '{}'", resourceLocation)); // EMC
             } else {
                 LOGGER.warn("Ignoring malformed attribute '{}'", string);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/ZombieVillager.java b/src/main/java/net/minecraft/world/entity/monster/ZombieVillager.java
index e0dabbf6d7a87b8722769c78ef0d2ba4353ed2cb..cdfb154121be6b4e080e574f49656b105ad75755 100644
--- a/src/main/java/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/src/main/java/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -132,7 +132,7 @@ public class ZombieVillager extends Zombie implements VillagerDataHolder {
             Logger logger1 = ZombieVillager.LOGGER;
 
             Objects.requireNonNull(logger1);
-            dataresult1.resultOrPartial(Util.prefix("Failed to load offers: ", logger1::warn)).ifPresent((merchantrecipelist) -> {
+            dataresult1.resultOrPartial(Util.prefix("Failed to load offers: ", logger1::debug)).ifPresent((merchantrecipelist) -> { // EMC
                 this.tradeOffers = merchantrecipelist;
             });
         }
diff --git a/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java b/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java
index d28ebcae036168dd65a5f3236d12ee416308c23f..75078f4d114aa801aa29ca37800e6288749cdb73 100644
--- a/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -216,7 +216,7 @@ public abstract class AbstractVillager extends AgeableMob implements InventoryCa
             Logger logger = AbstractVillager.LOGGER;
 
             Objects.requireNonNull(logger);
-            dataresult.resultOrPartial(Util.prefix("Failed to load offers: ", logger::warn)).ifPresent((merchantrecipelist) -> {
+            dataresult.resultOrPartial(Util.prefix("Failed to load offers: ", logger::debug)).ifPresent((merchantrecipelist) -> { // EMC
                 this.offers = merchantrecipelist;
             });
         }
