From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 29 Jul 2021 02:07:15 -0500
Subject: [PATCH] Capture item drops during block placement


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 8ab6f81632d848da222030e507759e195d4da3ab..2b02f1cd2215fd863aefa4fd277745388eea8346 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -351,6 +351,11 @@ public final class ItemStack {
                     world.captureTreeGeneration = true;
                 }
             }
+            // Paper start
+            world.captureDrops = new java.util.ArrayList<>();
+            org.bukkit.block.Block bblock = CraftBlock.at(world, blockposition);
+            org.bukkit.block.BlockState state = bblock.getState();
+            // Paper end
             Item item = this.getItem();
             InteractionResult enuminteractionresult = item.useOn(itemactioncontext);
             CompoundTag newData = this.getTagClone();
@@ -358,6 +363,10 @@ public final class ItemStack {
             this.setCount(oldCount);
             this.setTagClone(oldData);
             world.captureBlockStates = false;
+            // Paper start
+            List<ItemEntity> itemsToDrop = world.captureDrops;
+            world.captureDrops = null;
+            // Paper end
             if (enuminteractionresult.consumesAction() && world.captureTreeGeneration && world.capturedBlockStates.size() > 0) {
                 world.captureTreeGeneration = false;
                 Location location = new Location(world.getWorld(), blockposition.getX(), blockposition.getY(), blockposition.getZ());
@@ -377,6 +386,7 @@ public final class ItemStack {
                 org.bukkit.Bukkit.getPluginManager().callEvent(fertilizeEvent);
 
                 if (!fertilizeEvent.isCancelled()) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDropItemEvent(bblock, state, (ServerPlayer) entityhuman, itemsToDrop); // Paper
                     // Change the stack to its new contents if it hasn't been tampered with.
                     if (this.getCount() == oldCount && Objects.equals(this.tag, oldData)) {
                         this.setTag(newData);
@@ -421,6 +431,7 @@ public final class ItemStack {
                     }
                     SignItem.openSign = null; // SPIGOT-6758 - Reset on early return
                 } else {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDropItemEvent(bblock, state, (ServerPlayer) entityhuman, itemsToDrop); // Paper
                     // Change the stack to its new contents if it hasn't been tampered with.
                     if (this.getCount() == oldCount && Objects.equals(this.tag, oldData)) {
                         this.setTag(newData);
