From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 5 Jul 2021 22:01:39 -0500
Subject: [PATCH] DisperWaxBlockEvent


diff --git a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
index 54642b6236d257600a96c0613b01d7b7a17a7e35..0202ed2e4bdbc681593ad51feef5a77ae525642a 100644
--- a/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/src/main/java/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -1106,6 +1106,13 @@ public interface DispenseItemBehavior {
                 Optional<BlockState> optional = HoneycombItem.getWaxed(iblockdata);
 
                 if (optional.isPresent()) {
+                    // EMC start
+                    org.bukkit.block.Block block = worldserver.getWorld().getBlockAt(pointer.getPos().getX(), pointer.getPos().getY(), pointer.getPos().getZ());
+                    CraftItemStack craftItem = CraftItemStack.asCraftMirror(stack);
+                    if (!(new com.empireminecraft.customevents.DispenserWaxBlockEvent(MCUtil.toBukkitBlock(worldserver, blockposition), block, craftItem).callEvent())) {
+                        return stack;
+                    }
+                    // EMC end
                     worldserver.setBlockAndUpdate(blockposition, (BlockState) optional.get());
                     worldserver.levelEvent(3003, blockposition, 0);
                     stack.shrink(1);
