From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 31 Mar 2023 23:22:24 -0500
Subject: [PATCH] Revert mysql dependency version for now...


diff --git a/build.gradle.kts b/build.gradle.kts
index 861e687e5c0967a5eb9add1d20ac9607238cd653..faf9c277ea0aec4221f472198a8fd3e50fa6bed0 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -42,7 +42,7 @@ dependencies {
     implementation("commons-lang:commons-lang:2.6")
     implementation("net.fabricmc:mapping-io:0.5.0") // Paper - needed to read mappings for stacktrace deobfuscation
     runtimeOnly("org.xerial:sqlite-jdbc:3.42.0.1")
-    runtimeOnly("com.mysql:mysql-connector-j:8.2.0")
+    runtimeOnly("mysql:mysql-connector-java:8.0.29") // EMC
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
     // Paper start - Use Velocity cipher
     implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
