From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 5 Jul 2021 22:48:17 -0500
Subject: [PATCH] Add PlayerCandleCakeEvent


diff --git a/src/main/java/net/minecraft/world/level/block/CakeBlock.java b/src/main/java/net/minecraft/world/level/block/CakeBlock.java
index c95fe722c807b81b6300f27fe25e042e754b8e1a..61ae2ad2bba20cb28e3c73e45fd86cf7478a26c4 100644
--- a/src/main/java/net/minecraft/world/level/block/CakeBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CakeBlock.java
@@ -2,6 +2,7 @@ package net.minecraft.world.level.block;
 
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.server.MCUtil;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.stats.Stats;
@@ -54,6 +55,7 @@ public class CakeBlock extends Block {
             Block block = Block.byItem(item);
 
             if (block instanceof CandleBlock) {
+                new com.empireminecraft.customevents.PlayerCandleCakeEvent((org.bukkit.entity.Player) player.getBukkitEntity(), MCUtil.toLocation(world, pos), itemstack.getBukkitStack()).callEvent(); // EMC
                 if (!player.isCreative()) {
                     itemstack.shrink(1);
                 }
