From 8e930b52438f407ef468358ccb1ec415d63614d7 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 29 Dec 2016 17:37:46 -0500
Subject: [PATCH] Convert Horse Eggs with EMC Metadata

---
 .../craftbukkit/inventory/CraftMetaSpawnEgg.java   | 56 +++++++++++++++++-----
 1 file changed, 45 insertions(+), 11 deletions(-)

diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
index 4fa0f1f14..70f929cb5 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
@@ -68,23 +68,57 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
         if (type == null) {
             type = 0;
         }
+        // EMC start
+        String result;
         switch (type) {
-            case 0:
-            default:
-                return "Horse";
-
             case 1:
-                return "Donkey";
-
+                result = "Donkey";
+                break;
             case 2:
-                return "Mule";
-
+                result = "Mule";
+                break;
             case 3:
-                return "ZombieHorse";
-
+                result = "ZombieHorse";
+                break;
             case 4:
-                return "SkeletonHorse";
+                result = "SkeletonHorse";
+                break;
+            case 0:
+            default:
+                result = "Horse";
+                break;
+        }
+        if (this.hasLore()) {
+            java.util.List<String> lore = this.getLore();
+                if (31< lore.size()) {
+                    for (int i = 31; i + 1 < lore.size(); i += 2) {
+                        if ("var".equalsIgnoreCase(lore.get(i))) {
+                            String variant = lore.get(i + 1);
+                            if (variant != null) {
+                                switch(variant) {
+                                    case "DONKEY":
+                                        result = "Donkey";
+                                        break;
+                                    case "MULE":
+                                        result = "Mule";
+                                        break;
+                                    case "UNDEAD_HORSE":
+                                        result = "ZombieHorse";
+                                        break;
+                                    case "SKELETON_HORSE":
+                                        result = "SkeletonHorse";
+                                        break;
+                                    case "HORSE":
+                                    default:
+                                        result = "Horse";
+                            }
+                        }
+                    }
+                }
+            }
         }
+        return result;
+        // EMC end
     }
     private String convert(String id) {
         if (id == null) {
-- 
2.11.0

