From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 14 Mar 2022 10:10:03 -0500
Subject: [PATCH] Adjust maximum horse speed modifiers

This change is for EMC-specific motivation

diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
index 9b35a50eb99551f28d45fd5114e0401b4c54646e..31f4e4a2dacbf537567490b97e1ba3d3489a28c7 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -1095,7 +1095,7 @@ public abstract class AbstractHorse extends Animal implements ContainerListener,
     }
 
     protected static double generateSpeed(DoubleSupplier randomDoubleGetter) {
-        return (0.44999998807907104D + randomDoubleGetter.getAsDouble() * 0.3D + randomDoubleGetter.getAsDouble() * 0.3D + randomDoubleGetter.getAsDouble() * 0.3D) * 0.25D;
+        return (0.44999998807907104D + randomDoubleGetter.getAsDouble() * 0.315D + randomDoubleGetter.getAsDouble() * 0.315D + randomDoubleGetter.getAsDouble() * 0.315D) * 0.25D; // EMC - increase factor multiplier by 0.015
     }
 
     @Override
