From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 14 Mar 2022 10:10:03 -0500
Subject: [PATCH] Adjust maximum horse speed modifiers

This change is for EMC-specific motivation

diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
index 47cd69f91bbc2e2be9ec970674adc522e21593c8..87604afd5858b706810299e27b1916071fad3449 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/AbstractHorse.java
@@ -1099,7 +1099,7 @@ public abstract class AbstractHorse extends Animal implements ContainerListener,
     }
 
     protected static double generateSpeed(DoubleSupplier randomDoubleGetter) {
-        return (0.44999998807907104D + randomDoubleGetter.getAsDouble() * 0.3D + randomDoubleGetter.getAsDouble() * 0.3D + randomDoubleGetter.getAsDouble() * 0.3D) * 0.25D;
+        return (0.44999998807907104D + randomDoubleGetter.getAsDouble() * 0.315D + randomDoubleGetter.getAsDouble() * 0.315D + randomDoubleGetter.getAsDouble() * 0.315D) * 0.25D; // EMC - increase factor multiplier by 0.015
     }
 
     @Override
