From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 10 Jul 2021 23:31:59 -0500
Subject: [PATCH] VegetationPatchPlaceEvent


diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java b/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
index 0a89431e524c9cf999acb21f3d1018be01e585de..2b661e712a714ee176a24790bda62333dd2aad80 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
@@ -7,6 +7,7 @@ import java.util.Set;
 import java.util.function.Predicate;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.server.MCUtil;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.Tag;
 import net.minecraft.world.level.WorldGenLevel;
@@ -15,6 +16,7 @@ import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.configurations.VegetationPatchConfiguration;
+import org.bukkit.Bukkit;
 
 public class VegetationPatchFeature extends Feature<VegetationPatchConfiguration> {
     public VegetationPatchFeature(Codec<VegetationPatchConfiguration> configCodec) {
@@ -68,6 +70,15 @@ public class VegetationPatchFeature extends Feature<VegetationPatchConfiguration
                     if (world.isEmptyBlock(mutableBlockPos) && blockState.isFaceSturdy(world, mutableBlockPos2, config.surface.getDirection().getOpposite())) {
                         int l = config.depth.sample(random) + (config.extraBottomBlockChance > 0.0F && random.nextFloat() < config.extraBottomBlockChance ? 1 : 0);
                         BlockPos blockPos = mutableBlockPos2.immutable();
+                        // EMC start
+                        if (Bukkit.isPrimaryThread()) {
+                            org.bukkit.block.Block origin = MCUtil.toBukkitBlock(world.getLevel(), pos);
+                            org.bukkit.block.Block place = MCUtil.toBukkitBlock(world.getLevel(), blockPos);
+                            if (!(new com.empireminecraft.customevents.VegetationPatchPlaceEvent(place, origin).callEvent())) {
+                                continue;
+                            }
+                        }
+                        // EMC end
                         boolean bl6 = this.placeGround(world, config, replaceable, random, mutableBlockPos2, l);
                         if (bl6) {
                             set.add(blockPos);
@@ -83,6 +94,15 @@ public class VegetationPatchFeature extends Feature<VegetationPatchConfiguration
     protected void distributeVegetation(FeaturePlaceContext<VegetationPatchConfiguration> context, WorldGenLevel world, VegetationPatchConfiguration config, Random random, Set<BlockPos> positions, int radiusX, int radiusZ) {
         for(BlockPos blockPos : positions) {
             if (config.vegetationChance > 0.0F && random.nextFloat() < config.vegetationChance) {
+                // EMC start
+                if (Bukkit.isPrimaryThread()) {
+                    org.bukkit.block.Block origin = MCUtil.toBukkitBlock(context.level().getLevel(), context.origin());
+                    org.bukkit.block.Block place = MCUtil.toBukkitBlock(context.level().getLevel(), blockPos);
+                    if (!(new com.empireminecraft.customevents.VegetationPatchPlaceEvent(place, origin).callEvent())) {
+                        continue;
+                    }
+                }
+                // EMC end
                 this.placeVegetation(world, config, context.chunkGenerator(), random, blockPos);
             }
         }
