From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 5 Jul 2021 22:48:17 -0500
Subject: [PATCH] Add PlayerCandleCakeEvent


diff --git a/src/main/java/net/minecraft/world/level/block/CakeBlock.java b/src/main/java/net/minecraft/world/level/block/CakeBlock.java
index d2c78a0daaabc35f72a746ad0d60472383759b6f..07292e6dc2836e99620a4f4639fce4f3bb756e4e 100644
--- a/src/main/java/net/minecraft/world/level/block/CakeBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CakeBlock.java
@@ -2,6 +2,7 @@ package net.minecraft.world.level.block;
 
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.server.MCUtil;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.stats.Stats;
@@ -56,6 +57,7 @@ public class CakeBlock extends Block {
             Block block = Block.byItem(item);
 
             if (block instanceof CandleBlock) {
+                new com.empireminecraft.customevents.PlayerCandleCakeEvent((org.bukkit.entity.Player) player.getBukkitEntity(), MCUtil.toLocation(world, pos), itemstack.getBukkitStack()).callEvent(); // EMC
                 if (!player.isCreative()) {
                     itemstack.shrink(1);
                 }
