From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 2 Jan 2025 10:24:45 -0600
Subject: [PATCH] Improve loading old block entity names


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
index cffeaff83ec5037c21ff33ac32287017c1263b67..14befb3b020de18a15445d0288fa38f7c450776e 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.level.block.entity;
 
+import com.google.gson.JsonParseException;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Codec;
 import java.util.HashSet;
@@ -378,7 +379,18 @@ public abstract class BlockEntity {
     @Nullable
     public static Component parseCustomNameSafe(String json, HolderLookup.Provider registryLookup) {
         try {
-            return Component.Serializer.fromJson(json, registryLookup);
+            // EMC start
+            if (json.isEmpty()) {
+                return null;
+            }
+            if (json.startsWith("{") && json.endsWith("}")) {
+                try {
+                    return Component.Serializer.fromJson(json, registryLookup);
+                } catch (JsonParseException ignored) {
+                }
+            }
+            return Component.Serializer.fromJsonLenient(json, registryLookup);
+            // EMC end
         } catch (Exception exception) {
             BlockEntity.LOGGER.warn("Failed to parse custom name from string '{}', discarding", json, exception);
             return null;
