From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 11 Jun 2022 20:16:27 -0500
Subject: [PATCH] getInstrumentId implementation - temporary patch to get goat
 horn type


diff --git a/src/main/java/com/empireminecraft/api/CraftEAPI_Misc.java b/src/main/java/com/empireminecraft/api/CraftEAPI_Misc.java
index 453102c95350f2955b9c672fdb2762b5e21d776b..4beb57ed0dc5b79b7ad87f0843d6c0aa9ca62ade 100644
--- a/src/main/java/com/empireminecraft/api/CraftEAPI_Misc.java
+++ b/src/main/java/com/empireminecraft/api/CraftEAPI_Misc.java
@@ -30,8 +30,7 @@ import io.papermc.paper.util.MCUtil;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.ListTag;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.item.MapItem;
+import net.minecraft.world.item.*;
 import net.minecraft.world.level.saveddata.maps.MapDecoration;
 import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
 import org.bukkit.Location;
@@ -39,6 +38,7 @@ import org.bukkit.StructureType;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.craftbukkit.util.CraftMagicNumbers.NBT;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -88,6 +88,25 @@ public class CraftEAPI_Misc implements EAPI_Misc {
         return items;
     }
 
+    @Override
+    @Deprecated
+    public @Nullable String getInstrumentId(@NotNull org.bukkit.inventory.ItemStack itemStack) {
+        final CraftItemStack craft;
+        if (itemStack instanceof CraftItemStack) {
+            craft = ((CraftItemStack) itemStack);
+        } else {
+            craft = CraftItemStack.asCraftCopy(itemStack);
+        }
+        if (!(craft.handle.getItem() instanceof InstrumentItem)) {
+            return null;
+        }
+        CompoundTag compoundTag = craft.handle.getTag();
+        if (compoundTag != null) {
+            return compoundTag.getString("instrument");
+        }
+        return null;
+    }
+
     private static ListTag getNBTItems(@NotNull org.bukkit.inventory.ItemStack shulker) {
         if (!MaterialTags.SHULKER_BOXES.isTagged(shulker)) {
             throw new IllegalArgumentException("Not a shulker box");
