From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 16 Jun 2022 09:13:17 -0500
Subject: [PATCH] EntityTriggerSculkCatalystEvent


diff --git a/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
index 6c378e3485fef4e8b8906c33a45a23e6f77e9cdd..d6b7fca44fbc3ae85956b05543e6c357f029df3c 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -4,6 +4,7 @@ import com.google.common.annotations.VisibleForTesting;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.core.Position;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.server.level.ServerLevel;
@@ -94,6 +95,12 @@ public class SculkCatalystBlockEntity extends BlockEntity implements GameEventLi
 
                 if (entity instanceof LivingEntity) {
                     LivingEntity entityliving = (LivingEntity) entity;
+                    // EMC start
+                    Position pos = this.positionSource.getPosition(world).orElse(null);
+                    if (pos != null && !(new com.empireminecraft.customevents.EntityTriggerSculkCatalystEvent(entityliving.getBukkitLivingEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, BlockPos.containing(pos))).callEvent())) {
+                        return false;
+                    }
+                    // EMC end
 
                     if (!entityliving.wasExperienceConsumed()) {
                         int i = entityliving.getExperienceReward();
