From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 16 Jun 2022 09:13:17 -0500
Subject: [PATCH] EntityTriggerSculkCatalystEvent


diff --git a/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
index 163e63e3c538c7c1c75ed634896db9d8c00416f3..08480163d681531177b8f403c8052e2ec2c66bad 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -53,6 +53,11 @@ public class SculkCatalystBlockEntity extends BlockEntity implements GameEventLi
 
             if (entity instanceof LivingEntity) {
                 LivingEntity entityliving = (LivingEntity) entity;
+                    // EMC start
+                    if (!(new com.empireminecraft.customevents.EntityTriggerSculkCatalystEvent(entityliving.getBukkitLivingEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, this.getBlockPos())).callEvent())) {
+                        return false;
+                    }
+                    // EMC end
 
                 if (!entityliving.wasExperienceConsumed()) {
                     int i = entityliving.getExperienceReward();
