From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 7 Aug 2021 11:36:09 -0500
Subject: [PATCH] Change vanilla structure error to info log level


diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/StructureFeature.java b/src/main/java/net/minecraft/world/level/levelgen/feature/StructureFeature.java
index 461464a4208bbb09800d87bcdb54978758863c7b..3dc3f4deb4393cb802ff3e75dfaec7f34c89b0ac 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/feature/StructureFeature.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/feature/StructureFeature.java
@@ -133,7 +133,7 @@ public class StructureFeature<C extends FeatureConfiguration> {
 
                     return new StructureStart<>(structureFeature, chunkPos, i, piecesContainer);
                 } catch (Exception var10) {
-                    LOGGER.error("Failed Start with id {}", string, var10);
+                    LOGGER.info("Failed Start with id {}", string, var10); // EMC
                     return null;
                 }
             }
diff --git a/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java b/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java
index 6c87012d91639687e691016d9413f3a6841827ee..9332bcf16d00bca2b2bdb8d77c5d844713ef7dc1 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java
@@ -48,7 +48,7 @@ public class PoolElementStructurePiece extends StructurePiece {
         this.position = new BlockPos(nbt.getInt("PosX"), nbt.getInt("PosY"), nbt.getInt("PosZ"));
         this.groundLevelDelta = nbt.getInt("ground_level_delta");
         RegistryReadOps<Tag> registryReadOps = RegistryReadOps.create(NbtOps.INSTANCE, context.resourceManager(), context.registryAccess());
-        this.element = StructurePoolElement.CODEC.parse(registryReadOps, nbt.getCompound("pool_element")).resultOrPartial(LOGGER::error).orElseThrow(() -> {
+        this.element = StructurePoolElement.CODEC.parse(registryReadOps, nbt.getCompound("pool_element")).resultOrPartial(LOGGER::info).orElseThrow(() -> { // EMC
             return new IllegalStateException("Invalid pool element found");
         });
         this.rotation = Rotation.valueOf(nbt.getString("rotation"));
