From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 26 Jun 2023 18:41:34 -0500
Subject: [PATCH] Add ShulkerBoxDispensedEvent


diff --git a/src/main/java/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java b/src/main/java/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java
index cb308808906a8cdb127df8284e106e00553473ca..64ee566b768d4b3b5fabc123d54032b3b5f02618 100644
--- a/src/main/java/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java
+++ b/src/main/java/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java
@@ -1,5 +1,6 @@
 package net.minecraft.core.dispenser;
 
+import com.empireminecraft.customevents.ShulkerBoxDispensedEvent;
 import com.mojang.logging.LogUtils;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -66,6 +67,11 @@ public class ShulkerBoxDispenseBehavior extends OptionalDispenseItemBehavior {
             } catch (Exception exception) {
                 ShulkerBoxDispenseBehavior.LOGGER.error("Error trying to place shulker box at {}", blockposition, exception);
             }
+            // EMC start
+            if (this.isSuccess()) {
+                (new ShulkerBoxDispensedEvent(bukkitBlock, craftItem.clone())).callEvent();
+            }
+            // EMC end
         }
 
         return stack;
