From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 16 Jun 2022 09:13:17 -0500
Subject: [PATCH] EntityTriggerSculkCatalystEvent


diff --git a/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
index 551a062a578d9ec7d76261914f24db41f83cab43..4f26e90939365d6c7bd316f70968e658ee7d1725 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -49,9 +49,14 @@ public class SculkCatalystBlockEntity extends BlockEntity implements GameEventLi
         } else {
             GameEvent.Context context = event.context();
             if (event.gameEvent() == GameEvent.ENTITY_DIE) {
-                Entity i = context.sourceEntity();
-                if (i instanceof LivingEntity) {
-                    LivingEntity livingEntity = (LivingEntity)i;
+                Entity e = context.sourceEntity(); // Paper - decompile fix
+                if (e instanceof LivingEntity) { // Paper - decompile fix
+                    LivingEntity livingEntity = (LivingEntity)e; // Paper - decompile fix
+                    // EMC start
+                    if (!(new com.empireminecraft.customevents.EntityTriggerSculkCatalystEvent(livingEntity.getBukkitLivingEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(world, this.getBlockPos())).callEvent())) {
+                        return false;
+                    }
+                    // EMC end
                     if (!livingEntity.wasExperienceConsumed()) {
                         int i = livingEntity.getExperienceReward();
                         if (livingEntity.shouldDropExperience() && i > 0) {
