From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 16 Mar 2024 17:30:53 -0500
Subject: [PATCH] Parrot API


diff --git a/src/main/java/net/minecraft/world/entity/animal/ShoulderRidingEntity.java b/src/main/java/net/minecraft/world/entity/animal/ShoulderRidingEntity.java
index fdc7106d64949294cdfdfd60b51042bc9c6c767b..be4613b61002b2fe30765b15fa80055f1eaf4127 100644
--- a/src/main/java/net/minecraft/world/entity/animal/ShoulderRidingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/animal/ShoulderRidingEntity.java
@@ -38,6 +38,12 @@ public abstract class ShoulderRidingEntity extends TamableAnimal {
     }
 
     public boolean canSitOnShoulder() {
-        return this.rideCooldownCounter > 100;
+        // EMC start
+        return this.rideCooldownCounter > 100 && canSitOnShoulders; // EMC - add && canSitOnShoulders
     }
+    private boolean canSitOnShoulders = true;
+    public void setCanSitOnShoulders(boolean canSitOnShoulders) {
+        this.canSitOnShoulders = canSitOnShoulders;
+    }
+    // EMC end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftParrot.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftParrot.java
index 04d6cf6a1f3ae8316e3b2862c2d1b04e84a3b20a..f9be711dada3adf945fc75a68d20f9e13fbd18a2 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftParrot.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftParrot.java
@@ -37,4 +37,10 @@ public class CraftParrot extends CraftTameableAnimal implements Parrot {
     public boolean isDancing() {
         return this.getHandle().isPartyParrot();
     }
+    // EMC start
+    @Override
+    public void setCanSitOnShoulders(boolean canSitOnShoulders) {
+        this.getHandle().setCanSitOnShoulders(canSitOnShoulders);
+    }
+    // EMC end
 }
