From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 31 Mar 2023 23:22:24 -0500
Subject: [PATCH] Revert mysql dependency version for now...


diff --git a/build.gradle.kts b/build.gradle.kts
index 09b032e8793a57351bf45e833e27895f24a5645b..723c1a52adccb2caba6984a24277089b0efe4163 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -34,7 +34,7 @@ dependencies {
     implementation("commons-lang:commons-lang:2.6")
     implementation("net.fabricmc:mapping-io:0.3.0") // Paper - needed to read mappings for stacktrace deobfuscation
     runtimeOnly("org.xerial:sqlite-jdbc:3.42.0.0")
-    runtimeOnly("com.mysql:mysql-connector-j:8.0.33")
+    runtimeOnly("mysql:mysql-connector-java:8.0.29")
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
     // Paper start - Use Velocity cipher
     implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
