From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 26 Jun 2023 18:41:34 -0500
Subject: [PATCH] Add ShulkerBoxDispensedEvent


diff --git a/src/main/java/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java b/src/main/java/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java
index 6f2adf2334e35e8a617a4ced0c1af2abf32bbd8d..d01bb68bb0cd3c269d799c985ea02d086f574de2 100644
--- a/src/main/java/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java
+++ b/src/main/java/net/minecraft/core/dispenser/ShulkerBoxDispenseBehavior.java
@@ -1,5 +1,6 @@
 package net.minecraft.core.dispenser;
 
+import com.empireminecraft.customevents.ShulkerBoxDispensedEvent;
 import com.mojang.logging.LogUtils;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -61,6 +62,11 @@ public class ShulkerBoxDispenseBehavior extends OptionalDispenseItemBehavior {
             } catch (Exception exception) {
                 ShulkerBoxDispenseBehavior.LOGGER.error("Error trying to place shulker box at {}", blockposition, exception);
             }
+            // EMC start
+            if (this.isSuccess()) {
+                (new ShulkerBoxDispensedEvent(bukkitBlock, craftItem.clone())).callEvent();
+            }
+            // EMC end
         }
 
         return stack;
