From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 27 May 2023 16:03:16 -0500
Subject: [PATCH] PortalFrameCreateBlockCheckEvent


diff --git a/src/main/java/net/minecraft/world/level/portal/PortalForcer.java b/src/main/java/net/minecraft/world/level/portal/PortalForcer.java
index 42212d4533ce25d1cfcf4c58f1fc88791d546cff..2a423960305655a0994d0c8b8e9a443b85927d4b 100644
--- a/src/main/java/net/minecraft/world/level/portal/PortalForcer.java
+++ b/src/main/java/net/minecraft/world/level/portal/PortalForcer.java
@@ -256,6 +256,11 @@ public class PortalForcer {
                 if (k >= 0 && !this.canPortalReplaceBlock(temp)) {
                     return false;
                 }
+                // EMC start
+                if (!(new com.empireminecraft.customevents.PortalFrameCreateBlockCheckEvent(org.bukkit.craftbukkit.block.CraftBlock.at(this.level, temp))).callEvent()) {
+                    return false;
+                }
+                // EMC end
             }
         }
 
