From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Wed, 23 Jun 2021 12:24:46 -0500
Subject: [PATCH] implement RangedEntity in CraftStray


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftStray.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftStray.java
index e71314e71ea97024807b0e282d74e950d795c31f..d263159e324da2989ff67f9d5c71c4fd992e42bc 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftStray.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftStray.java
@@ -5,7 +5,7 @@ import org.bukkit.entity.EntityType;
 import org.bukkit.entity.Skeleton.SkeletonType;
 import org.bukkit.entity.Stray;
 
-public class CraftStray extends CraftAbstractSkeleton implements Stray {
+public class CraftStray extends CraftAbstractSkeleton implements Stray, com.destroystokyo.paper.entity.CraftRangedEntity<net.minecraft.world.entity.monster.AbstractSkeleton> { // Paper
 
     public CraftStray(CraftServer server, net.minecraft.world.entity.monster.Stray entity) {
         super(server, entity);
@@ -25,4 +25,10 @@ public class CraftStray extends CraftAbstractSkeleton implements Stray {
     public SkeletonType getSkeletonType() {
         return SkeletonType.STRAY;
     }
+    // Paper start
+    @Override
+    public net.minecraft.world.entity.monster.Stray getHandle() {
+        return (net.minecraft.world.entity.monster.Stray) entity;
+    }
+    // Paper end
 }
