From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 16 Jan 2021 22:44:20 -0600
Subject: [PATCH] Convert EntityTag attributes on creature spawn

---
 .../minecraft/world/entity/EntityTypes.java   | 29 ++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/src/main/java/net/minecraft/world/entity/EntityTypes.java b/src/main/java/net/minecraft/world/entity/EntityTypes.java
index f83315937cda7eaabc857129aef3473833b387b4..8bcba2b878ec535b6938e393cd95b07d4189ee17 100644
--- a/src/main/java/net/minecraft/world/entity/EntityTypes.java
+++ b/src/main/java/net/minecraft/world/entity/EntityTypes.java
@@ -348,6 +348,7 @@ public class EntityTypes<T extends Entity> {
             }
         }
         // Paper end
+        convertNBT(nbttagcompound); // EMC
         T t0 = this.createCreature(worldserver, nbttagcompound, ichatbasecomponent, entityhuman, blockposition, enummobspawn, flag, flag1);
         // EMC start - if false the spawn was cancelled, add new event
         if ((spawnReason == CreatureSpawnEvent.SpawnReason.SPAWNER_EGG || spawnReason == CreatureSpawnEvent.SpawnReason.DISPENSE_EGG) && itemstack != null && t0 != null) {
@@ -372,7 +373,33 @@ public class EntityTypes<T extends Entity> {
 
         return t0;
     }
-
+    // EMC start
+    private static final Map<String, String> attributeConversion = new com.google.common.collect.ImmutableMap.Builder<String, String>()
+        .put("generic.maxHealth", "generic.max_health")
+        .put("horse.jumpStrength", "horse.jump_strength")
+        .put("generic.followRange", "generic.follow_range")
+        .put("generic.knockbackResistance", "generic.knockback_resistance")
+        .put("generic.movementSpeed", "generic.movement_speed")
+        .put("generic.flyingSpeed", "generic.flying_speed")
+        .put("generic.attackDamage", "generic.attack_damage")
+        .put("generic.attackKnockback", "generic.attack_knockback")
+        .put("generic.attackSpeed", "generic.attack_speed")
+        .put("generic.armorToughness", "generic.armor_toughness").build();
+    private static void convertNBT(@Nullable NBTTagCompound nbttagcompound) {
+        if (nbttagcompound != null && nbttagcompound.hasKeyOfType("EntityTag", 10)) {
+            NBTTagCompound entityTag = nbttagcompound.getCompound("EntityTag");
+            if (entityTag.hasKeyOfType("Attributes", 9)) {
+                NBTTagList nbttaglist = entityTag.getList("Attributes", 10);
+                for (int i = 0; i < nbttaglist.size(); i++) {
+                    String value = attributeConversion.get(nbttaglist.getCompound(i).getString("Name"));
+                    if (value != null) {
+                        nbttaglist.getCompound(i).setString("Name", value);
+                    }
+                }
+            }
+        }
+    }
+    // EMC end
     @Nullable
     public T createCreature(WorldServer worldserver, @Nullable NBTTagCompound nbttagcompound, @Nullable IChatBaseComponent ichatbasecomponent, @Nullable EntityHuman entityhuman, BlockPosition blockposition, EnumMobSpawn enummobspawn, boolean flag, boolean flag1) {
         T t0 = this.a((World) worldserver);
