From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 31 Mar 2023 23:22:24 -0500
Subject: [PATCH] Revert mysql dependency version for now...


diff --git a/build.gradle.kts b/build.gradle.kts
index e84738f953bcc31aa59a2cfa5a5c3ac3936f659a..e0c7afbd9c9b6fe393a5350f2449f30eb0646caf 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -34,7 +34,7 @@ dependencies {
     implementation("commons-lang:commons-lang:2.6")
     implementation("net.fabricmc:mapping-io:0.3.0") // Paper - needed to read mappings for stacktrace deobfuscation
     runtimeOnly("org.xerial:sqlite-jdbc:3.41.0.0")
-    runtimeOnly("com.mysql:mysql-connector-j:8.0.32")
+    runtimeOnly("mysql:mysql-connector-java:8.0.29")
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
     // Paper start - Use Velocity cipher
     implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
