From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 21 Mar 2022 16:06:35 -0500
Subject: [PATCH] Promote caught connection exceptions to warn instead of debug


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index ff82fb1a9077cac31dc9369f25e489152fb16512..557f23f786794d883402f6583c95d5f7d9a19873 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -236,7 +236,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                     if (player != null) player.quitReason = org.bukkit.event.player.PlayerQuitEvent.QuitReason.ERRONEOUS_STATE; // Paper
                     if (flag) {
-                        Connection.LOGGER.debug("Failed to sent packet", throwable);
+                        Connection.LOGGER.warn("Failed to sent packet", throwable);
                         ConnectionProtocol enumprotocol = this.getCurrentProtocol();
                         Packet<?> packet = enumprotocol == ConnectionProtocol.LOGIN ? new ClientboundLoginDisconnectPacket(chatmessage) : new ClientboundDisconnectPacket(chatmessage);
 
@@ -245,7 +245,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
                         });
                         this.setReadOnly();
                     } else {
-                        Connection.LOGGER.debug("Double fault", throwable);
+                        Connection.LOGGER.warn("Double fault", throwable);
                         this.disconnect(chatmessage);
                     }
                 }
