From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 10 Jul 2021 23:31:59 -0500
Subject: [PATCH] VegetationPatchPlaceEvent


diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java b/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
index 62165c0f5d1eba8c2734356bec33998d9dadbd1b..6d1347c6ec6c2e41a9b81886253c38562a624190 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/feature/VegetationPatchFeature.java
@@ -7,6 +7,7 @@ import java.util.Set;
 import java.util.function.Predicate;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.server.MCUtil;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.tags.Tag;
 import net.minecraft.world.level.WorldGenLevel;
@@ -83,6 +84,13 @@ public class VegetationPatchFeature extends Feature<VegetationPatchConfiguration
     protected void distributeVegetation(FeaturePlaceContext<VegetationPatchConfiguration> context, WorldGenLevel world, VegetationPatchConfiguration config, Random random, Set<BlockPos> positions, int radiusX, int radiusZ) {
         for(BlockPos blockPos : positions) {
             if (config.vegetationChance > 0.0F && random.nextFloat() < config.vegetationChance) {
+                // EMC start
+                org.bukkit.block.Block origin = MCUtil.toBukkitBlock(context.level().getLevel(), context.origin());
+                org.bukkit.block.Block place = MCUtil.toBukkitBlock(context.level().getLevel(), context.origin());
+                if (!(new com.empireminecraft.customevents.VegetationPatchPlaceEvent(place, origin).callEvent())) {
+                    continue;
+                }
+                // EMC end
                 this.placeVegetation(world, config, context.chunkGenerator(), random, blockPos);
             }
         }
