From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Fri, 1 Jul 2022 20:15:55 -0500
Subject: [PATCH] Sheep should not regrow wool if grass eating is cancelled


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/EatBlockGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
index abf796c3369da6b73c8587dfc05f56d0b8933fde..d5261a833918c4b7d66fe6fdc194beabc0a96a4c 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
@@ -71,9 +71,10 @@ public class EatBlockGoal extends Goal {
                 // CraftBukkit
                 if (!CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockposition, Blocks.AIR.defaultBlockState(), !this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)).isCancelled()) {
                     this.level.destroyBlock(blockposition, false);
+                    this.mob.ate(); // EMC - moved from below
                 }
 
-                this.mob.ate();
+                //this.mob.ate(); // EMC - move up
             } else {
                 BlockPos blockposition1 = blockposition.below();
 
@@ -82,9 +83,10 @@ public class EatBlockGoal extends Goal {
                     if (!CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockposition1, Blocks.DIRT.defaultBlockState(), !this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)).isCancelled()) { // Paper - Fix wrong block state
                         this.level.levelEvent(2001, blockposition1, Block.getId(Blocks.GRASS_BLOCK.defaultBlockState()));
                         this.level.setBlock(blockposition1, Blocks.DIRT.defaultBlockState(), 2);
+                        this.mob.ate(); // EMC - moved from below
                     }
 
-                    this.mob.ate();
+                    //this.mob.ate(); // EMC - move up
                 }
             }
 
