From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Thu, 4 Jul 2024 07:44:22 -0500
Subject: [PATCH] Incomplete play launch sound control for projectiles


diff --git a/src/main/java/net/minecraft/world/item/SnowballItem.java b/src/main/java/net/minecraft/world/item/SnowballItem.java
index bc8186a5bc3a98b35fad570729dd4ba52efab238..3971bd941c86eb8d3102518ead328dc88b87948f 100644
--- a/src/main/java/net/minecraft/world/item/SnowballItem.java
+++ b/src/main/java/net/minecraft/world/item/SnowballItem.java
@@ -37,8 +37,10 @@ public class SnowballItem extends Item {
                     ((net.minecraft.server.level.ServerPlayer) user).getBukkitEntity().updateInventory();  // Paper - PlayerLaunchProjectileEvent
                 }
 
+                if (event.shouldPlayLaunchSound()) { // EMC
                 world.playSound((Player) null, user.getX(), user.getY(), user.getZ(), SoundEvents.SNOWBALL_THROW, SoundSource.NEUTRAL, 0.5F, 0.4F / (world.getRandom().nextFloat() * 0.4F + 0.8F));
-            } else { // Paper - PlayerLaunchProjectileEvent
+                } // EMC
+                } else { // Paper - PlayerLaunchProjectileEvent
                 if (user instanceof net.minecraft.server.level.ServerPlayer) ((net.minecraft.server.level.ServerPlayer) user).getBukkitEntity().updateInventory(); // Paper - PlayerLaunchProjectileEvent
                 return InteractionResultHolder.fail(itemstack); // Paper - PlayerLaunchProjectileEvent
             }
