From 84d2271afc507fc6bf3436ef9b5537bb02e5e8a7 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Tue, 17 Mar 2020 18:59:20 -0500
Subject: [PATCH] Properly remove empty entityTags from SpawnEggs

---
 .../inventory/CraftMetaSpawnEgg.java          | 23 ++++++++++++++++++-
 1 file changed, 22 insertions(+), 1 deletion(-)

diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
index 0ee50e39df..455a47ed17 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
@@ -43,6 +43,18 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
 
         if (tag.hasKey(ENTITY_TAG.NBT)) {
             entityTag = tag.getCompound(ENTITY_TAG.NBT);
+            // EMC start
+            if (entityTag != null) {
+                String entityType = entityTag.getString(ENTITY_ID.NBT);
+                if (entityType != null) {
+                    this.spawnedType = EntityType.fromName(entityType);
+                    entityTag.remove(ENTITY_ID.NBT);
+                }
+                if (entityTag != null && entityTag.isEmpty()) {
+                    entityTag = null;
+                }
+            }
+            // EMC end
         }
     }
 
@@ -86,7 +98,15 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             // See if we can read a converted ID tag
             if (entityTag.hasKey(ENTITY_ID.NBT)) {
                 this.spawnedType = EntityType.fromName(new MinecraftKey(entityTag.getString(ENTITY_ID.NBT)).getKey());
+                // EMC start
+                if (spawnedType != null) {
+                    entityTag.remove(ENTITY_ID.NBT);
+                }
+            }
+            if (entityTag.isEmpty()) {
+                entityTag = null;
             }
+            // EMC end
         }
     }
 
@@ -228,7 +248,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             CraftMetaSpawnEgg that = (CraftMetaSpawnEgg) meta;
 
             return hasSpawnedType() ? that.hasSpawnedType() && this.spawnedType.equals(that.spawnedType) : !that.hasSpawnedType()
-                    && entityTag != null ? that.entityTag != null && this.entityTag.equals(that.entityTag) : entityTag == null;
+                    && (entityTag != null ? that.entityTag != null && this.entityTag.equals(that.entityTag) : that.entityTag == null); // EMC - fix logic
         }
         return true;
     }
@@ -283,6 +303,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             if (entityTag != null) {
                 // Remove ID tag as it is now in the material
                 entityTag.remove(ENTITY_ID.NBT);
+                if (entityTag.isEmpty()) entityTag = null; // EMC
             }
 
             return CraftLegacy.fromLegacy(new MaterialData(Material.LEGACY_MONSTER_EGG, (byte) spawnedType.getTypeId()));
-- 
2.25.1

