From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sat, 7 Aug 2021 11:36:09 -0500
Subject: [PATCH] Change vanilla structure error to info log level


diff --git a/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java b/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java
index 663b37494d664c5ecb59784737bd660fdc9a14b8..ff9c6c21d105e51b25e7f16d53ff376ffa8ee11a 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/structure/PoolElementStructurePiece.java
@@ -48,7 +48,7 @@ public class PoolElementStructurePiece extends StructurePiece {
         this.position = new BlockPos(nbt.getInt("PosX"), nbt.getInt("PosY"), nbt.getInt("PosZ"));
         this.groundLevelDelta = nbt.getInt("ground_level_delta");
         RegistryReadOps<Tag> registryReadOps = RegistryReadOps.create(NbtOps.INSTANCE, world.getServer().getResourceManager(), world.getServer().registryAccess());
-        this.element = StructurePoolElement.CODEC.parse(registryReadOps, nbt.getCompound("pool_element")).resultOrPartial(LOGGER::error).orElseThrow(() -> {
+        this.element = StructurePoolElement.CODEC.parse(registryReadOps, nbt.getCompound("pool_element")).resultOrPartial(LOGGER::info).orElseThrow(() -> {
             return new IllegalStateException("Invalid pool element found");
         });
         this.rotation = Rotation.valueOf(nbt.getString("rotation"));
