From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 26 Sep 2021 11:59:40 -0500
Subject: [PATCH] Log error of item conversion failure instead of throwing an
 exception


diff --git a/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java b/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
index 4693bb22249bc5b91b387d9631f3df089f5b7e9c..c772022c9fdb033c3cba58b4f5234927bce53fa7 100644
--- a/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
+++ b/src/main/java/com/empireminecraft/api/meta/MetaApiAccessor.java
@@ -421,7 +421,12 @@ public final class MetaApiAccessor {
                 cmp.remove(META_TYPE);
                 if (ITEM.equals(type)) {
                     int version = cmp.contains("DataVersion") ? cmp.getInt("DataVersion") : -1;
-                    cmp = (CompoundTag) MinecraftServer.getServer().fixerUpper.update(References.ITEM_STACK, new Dynamic(NbtOps.INSTANCE, cmp), version, SharedConstants.getCurrentVersion().getWorldVersion()).getValue();
+                    try {
+                        cmp = (CompoundTag) MinecraftServer.getServer().fixerUpper.update(References.ITEM_STACK, new Dynamic(NbtOps.INSTANCE, cmp), version, SharedConstants.getCurrentVersion().getWorldVersion()).getValue();
+                    } catch (IllegalArgumentException e) {
+                        MinecraftServer.LOGGER.error("Error converting item: " + cmp, e);
+                        return null;
+                    }
                     return CraftItemStack.asCraftMirror(ItemStack.of(cmp));
                 } else if (META_MAP.equals(type)) {
                     final PersistentMetaMap metaMap = getMetaMapFromCompound(cmp);
