From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 6 Sep 2021 01:57:52 -0500
Subject: [PATCH] Add 1.17 mobs to CraftMetaSpawnEgg


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
index b4c64b9d3d32a30543754ede8f18540cee483be0..00ca9d7080ed81b931b57cb9788fcb3e262fd017 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
@@ -133,6 +133,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
     @Override
     boolean applicableTo(Material type) {
         switch (type) {
+            case AXOLOTL_SPAWN_EGG: // EMC
             case BAT_SPAWN_EGG:
             case BEE_SPAWN_EGG:
             case BLAZE_SPAWN_EGG:
@@ -151,6 +152,8 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             case EVOKER_SPAWN_EGG:
             case FOX_SPAWN_EGG:
             case GHAST_SPAWN_EGG:
+            case GLOW_SQUID_SPAWN_EGG: // EMC
+            case GOAT_SPAWN_EGG: // EMC
             case GUARDIAN_SPAWN_EGG:
             case HOGLIN_SPAWN_EGG:
             case HORSE_SPAWN_EGG:
