From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Mon, 6 Sep 2021 01:57:52 -0500
Subject: [PATCH] Add 1.17 mobs to CraftMetaSpawnEgg


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
index d9369c19612aed880cf0a16f8f60e5a4112500ae..653da934273975cb9fa1e569198be6e52f41a5b5 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemFactory.java
@@ -124,6 +124,7 @@ public final class CraftItemFactory implements ItemFactory {
         case YELLOW_BANNER:
         case YELLOW_WALL_BANNER:
             return meta instanceof CraftMetaBanner ? meta : new CraftMetaBanner(meta);
+        case AXOLOTL_SPAWN_EGG: // EMC
         case BAT_SPAWN_EGG:
         case BEE_SPAWN_EGG:
         case BLAZE_SPAWN_EGG:
@@ -142,6 +143,8 @@ public final class CraftItemFactory implements ItemFactory {
         case EVOKER_SPAWN_EGG:
         case FOX_SPAWN_EGG:
         case GHAST_SPAWN_EGG:
+        case GLOW_SQUID_SPAWN_EGG: // EMC
+        case GOAT_SPAWN_EGG: // EMC
         case GUARDIAN_SPAWN_EGG:
         case HOGLIN_SPAWN_EGG:
         case HORSE_SPAWN_EGG:
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index bfed3e4d485c3dfaa86d16fb2057acd6145bf7e2..3e4712fe745fbd7b1a369a7b2e8502465ae73dde 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -358,6 +358,7 @@ public final class CraftItemStack extends ItemStack {
             case YELLOW_BANNER:
             case YELLOW_WALL_BANNER:
                 return new CraftMetaBanner(item.getTag());
+            case AXOLOTL_SPAWN_EGG: // EMC
             case BAT_SPAWN_EGG:
             case BEE_SPAWN_EGG:
             case BLAZE_SPAWN_EGG:
@@ -376,6 +377,8 @@ public final class CraftItemStack extends ItemStack {
             case EVOKER_SPAWN_EGG:
             case FOX_SPAWN_EGG:
             case GHAST_SPAWN_EGG:
+            case GLOW_SQUID_SPAWN_EGG: // EMC
+            case GOAT_SPAWN_EGG: // EMC
             case GUARDIAN_SPAWN_EGG:
             case HOGLIN_SPAWN_EGG:
             case HORSE_SPAWN_EGG:
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
index b4c64b9d3d32a30543754ede8f18540cee483be0..00ca9d7080ed81b931b57cb9788fcb3e262fd017 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSpawnEgg.java
@@ -133,6 +133,7 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
     @Override
     boolean applicableTo(Material type) {
         switch (type) {
+            case AXOLOTL_SPAWN_EGG: // EMC
             case BAT_SPAWN_EGG:
             case BEE_SPAWN_EGG:
             case BLAZE_SPAWN_EGG:
@@ -151,6 +152,8 @@ public class CraftMetaSpawnEgg extends CraftMetaItem implements SpawnEggMeta {
             case EVOKER_SPAWN_EGG:
             case FOX_SPAWN_EGG:
             case GHAST_SPAWN_EGG:
+            case GLOW_SQUID_SPAWN_EGG: // EMC
+            case GOAT_SPAWN_EGG: // EMC
             case GUARDIAN_SPAWN_EGG:
             case HOGLIN_SPAWN_EGG:
             case HORSE_SPAWN_EGG:
