--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/block/BlockBreakNaturallyEvent.java
@@ -1,0 +_,40 @@
+package com.empireminecraft.empirecraft.event.block;
+
+import org.bukkit.Location;
+import org.bukkit.entity.Item;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.block.BlockEvent;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Collections;
+
+public class BlockBreakNaturallyEvent extends BlockEvent {
+    private final Collection<Item> items;
+
+    public BlockBreakNaturallyEvent(@NotNull Location pos, @NotNull Collection<Item> items) {
+        super(pos.getBlock());
+        this.items = items;
+    }
+
+    public BlockBreakNaturallyEvent(@NotNull Location pos, @NotNull Item item) {
+        this(pos, Collections.singleton(item));
+    }
+
+    @NotNull
+    public Collection<Item> getItems() {
+        return items;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
