--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/block/BlockBreakNaturallyEvent.java
@@ -1,0 +_,38 @@
+package com.empireminecraft.empirecraft.event.block;
+
+import org.bukkit.Location;
+import org.bukkit.entity.Item;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.block.BlockEvent;
+import org.jspecify.annotations.NullMarked;
+
+import java.util.Collection;
+import java.util.Collections;
+
+@NullMarked
+public class BlockBreakNaturallyEvent extends BlockEvent {
+    private final Collection<Item> items;
+
+    public BlockBreakNaturallyEvent(Location pos, Collection<Item> items) {
+        super(pos.getBlock());
+        this.items = items;
+    }
+
+    public BlockBreakNaturallyEvent(Location pos, Item item) {
+        this(pos, Collections.singleton(item));
+    }
+
+    public Collection<Item> getItems() {
+        return items;
+    }
+
+    private static final HandlerList handlers = new HandlerList();
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
