--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/block/ConduitNewTargetEvent.java
@@ -1,0 +_,41 @@
+package com.empireminecraft.empirecraft.event.block;
+
+import org.bukkit.block.Block;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.block.BlockEvent;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+public class ConduitNewTargetEvent extends BlockEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private final List<LivingEntity> entities;
+
+    public ConduitNewTargetEvent(@NotNull List<LivingEntity> entities, @NotNull Block theBlock) {
+        super(theBlock);
+        this.entities = entities;
+    }
+
+    /**
+     * Remove entities from this list to remove them from the potential targets of the conduit.
+     * Implementation will not use any additions to this list.
+     *
+     * @return
+     */
+    @NotNull
+    public List<LivingEntity> potentialTargets() {
+        return entities;
+    }
+
+    @Override
+    @NotNull
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
