--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/block/ShulkerBoxDispensedEvent.java
@@ -1,0 +_,29 @@
+package com.empireminecraft.empirecraft.event.block;
+
+import org.bukkit.block.Block;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.block.BlockEvent;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+public class ShulkerBoxDispensedEvent extends BlockEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private final ItemStack item;
+
+    public ShulkerBoxDispensedEvent(@NotNull Block block, @NotNull ItemStack item) {
+        super(block);
+        this.item = item;
+    }
+
+    public @NotNull ItemStack getItem() {
+        return item;
+    }
+
+    public @NotNull HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static @NotNull HandlerList getHandlerList() {
+        return handlers;
+    }
+}
