--- /dev/null
+++ b/src/main/java/com/empireminecraft/empirecraft/event/block/ShulkerBoxDispensedEvent.java
@@ -1,0 +_,30 @@
+package com.empireminecraft.empirecraft.event.block;
+
+import org.bukkit.block.Block;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.block.BlockEvent;
+import org.bukkit.inventory.ItemStack;
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class ShulkerBoxDispensedEvent extends BlockEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private final ItemStack item;
+
+    public ShulkerBoxDispensedEvent(Block block, ItemStack item) {
+        super(block);
+        this.item = item;
+    }
+
+    public ItemStack getItem() {
+        return item;
+    }
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
