--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -455,14 +_,14 @@
             level.recipeAccess().byKey(entry.getKey()).ifPresent(recipe -> {
                 if (!(recipe.value() instanceof AbstractCookingRecipe)) return; // Paper - don't process non-cooking recipes
                 recipesToAward.add((RecipeHolder<?>)recipe);
-                createExperience(level, position, entry.getIntValue(), ((AbstractCookingRecipe)recipe.value()).experience(), blockPos, serverPlayer, itemStack, amount); // Paper - don't process non-cooking recipes
+                    createExperience(level, position, entry.getIntValue(), ((AbstractCookingRecipe)recipe.value()).experience(), blockPos, serverPlayer, itemStack, amount, ((AbstractCookingRecipe) recipe.value()).result().create()); // Paper - don't process non-cooking recipes // EMC - add result item
             });
         }
 
         return recipesToAward;
     }
 
-    private static void createExperience(final ServerLevel level, final Vec3 position, final int amount, final float value, final BlockPos blockPos, final ServerPlayer player, final ItemStack itemStack, final int removeCount) {
+    private static void createExperience(final ServerLevel level, final Vec3 position, final int amount, final float value, final BlockPos blockPos, final ServerPlayer player, final ItemStack itemStack, final int removeCount, ItemStack result) { // CraftBukkit // EMC - add result item
         int xpReward = Mth.floor(amount * value);
         float xpFraction = Mth.frac(amount * value);
         if (xpFraction != 0.0F && level.getRandom().nextFloat() < xpFraction) {
@@ -471,13 +_,17 @@
 
         // CraftBukkit start - fire FurnaceExtractEvent / BlockExpEvent
         org.bukkit.event.block.BlockExpEvent event;
-        if (removeCount != 0) {
+        // EMC start
+        if (true || removeCount != 0) {
             event = new org.bukkit.event.inventory.FurnaceExtractEvent(
-                player.getBukkitEntity(),
+                player == null ? null : player.getBukkitEntity(),
                 org.bukkit.craftbukkit.block.CraftBlock.at(level, blockPos),
-                itemStack.asBukkitCopy(),
+                itemStack == null ? null : itemStack.asBukkitCopy(),
                 removeCount,
-                xpReward
+                xpReward,
+                result == null ? null : result.asBukkitCopy(),
+                amount
+                // EMC end
             );
         } else {
             event = new org.bukkit.event.block.BlockExpEvent(
