--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -477,14 +_,14 @@
             level.recipeAccess().byKey(entry.getKey()).ifPresent(recipe -> {
                 if (!(recipe.value() instanceof AbstractCookingRecipe)) return; // Paper - don't process non-cooking recipes
                 list.add((RecipeHolder<?>)recipe);
-                createExperience(level, popVec, entry.getIntValue(), ((AbstractCookingRecipe)recipe.value()).experience(), blockPos, serverPlayer, itemStack, amount);
+                createExperience(level, popVec, entry.getIntValue(), ((AbstractCookingRecipe)recipe.value()).experience(), blockPos, serverPlayer, itemStack, amount, ((AbstractCookingRecipe) recipe.value()).result()); // CraftBukkit // EMC - add result item
             });
         }
 
         return list;
     }
 
-    private static void createExperience(ServerLevel level, Vec3 popVec, int recipeIndex, float experience, BlockPos blockPos, ServerPlayer serverPlayer, ItemStack itemStack, int amount) { // CraftBukkit
+    private static void createExperience(ServerLevel level, Vec3 popVec, int recipeIndex, float experience, BlockPos blockPos, ServerPlayer serverPlayer, ItemStack itemStack, int amount, ItemStack result) { // CraftBukkit // EMC - add result item
         int floor = Mth.floor(recipeIndex * experience);
         float fraction = Mth.frac(recipeIndex * experience);
         if (fraction != 0.0F && Math.random() < fraction) {
@@ -493,13 +_,13 @@
 
         // CraftBukkit start - fire FurnaceExtractEvent / BlockExpEvent
         org.bukkit.event.block.BlockExpEvent event;
-        if (amount != 0) {
+        if (true || amount != 0) { // EMC
             event = new org.bukkit.event.inventory.FurnaceExtractEvent(
-                serverPlayer.getBukkitEntity(),
+                    serverPlayer == null ? null : serverPlayer.getBukkitEntity(),
                 org.bukkit.craftbukkit.block.CraftBlock.at(level, blockPos),
                 org.bukkit.craftbukkit.inventory.CraftItemType.minecraftToBukkit(itemStack.getItem()),
                 amount,
-                floor
+                floor, org.bukkit.craftbukkit.inventory.CraftItemType.minecraftToBukkit(result.getItem()), recipeIndex // EMC - add == null for entityhuman and itemstack and add resultType/resultAmount
             );
         } else {
             event = new org.bukkit.event.block.BlockExpEvent(
