--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -377,7 +_,7 @@
 
     // CraftBukkit start - add method
     public org.bukkit.inventory.InventoryHolder getOwner() {
-        return getOwner(true);
+        return getOwner(false); // EMC - don't use snapshots for TileEntity getOwner
     }
 
     public org.bukkit.inventory.InventoryHolder getOwner(boolean useSnapshot) {
