--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -425,6 +_,13 @@
             Direction opposite = blockEntity.facing.getOpposite();
             if (isFullContainer(attachedContainer, opposite)) {
                 return false;
+                // EMC start - HopperFillEvent
+            }
+            com.empireminecraft.empirecraft.event.inventory.HopperFillEvent fillEvent = new com.empireminecraft.empirecraft.event.inventory.HopperFillEvent(blockEntity.getOwner(false).getInventory(), getInventory(attachedContainer));
+            org.bukkit.Bukkit.getPluginManager().callEvent(fillEvent);
+            if (fillEvent.isCancelled()) {
+                return false;
+                // EMC end - HopperFillEvent
             } else {
                 // Paper start - Perf: Optimize Hoppers
                 return hopperPush(level, attachedContainer, opposite, blockEntity);
@@ -533,6 +_,13 @@
         BlockState blockState = level.getBlockState(blockPos);
         Container sourceContainer = getSourceContainer(level, hopper, blockPos, blockState);
         if (sourceContainer != null) {
+            // EMC start - HopperDrainEvent
+            com.empireminecraft.empirecraft.event.inventory.HopperDrainEvent drainEvent = new com.empireminecraft.empirecraft.event.inventory.HopperDrainEvent(new org.bukkit.craftbukkit.inventory.CraftInventory(hopper), getInventory(sourceContainer));
+            org.bukkit.Bukkit.getPluginManager().callEvent(drainEvent);
+            if (drainEvent.isCancelled()) {
+                return false;
+            }
+            // EMC end - HopperDrainEvent
             Direction direction = Direction.DOWN;
             skipPullModeEventFire = skipHopperEvents; // Paper - Perf: Optimize Hoppers
 
