--- a/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -96,6 +_,12 @@
         @Override
         public boolean handleGameEvent(ServerLevel level, Holder<GameEvent> gameEvent, GameEvent.Context context, Vec3 pos) {
             if (gameEvent.is(GameEvent.ENTITY_DIE) && context.sourceEntity() instanceof LivingEntity livingEntity) {
+                // EMC start - EntityTriggerSculkCatalystEvent
+                net.minecraft.core.Position blockPos = this.positionSource.getPosition(level).orElse(null);
+                if (blockPos != null && !(new com.empireminecraft.empirecraft.event.entity.EntityTriggerSculkCatalystEvent(livingEntity.getBukkitLivingEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, BlockPos.containing(pos))).callEvent())) {
+                    return false;
+                }
+                // EMC end - EntityTriggerSculkCatalystEvent
                 if (!livingEntity.wasExperienceConsumed()) {
                     DamageSource lastDamageSource = livingEntity.getLastDamageSource();
                     int experienceReward = livingEntity.getExperienceReward(level, Optionull.map(lastDamageSource, DamageSource::getEntity));
