From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 4 Mar 2013 23:35:02 -0500
Subject: [PATCH] Temporary Meta API


diff --git a/src/main/java/io/papermc/paper/world/saveddata/PaperLevelOverrides.java b/src/main/java/io/papermc/paper/world/saveddata/PaperLevelOverrides.java
index fb575b634de6e4fe94f3a59b288074266028e0ce..618dadcc6ee0b8991e837f8bb492f842069c6f22 100644
--- a/src/main/java/io/papermc/paper/world/saveddata/PaperLevelOverrides.java
+++ b/src/main/java/io/papermc/paper/world/saveddata/PaperLevelOverrides.java
@@ -44,6 +44,7 @@ public final class PaperLevelOverrides extends SavedData implements ServerLevelD
     private LevelSettings.DifficultySettings difficultySettings;
     private transient @Nullable PrimaryLevelData rootData;
     private transient ResourceKey<Level> dimension = Level.OVERWORLD;
+    public final java.util.Map<org.bukkit.NamespacedKey, Object> tempMeta = com.google.common.collect.Maps.newHashMap(); // EMC - Temporary Meta API
 
     private PaperLevelOverrides() {
         this(RespawnData.DEFAULT, 0L, false, GameType.SURVIVAL, LevelSettings.DifficultySettings.DEFAULT);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index c8d179d936d57e8222d87b7fd943dbddfa8446a0..1ec68fdbd6dae07f09d8065fbb6bd0c120762cf3 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -89,6 +89,11 @@ public class CraftChunk implements Chunk {
 
         return chunkAccess;
     }
+    // EMC start - Temporary Meta API
+    public java.util.@org.jspecify.annotations.NonNull Map<org.bukkit.NamespacedKey, Object> getTemporaryMeta() {
+        return getHandle(ChunkStatus.FULL).tempMeta;
+    }
+    // EMC end - Temporary Meta API
 
     @Override
     public int getX() {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index f77cb6d85c8947a96d41b96b31814c551dfd5258..768b14d7cad073e6a14ed9412d94c8c0a34ce505 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -300,6 +300,11 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         ).isValid();
     }
     // Paper end
+    // EMC start - Temporary Meta API
+    public java.util.@org.jspecify.annotations.NonNull Map<NamespacedKey, Object> getTemporaryMeta() {
+        return getHandle().serverLevelData.tempMeta;
+    }
+    // EMC end - Temporary Meta API
 
     private static final Random rand = new Random();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 2f40476da09c0c679047978437de07a5d162fee4..5738c1a96d50bd8d68611346c98e12edfc67274d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -231,6 +231,11 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
         return false;
     }
+    // EMC start
+    public java.util.@org.jspecify.annotations.NonNull Map<org.bukkit.NamespacedKey, Object> getTemporaryMeta() {
+        return getHandle().tempMeta;
+    }
+    // EMC end
 
     @Override
     public double getHeight() {
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 7a42d1810f74a041318c8a3c53793106e872081f..76b238699bd7e0eab732dbe2046a2ba41bdad7a2 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1610,6 +1610,11 @@ public class CraftEventFactory {
         InventoryCloseEvent event = new InventoryCloseEvent(human.containerMenu.getBukkitView(), reason); // Paper
         human.level().getCraftServer().getPluginManager().callEvent(event);
         human.containerMenu.transferTo(human.inventoryMenu, human.getBukkitEntity());
+        // EMC start - Temporary Meta API
+        if (event.getInventory().getViewers().isEmpty()) {
+            org.bukkit.craftbukkit.inventory.CraftInventory.meta.remove(((org.bukkit.craftbukkit.inventory.CraftInventory) event.getInventory()).getInventory());
+        }
+        // EMC end - Temporary Meta API
     }
 
     public static ItemStack handleEditBookEvent(ServerPlayer player, int itemInHandIndex, ItemStack itemInHand, ItemStack newBookItem) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
index 9d5fe6fc727948955d89f9ed5e181185393a9611..c821b74bfe98a9684580ba94d37a719f07422fef 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
@@ -448,6 +448,12 @@ public class CraftInventory implements Inventory {
         return count;
     }
     // Paper end
+    // EMC start - Temporary Meta API
+    public static final java.util.WeakHashMap<Container, java.util.Map<org.bukkit.NamespacedKey, Object>> meta = new java.util.WeakHashMap<>(64);
+    public @org.jetbrains.annotations.NotNull java.util.Map<org.bukkit.NamespacedKey, Object> getTemporaryMeta() {
+        return meta.computeIfAbsent(inventory, k -> new HashMap<>());
+    }
+    // EMC end - Temporary Meta API
 
     @Override
     public ListIterator<ItemStack> iterator() {
