From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 4 Mar 2013 23:35:02 -0500
Subject: [PATCH] Temporary Meta API


diff --git a/src/main/java/io/papermc/paper/world/saveddata/PaperLevelOverrides.java b/src/main/java/io/papermc/paper/world/saveddata/PaperLevelOverrides.java
index 0aba139ace05262464ed79bdbf6a7772964baebf..858ee60a7db04b41c2ef9eb426d806834f2f5322 100644
--- a/src/main/java/io/papermc/paper/world/saveddata/PaperLevelOverrides.java
+++ b/src/main/java/io/papermc/paper/world/saveddata/PaperLevelOverrides.java
@@ -44,6 +44,7 @@ public final class PaperLevelOverrides extends SavedData implements ServerLevelD
     private LevelSettings.DifficultySettings difficultySettings;
     private transient @Nullable PrimaryLevelData rootData;
     private transient ResourceKey<Level> dimension = Level.OVERWORLD;
+    public final java.util.Map<org.bukkit.NamespacedKey, Object> tempMeta = com.google.common.collect.Maps.newHashMap(); // EMC - Temporary Meta API
 
     private PaperLevelOverrides() {
         this(RespawnData.DEFAULT, 0L, false, GameType.SURVIVAL, LevelSettings.DifficultySettings.DEFAULT);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index f020f310c4bca49a23de91bea72ccd9941c0b03c..c28db305537ad1145971b0bce4bf4713584eb26e 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -87,6 +87,11 @@ public class CraftChunk implements Chunk {
 
         return chunkAccess;
     }
+    // EMC start - Temporary Meta API
+    public java.util.@org.jspecify.annotations.NonNull Map<org.bukkit.NamespacedKey, Object> getTemporaryMeta() {
+        return getHandle(ChunkStatus.FULL).tempMeta;
+    }
+    // EMC end - Temporary Meta API
 
     @Override
     public int getX() {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index f3f068011abd8c6585463169d50634bf42a09497..7a69240f629b4ce94799ef40500a14aad5a6f29d 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -292,6 +292,11 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         ).isValid();
     }
     // Paper end
+    // EMC start - Temporary Meta API
+    public java.util.@org.jspecify.annotations.NonNull Map<NamespacedKey, Object> getTemporaryMeta() {
+        return getHandle().serverLevelData.tempMeta;
+    }
+    // EMC end - Temporary Meta API
 
     private static final Random rand = new Random();
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index 0965a503160ae11aa20963a8871c3838c50d0c6a..2db55dab8fd645ce19e7023fe5dee55ae01325c3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -236,6 +236,11 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
         return false;
     }
+    // EMC start
+    public java.util.@org.jspecify.annotations.NonNull Map<org.bukkit.NamespacedKey, Object> getTemporaryMeta() {
+        return getHandle().tempMeta;
+    }
+    // EMC end
 
     @Override
     public double getHeight() {
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index a5ea91906abebf44cfbfa120c535b8c391ab1acd..7ac87e02e9d30d9351b3d3e8da3a646adec613f7 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1602,6 +1602,11 @@ public class CraftEventFactory {
         InventoryCloseEvent event = new InventoryCloseEvent(human.containerMenu.getBukkitView(), reason); // Paper
         human.level().getCraftServer().getPluginManager().callEvent(event);
         human.containerMenu.transferTo(human.inventoryMenu, human.getBukkitEntity());
+        // EMC start - Temporary Meta API
+        if (event.getInventory().getViewers().isEmpty()) {
+            org.bukkit.craftbukkit.inventory.CraftInventory.meta.remove(((org.bukkit.craftbukkit.inventory.CraftInventory) event.getInventory()).getInventory());
+        }
+        // EMC end - Temporary Meta API
     }
 
     public static ItemStack handleEditBookEvent(ServerPlayer player, int itemInHandIndex, ItemStack itemInHand, ItemStack newBookItem) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
index 9d5fe6fc727948955d89f9ed5e181185393a9611..c821b74bfe98a9684580ba94d37a719f07422fef 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
@@ -448,6 +448,12 @@ public class CraftInventory implements Inventory {
         return count;
     }
     // Paper end
+    // EMC start - Temporary Meta API
+    public static final java.util.WeakHashMap<Container, java.util.Map<org.bukkit.NamespacedKey, Object>> meta = new java.util.WeakHashMap<>(64);
+    public @org.jetbrains.annotations.NotNull java.util.Map<org.bukkit.NamespacedKey, Object> getTemporaryMeta() {
+        return meta.computeIfAbsent(inventory, k -> new HashMap<>());
+    }
+    // EMC end - Temporary Meta API
 
     @Override
     public ListIterator<ItemStack> iterator() {
