From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 18 Dec 2014 22:48:46 -0500
Subject: [PATCH] Add BlockBreakNaturally Event

To give reliable control over all blocks dropping to world to restore custom item meta

diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 2779d819eb310574e456725f47bc5ced457024b9..cac5f323e30cccbf0fae7f2a149b57ff6fef78a5 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -410,6 +410,19 @@ public class CraftEventFactory {
         Bukkit.getPluginManager().callEvent(event);
 
         if (!event.isCancelled()) {
+            // EMC start - BlockBreakNaturallyEvent
+            com.google.common.collect.Multimap<org.bukkit.Location, org.bukkit.entity.Item> drops = com.google.common.collect.ArrayListMultimap.create();
+            for (Item item : list) {
+                ItemEntity entityItem = (ItemEntity) ((CraftEntity) item).getHandle();
+                if (entityItem.droppedPosition == null) {
+                    entityItem.droppedPosition = new BlockPos(block.getX(), block.getY(), block.getZ());
+                }
+                drops.put(new Location(entityItem.level().getWorld(), entityItem.droppedPosition.getX(), entityItem.droppedPosition.getY(), entityItem.droppedPosition.getZ()), (org.bukkit.entity.Item) entityItem.getBukkitEntity());
+            }
+            for (org.bukkit.Location location : drops.keySet()) {
+                new com.empireminecraft.empirecraft.event.block.BlockBreakNaturallyEvent(location, drops.get(location)).callEvent();
+            }
+            // EMC end - BlockBreakNaturallyEvent
             for (Item bukkit : list) {
                 if (!bukkit.isValid()) {
                     Entity item = ((org.bukkit.craftbukkit.entity.CraftItem) bukkit).getHandle();
