From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 18 Dec 2014 22:48:46 -0500
Subject: [PATCH] Add BlockBreakNaturally Event

To give reliable control over all blocks dropping to world to restore custom item meta

diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 585df7aaf000f5867f6f5afa9a663a767b0e4ba9..a8af557b4a175df8f54538af7802c134f09a82c2 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -471,6 +471,19 @@ public class CraftEventFactory {
         Bukkit.getPluginManager().callEvent(event);
 
         if (!event.isCancelled()) {
+            // EMC start - BlockBreakNaturallyEvent
+            com.google.common.collect.Multimap<org.bukkit.Location, org.bukkit.entity.Item> drops = com.google.common.collect.ArrayListMultimap.create();
+            for (Item item : list) {
+                ItemEntity entityItem = (ItemEntity) ((CraftEntity) item).getHandle();
+                if (entityItem.droppedPosition == null) {
+                    entityItem.droppedPosition = new BlockPos(block.getX(), block.getY(), block.getZ());
+                }
+                drops.put(io.papermc.paper.util.MCUtil.toLocation(entityItem.level(), entityItem.droppedPosition), (org.bukkit.entity.Item) entityItem.getBukkitEntity());
+            }
+            for (org.bukkit.Location location : drops.keySet()) {
+                new com.empireminecraft.empirecraft.event.block.BlockBreakNaturallyEvent(location, drops.get(location)).callEvent();
+            }
+            // EMC end - BlockBreakNaturallyEvent
             // Paper start - Allow adding items to BlockDropItemEvent
             for (Item bukkit : list) {
                 if (!bukkit.isValid()) {
