From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 29 Dec 2016 23:27:50 -0500
Subject: [PATCH] Make CraftMetaItem not crash on HideFlags error


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/SerializableMeta.java b/src/main/java/org/bukkit/craftbukkit/inventory/SerializableMeta.java
index 1fc7af0f8e32bc7c07e187241d78220e8b017945..474264f8822a9321fcc4b363b19ba17faecf9919 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/SerializableMeta.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/SerializableMeta.java
@@ -136,7 +136,13 @@ public final class SerializableMeta implements ConfigurationSerializable {
             }
             return null;
         }
-        throw new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz);
+        // EMC start
+        if (!nullable) {
+            throw new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz);
+        }
+        new IllegalArgumentException(field + "(" + object + ") is not a valid " + clazz).printStackTrace();
+        return null;
+        // EMC end
     }
 
     public static <T> List<T> getList(Class<T> clazz, Map<?, ?> map, Object field) {
