From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: chickeneer <emcchickeneer@gmail.com>
Date: Sun, 29 Dec 2024 02:02:47 -0600
Subject: [PATCH] Fix deserialization issue with leather armor


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaColorableArmor.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaColorableArmor.java
index 51d7263cdd34359d9cdf72cc01ba654b519f838d..8c7a63696862f40d7932edee7b3a5cf8da95ab6d 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaColorableArmor.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaColorableArmor.java
@@ -7,6 +7,7 @@ import net.minecraft.core.component.DataComponentPatch;
 import org.bukkit.Color;
 import org.bukkit.configuration.serialization.DelegateDeserialization;
 import org.bukkit.inventory.meta.ColorableArmorMeta;
+import org.bukkit.inventory.meta.LeatherArmorMeta;
 
 @DelegateDeserialization(SerializableMeta.class)
 public class CraftMetaColorableArmor extends CraftMetaArmor implements ColorableArmorMeta {
@@ -17,6 +18,13 @@ public class CraftMetaColorableArmor extends CraftMetaArmor implements Colorable
         super(meta);
         // Paper start
         if (!(meta instanceof CraftMetaColorableArmor armorMeta)) {
+            // EMC start
+            if (meta instanceof CraftMetaLeatherArmor armorMeta) {
+                if (armorMeta.hasColor()) {
+                    this.color = armorMeta.getColor().asRGB();
+                }
+            }
+            // EMC end
             return;
         }
 
