From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 30 Apr 2015 00:46:26 -0400
Subject: [PATCH] Don't require flag for sign conversion

Also reverts part of Paper's old sign conversion patch

diff --git a/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
index 4da4edae517a0efec6e03a719ec47b700509dab1..2876485b5b7f749b1670aada24ac8f580447c658 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -26,7 +26,7 @@ import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 
 public class SignBlockEntity extends BlockEntity implements CommandSource { // CraftBukkit - implements
-    private static final boolean CONVERT_LEGACY_SIGNS = Boolean.getBoolean("convertLegacySigns"); // Paper
+    private static final boolean CONVERT_LEGACY_SIGNS = true; // EMC
 
     public static final int LINES = 4;
     private static final int MAX_TEXT_LINE_WIDTH = 90;
@@ -104,7 +104,7 @@ public class SignBlockEntity extends BlockEntity implements CommandSource { // C
         // CraftBukkit start - Add an option to convert signs correctly
         // This is done with a flag instead of all the time because
         // we have no way to tell whether a sign is from 1.7.10 or 1.8
-        boolean oldSign = Boolean.getBoolean("convertLegacySigns") && !nbt.getBoolean("Bukkit.isConverted");
+        boolean oldSign = !nbt.getBoolean("Bukkit.isConverted"); // EMC - don't use flag
         // CraftBukkit end
 
         for (int i = 0; i < 4; ++i) {
