--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -60,6 +_,27 @@
     }
 
     public boolean isNearPlayer(Level level, BlockPos pos) {
+        // EMC start - Add SpawnerInitiateEvent
+        SpawnData spawnData = this.nextSpawnData;
+        if (spawnData == null) {
+            return false;
+        }
+        org.bukkit.entity.EntityType entityType;
+        try (ProblemReporter.ScopedCollector scopedCollector = new ProblemReporter.ScopedCollector(() -> "spawner@" + pos, LOGGER)) {
+            ValueInput valueInput = TagValueInput.create(scopedCollector, level.registryAccess(), spawnData.entityToSpawn());
+            Optional<net.minecraft.world.entity.EntityType<?>> type = net.minecraft.world.entity.EntityType.by(valueInput);
+            entityType = type.map(org.bukkit.craftbukkit.entity.CraftEntityType::minecraftToBukkit).orElse(null);
+        }
+        if (entityType == null) {
+            return false;
+        }
+
+        if (!(new com.empireminecraft.empirecraft.event.SpawnerInitiateEvent(entityType,
+            level.getWorld(),
+            org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level)).callEvent())) {
+            return false;
+        }
+        // EMC end - Add SpawnerInitiateEvent
         return level.hasNearbyAlivePlayerThatAffectsSpawning(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, this.requiredPlayerRange); // Paper - Affects Spawning API
     }
 
