--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -52,6 +_,25 @@
     }
 
     public boolean isNearPlayer(Level level, BlockPos pos) {
+        // EMC start - Add SpawnerInitiateEvent
+        if (this.nextSpawnData == null) {
+            return false;
+        }
+        Optional<net.minecraft.world.entity.EntityType<?>> type = net.minecraft.world.entity.EntityType.by(this.nextSpawnData.getEntityToSpawn());
+        if (type.isEmpty()) {
+            return false;
+        }
+        final int x = pos.getX();
+        final int y = pos.getY();
+        final int z = pos.getZ();
+        org.bukkit.Location loc = new org.bukkit.Location(level.getWorld(), x, y, z);
+
+        if (!(new com.empireminecraft.empirecraft.event.SpawnerInitiateEvent(
+                org.bukkit.entity.EntityType.fromName(net.minecraft.world.entity.EntityType.getKey(type.get()).getPath()),
+                level.getWorld(), loc).callEvent())) {
+            return false;
+        }
+        // EMC end - Add SpawnerInitiateEvent
         return level.hasNearbyAlivePlayerThatAffectsSpawning(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, this.requiredPlayerRange); // Paper - Affects Spawning API
     }
 
