--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -141,7 +_,7 @@
     // CraftBukkit start
     public final io.papermc.paper.antixray.ChunkPacketBlockController chunkPacketBlockController; // Paper - Anti-Xray
     private final CraftWorld world;
-    public net.kyori.adventure.util.TriState pvpMode = net.kyori.adventure.util.TriState.NOT_SET;
+    public net.kyori.adventure.util.TriState pvpMode = net.kyori.adventure.util.TriState.TRUE; // EMC - Force PVP on
     public @Nullable org.bukkit.generator.ChunkGenerator generator;
 
     public boolean captureBlockStates = false;
