--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -434,6 +_,11 @@
 
     public static boolean isInNetherFortressBounds(BlockPos pos, ServerLevel level, MobCategory category, StructureManager structureManager) {
         if (category == MobCategory.MONSTER && level.getBlockState(pos.below()).is(Blocks.NETHER_BRICKS)) {
+            // EMC start - add nether check and skip structure requirement
+            if (level.getWorld().getEnvironment() == org.bukkit.World.Environment.NETHER) {
+                return true;
+            }
+            // EMC end
             Structure structure = structureManager.registryAccess().lookupOrThrow(Registries.STRUCTURE).getValue(BuiltinStructures.FORTRESS);
             return structure != null && structureManager.getStructureAt(pos, structure).isValid();
         } else {
