--- a/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
@@ -69,18 +_,20 @@
             if (IS_EDIBLE.test(blockState)) { // Paper - fix wrong block state
                 if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockPos, blockState.getFluidState().createLegacyBlock(), !getServerLevel(this.level).getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))) { // CraftBukkit // Paper - fix wrong block state
                     this.level.destroyBlock(blockPos, false);
+                    this.mob.ate(); // EMC - moved from below - Sheep should not regrow wool if grass eating is cancelled
                 }
 
-                this.mob.ate();
+                //this.mob.ate(); // EMC - move up - Sheep should not regrow wool if grass eating is cancelled
             } else {
                 BlockPos blockPos1 = blockPos.below();
                 if (this.level.getBlockState(blockPos1).is(Blocks.GRASS_BLOCK)) {
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockPos1, Blocks.DIRT.defaultBlockState(), !getServerLevel(this.level).getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))) { // CraftBukkit // Paper - Fix wrong block state
                         this.level.levelEvent(2001, blockPos1, Block.getId(Blocks.GRASS_BLOCK.defaultBlockState()));
                         this.level.setBlock(blockPos1, Blocks.DIRT.defaultBlockState(), 2);
+                        this.mob.ate(); // EMC - move from below - Sheep should not regrow wool if grass eating is cancelled
                     }
 
-                    this.mob.ate();
+                    //this.mob.ate(); // EMC - move up - Sheep should not regrow wool if grass eating is cancelled
                 }
             }
         }
