--- a/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
@@ -66,18 +_,20 @@
             if (IS_EDIBLE.test(blockState)) { // Paper - fix wrong block state
                 if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.mob, pos, blockState.getFluidState().createLegacyBlock(), !getServerLevel(this.level).getGameRules().get(GameRules.MOB_GRIEFING))) { // CraftBukkit // Paper - fix wrong block state
                     this.level.destroyBlock(pos, false);
+                    this.mob.ate(); // EMC - moved from below - Sheep should not regrow wool if grass eating is cancelled
                 }
 
-                this.mob.ate();
+                //this.mob.ate(); // EMC - move up - Sheep should not regrow wool if grass eating is cancelled
             } else {
                 BlockPos below = pos.below();
                 if (this.level.getBlockState(below).is(Blocks.GRASS_BLOCK)) {
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.mob, below, Blocks.DIRT.defaultBlockState(), !getServerLevel(this.level).getGameRules().get(GameRules.MOB_GRIEFING))) { // CraftBukkit // Paper - Fix wrong block state
                         this.level.levelEvent(LevelEvent.PARTICLES_DESTROY_BLOCK, below, Block.getId(Blocks.GRASS_BLOCK.defaultBlockState()));
                         this.level.setBlock(below, Blocks.DIRT.defaultBlockState(), Block.UPDATE_CLIENTS);
+                        this.mob.ate(); // EMC - moved from below - Sheep should not regrow wool if grass eating is cancelled
                     }
 
-                    this.mob.ate();
+                    //this.mob.ate(); // EMC - move up - Sheep should not regrow wool if grass eating is cancelled
                 }
             }
         }
