--- a/net/minecraft/world/entity/monster/spider/Spider.java
+++ b/net/minecraft/world/entity/monster/spider/Spider.java
@@ -178,6 +_,7 @@
         return vehicle.getBbWidth() <= this.getBbWidth() ? new Vec3(0.0, 0.3125 * this.getScale(), 0.0) : super.getVehicleAttachmentPoint(vehicle);
     }
 
+    public boolean attackDuringDay = false; // EMC - Attack During Day API
     private static class SpiderAttackGoal extends MeleeAttackGoal {
         public SpiderAttackGoal(final Spider mob) {
             super(mob, 1.0, true);
@@ -191,6 +_,7 @@
         @Override
         public boolean canContinueToUse() {
             float br = this.mob.getLightLevelDependentMagicValue();
+            if (((Spider) this.mob).attackDuringDay) { br = 0.0F; } // EMC - Attack During Day API
             if (br >= 0.5F && this.mob.getRandom().nextInt(100) == 0) {
                 this.mob.setTarget(null);
                 return false;
@@ -225,6 +_,7 @@
         @Override
         public boolean canUse() {
             float br = this.mob.getLightLevelDependentMagicValue();
+            if (((Spider) this.mob).attackDuringDay) { br = 0.0F; } // EMC - Attack During Day API
             return !(br >= 0.5F) && super.canUse();
         }
     }
